// @taiga-ui/core/esm2015/components/link/link.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core/components/svg';
import { TuiLinkComponent } from './link.component';
import * as i0 from "@angular/core";
export class TuiLinkModule {
}
TuiLinkModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLinkModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLinkModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLinkModule, declarations: [TuiLinkComponent], imports: [CommonModule, TuiSvgModule], exports: [TuiLinkComponent] });
TuiLinkModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLinkModule, imports: [[CommonModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLinkModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule],
                    declarations: [TuiLinkComponent],
                    exports: [TuiLinkComponent],
                }]
        }] });
