// @taiga-ui/kit/esm2015/components/input-year/input-year.module.js
import { NgModule } from '@angular/core';
import { TuiPreventDefaultModule } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiPrimitiveYearPickerModule, TuiTextfieldComponent, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TextMaskModule } from '@taiga-ui/kit/directives';
import { TuiToYearPipeModule } from '@taiga-ui/kit/pipes';
import { TuiInputYearComponent } from './input-year.component';
import { TuiInputYearDirective } from './input-year.directive';
import * as i0 from "@angular/core";
export class TuiInputYearModule {
}
TuiInputYearModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputYearModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputYearModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputYearModule, declarations: [TuiInputYearComponent, TuiInputYearDirective], imports: [TextMaskModule,
        TuiHostedDropdownModule,
        TuiPrimitiveTextfieldModule,
        TuiPrimitiveYearPickerModule,
        TuiPreventDefaultModule,
        TuiTextfieldControllerModule,
        TuiToYearPipeModule], exports: [TuiInputYearComponent, TuiInputYearDirective, TuiTextfieldComponent] });
TuiInputYearModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputYearModule, imports: [[
            TextMaskModule,
            TuiHostedDropdownModule,
            TuiPrimitiveTextfieldModule,
            TuiPrimitiveYearPickerModule,
            TuiPreventDefaultModule,
            TuiTextfieldControllerModule,
            TuiToYearPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputYearModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        TextMaskModule,
                        TuiHostedDropdownModule,
                        TuiPrimitiveTextfieldModule,
                        TuiPrimitiveYearPickerModule,
                        TuiPreventDefaultModule,
                        TuiTextfieldControllerModule,
                        TuiToYearPipeModule,
                    ],
                    declarations: [TuiInputYearComponent, TuiInputYearDirective],
                    exports: [TuiInputYearComponent, TuiInputYearDirective, TuiTextfieldComponent],
                }]
        }] });
