// @taiga-ui/kit/esm2015/components/stepper/stepper.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiStepComponent } from './step/step.component';
import { TuiStepperComponent } from './stepper.component';
import * as i0 from "@angular/core";
export class TuiStepperModule {
}
TuiStepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiStepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStepperModule, declarations: [TuiStepperComponent, TuiStepComponent], imports: [CommonModule, TuiSvgModule], exports: [TuiStepperComponent, TuiStepComponent] });
TuiStepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStepperModule, imports: [[CommonModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule],
                    declarations: [TuiStepperComponent, TuiStepComponent],
                    exports: [TuiStepperComponent, TuiStepComponent],
                }]
        }] });
