// @taiga-ui/kit/esm2015/components/select-option/select-option.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiScrollIntoViewModule, TuiSvgModule } from '@taiga-ui/core';
import { TuiSelectOptionComponent } from './select-option.component';
import * as i0 from "@angular/core";
export class TuiSelectOptionModule {
}
TuiSelectOptionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectOptionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSelectOptionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectOptionModule, declarations: [TuiSelectOptionComponent], imports: [CommonModule, TuiSvgModule, TuiScrollIntoViewModule], exports: [TuiSelectOptionComponent] });
TuiSelectOptionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectOptionModule, imports: [[CommonModule, TuiSvgModule, TuiScrollIntoViewModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectOptionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiScrollIntoViewModule],
                    declarations: [TuiSelectOptionComponent],
                    exports: [TuiSelectOptionComponent],
                }]
        }] });
