// @taiga-ui/cdk/esm2015/abstract/portal-host.js
import { ComponentFactoryResolver, Directive, ElementRef, Inject, INJECTOR, Injector, ViewChild, ViewContainerRef, } from '@angular/core';
// TODO: find the best way for prevent cycle
// eslint-disable-next-line import/no-cycle
import { AbstractTuiPortalService } from './portal-service';
import * as i0 from "@angular/core";
import * as i1 from "./portal-service";
/**
 * Abstract class for host element for dynamically created portals.
 */
export class AbstractTuiPortalHostComponent {
    constructor(injector, el, portalService) {
        this.injector = injector;
        this.el = el;
        portalService.attach(this);
    }
    get clientRect() {
        return this.el.nativeElement.getBoundingClientRect();
    }
    addComponentChild(component) {
        const parent = component.createInjector(this.injector);
        const resolver = parent.get(ComponentFactoryResolver);
        const factory = resolver.resolveComponentFactory(component.component);
        const providers = [{ provide: AbstractTuiPortalHostComponent, useValue: this }];
        const injector = Injector.create({ parent, providers });
        const ref = this.vcr.createComponent(factory, undefined, injector);
        ref.changeDetectorRef.detectChanges();
        return ref;
    }
    addTemplateChild(templateRef, context) {
        return this.vcr.createEmbeddedView(templateRef, context);
    }
}
AbstractTuiPortalHostComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiPortalHostComponent, deps: [{ token: INJECTOR }, { token: ElementRef }, { token: AbstractTuiPortalService }], target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiPortalHostComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiPortalHostComponent, viewQueries: [{ propertyName: "vcr", first: true, predicate: ["viewContainer"], descendants: true, read: ViewContainerRef }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiPortalHostComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.Injector, decorators: [{
                    type: Inject,
                    args: [INJECTOR]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i1.AbstractTuiPortalService, decorators: [{
                    type: Inject,
                    args: [AbstractTuiPortalService]
                }] }]; }, propDecorators: { vcr: [{
                type: ViewChild,
                args: [`viewContainer`, { read: ViewContainerRef }]
            }] } });
