// @taiga-ui/cdk/esm2015/abstract/theme-switcher.js
import { DOCUMENT } from '@angular/common';
import { Directive, Inject } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Use this abstract class to create your own toggleable themes.
 * A component extending this class must have CSS variables definitions
 * and have ViewEncapsulation set to NONE. A boolean input allows to
 * switch theme on or off.
 */
export class AbstractTuiThemeSwitcher {
    constructor(doc) {
        this.doc = doc;
        if (this.style !== null) {
            this.addTheme();
            return;
        }
        const styles = this.doc.head.querySelectorAll(`style`);
        this.constructor.style =
            styles[styles.length - 1];
    }
    get style() {
        return this.constructor.style;
    }
    ngOnDestroy() {
        this.removeTheme();
    }
    addTheme() {
        if (this.style) {
            this.doc.head.appendChild(this.style);
        }
    }
    removeTheme() {
        if (this.style && this.doc.head.contains(this.style)) {
            this.doc.head.removeChild(this.style);
        }
    }
}
AbstractTuiThemeSwitcher.style = null;
AbstractTuiThemeSwitcher.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiThemeSwitcher, deps: [{ token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiThemeSwitcher.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiThemeSwitcher, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiThemeSwitcher, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; } });
