// @taiga-ui/kit/esm2015/components/slider/slider.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSliderKeyStepsDirective } from './helpers/slider-key-steps.directive';
import { TuiSliderReadonlyDirective } from './helpers/slider-readonly.directive';
import { TuiSliderThumbLabelComponent } from './helpers/slider-thumb-label/slider-thumb-label.component';
import { TuiSliderComponent } from './slider.component';
import * as i0 from "@angular/core";
export class TuiSliderModule {
}
TuiSliderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSliderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSliderModule, declarations: [TuiSliderComponent,
        TuiSliderThumbLabelComponent,
        TuiSliderKeyStepsDirective,
        TuiSliderReadonlyDirective], imports: [CommonModule], exports: [TuiSliderComponent,
        TuiSliderThumbLabelComponent,
        TuiSliderKeyStepsDirective,
        TuiSliderReadonlyDirective] });
TuiSliderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSliderModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSliderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        TuiSliderComponent,
                        TuiSliderThumbLabelComponent,
                        TuiSliderKeyStepsDirective,
                        TuiSliderReadonlyDirective,
                    ],
                    exports: [
                        TuiSliderComponent,
                        TuiSliderThumbLabelComponent,
                        TuiSliderKeyStepsDirective,
                        TuiSliderReadonlyDirective,
                    ],
                }]
        }] });
