// @taiga-ui/kit/esm2015/components/input-range/input-range.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiActiveZoneModule, TuiPressedModule } from '@taiga-ui/cdk';
import { TuiTextfieldControllerModule, TuiWrapperModule } from '@taiga-ui/core';
import { TuiInputNumberModule } from '@taiga-ui/kit/components/input-number';
import { TuiRangeModule } from '@taiga-ui/kit/components/range';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputRangeComponent } from './input-range.component';
import * as i0 from "@angular/core";
export class TuiInputRangeModule {
}
TuiInputRangeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputRangeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputRangeModule, declarations: [TuiInputRangeComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiActiveZoneModule,
        TuiInputNumberModule,
        TuiPressedModule,
        TuiRangeModule,
        TuiWrapperModule,
        TuiTextfieldControllerModule], exports: [TuiInputRangeComponent] });
TuiInputRangeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputRangeModule, imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiActiveZoneModule,
            TuiInputNumberModule,
            TuiPressedModule,
            TuiRangeModule,
            TuiWrapperModule,
            TuiTextfieldControllerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiActiveZoneModule,
                        TuiInputNumberModule,
                        TuiPressedModule,
                        TuiRangeModule,
                        TuiWrapperModule,
                        TuiTextfieldControllerModule,
                    ],
                    declarations: [TuiInputRangeComponent],
                    exports: [TuiInputRangeComponent],
                }]
        }] });
