// @taiga-ui/kit/esm2015/components/rating/rating.module.js
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { TuiFocusableModule, TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiRatingComponent } from './rating.component';
import * as i0 from "@angular/core";
export class TuiRatingModule {
}
TuiRatingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRatingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, declarations: [TuiRatingComponent], imports: [TuiSvgModule,
        TuiFocusableModule,
        TuiRepeatTimesModule,
        FormsModule,
        ReactiveFormsModule], exports: [TuiRatingComponent] });
TuiRatingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, imports: [[
            TuiSvgModule,
            TuiFocusableModule,
            TuiRepeatTimesModule,
            FormsModule,
            ReactiveFormsModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        TuiSvgModule,
                        TuiFocusableModule,
                        TuiRepeatTimesModule,
                        FormsModule,
                        ReactiveFormsModule,
                    ],
                    declarations: [TuiRatingComponent],
                    exports: [TuiRatingComponent],
                }]
        }] });
