// @taiga-ui/cdk/schematics/utils/colored-log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUCCESS_SYMBOL = exports.PROCESSING_SYMBOL = exports.REPLACE_SYMBOL = exports.FINISH_SYMBOL = exports.START_SYMBOL = exports.SMALL_TAB_SYMBOL = exports.titleLog = exports.infoLog = exports.successLog = exports.errorLog = exports.processLog = void 0;
function processLog(message) {
    console.info(`\x1B[36m%s\x1B[0m`, message);
}
exports.processLog = processLog;
function errorLog(message) {
    console.info(`\x1B[31m%s\x1B[0m`, message);
}
exports.errorLog = errorLog;
function successLog(message) {
    console.info(`\x1B[32m%s\x1B[0m`, message);
}
exports.successLog = successLog;
function infoLog(message) {
    console.info(`\x1B[34m%s\x1B[0m`, message);
}
exports.infoLog = infoLog;
function titleLog(message) {
    console.info(`\x1B[35m`, message);
}
exports.titleLog = titleLog;
exports.SMALL_TAB_SYMBOL = `  `; // @note: if you use \t then we have big gaps
exports.START_SYMBOL = `ðŸš€`;
exports.FINISH_SYMBOL = `ðŸ†`;
exports.REPLACE_SYMBOL = `âš¡ï¸`;
exports.PROCESSING_SYMBOL = `> `;
exports.SUCCESS_SYMBOL = `âœ… `;
