// @taiga-ui/kit/esm2015/components/action/action.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusableModule, TuiFocusVisibleModule } from '@taiga-ui/cdk';
import { TuiMarkerIconModule } from '@taiga-ui/kit/components/marker-icon';
import { TuiActionComponent } from './action.component';
import * as i0 from "@angular/core";
export class TuiActionModule {
}
TuiActionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiActionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionModule, declarations: [TuiActionComponent], imports: [CommonModule,
        TuiFocusableModule,
        TuiFocusVisibleModule,
        TuiMarkerIconModule], exports: [TuiActionComponent] });
TuiActionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionModule, imports: [[
            CommonModule,
            TuiFocusableModule,
            TuiFocusVisibleModule,
            TuiMarkerIconModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiFocusableModule,
                        TuiFocusVisibleModule,
                        TuiMarkerIconModule,
                    ],
                    declarations: [TuiActionComponent],
                    exports: [TuiActionComponent],
                }]
        }] });
