// @taiga-ui/cdk/esm2015/directives/copy-processor/copy-processor.directive.js
import { __decorate } from "tslib";
import { Directive, HostListener, Inject, Input } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { tuiDefaultProp } from '@taiga-ui/cdk/decorators';
import { tuiGetSelectedText } from '@taiga-ui/cdk/utils';
import { identity } from 'rxjs';
import * as i0 from "@angular/core";
export class TuiCopyProcessorDirective {
    constructor(win) {
        this.win = win;
        this.tuiCopyProcessor = identity;
    }
    onCopy(event) {
        var _a;
        const text = tuiGetSelectedText(this.win);
        if (text) {
            (_a = event.clipboardData) === null || _a === void 0 ? void 0 : _a.setData('text/plain', this.tuiCopyProcessor(text));
        }
    }
}
TuiCopyProcessorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCopyProcessorDirective, deps: [{ token: WINDOW }], target: i0.ɵɵFactoryTarget.Directive });
TuiCopyProcessorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCopyProcessorDirective, selector: "[tuiCopyProcessor]", inputs: { tuiCopyProcessor: "tuiCopyProcessor" }, host: { listeners: { "copy.prevent": "onCopy($event)" } }, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiCopyProcessorDirective.prototype, "tuiCopyProcessor", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCopyProcessorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCopyProcessor]',
                }]
        }], ctorParameters: function () { return [{ type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }]; }, propDecorators: { tuiCopyProcessor: [{
                type: Input
            }], onCopy: [{
                type: HostListener,
                args: ['copy.prevent', ['$event']]
            }] } });
