// @taiga-ui/kit/esm2015/components/toggle/toggle.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiCheckedModule, TuiFocusableModule, TuiFocusedModule, TuiFocusVisibleModule, TuiPressedModule, } from '@taiga-ui/cdk';
import { TuiLoaderModule, TuiSvgModule, TuiWrapperModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiToggleComponent } from './toggle.component';
import * as i0 from "@angular/core";
export class TuiToggleModule {
}
TuiToggleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToggleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiToggleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToggleModule, declarations: [TuiToggleComponent], imports: [CommonModule,
        TuiFocusedModule,
        TuiFocusableModule,
        TuiPressedModule,
        TuiFocusVisibleModule,
        TuiCheckedModule,
        TuiWrapperModule,
        TuiSvgModule,
        TuiLoaderModule,
        PolymorpheusModule], exports: [TuiToggleComponent] });
TuiToggleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToggleModule, imports: [[
            CommonModule,
            TuiFocusedModule,
            TuiFocusableModule,
            TuiPressedModule,
            TuiFocusVisibleModule,
            TuiCheckedModule,
            TuiWrapperModule,
            TuiSvgModule,
            TuiLoaderModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToggleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiFocusedModule,
                        TuiFocusableModule,
                        TuiPressedModule,
                        TuiFocusVisibleModule,
                        TuiCheckedModule,
                        TuiWrapperModule,
                        TuiSvgModule,
                        TuiLoaderModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiToggleComponent],
                    exports: [TuiToggleComponent],
                }]
        }] });
