// @taiga-ui/core/esm2015/pipes/order-week-days/order-week-days.pipe.js
import { Inject, Pipe } from '@angular/core';
import { TUI_FIRST_DAY_OF_WEEK } from '@taiga-ui/core/tokens';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiOrderWeekDaysPipe {
    constructor(firstDayOfWeekIndex) {
        this.firstDayOfWeekIndex = firstDayOfWeekIndex;
    }
    transform(mondayFirstWeekDays$) {
        return mondayFirstWeekDays$.pipe(map(convertToSundayFirstWeekFormat), map(weekDays => [
            ...weekDays.slice(this.firstDayOfWeekIndex),
            ...weekDays.slice(0, this.firstDayOfWeekIndex),
        ]));
    }
}
TuiOrderWeekDaysPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiOrderWeekDaysPipe, deps: [{ token: TUI_FIRST_DAY_OF_WEEK }], target: i0.ɵɵFactoryTarget.Pipe });
TuiOrderWeekDaysPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiOrderWeekDaysPipe, name: "tuiOrderWeekDays" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiOrderWeekDaysPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: `tuiOrderWeekDays`,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_FIRST_DAY_OF_WEEK]
                }] }]; } });
function convertToSundayFirstWeekFormat(weekDaysNames) {
    const sundayIndex = weekDaysNames.length - 1;
    return [
        weekDaysNames[sundayIndex],
        ...weekDaysNames.slice(0, sundayIndex),
    ];
}
