// @tinkoff/ng-polymorpheus/esm2015/directives/template.js
import { Directive, Self } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * ng-template wrapper directive also stores {@link ChangeDetectorRef} to properly handle change detection.
 */
export class PolymorpheusTemplate {
    constructor(template, cdr) {
        this.template = template;
        this.cdr = cdr;
        this.polymorpheus = '';
    }
    check() {
        this.cdr.markForCheck();
    }
    static ngTemplateContextGuard(_dir, _ctx) {
        return true;
    }
}
/** @nocollapse */ PolymorpheusTemplate.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0, type: PolymorpheusTemplate, deps: [{ token: i0.TemplateRef, self: true }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
/** @nocollapse */ PolymorpheusTemplate.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.0", type: PolymorpheusTemplate, selector: "ng-template[polymorpheus]", inputs: { polymorpheus: "polymorpheus" }, exportAs: ["polymorpheus"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0, type: PolymorpheusTemplate, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[polymorpheus]',
                    exportAs: 'polymorpheus',
                    inputs: ['polymorpheus'],
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Self
                }] }, { type: i0.ChangeDetectorRef }]; } });
