// @taiga-ui/kit/esm2015/components/calendar-month/calendar-month.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusableModule, TuiHoveredModule, TuiLetModule, TuiPressedModule, } from '@taiga-ui/cdk';
import { TuiLinkModule, TuiPrimitiveSpinButtonModule, TuiPrimitiveYearPickerModule, TuiScrollbarModule, } from '@taiga-ui/core';
import { TuiCalendarMonthComponent } from './calendar-month.component';
import * as i0 from "@angular/core";
export class TuiCalendarMonthModule {
}
TuiCalendarMonthModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarMonthModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCalendarMonthModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarMonthModule, declarations: [TuiCalendarMonthComponent], imports: [CommonModule,
        TuiPrimitiveYearPickerModule,
        TuiPrimitiveSpinButtonModule,
        TuiScrollbarModule,
        TuiLinkModule,
        TuiLetModule,
        TuiHoveredModule,
        TuiPressedModule,
        TuiFocusableModule], exports: [TuiCalendarMonthComponent] });
TuiCalendarMonthModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarMonthModule, imports: [[
            CommonModule,
            TuiPrimitiveYearPickerModule,
            TuiPrimitiveSpinButtonModule,
            TuiScrollbarModule,
            TuiLinkModule,
            TuiLetModule,
            TuiHoveredModule,
            TuiPressedModule,
            TuiFocusableModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarMonthModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiPrimitiveYearPickerModule,
                        TuiPrimitiveSpinButtonModule,
                        TuiScrollbarModule,
                        TuiLinkModule,
                        TuiLetModule,
                        TuiHoveredModule,
                        TuiPressedModule,
                        TuiFocusableModule,
                    ],
                    declarations: [TuiCalendarMonthComponent],
                    exports: [TuiCalendarMonthComponent],
                }]
        }] });
