// @taiga-ui/kit/esm2015/components/radio/radio.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiCheckedModule, TuiFocusableModule, TuiFocusedModule, TuiFocusVisibleModule, } from '@taiga-ui/cdk';
import { TuiWrapperModule } from '@taiga-ui/core';
import { TuiRadioComponent } from './radio.component';
import * as i0 from "@angular/core";
export class TuiRadioModule {
}
TuiRadioModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRadioModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioModule, declarations: [TuiRadioComponent], imports: [CommonModule,
        TuiCheckedModule,
        TuiFocusableModule,
        TuiFocusedModule,
        TuiFocusVisibleModule,
        TuiWrapperModule], exports: [TuiRadioComponent] });
TuiRadioModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioModule, imports: [[
            CommonModule,
            TuiCheckedModule,
            TuiFocusableModule,
            TuiFocusedModule,
            TuiFocusVisibleModule,
            TuiWrapperModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiCheckedModule,
                        TuiFocusableModule,
                        TuiFocusedModule,
                        TuiFocusVisibleModule,
                        TuiWrapperModule,
                    ],
                    declarations: [TuiRadioComponent],
                    exports: [TuiRadioComponent],
                }]
        }] });
