// @taiga-ui/kit/esm2015/internal/primitive-calendar-range/primitive-calendar-range.module.js
import { NgModule } from '@angular/core';
import { TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiCalendarModule, TuiScrollbarModule } from '@taiga-ui/core';
import { TuiPrimitiveCalendarRangeComponent } from './primitive-calendar-range.component';
import * as i0 from "@angular/core";
/**
 * @internal
 */
export class TuiPrimitiveCalendarRangeModule {
}
TuiPrimitiveCalendarRangeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveCalendarRangeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarRangeModule, declarations: [TuiPrimitiveCalendarRangeComponent], imports: [TuiMapperPipeModule, TuiScrollbarModule, TuiCalendarModule], exports: [TuiPrimitiveCalendarRangeComponent] });
TuiPrimitiveCalendarRangeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarRangeModule, imports: [[TuiMapperPipeModule, TuiScrollbarModule, TuiCalendarModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiMapperPipeModule, TuiScrollbarModule, TuiCalendarModule],
                    declarations: [TuiPrimitiveCalendarRangeComponent],
                    exports: [TuiPrimitiveCalendarRangeComponent],
                }]
        }] });
