// @taiga-ui/kit/esm2015/components/input-slider/input-slider.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiFocusableModule } from '@taiga-ui/cdk';
import { TuiTextfieldControllerModule } from '@taiga-ui/core';
import { TuiInputNumberModule } from '@taiga-ui/kit/components/input-number';
import { TuiSliderModule } from '@taiga-ui/kit/components/slider';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputSliderComponent } from './input-slider.component';
import * as i0 from "@angular/core";
export class TuiInputSliderModule {
}
TuiInputSliderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputSliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputSliderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputSliderModule, declarations: [TuiInputSliderComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiFocusableModule,
        TuiInputNumberModule,
        TuiSliderModule,
        TuiTextfieldControllerModule], exports: [TuiInputSliderComponent] });
TuiInputSliderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputSliderModule, imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiFocusableModule,
            TuiInputNumberModule,
            TuiSliderModule,
            TuiTextfieldControllerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputSliderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiFocusableModule,
                        TuiInputNumberModule,
                        TuiSliderModule,
                        TuiTextfieldControllerModule,
                    ],
                    declarations: [TuiInputSliderComponent],
                    exports: [TuiInputSliderComponent],
                }]
        }] });
