// @taiga-ui/kit/esm2015/components/badge/badge.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiBadgeComponent } from './badge.component';
import * as i0 from "@angular/core";
export class TuiBadgeModule {
}
TuiBadgeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBadgeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, declarations: [TuiBadgeComponent], imports: [CommonModule], exports: [TuiBadgeComponent] });
TuiBadgeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBadgeComponent],
                    exports: [TuiBadgeComponent],
                }]
        }] });
