// @taiga-ui/kit/esm2015/components/push/push-alert.directive.js
import { ChangeDetectorRef, Directive, forwardRef, Inject, Input, Self, TemplateRef, } from '@angular/core';
import { TuiDestroyService, tuiIfMap } from '@taiga-ui/cdk';
import { PolymorpheusTemplate } from '@tinkoff/ng-polymorpheus';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
// TODO: find the best way for prevent cycle
// eslint-disable-next-line import/no-cycle
import { TuiPushService } from './push.service';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
import * as i2 from "./push.service";
export class TuiPushAlertDirective extends PolymorpheusTemplate {
    constructor(template, cdr, destroy$, push) {
        super(template, cdr);
        this.show$ = new Subject();
        this.show$
            .pipe(tuiIfMap(() => push.open(this)), takeUntil(destroy$))
            .subscribe();
    }
    set tuiPush(show) {
        this.show$.next(show);
    }
}
TuiPushAlertDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushAlertDirective, deps: [{ token: TemplateRef }, { token: ChangeDetectorRef }, { token: TuiDestroyService, self: true }, { token: forwardRef(() => TuiPushService) }], target: i0.ɵɵFactoryTarget.Directive });
TuiPushAlertDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPushAlertDirective, selector: "[tuiPush]", inputs: { tuiPush: "tuiPush" }, providers: [TuiDestroyService], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushAlertDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPush]',
                    providers: [TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }, { type: i0.ChangeDetectorRef, decorators: [{
                    type: Inject,
                    args: [ChangeDetectorRef]
                }] }, { type: i1.Observable, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }, { type: i2.TuiPushService, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => TuiPushService)]
                }] }]; }, propDecorators: { tuiPush: [{
                type: Input
            }] } });
