// @taiga-ui/kit/esm2015/components/push/push.directive.js
import { Directive, Inject, Optional, SkipSelf } from '@angular/core';
import { TUI_BUTTON_OPTIONS, TuiModeDirective } from '@taiga-ui/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
export class TuiPushDirective extends TuiModeDirective {
    constructor(modeDirective) {
        var _a;
        super();
        this.modeDirective = modeDirective;
        this.size = 's';
        this.shape = null;
        this.change$ = ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.change$) || new Subject();
    }
    get appearance() {
        var _a;
        return ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.mode) === 'onDark' ? 'accent' : 'secondary';
    }
}
TuiPushDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushDirective, deps: [{ token: TuiModeDirective, optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.Directive });
TuiPushDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPushDirective, selector: "tui-push", providers: [
        {
            provide: TuiModeDirective,
            useExisting: TuiPushDirective,
        },
        {
            provide: TUI_BUTTON_OPTIONS,
            useExisting: TuiPushDirective,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-push',
                    providers: [
                        {
                            provide: TuiModeDirective,
                            useExisting: TuiPushDirective,
                        },
                        {
                            provide: TUI_BUTTON_OPTIONS,
                            useExisting: TuiPushDirective,
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiModeDirective, decorators: [{
                    type: Optional
                }, {
                    type: SkipSelf
                }, {
                    type: Inject,
                    args: [TuiModeDirective]
                }] }]; } });
