// @taiga-ui/core/esm2015/internal/primitive-year-month-pagination/primitive-year-month-pagination.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusableModule } from '@taiga-ui/cdk';
import { TuiLinkModule } from '@taiga-ui/core/components/link';
import { TuiPrimitiveSpinButtonModule } from '@taiga-ui/core/components/primitive-spin-button';
import { TuiMonthPipeModule } from '@taiga-ui/core/pipes';
import { TuiPrimitiveYearMonthPaginationComponent } from './primitive-year-month-pagination.component';
import * as i0 from "@angular/core";
export class TuiPrimitiveYearMonthPaginationModule {
}
TuiPrimitiveYearMonthPaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearMonthPaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveYearMonthPaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearMonthPaginationModule, declarations: [TuiPrimitiveYearMonthPaginationComponent], imports: [CommonModule,
        TuiFocusableModule,
        TuiPrimitiveSpinButtonModule,
        TuiLinkModule,
        TuiMonthPipeModule], exports: [TuiPrimitiveYearMonthPaginationComponent] });
TuiPrimitiveYearMonthPaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearMonthPaginationModule, imports: [[
            CommonModule,
            TuiFocusableModule,
            TuiPrimitiveSpinButtonModule,
            TuiLinkModule,
            TuiMonthPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearMonthPaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiFocusableModule,
                        TuiPrimitiveSpinButtonModule,
                        TuiLinkModule,
                        TuiMonthPipeModule,
                    ],
                    declarations: [TuiPrimitiveYearMonthPaginationComponent],
                    exports: [TuiPrimitiveYearMonthPaginationComponent],
                }]
        }] });
