// @taiga-ui/kit/esm2015/components/progress/progress-label/progress-label.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiProgressLabelComponent {
}
TuiProgressLabelComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressLabelComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiProgressLabelComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiProgressLabelComponent, selector: "label[tuiProgressLabel]", ngImport: i0, template: "<ng-content select=\"progress\"></ng-content>\n<ng-content select=\"tui-progress-circle\"></ng-content>\n<span class=\"t-label\">\n    <ng-content></ng-content>\n</span>\n", styles: [":host{position:relative;display:inline-block}.t-label{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;font:var(--tui-font-text-s);color:var(--tui-text-01);flex-direction:column;justify-content:center;align-items:center}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressLabelComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'label[tuiProgressLabel]',
                    templateUrl: './progress-label.template.html',
                    styleUrls: ['./progress-label.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
