// @taiga-ui/cdk/esm2015/directives/hovered/hovered.service.js
import { ElementRef, Inject, Injectable, NgZone } from '@angular/core';
import { ALWAYS_FALSE_HANDLER, ALWAYS_TRUE_HANDLER } from '@taiga-ui/cdk/constants';
import { tuiTypedFromEvent, tuiZoneOptimized } from '@taiga-ui/cdk/observables';
import { tuiIsElement } from '@taiga-ui/cdk/utils';
import { merge, Observable } from 'rxjs';
import { distinctUntilChanged, filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
function movedOut({ currentTarget, relatedTarget }) {
    return (!tuiIsElement(relatedTarget) ||
        !tuiIsElement(currentTarget) ||
        !currentTarget.contains(relatedTarget));
}
export class TuiHoveredService extends Observable {
    constructor(el, ngZone) {
        super(subscriber => this.stream$.subscribe(subscriber));
        this.el = el;
        this.ngZone = ngZone;
        this.stream$ = merge(tuiTypedFromEvent(this.el.nativeElement, `mouseenter`).pipe(map(ALWAYS_TRUE_HANDLER)), tuiTypedFromEvent(this.el.nativeElement, `mouseleave`).pipe(map(ALWAYS_FALSE_HANDLER)), 
        // Hello, Safari
        tuiTypedFromEvent(this.el.nativeElement, `mouseout`).pipe(filter(movedOut), map(ALWAYS_FALSE_HANDLER))).pipe(distinctUntilChanged(), tuiZoneOptimized(this.ngZone));
    }
}
TuiHoveredService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHoveredService, deps: [{ token: ElementRef }, { token: NgZone }], target: i0.ɵɵFactoryTarget.Injectable });
TuiHoveredService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHoveredService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHoveredService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }]; } });
