// @taiga-ui/core/esm2015/components/primitive-textfield/primitive-textfield.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiAutofilledModule, TuiFocusableModule } from '@taiga-ui/cdk';
import { TuiSvgModule } from '@taiga-ui/core/components/svg';
import { TuiTooltipModule } from '@taiga-ui/core/components/tooltip';
import { TuiMaskAccessorModule } from '@taiga-ui/core/directives/mask-accessor';
import { TuiWrapperModule } from '@taiga-ui/core/directives/wrapper';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPrimitiveTextfieldComponent } from './primitive-textfield.component';
import { TuiPrimitiveTextfieldDirective } from './primitive-textfield.directive';
import { TuiTextfieldComponent } from './textfield/textfield.component';
import { TuiValueDecorationComponent } from './value-decoration/value-decoration.component';
import * as i0 from "@angular/core";
export class TuiPrimitiveTextfieldModule {
}
TuiPrimitiveTextfieldModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveTextfieldModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveTextfieldModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveTextfieldModule, declarations: [TuiPrimitiveTextfieldComponent,
        TuiPrimitiveTextfieldDirective,
        TuiTextfieldComponent,
        TuiValueDecorationComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiMaskAccessorModule,
        TuiFocusableModule,
        TuiWrapperModule,
        TuiSvgModule,
        TuiTooltipModule,
        TuiAutofilledModule], exports: [TuiPrimitiveTextfieldComponent,
        TuiPrimitiveTextfieldDirective,
        TuiTextfieldComponent] });
TuiPrimitiveTextfieldModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveTextfieldModule, imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiMaskAccessorModule,
            TuiFocusableModule,
            TuiWrapperModule,
            TuiSvgModule,
            TuiTooltipModule,
            TuiAutofilledModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveTextfieldModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiMaskAccessorModule,
                        TuiFocusableModule,
                        TuiWrapperModule,
                        TuiSvgModule,
                        TuiTooltipModule,
                        TuiAutofilledModule,
                    ],
                    declarations: [
                        TuiPrimitiveTextfieldComponent,
                        TuiPrimitiveTextfieldDirective,
                        TuiTextfieldComponent,
                        TuiValueDecorationComponent,
                    ],
                    exports: [
                        TuiPrimitiveTextfieldComponent,
                        TuiPrimitiveTextfieldDirective,
                        TuiTextfieldComponent,
                    ],
                }]
        }] });
