// @taiga-ui/core/esm2015/components/primitive-textfield/primitive-textfield.directive.js
import { Directive, Inject } from '@angular/core';
import { AbstractTuiTextfieldHost } from '@taiga-ui/core/abstract';
import { tuiAsTextfieldHost } from '@taiga-ui/core/tokens';
import { TuiPrimitiveTextfieldComponent } from './primitive-textfield.component';
import * as i0 from "@angular/core";
import * as i1 from "./primitive-textfield.component";
export class TuiPrimitiveTextfieldDirective extends AbstractTuiTextfieldHost {
    constructor(textfield) {
        /**
         * TuiPrimitiveTextfieldComponent satisfies all required parts of
         * AbstractTuiControl interface. A new interface is not introduced
         * so AbstractTuiControl is automatically injected for all other controls.
         */
        super(textfield);
        this.textfield = textfield;
    }
    get readOnly() {
        return this.textfield.readOnly || !this.textfield.editable;
    }
    onValueChange(value) {
        this.textfield.onModelChange(value);
    }
}
TuiPrimitiveTextfieldDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveTextfieldDirective, deps: [{ token: TuiPrimitiveTextfieldComponent }], target: i0.ɵɵFactoryTarget.Directive });
TuiPrimitiveTextfieldDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield", providers: [tuiAsTextfieldHost(TuiPrimitiveTextfieldDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveTextfieldDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-primitive-textfield',
                    providers: [tuiAsTextfieldHost(TuiPrimitiveTextfieldDirective)],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiPrimitiveTextfieldComponent, decorators: [{
                    type: Inject,
                    args: [TuiPrimitiveTextfieldComponent]
                }] }]; } });
