// @taiga-ui/kit/esm2015/components/tree/tree.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLetModule } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiExpandModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTreeComponent } from './components/tree/tree.component';
import { TuiTreeItemComponent } from './components/tree-item/tree-item.component';
import { TuiTreeItemContentComponent } from './components/tree-item-content/tree-item-content.component';
import { TuiTreeChildrenDirective } from './directives/tree-children.directive';
import { TuiTreeControllerDirective } from './directives/tree-controller.directive';
import { TuiTreeItemControllerDirective } from './directives/tree-item-controller.directive';
import { TuiTreeNodeDirective } from './directives/tree-node.directive';
import * as i0 from "@angular/core";
export class TuiTreeModule {
}
TuiTreeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTreeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeModule, declarations: [TuiTreeComponent,
        TuiTreeItemComponent,
        TuiTreeItemContentComponent,
        TuiTreeChildrenDirective,
        TuiTreeItemControllerDirective,
        TuiTreeControllerDirective,
        TuiTreeNodeDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiExpandModule,
        TuiLetModule,
        TuiButtonModule], exports: [TuiTreeComponent,
        TuiTreeItemComponent,
        TuiTreeChildrenDirective,
        TuiTreeItemControllerDirective,
        TuiTreeControllerDirective] });
TuiTreeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiExpandModule,
            TuiLetModule,
            TuiButtonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiExpandModule,
                        TuiLetModule,
                        TuiButtonModule,
                    ],
                    declarations: [
                        TuiTreeComponent,
                        TuiTreeItemComponent,
                        TuiTreeItemContentComponent,
                        TuiTreeChildrenDirective,
                        TuiTreeItemControllerDirective,
                        TuiTreeControllerDirective,
                        TuiTreeNodeDirective,
                    ],
                    exports: [
                        TuiTreeComponent,
                        TuiTreeItemComponent,
                        TuiTreeChildrenDirective,
                        TuiTreeItemControllerDirective,
                        TuiTreeControllerDirective,
                    ],
                }]
        }] });
