// @taiga-ui/core/esm2015/components/primitive-year-picker/primitive-year-picker.module.js
import { NgModule } from '@angular/core';
import { TuiHoveredModule, TuiLetModule, TuiPressedModule, TuiRepeatTimesModule, } from '@taiga-ui/cdk';
import { TuiScrollIntoViewModule } from '@taiga-ui/core/directives/scroll-into-view';
import { TuiPrimitiveYearPickerComponent } from './primitive-year-picker.component';
import * as i0 from "@angular/core";
export class TuiPrimitiveYearPickerModule {
}
TuiPrimitiveYearPickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearPickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveYearPickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearPickerModule, declarations: [TuiPrimitiveYearPickerComponent], imports: [TuiHoveredModule,
        TuiPressedModule,
        TuiRepeatTimesModule,
        TuiLetModule,
        TuiScrollIntoViewModule], exports: [TuiPrimitiveYearPickerComponent] });
TuiPrimitiveYearPickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearPickerModule, imports: [[
            TuiHoveredModule,
            TuiPressedModule,
            TuiRepeatTimesModule,
            TuiLetModule,
            TuiScrollIntoViewModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveYearPickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        TuiHoveredModule,
                        TuiPressedModule,
                        TuiRepeatTimesModule,
                        TuiLetModule,
                        TuiScrollIntoViewModule,
                    ],
                    declarations: [TuiPrimitiveYearPickerComponent],
                    exports: [TuiPrimitiveYearPickerComponent],
                }]
        }] });
