// @taiga-ui/core/esm2015/components/primitive-checkbox/primitive-checkbox.module.js
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core/components/svg';
import { TuiWrapperModule } from '@taiga-ui/core/directives/wrapper';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPrimitiveCheckboxComponent } from './primitive-checkbox.component';
import * as i0 from "@angular/core";
export class TuiPrimitiveCheckboxModule {
}
TuiPrimitiveCheckboxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCheckboxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveCheckboxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCheckboxModule, declarations: [TuiPrimitiveCheckboxComponent], imports: [TuiSvgModule, TuiWrapperModule, PolymorpheusModule], exports: [TuiPrimitiveCheckboxComponent] });
TuiPrimitiveCheckboxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCheckboxModule, imports: [[TuiSvgModule, TuiWrapperModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCheckboxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiSvgModule, TuiWrapperModule, PolymorpheusModule],
                    declarations: [TuiPrimitiveCheckboxComponent],
                    exports: [TuiPrimitiveCheckboxComponent],
                }]
        }] });
