// @taiga-ui/kit/esm2015/components/multi-select/hide-selected.pipe.js
import { __decorate } from "tslib";
import { Inject, Pipe } from '@angular/core';
import { tuiPure } from '@taiga-ui/cdk';
import { tuiIsFlat } from '@taiga-ui/kit/utils';
import { TuiMultiSelectComponent } from './multi-select.component';
import * as i0 from "@angular/core";
import * as i1 from "./multi-select.component";
export class TuiHideSelectedPipe {
    constructor(component) {
        this.component = component;
    }
    transform(items) {
        const { value, identityMatcher } = this.component;
        if (!items) {
            return null;
        }
        return tuiIsFlat(items)
            ? this.filter(items, value, identityMatcher)
            : this.filter2d(items, value, identityMatcher);
    }
    filter2d(items, value, matcher) {
        return items.map(subItems => this.filter(subItems, value, matcher));
    }
    filter(items, value, matcher) {
        return items.filter(item => value.every(selected => !matcher(selected, item)));
    }
}
TuiHideSelectedPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHideSelectedPipe, deps: [{ token: TuiMultiSelectComponent }], target: i0.ɵɵFactoryTarget.Pipe });
TuiHideSelectedPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHideSelectedPipe, name: "tuiHideSelected", pure: false });
__decorate([
    tuiPure
], TuiHideSelectedPipe.prototype, "filter2d", null);
__decorate([
    tuiPure
], TuiHideSelectedPipe.prototype, "filter", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHideSelectedPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: `tuiHideSelected`,
                    pure: false,
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiMultiSelectComponent, decorators: [{
                    type: Inject,
                    args: [TuiMultiSelectComponent]
                }] }]; }, propDecorators: { filter2d: [], filter: [] } });
