// @taiga-ui/core/esm2015/components/primitive-calendar/primitive-calendar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiHoveredModule, TuiLetModule, TuiMapperPipeModule, TuiPressedModule, TuiRepeatTimesModule, } from '@taiga-ui/cdk';
import { TuiCalendarSheetPipeModule, TuiOrderWeekDaysPipeModule, } from '@taiga-ui/core/pipes';
import { TuiPrimitiveCalendarComponent } from './primitive-calendar.component';
import * as i0 from "@angular/core";
export class TuiPrimitiveCalendarModule {
}
TuiPrimitiveCalendarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveCalendarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarModule, declarations: [TuiPrimitiveCalendarComponent], imports: [CommonModule,
        TuiLetModule,
        TuiMapperPipeModule,
        TuiRepeatTimesModule,
        TuiHoveredModule,
        TuiPressedModule,
        TuiCalendarSheetPipeModule,
        TuiOrderWeekDaysPipeModule], exports: [TuiPrimitiveCalendarComponent] });
TuiPrimitiveCalendarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarModule, imports: [[
            CommonModule,
            TuiLetModule,
            TuiMapperPipeModule,
            TuiRepeatTimesModule,
            TuiHoveredModule,
            TuiPressedModule,
            TuiCalendarSheetPipeModule,
            TuiOrderWeekDaysPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiLetModule,
                        TuiMapperPipeModule,
                        TuiRepeatTimesModule,
                        TuiHoveredModule,
                        TuiPressedModule,
                        TuiCalendarSheetPipeModule,
                        TuiOrderWeekDaysPipeModule,
                    ],
                    declarations: [TuiPrimitiveCalendarComponent],
                    exports: [TuiPrimitiveCalendarComponent],
                }]
        }] });
