// @taiga-ui/cdk/esm2015/directives/auto-focus/autofocus.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { tuiCoerceBooleanProperty } from '@taiga-ui/cdk/coercion';
import { TUI_AUTOFOCUS_HANDLER, TUI_AUTOFOCUS_PROVIDERS, } from './autofocus.options';
import * as i0 from "@angular/core";
export class TuiAutoFocusDirective {
    constructor(handler) {
        this.handler = handler;
        this.autoFocus = true;
    }
    ngAfterViewInit() {
        if (tuiCoerceBooleanProperty(this.autoFocus)) {
            this.focus();
        }
    }
    focus() {
        void Promise.resolve().then(() => this.handler.setFocus());
    }
}
TuiAutoFocusDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoFocusDirective, deps: [{ token: TUI_AUTOFOCUS_HANDLER }], target: i0.ɵɵFactoryTarget.Directive });
TuiAutoFocusDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAutoFocusDirective, selector: "[tuiAutoFocus]", inputs: { autoFocus: ["tuiAutoFocus", "autoFocus"] }, providers: TUI_AUTOFOCUS_PROVIDERS, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoFocusDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiAutoFocus]',
                    providers: TUI_AUTOFOCUS_PROVIDERS,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_AUTOFOCUS_HANDLER]
                }] }]; }, propDecorators: { autoFocus: [{
                type: Input,
                args: ['tuiAutoFocus']
            }] } });
