// @taiga-ui/kit/esm2015/components/radio-list/radio-list.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiRadioGroupModule } from '@taiga-ui/kit/components/radio-group';
import { TuiRadioLabeledModule } from '@taiga-ui/kit/components/radio-labeled';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiRadioListComponent } from './radio-list.component';
import * as i0 from "@angular/core";
export class TuiRadioListModule {
}
TuiRadioListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRadioListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioListModule, declarations: [TuiRadioListComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiRadioGroupModule,
        TuiRadioLabeledModule], exports: [TuiRadioListComponent] });
TuiRadioListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioListModule, imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiRadioGroupModule,
            TuiRadioLabeledModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiRadioGroupModule,
                        TuiRadioLabeledModule,
                    ],
                    declarations: [TuiRadioListComponent],
                    exports: [TuiRadioListComponent],
                }]
        }] });
