// @taiga-ui/kit/esm2015/pipes/sort-countries/sort-countries.pipe.js
import { Inject, Pipe } from '@angular/core';
import { TUI_COUNTRIES } from '@taiga-ui/kit/tokens';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiSortCountriesPipe {
    constructor(countriesNames$) {
        this.countriesNames$ = countriesNames$;
    }
    transform(countries) {
        return this.countriesNames$.pipe(map(names => [...countries].sort((a, b) => names[a].localeCompare(names[b]))));
    }
}
TuiSortCountriesPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortCountriesPipe, deps: [{ token: TUI_COUNTRIES }], target: i0.ɵɵFactoryTarget.Pipe });
TuiSortCountriesPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortCountriesPipe, name: "tuiSortCountries" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortCountriesPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: `tuiSortCountries`,
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_COUNTRIES]
                }] }]; } });
