// @taiga-ui/core/esm2015/components/data-list/data-list.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core/components/svg';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDataListComponent } from './data-list.component';
import { TuiDataListDirective } from './data-list.directive';
import { TuiOptGroupDirective } from './opt-group.directive';
import { TuiOptionComponent } from './option/option.component';
import * as i0 from "@angular/core";
export class TuiDataListModule {
}
TuiDataListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDataListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListModule, declarations: [TuiDataListComponent,
        TuiOptionComponent,
        TuiOptGroupDirective,
        TuiDataListDirective], imports: [CommonModule, TuiSvgModule, PolymorpheusModule], exports: [TuiDataListComponent,
        TuiOptionComponent,
        TuiOptGroupDirective,
        TuiDataListDirective] });
TuiDataListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListModule, imports: [[CommonModule, TuiSvgModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, PolymorpheusModule],
                    declarations: [
                        TuiDataListComponent,
                        TuiOptionComponent,
                        TuiOptGroupDirective,
                        TuiDataListDirective,
                    ],
                    exports: [
                        TuiDataListComponent,
                        TuiOptionComponent,
                        TuiOptGroupDirective,
                        TuiDataListDirective,
                    ],
                }]
        }] });
