// @taiga-ui/kit/esm2015/components/checkbox-labeled/checkbox-labeled.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiCheckboxModule } from '@taiga-ui/kit/components/checkbox';
import { TuiCheckboxLabeledComponent } from './checkbox-labeled.component';
import * as i0 from "@angular/core";
export class TuiCheckboxLabeledModule {
}
TuiCheckboxLabeledModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxLabeledModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCheckboxLabeledModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxLabeledModule, declarations: [TuiCheckboxLabeledComponent], imports: [CommonModule, FormsModule, TuiCheckboxModule], exports: [TuiCheckboxLabeledComponent] });
TuiCheckboxLabeledModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxLabeledModule, imports: [[CommonModule, FormsModule, TuiCheckboxModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxLabeledModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, TuiCheckboxModule],
                    declarations: [TuiCheckboxLabeledComponent],
                    exports: [TuiCheckboxLabeledComponent],
                }]
        }] });
