// @taiga-ui/kit/esm2015/components/input-date-time/native-date-time/native-date-time.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiDay, TuiTime } from '@taiga-ui/cdk';
import { TUI_TEXTFIELD_HOST } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiNativeDateTimeDirective {
    constructor(host) {
        this.host = host;
    }
    get value() {
        var _a, _b;
        if (!this.host.rawValue[0] || !this.host.rawValue[1]) {
            return '';
        }
        return `${(_a = this.host.rawValue[0]) === null || _a === void 0 ? void 0 : _a.toString('YMD', '-')}T${(_b = this.host.rawValue[1]) === null || _b === void 0 ? void 0 : _b.toString('HH:MM')}`;
    }
    onChange(value) {
        if (!value) {
            this.host.writeValue([null, null]);
            return;
        }
        const date = new Date(value);
        const day = TuiDay.fromLocalNativeDate(date);
        const time = TuiTime.fromLocalNativeDate(date);
        this.host.writeValue([day, time]);
    }
}
TuiNativeDateTimeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNativeDateTimeDirective, deps: [{ token: TUI_TEXTFIELD_HOST }], target: i0.ɵɵFactoryTarget.Directive });
TuiNativeDateTimeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiNativeDateTimeDirective, selector: "input[tuiDateTime]", host: { attributes: { "type": "datetime-local" }, listeners: { "change.stop": "onChange($event.target.value)", "input.stop": "onChange($event.target.value)", "click.stop.silent": "0", "mousedown.stop.silent": "0" }, properties: { "tabIndex": "-1", "value": "value" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNativeDateTimeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[tuiDateTime]',
                    host: {
                        type: 'datetime-local',
                        '[tabIndex]': '-1',
                        '[value]': 'value',
                        '(change.stop)': 'onChange($event.target.value)',
                        '(input.stop)': 'onChange($event.target.value)',
                        '(click.stop.silent)': '0',
                        '(mousedown.stop.silent)': '0',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_TEXTFIELD_HOST]
                }] }]; } });
