// @taiga-ui/kit/esm2015/components/tiles/tiles.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiTileComponent } from './tile.component';
import { TuiTileHandleDirective } from './tile-handle.directive';
import { TuiTilesComponent } from './tiles.component';
import * as i0 from "@angular/core";
export class TuiTilesModule {
}
TuiTilesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTilesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTilesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTilesModule, declarations: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective], imports: [CommonModule], exports: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective] });
TuiTilesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTilesModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTilesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective],
                    exports: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective],
                }]
        }] });
