// @taiga-ui/kit/esm2015/components/input-date-time/input-date-time.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost, tuiAsTextfieldHost } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiInputDateTimeDirective extends AbstractTuiTextfieldHost {
    get value() {
        return this.host.computedValue;
    }
    get rawValue() {
        return this.host.value;
    }
    onValueChange(value) {
        this.host.onValueChange(value);
    }
    writeValue(value) {
        this.host.writeValue(value);
    }
    process(input) {
        input.inputMode = 'numeric';
    }
}
TuiInputDateTimeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateTimeDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiInputDateTimeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputDateTimeDirective, selector: "tui-input-date-time", providers: [tuiAsTextfieldHost(TuiInputDateTimeDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateTimeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-input-date-time',
                    providers: [tuiAsTextfieldHost(TuiInputDateTimeDirective)],
                }]
        }] });
