// @taiga-ui/core/fesm2015/taiga-ui-core-pipes-format-phone.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';
import { tuiAssert, CHAR_PLUS } from '@taiga-ui/cdk';
import { tuiFormatPhone } from '@taiga-ui/core/utils/format';

class TuiFormatPhonePipe {
    /**
     * Formats phone number string of +7XXXXXXXXXX format with adding separator symbols
     * @param value
     * @param countryCode
     * @param phoneMask
     * @return formatted phone number string +7(XXX)XXX-XX-XX
     */
    transform(value, countryCode = `${CHAR_PLUS}7`, phoneMask = `(###) ###-##-##`) {
        ngDevMode &&
            tuiAssert.assert(value !== undefined, `Undefined was passed to tuiFormatPhone pipe`);
        return tuiFormatPhone(value, countryCode, phoneMask);
    }
}
TuiFormatPhonePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiFormatPhonePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipe, name: "tuiFormatPhone" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipe, decorators: [{
            type: Pipe,
            args: [{ name: `tuiFormatPhone` }]
        }] });

class TuiFormatPhonePipeModule {
}
TuiFormatPhonePipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiFormatPhonePipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipeModule, declarations: [TuiFormatPhonePipe], exports: [TuiFormatPhonePipe] });
TuiFormatPhonePipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatPhonePipeModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [TuiFormatPhonePipe],
                    declarations: [TuiFormatPhonePipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFormatPhonePipe, TuiFormatPhonePipeModule };
