// @taiga-ui/core/fesm2015/taiga-ui-core-pipes-month.js
import * as i0 from '@angular/core';
import { Pipe, Inject, NgModule } from '@angular/core';
import { TUI_MONTHS } from '@taiga-ui/core/tokens';
import { map } from 'rxjs/operators';
import * as i1 from 'rxjs';

class TuiMonthPipe {
    constructor(months$) {
        this.months$ = months$;
    }
    transform({ month }) {
        return this.months$.pipe(map(months => months[month]));
    }
}
TuiMonthPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipe, deps: [{ token: TUI_MONTHS }], target: i0.ɵɵFactoryTarget.Pipe });
TuiMonthPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipe, name: "tuiMonth" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipe, decorators: [{
            type: Pipe,
            args: [{ name: `tuiMonth` }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_MONTHS]
                }] }]; } });

class TuiMonthPipeModule {
}
TuiMonthPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMonthPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipeModule, declarations: [TuiMonthPipe], exports: [TuiMonthPipe] });
TuiMonthPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMonthPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [TuiMonthPipe],
                    declarations: [TuiMonthPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMonthPipe, TuiMonthPipeModule };
