// @taiga-ui/core/esm2015/services/breakpoint.service.js
import { Inject, Injectable } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { TUI_MEDIA } from '@taiga-ui/core/tokens';
import { fromEvent, merge, Observable } from 'rxjs';
import { map, shareReplay, startWith } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * Service to provide the current breakpoint based on Taiga UI's media queries
 */
export class TuiBreakpointService extends Observable {
    constructor(media, win) {
        const breakpoints = getBreakpoints(media);
        const events$ = breakpoints.map(({ query }) => fromEvent(win.matchMedia(query), `change`));
        const media$ = merge(...events$).pipe(map(() => currentBreakpoint(breakpoints, win.innerWidth).name), startWith(currentBreakpoint(breakpoints, win.innerWidth).name), shareReplay({ bufferSize: 1, refCount: true }));
        super(subscriber => media$.subscribe(subscriber));
    }
}
TuiBreakpointService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreakpointService, deps: [{ token: TUI_MEDIA }, { token: WINDOW }], target: i0.ɵɵFactoryTarget.Injectable });
TuiBreakpointService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreakpointService, providedIn: `root` });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreakpointService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: `root`,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_MEDIA]
                }] }, { type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }]; } });
function getBreakpoints(media) {
    return Object.entries(media).map(([name, width]) => ({
        name: name,
        /**
         * @note:
         * min-width query in css is inclusive, but in window.matchMedia it is exclusive
         * so we need to subtract 1px to get the same result
         */
        query: `(max-width: ${width - 1}px)`,
        width,
    }));
}
function currentBreakpoint(breakpoints, innerWidth) {
    var _a;
    return (_a = breakpoints.find(({ width }) => innerWidth < width)) !== null && _a !== void 0 ? _a : breakpoints.slice(-1)[0];
}
