// @taiga-ui/core/esm2015/services/visual-viewport.service.js
import { Inject, Injectable } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { TUI_IS_WEBKIT } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class TuiVisualViewportService {
    constructor(win, isWebkit) {
        this.win = win;
        this.isWebkit = isWebkit;
    }
    // https://bugs.webkit.org/show_bug.cgi?id=207089
    correct(point) {
        var _a, _b, _c, _d;
        return this.isWebkit
            ? [
                point[0] + ((_b = (_a = this.win.visualViewport) === null || _a === void 0 ? void 0 : _a.offsetTop) !== null && _b !== void 0 ? _b : 0),
                point[1] + ((_d = (_c = this.win.visualViewport) === null || _c === void 0 ? void 0 : _c.offsetLeft) !== null && _d !== void 0 ? _d : 0),
            ]
            : point;
    }
}
TuiVisualViewportService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiVisualViewportService, deps: [{ token: WINDOW }, { token: TUI_IS_WEBKIT }], target: i0.ɵɵFactoryTarget.Injectable });
TuiVisualViewportService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiVisualViewportService, providedIn: `root` });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiVisualViewportService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: `root`,
                }]
        }], ctorParameters: function () { return [{ type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_WEBKIT]
                }] }]; } });
