// @taiga-ui/cdk/schematics/ng-update/v3/constants/services.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SERVICES_TO_REPLACE = void 0;
exports.SERVICES_TO_REPLACE = [
    {
        from: {
            name: `TuiPortalService`,
        },
        to: {
            name: `TuiDropdownPortalService`,
            moduleSpecifier: `@taiga-ui/cdk`,
        },
    },
    {
        from: {
            name: `TuiNotificationsService`,
        },
        to: {
            name: `TuiAlertService`,
            moduleSpecifier: `@taiga-ui/core`,
        },
        replaceMethods: [
            {
                from: `show`,
                to: `open`,
            },
        ],
    },
    {
        from: {
            name: `TuiCodeEditor`,
        },
        to: {
            name: `TuiCodeEditor`,
            moduleSpecifier: `@taiga-ui/addon-doc`,
        },
        replaceMethods: [
            {
                from: `open`,
                to: `edit`,
            },
        ],
    },
    {
        from: {
            name: `PreviewDialogService`,
        },
        to: {
            name: `TuiPreviewDialogService`,
            moduleSpecifier: `@taiga-ui/addon-preview`,
        },
    },
];
