// @taiga-ui/cdk/esm2015/directives/obscured/obscured.directive.js
import { __decorate } from "tslib";
import { Directive, Inject, Input, Optional, Output, Self } from '@angular/core';
import { tuiRequiredSetter } from '@taiga-ui/cdk/decorators';
import { TuiActiveZoneDirective } from '@taiga-ui/cdk/directives/active-zone';
import { tuiIfMap } from '@taiga-ui/cdk/observables';
import { TuiDestroyService, TuiObscuredService, TuiParentsScrollService, } from '@taiga-ui/cdk/services';
import { Subject } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/directives/active-zone";
import * as i2 from "@taiga-ui/cdk/services";
/**
 * Directive that monitors element visibility
 */
export class TuiObscuredDirective {
    constructor(activeZone, obscured$) {
        this.enabled$ = new Subject();
        const mapped$ = obscured$.pipe(map(obscuredBy => !!obscuredBy &&
            (!activeZone ||
                !obscuredBy.length ||
                obscuredBy.every(element => !activeZone.contains(element)))));
        this.tuiObscured = this.enabled$.pipe(tuiIfMap(() => mapped$));
    }
    set tuiObscuredEnabled(enabled) {
        this.enabled$.next(enabled);
    }
}
TuiObscuredDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiObscuredDirective, deps: [{ token: TuiActiveZoneDirective, optional: true }, { token: TuiObscuredService, self: true }], target: i0.ɵɵFactoryTarget.Directive });
TuiObscuredDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiObscuredDirective, selector: "[tuiObscured]", inputs: { tuiObscuredEnabled: "tuiObscuredEnabled" }, outputs: { tuiObscured: "tuiObscured" }, providers: [TuiObscuredService, TuiParentsScrollService, TuiDestroyService], ngImport: i0 });
__decorate([
    tuiRequiredSetter()
], TuiObscuredDirective.prototype, "tuiObscuredEnabled", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiObscuredDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiObscured]',
                    providers: [TuiObscuredService, TuiParentsScrollService, TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiActiveZoneDirective, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TuiActiveZoneDirective]
                }] }, { type: i2.TuiObscuredService, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiObscuredService]
                }] }]; }, propDecorators: { tuiObscuredEnabled: [{
                type: Input
            }], tuiObscured: [{
                type: Output
            }] } });
