// @taiga-ui/core/bundles/taiga-ui-core-directives-mask-accessor.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/core/tokens')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/directives/mask-accessor', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/core/tokens'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.directives = global["taiga-ui"].core.directives || {}, global["taiga-ui"].core.directives["mask-accessor"] = {}), global.ng.core, global.ng.forms, global["taiga-ui"].core.tokens));
})(this, (function (exports, i0, forms, tokens) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiMaskAccessorDirective = /** @class */ (function () {
        function TuiMaskAccessorDirective() {
        }
        return TuiMaskAccessorDirective;
    }());
    TuiMaskAccessorDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMaskAccessorDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiMaskAccessorDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiMaskAccessorDirective, selector: "input[tuiMaskAccessor]", providers: [
            {
                provide: forms.NG_VALUE_ACCESSOR,
                deps: [[new i0.Optional(), tokens.TUI_VALUE_ACCESSOR], forms.DefaultValueAccessor],
                multi: true,
                useFactory: function (accessor, fallback) { return (accessor ? accessor[0] : fallback); },
            },
        ], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMaskAccessorDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'input[tuiMaskAccessor]',
                        providers: [
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                deps: [[new i0.Optional(), tokens.TUI_VALUE_ACCESSOR], forms.DefaultValueAccessor],
                                multi: true,
                                useFactory: function (accessor, fallback) { return (accessor ? accessor[0] : fallback); },
                            },
                        ],
                    }]
            }] });

    /**
     * @internal
     */
    var TuiMaskAccessorModule = /** @class */ (function () {
        function TuiMaskAccessorModule() {
        }
        return TuiMaskAccessorModule;
    }());
    TuiMaskAccessorModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMaskAccessorModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiMaskAccessorModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMaskAccessorModule, declarations: [TuiMaskAccessorDirective], exports: [TuiMaskAccessorDirective] });
    TuiMaskAccessorModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMaskAccessorModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMaskAccessorModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiMaskAccessorDirective],
                        exports: [TuiMaskAccessorDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiMaskAccessorDirective = TuiMaskAccessorDirective;
    exports.TuiMaskAccessorModule = TuiMaskAccessorModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
