// @taiga-ui/core/bundles/taiga-ui-core-internal-svg-defs-host.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/services'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/internal/svg-defs-host', ['exports', '@angular/common', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/services', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.internal = global["taiga-ui"].core.internal || {}, global["taiga-ui"].core.internal["svg-defs-host"] = {}), global.ng.common, global.ng.core, global.i3, global["taiga-ui"].core.services, global.rxjs.operators));
})(this, (function (exports, i1, i0, i3, i2, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var TuiSvgDefsHostComponent = /** @class */ (function () {
        function TuiSvgDefsHostComponent(svgService, cdr, destroy$, platformId) {
            this.svgService = svgService;
            this.cdr = cdr;
            this.destroy$ = destroy$;
            this.isBrowser = true;
            this.isBrowser = !i1.isPlatformServer(platformId);
        }
        // @bad TODO: Looks like it could be async piped but it was probably written like that for a reason
        TuiSvgDefsHostComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.svgService.items$.pipe(operators.takeUntil(this.destroy$)).subscribe(function (defsMap) {
                _this.items = defsMap.values();
                _this.cdr.detectChanges();
            });
        };
        return TuiSvgDefsHostComponent;
    }());
    TuiSvgDefsHostComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgDefsHostComponent, deps: [{ token: i2.TuiSvgService }, { token: i0.ChangeDetectorRef }, { token: i3.TuiDestroyService, self: true }, { token: i0.PLATFORM_ID }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSvgDefsHostComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSvgDefsHostComponent, selector: "tui-svg-defs-host", providers: [i3.TuiDestroyService], ngImport: i0__namespace, template: "<svg:svg\n    version=\"1.1\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n    xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n>\n    <ng-container *ngIf=\"isBrowser; else server\">\n        <defs\n            *ngFor=\"let def of items\"\n            [innerHTML]=\"def\"\n        ></defs>\n    </ng-container>\n    <ng-template #server>\n        <!--svg[outerHTML] - to support SSR, because there is no support of innerHTML for for defs element in domino-->\n        <defs *ngFor=\"let def of items\">\n            <svg [outerHTML]=\"def\"></svg>\n        </defs>\n    </ng-template>\n</svg:svg>\n", styles: [":host{display:none}\n"], directives: [{ type: i1__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgDefsHostComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-svg-defs-host',
                        templateUrl: './svg-defs-host.template.html',
                        styleUrls: ['./svg-defs-host.style.less'],
                        providers: [i3.TuiDestroyService],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i2__namespace.TuiSvgService, decorators: [{
                            type: i0.Inject,
                            args: [i2.TuiSvgService]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i3__namespace.TuiDestroyService, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDestroyService]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i0.PLATFORM_ID]
                        }] }];
        } });

    var TuiSvgDefsHostModule = /** @class */ (function () {
        function TuiSvgDefsHostModule() {
        }
        return TuiSvgDefsHostModule;
    }());
    TuiSvgDefsHostModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgDefsHostModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSvgDefsHostModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgDefsHostModule, declarations: [TuiSvgDefsHostComponent], imports: [i1.CommonModule, i3.TuiLetModule], exports: [TuiSvgDefsHostComponent] });
    TuiSvgDefsHostModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgDefsHostModule, imports: [[i1.CommonModule, i3.TuiLetModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgDefsHostModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule, i3.TuiLetModule],
                        declarations: [TuiSvgDefsHostComponent],
                        exports: [TuiSvgDefsHostComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiSvgDefsHostComponent = TuiSvgDefsHostComponent;
    exports.TuiSvgDefsHostModule = TuiSvgDefsHostModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
