// @taiga-ui/core/bundles/taiga-ui-core-components-hints-host.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/services'), require('@angular/common'), require('@tinkoff/ng-polymorpheus'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/hints-host', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/services', '@angular/common', '@tinkoff/ng-polymorpheus', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components["hints-host"] = {}), global.ng.core, global.i2, global["taiga-ui"].core.services, global.ng.common, global.i3, global.rxjs));
})(this, (function (exports, i0, i2, services, i1, i3, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    var TuiHintsHostComponent = /** @class */ (function () {
        function TuiHintsHostComponent(hints$) {
            this.hints$ = hints$;
        }
        return TuiHintsHostComponent;
    }());
    TuiHintsHostComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintsHostComponent, deps: [{ token: services.TuiHintService }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiHintsHostComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintsHostComponent, selector: "tui-hints-host", host: { attributes: { "aria-live": "polite" } }, ngImport: i0__namespace, template: "<div\n    *ngFor=\"let hint of hints$ | async\"\n    role=\"tooltip\"\n    @tuiParentAnimation\n    [tuiActiveZoneParent]=\"hint.activeZone || null\"\n>\n    <ng-container *polymorpheusOutlet=\"hint.component; context: {$implicit: hint}\"></ng-container>\n</div>\n", styles: [":host{position:fixed;top:0;left:0;width:100%;height:0}\n"], directives: [{ type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i3__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i1__namespace.AsyncPipe }, animations: [i2.TUI_PARENT_ANIMATION], changeDetection: i0__namespace.ChangeDetectionStrategy.Default });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintsHostComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-hints-host',
                        templateUrl: './hints-host.template.html',
                        styleUrls: ['./hints-host.style.less'],
                        // So that we do not force OnPush on custom hints
                        // eslint-disable-next-line @angular-eslint/prefer-on-push-component-change-detection
                        changeDetection: i0.ChangeDetectionStrategy.Default,
                        animations: [i2.TUI_PARENT_ANIMATION],
                        host: {
                            'aria-live': 'polite',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i4__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [services.TuiHintService]
                        }] }];
        } });

    var TuiHintsHostModule = /** @class */ (function () {
        function TuiHintsHostModule() {
        }
        return TuiHintsHostModule;
    }());
    TuiHintsHostModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintsHostModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiHintsHostModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintsHostModule, declarations: [TuiHintsHostComponent], imports: [i1.CommonModule, i3.PolymorpheusModule, i2.TuiActiveZoneModule], exports: [TuiHintsHostComponent] });
    TuiHintsHostModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintsHostModule, imports: [[i1.CommonModule, i3.PolymorpheusModule, i2.TuiActiveZoneModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintsHostModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule, i3.PolymorpheusModule, i2.TuiActiveZoneModule],
                        declarations: [TuiHintsHostComponent],
                        exports: [TuiHintsHostComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiHintsHostComponent = TuiHintsHostComponent;
    exports.TuiHintsHostModule = TuiHintsHostModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
