// @taiga-ui/core/bundles/taiga-ui-core-services.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/core/tokens'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/cdk'), require('@taiga-ui/core/abstract'), require('@angular/router'), require('@angular/platform-browser'), require('@taiga-ui/core/utils/dom')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/services', ['exports', '@angular/core', '@ng-web-apis/common', '@taiga-ui/core/tokens', 'rxjs', 'rxjs/operators', '@taiga-ui/cdk', '@taiga-ui/core/abstract', '@angular/router', '@angular/platform-browser', '@taiga-ui/core/utils/dom'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.services = {}), global.ng.core, global.common, global["taiga-ui"].core.tokens, global.rxjs, global.rxjs.operators, global.i3, global["taiga-ui"].core.abstract, global.ng.router, global.ng.platformBrowser, global["taiga-ui"].core.utils.dom));
})(this, (function (exports, i0, common, tokens, i1, operators, i3, i2, i1$1, i1$2, dom) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i1__namespace$2 = /*#__PURE__*/_interopNamespace(i1$2);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    /**
     * Service to provide the current breakpoint based on Taiga UI's media queries
     */
    var TuiBreakpointService = /** @class */ (function (_super) {
        __extends(TuiBreakpointService, _super);
        function TuiBreakpointService(media, win) {
            var _this = this;
            var breakpoints = getBreakpoints(media);
            var events$ = breakpoints.map(function (_b) {
                var query = _b.query;
                return i1.fromEvent(win.matchMedia(query), "change");
            });
            var media$ = i1.merge.apply(void 0, __spreadArray([], __read(events$))).pipe(operators.map(function () { return currentBreakpoint(breakpoints, win.innerWidth).name; }), operators.startWith(currentBreakpoint(breakpoints, win.innerWidth).name), operators.shareReplay({ bufferSize: 1, refCount: true }));
            _this = _super.call(this, function (subscriber) { return media$.subscribe(subscriber); }) || this;
            return _this;
        }
        return TuiBreakpointService;
    }(i1.Observable));
    TuiBreakpointService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBreakpointService, deps: [{ token: tokens.TUI_MEDIA }, { token: common.WINDOW }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiBreakpointService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBreakpointService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBreakpointService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_MEDIA]
                        }] }, { type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }];
        } });
    function getBreakpoints(media) {
        return Object.entries(media).map(function (_b) {
            var _c = __read(_b, 2), name = _c[0], width = _c[1];
            return ({
                name: name,
                /**
                 * @note:
                 * min-width query in css is inclusive, but in window.matchMedia it is exclusive
                 * so we need to subtract 1px to get the same result
                 */
                query: "(max-width: " + (width - 1) + "px)",
                width: width,
            });
        });
    }
    function currentBreakpoint(breakpoints, innerWidth) {
        var _a;
        return (_a = breakpoints.find(function (_b) {
            var width = _b.width;
            return innerWidth < width;
        })) !== null && _a !== void 0 ? _a : breakpoints.slice(-1)[0];
    }

    var TuiFormatDateService = /** @class */ (function () {
        function TuiFormatDateService(locale) {
            this.locale = locale;
        }
        TuiFormatDateService.prototype.format = function (timestamp) {
            return i1.of(new Date(timestamp).toLocaleTimeString(this.locale, {
                hour: "numeric",
                minute: "2-digit",
            }));
        };
        return TuiFormatDateService;
    }());
    TuiFormatDateService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDateService, deps: [{ token: i0.LOCALE_ID }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiFormatDateService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDateService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDateService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i0.LOCALE_ID]
                        }] }];
        } });

    /**
     * Service for displaying hints/tooltips
     */
    var TuiHintService = /** @class */ (function (_super) {
        __extends(TuiHintService, _super);
        function TuiHintService() {
            return _super.call(this, []) || this;
        }
        TuiHintService.prototype.add = function (directive) {
            this.next(this.value.concat(directive));
        };
        TuiHintService.prototype.remove = function (directive) {
            if (this.value.includes(directive)) {
                this.next(this.value.filter(function (hint) { return hint !== directive; }));
            }
        };
        return TuiHintService;
    }(i1.BehaviorSubject));
    TuiHintService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintService, deps: [], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiHintService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHintService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }], ctorParameters: function () { return []; } });

    var TuiNightThemeService = /** @class */ (function (_super) {
        __extends(TuiNightThemeService, _super);
        function TuiNightThemeService(win) {
            var _this = this;
            var media = win.matchMedia("(prefers-color-scheme: dark)");
            var media$ = i1.fromEvent(media, "change").pipe(operators.startWith(null), operators.map(function () { return media.matches; }), operators.share());
            _this = _super.call(this, function (subscriber) { return media$.subscribe(subscriber); }) || this;
            return _this;
        }
        return TuiNightThemeService;
    }(i1.Observable));
    TuiNightThemeService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNightThemeService, deps: [{ token: common.WINDOW }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiNightThemeService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNightThemeService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNightThemeService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }], ctorParameters: function () {
            return [{ type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }];
        } });

    var TuiPositionService = /** @class */ (function (_super) {
        __extends(TuiPositionService, _super);
        function TuiPositionService(_a, animationFrame, ngZone, accessor) {
            var nativeElement = _a.nativeElement;
            return _super.call(this, function (subscriber) { return animationFrame
                .pipe(operators.map(function () { return nativeElement.getBoundingClientRect(); }), operators.map(function (rect) { return accessor.getPosition(rect); }), i3.tuiZonefree(ngZone))
                .subscribe(subscriber); }) || this;
        }
        return TuiPositionService;
    }(i1.Observable));
    TuiPositionService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPositionService, deps: [{ token: i0.ElementRef }, { token: common.ANIMATION_FRAME }, { token: i0.NgZone }, { token: i2.TuiPositionAccessor }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiPositionService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPositionService });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPositionService, decorators: [{
                type: i0.Injectable
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [common.ANIMATION_FRAME]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i2__namespace.TuiPositionAccessor, decorators: [{
                            type: i0.Inject,
                            args: [i2.TuiPositionAccessor]
                        }] }];
        } });

    // TODO: Remove when Angular is update and `RouterLinkActive` has output
    var TuiRouterLinkActiveService = /** @class */ (function (_super) {
        __extends(TuiRouterLinkActiveService, _super);
        function TuiRouterLinkActiveService(routerLinkActive, ngZone, animationFrame$, destroy$) {
            var _this = this;
            var stream$ = routerLinkActive
                ? i1.merge(i1.timer(0), // SSR (animationFrame$ never emits value during SSR)
                animationFrame$).pipe(operators.map(function () { return routerLinkActive.isActive; }), operators.distinctUntilChanged(), i3.tuiZoneOptimized(ngZone), operators.takeUntil(destroy$))
                : i1.EMPTY;
            _this = _super.call(this, function (subscriber) { return stream$.subscribe(subscriber); }) || this;
            return _this;
        }
        return TuiRouterLinkActiveService;
    }(i1.Observable));
    TuiRouterLinkActiveService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRouterLinkActiveService, deps: [{ token: i1$1.RouterLinkActive, optional: true }, { token: i0.NgZone }, { token: common.ANIMATION_FRAME }, { token: i3.TuiDestroyService, self: true }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiRouterLinkActiveService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRouterLinkActiveService });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRouterLinkActiveService, decorators: [{
                type: i0.Injectable
            }], ctorParameters: function () {
            return [{ type: i1__namespace$1.RouterLinkActive, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i1$1.RouterLinkActive]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [common.ANIMATION_FRAME]
                        }] }, { type: i3__namespace.TuiDestroyService, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDestroyService]
                        }] }];
        } });

    /**
     * Service for reusing SVGs without inlining each instance
     */
    var TuiSvgService = /** @class */ (function () {
        function TuiSvgService(tuiSanitizer, sanitizer, icons) {
            this.tuiSanitizer = tuiSanitizer;
            this.sanitizer = sanitizer;
            this.originals = {};
            this.items$ = new i1.BehaviorSubject(new Map());
            this.define(icons);
        }
        TuiSvgService.prototype.define = function (icons) {
            var _this = this;
            var value = this.items$.value;
            Object.keys(icons).forEach(function (key) {
                _this.defineIcon(key, icons[key], value);
            });
            this.items$.next(value);
        };
        TuiSvgService.prototype.getOriginal = function (name) {
            return this.originals[name] || null;
        };
        TuiSvgService.prototype.defineIcon = function (name, src, map) {
            var _a;
            if (map.has(name)) {
                return;
            }
            var parsed = this.parseSrc(name, src);
            if (!parsed) {
                ngDevMode && i3.tuiAssert.assert(false, "Unable to parse given SVG src");
                return;
            }
            map.set(name, parsed);
            this.originals = Object.assign(Object.assign({}, this.originals), (_a = {}, _a[name] = src, _a));
        };
        TuiSvgService.prototype.parseSrc = function (name, src) {
            return this.sanitize(dom.tuiProcessIcon(src, name));
        };
        TuiSvgService.prototype.sanitize = function (src) {
            return this.sanitizer.bypassSecurityTrustHtml((this.tuiSanitizer
                ? this.tuiSanitizer.sanitize(i0.SecurityContext.HTML, src)
                : this.sanitizer.sanitize(i0.SecurityContext.HTML, src)) || "");
        };
        return TuiSvgService;
    }());
    TuiSvgService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgService, deps: [{ token: tokens.TUI_SANITIZER, optional: true }, { token: i1$2.DomSanitizer }, { token: tokens.TUI_ICONS }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiSvgService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.Sanitizer, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_SANITIZER]
                        }] }, { type: i1__namespace$2.DomSanitizer, decorators: [{
                            type: i0.Inject,
                            args: [i1$2.DomSanitizer]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ICONS]
                        }] }];
        } });

    var TuiVisualViewportService = /** @class */ (function () {
        function TuiVisualViewportService(win, isWebkit) {
            this.win = win;
            this.isWebkit = isWebkit;
        }
        // https://bugs.webkit.org/show_bug.cgi?id=207089
        TuiVisualViewportService.prototype.correct = function (point) {
            var _a, _b, _c, _d;
            return this.isWebkit
                ? [
                    point[0] + ((_b = (_a = this.win.visualViewport) === null || _a === void 0 ? void 0 : _a.offsetTop) !== null && _b !== void 0 ? _b : 0),
                    point[1] + ((_d = (_c = this.win.visualViewport) === null || _c === void 0 ? void 0 : _c.offsetLeft) !== null && _d !== void 0 ? _d : 0),
                ]
                : point;
        };
        return TuiVisualViewportService;
    }());
    TuiVisualViewportService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiVisualViewportService, deps: [{ token: common.WINDOW }, { token: i3.TUI_IS_WEBKIT }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiVisualViewportService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiVisualViewportService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiVisualViewportService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }], ctorParameters: function () {
            return [{ type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_IS_WEBKIT]
                        }] }];
        } });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiBreakpointService = TuiBreakpointService;
    exports.TuiFormatDateService = TuiFormatDateService;
    exports.TuiHintService = TuiHintService;
    exports.TuiNightThemeService = TuiNightThemeService;
    exports.TuiPositionService = TuiPositionService;
    exports.TuiRouterLinkActiveService = TuiRouterLinkActiveService;
    exports.TuiSvgService = TuiSvgService;
    exports.TuiVisualViewportService = TuiVisualViewportService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
