// @taiga-ui/core/bundles/taiga-ui-core-pipes-month.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core/tokens'), require('rxjs/operators'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/pipes/month', ['exports', '@angular/core', '@taiga-ui/core/tokens', 'rxjs/operators', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.pipes = global["taiga-ui"].core.pipes || {}, global["taiga-ui"].core.pipes.month = {}), global.ng.core, global["taiga-ui"].core.tokens, global.rxjs.operators, global.rxjs));
})(this, (function (exports, i0, tokens, operators, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiMonthPipe = /** @class */ (function () {
        function TuiMonthPipe(months$) {
            this.months$ = months$;
        }
        TuiMonthPipe.prototype.transform = function (_a) {
            var month = _a.month;
            return this.months$.pipe(operators.map(function (months) { return months[month]; }));
        };
        return TuiMonthPipe;
    }());
    TuiMonthPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipe, deps: [{ token: tokens.TUI_MONTHS }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiMonthPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipe, name: "tuiMonth" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: "tuiMonth" }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_MONTHS]
                        }] }];
        } });

    var TuiMonthPipeModule = /** @class */ (function () {
        function TuiMonthPipeModule() {
        }
        return TuiMonthPipeModule;
    }());
    TuiMonthPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiMonthPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipeModule, declarations: [TuiMonthPipe], exports: [TuiMonthPipe] });
    TuiMonthPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMonthPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        exports: [TuiMonthPipe],
                        declarations: [TuiMonthPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiMonthPipe = TuiMonthPipe;
    exports.TuiMonthPipeModule = TuiMonthPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
