// @taiga-ui/core/bundles/taiga-ui-core-pipes-format-number.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/utils/format')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/pipes/format-number', ['exports', '@angular/core', '@taiga-ui/core/tokens', '@taiga-ui/core/utils/format'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.pipes = global["taiga-ui"].core.pipes || {}, global["taiga-ui"].core.pipes["format-number"] = {}), global.ng.core, global["taiga-ui"].core.tokens, global["taiga-ui"].core.utils.format));
})(this, (function (exports, i0, tokens, format) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiFormatNumberPipe = /** @class */ (function () {
        function TuiFormatNumberPipe(numberFormat) {
            this.numberFormat = numberFormat;
        }
        /**
         * Formats number adding thousand separators and correct decimal separator
         * padding decimal part with zeroes to given length
         * @param value number
         * @param settings See {@link TuiNumberFormatSettings}
         */
        TuiFormatNumberPipe.prototype.transform = function (value, settings) {
            if (settings === void 0) { settings = {}; }
            return format.tuiFormatNumber(value, Object.assign(Object.assign({}, this.numberFormat), settings));
        };
        return TuiFormatNumberPipe;
    }());
    TuiFormatNumberPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipe, deps: [{ token: tokens.TUI_NUMBER_FORMAT }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiFormatNumberPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipe, name: "tuiFormatNumber" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: "tuiFormatNumber" }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_NUMBER_FORMAT]
                        }] }];
        } });

    var TuiFormatNumberPipeModule = /** @class */ (function () {
        function TuiFormatNumberPipeModule() {
        }
        return TuiFormatNumberPipeModule;
    }());
    TuiFormatNumberPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFormatNumberPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipeModule, declarations: [TuiFormatNumberPipe], exports: [TuiFormatNumberPipe] });
    TuiFormatNumberPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatNumberPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        exports: [TuiFormatNumberPipe],
                        declarations: [TuiFormatNumberPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiFormatNumberPipe = TuiFormatNumberPipe;
    exports.TuiFormatNumberPipeModule = TuiFormatNumberPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
