// @taiga-ui/core/bundles/taiga-ui-core-components-hosted-dropdown.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core/abstract'), require('@taiga-ui/cdk'), require('@taiga-ui/core/directives/dropdown'), require('@taiga-ui/core/utils/miscellaneous'), require('@tinkoff/ng-event-plugins'), require('rxjs'), require('rxjs/operators'), require('@tinkoff/ng-polymorpheus'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/hosted-dropdown', ['exports', '@angular/core', '@taiga-ui/core/abstract', '@taiga-ui/cdk', '@taiga-ui/core/directives/dropdown', '@taiga-ui/core/utils/miscellaneous', '@tinkoff/ng-event-plugins', 'rxjs', 'rxjs/operators', '@tinkoff/ng-polymorpheus', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components["hosted-dropdown"] = {}), global.ng.core, global["taiga-ui"].core.abstract, global.i1, global["taiga-ui"].core.directives.dropdown, global["taiga-ui"].core.utils.miscellaneous, global.ngEventPlugins, global.rxjs, global.rxjs.operators, global.i4, global.ng.common));
})(this, (function (exports, i0, abstract, i1, i3, miscellaneous, ngEventPlugins, rxjs, operators, i4, i5) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiAccessorProxyDirective = /** @class */ (function (_super) {
        __extends(TuiAccessorProxyDirective, _super);
        function TuiAccessorProxyDirective() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.type = 'dummy';
            return _this;
        }
        TuiAccessorProxyDirective.prototype.getPosition = function () {
            return [0, 0];
        };
        return TuiAccessorProxyDirective;
    }(abstract.TuiPositionAccessor));
    TuiAccessorProxyDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAccessorProxyDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiAccessorProxyDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAccessorProxyDirective, selector: "[tuiAccessorProxy]", providers: [
            {
                provide: abstract.TuiPositionAccessor,
                multi: true,
                deps: [
                    [new i0.Optional(), new i0.SkipSelf(), TuiAccessorProxyDirective],
                    TuiAccessorProxyDirective,
                ],
                useFactory: function (accessor, fallback) { return accessor || fallback; },
            },
        ], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAccessorProxyDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiAccessorProxy]',
                        providers: [
                            {
                                provide: abstract.TuiPositionAccessor,
                                multi: true,
                                deps: [
                                    [new i0.Optional(), new i0.SkipSelf(), TuiAccessorProxyDirective],
                                    TuiAccessorProxyDirective,
                                ],
                                useFactory: function (accessor, fallback) { return accessor || fallback; },
                            },
                        ],
                    }]
            }] });

    var TuiHostedDropdownConnectorDirective = /** @class */ (function () {
        function TuiHostedDropdownConnectorDirective() {
        }
        return TuiHostedDropdownConnectorDirective;
    }());
    TuiHostedDropdownConnectorDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownConnectorDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiHostedDropdownConnectorDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHostedDropdownConnectorDirective, selector: "[tuiHostedDropdownHost]", ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownConnectorDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiHostedDropdownHost]',
                    }]
            }] });

    function shouldClose(event) {
        var _a;
        return ('key' in event &&
            event.key.toLowerCase() === 'escape' &&
            this.canOpen &&
            this.open &&
            !((_a = this.dropdown) === null || _a === void 0 ? void 0 : _a.nextElementSibling));
    }
    /* eslint-disable @typescript-eslint/member-ordering */
    var TuiHostedDropdownComponent = /** @class */ (function () {
        function TuiHostedDropdownComponent(hover$, el) {
            var _this = this;
            this.hover$ = hover$;
            this.el = el;
            /** TODO: rename in 4.0 */
            this.openChange = new rxjs.BehaviorSubject(false);
            this.sided = false;
            this.canOpen = true;
            this.open$ = rxjs.merge(this.openChange, this.hover$ || rxjs.EMPTY).pipe(operators.skip(1), operators.distinctUntilChanged());
            this.focusedChange = new i0.EventEmitter();
            this.close = function () { return _this.updateOpen(false); };
        }
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "open", {
            get: function () {
                return this.openChange.value;
            },
            set: function (open) {
                this.openChange.next(open);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "host", {
            get: function () {
                var _a;
                return ((_a = this.dropdownHost) === null || _a === void 0 ? void 0 : _a.nativeElement) || this.el.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "computedHost", {
            get: function () {
                var _a;
                return (((_a = this.dropdownHost) === null || _a === void 0 ? void 0 : _a.nativeElement) ||
                    this.nativeFocusableElement ||
                    this.el.nativeElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "dropdown", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.dropdownDirective) === null || _a === void 0 ? void 0 : _a.dropdownBoxRef) === null || _b === void 0 ? void 0 : _b.location.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return i1.tuiIsNativeKeyboardFocusable(this.host)
                    ? this.host
                    : i1.tuiGetClosestFocusable({
                        initial: this.host,
                        root: this.el.nativeElement,
                    });
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "focused", {
            get: function () {
                return (i1.tuiIsNativeFocusedIn(this.host) ||
                    (this.open &&
                        !!this.wrapper &&
                        i1.tuiIsNativeFocusedIn(this.wrapper.nativeElement)));
            },
            enumerable: false,
            configurable: true
        });
        TuiHostedDropdownComponent.prototype.onFocusIn = function (target) {
            if (!this.computedHost.contains(target)) {
                this.updateOpen(false);
            }
        };
        TuiHostedDropdownComponent.prototype.onClick = function (target) {
            var _a;
            if (!this.hostEditable &&
                this.computedHost.contains(target) &&
                !((_a = this.hover$) === null || _a === void 0 ? void 0 : _a.hovered)) {
                this.updateOpen(!this.open);
            }
        };
        TuiHostedDropdownComponent.prototype.onKeyDownEsc = function (event) {
            event.stopPropagation();
            this.closeDropdown();
        };
        TuiHostedDropdownComponent.prototype.onArrow = function (event, down) {
            this.focusDropdown(event, down);
        };
        TuiHostedDropdownComponent.prototype.onKeydown = function (_c) {
            var key = _c.key, target = _c.target, defaultPrevented = _c.defaultPrevented;
            if (!defaultPrevented &&
                miscellaneous.tuiIsEditingKey(key) &&
                this.hostEditable &&
                i1.tuiIsHTMLElement(target) &&
                !i1.tuiIsElementEditable(target)) {
                this.focusHost();
            }
        };
        TuiHostedDropdownComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
            if (!active) {
                this.updateOpen(false);
            }
        };
        TuiHostedDropdownComponent.prototype.onHostObscured = function (obscured) {
            if (obscured) {
                this.closeDropdown();
            }
        };
        TuiHostedDropdownComponent.prototype.updateOpen = function (open) {
            if (!open || this.canOpen) {
                this.open = open;
            }
        };
        Object.defineProperty(TuiHostedDropdownComponent.prototype, "hostEditable", {
            get: function () {
                return i1.tuiIsElementEditable(this.computedHost);
            },
            enumerable: false,
            configurable: true
        });
        TuiHostedDropdownComponent.prototype.focusDropdown = function (event, first) {
            var host = this.nativeFocusableElement;
            if (!host ||
                !i1.tuiIsHTMLElement(host) ||
                !i1.tuiIsElement(event.target) ||
                !host.contains(event.target)) {
                return;
            }
            if (!this.wrapper ||
                !this.open ||
                !this.dropdown ||
                !i1.tuiIsHTMLElement(this.wrapper.nativeElement.nextElementSibling)) {
                this.updateOpen(true);
                if (!i1.tuiIsElementEditable(host)) {
                    event.preventDefault();
                }
                return;
            }
            var initial = first
                ? this.wrapper.nativeElement
                : this.wrapper.nativeElement.nextElementSibling;
            var focusable = i1.tuiGetClosestFocusable({
                initial: initial,
                root: this.wrapper.nativeElement,
                previous: !first,
            });
            if (!focusable) {
                return;
            }
            focusable.focus();
            event.preventDefault();
        };
        TuiHostedDropdownComponent.prototype.closeDropdown = function () {
            if (this.focused) {
                this.focusHost();
            }
            this.updateOpen(false);
        };
        TuiHostedDropdownComponent.prototype.focusHost = function () {
            var host = this.nativeFocusableElement;
            if (host) {
                host.focus({ preventScroll: true });
            }
        };
        TuiHostedDropdownComponent.prototype.updateFocused = function (focused) {
            this.focusedChange.emit(focused);
        };
        return TuiHostedDropdownComponent;
    }());
    TuiHostedDropdownComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownComponent, deps: [{ token: i3.TuiDropdownHoverDirective, optional: true }, { token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiHostedDropdownComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: { content: "content", sided: "sided", canOpen: "canOpen", open: "open" }, outputs: { open$: "openChange", focusedChange: "focusedChange" }, host: { listeners: { "focusin": "onFocusIn($event.target)", "click": "onClick($event.target)", "document:keydown.silent": "onKeyDownEsc($event)", "keydown.arrowDown": "onArrow($event,true)", "keydown.arrowUp": "onArrow($event,false)" }, properties: { "class._hosted_dropdown_focused": "this.focused" } }, providers: [
            i1.tuiAsFocusableItemAccessor(TuiHostedDropdownComponent),
            {
                provide: TuiAccessorProxyDirective,
                deps: [[new i0.Optional(), new i0.Self(), abstract.TuiPositionAccessor]],
                useFactory: function (position) { return position === null || position === void 0 ? void 0 : position[0]; },
            },
        ], queries: [{ propertyName: "dropdownHost", first: true, predicate: TuiHostedDropdownConnectorDirective, descendants: true, read: i0.ElementRef }], viewQueries: [{ propertyName: "wrapper", first: true, predicate: ["wrapper"], descendants: true, read: i0.ElementRef }, { propertyName: "dropdownDirective", first: true, predicate: i3.TuiDropdownDirective, descendants: true }, { propertyName: "activeZone", first: true, predicate: i1.TuiActiveZoneDirective, descendants: true }], ngImport: i0__namespace, template: "<div\n    *tuiLet=\"!!(open$ | async) as isOpen\"\n    #activeZone=\"tuiActiveZone\"\n    tuiAccessorProxy\n    class=\"t-wrapper\"\n    [tuiDropdownSided]=\"sided\"\n    [tuiDropdown]=\"dropdown\"\n    [tuiDropdownManual]=\"isOpen && canOpen\"\n    [tuiObscuredEnabled]=\"isOpen\"\n    (tuiObscured)=\"onHostObscured($event)\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <ng-content></ng-content>\n    <ng-template\n        #dropdown=\"polymorpheus\"\n        [polymorpheus]=\"context\"\n    >\n        <div\n            #wrapper\n            (keydown)=\"onKeydown($event)\"\n        >\n            <ng-container\n                *polymorpheusOutlet=\"\n                    content as text;\n                    context: {\n                        $implicit: activeZone,\n                        close: close\n                    }\n                \"\n            >\n                {{ text }}\n            </ng-container>\n        </div>\n        <!--This DIV is here to start backwards TreeWalker for focusing last focusable item on ArrowUp-->\n        <div></div>\n    </ng-template>\n</div>\n", styles: [":host{display:inline-flex}.t-wrapper{border-radius:inherit;height:inherit;flex:1 1 auto;width:100%}\n"], directives: [{ type: i1__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: TuiAccessorProxyDirective, selector: "[tuiAccessorProxy]" }, { type: i3__namespace.TuiDropdownPositionSidedDirective, selector: "[tuiDropdownSided]", inputs: ["tuiDropdownSided", "tuiDropdownSidedOffset"] }, { type: i3__namespace.TuiDropdownDirective, selector: "[tuiDropdown]:not(ng-container)", inputs: ["tuiDropdown"], exportAs: ["tuiDropdown"] }, { type: i3__namespace.TuiDropdownDriverDirective, selector: "[tuiDropdown]" }, { type: i3__namespace.TuiDropdownManualDirective, selector: "[tuiDropdown][tuiDropdownManual]", inputs: ["tuiDropdownManual"] }, { type: i1__namespace.TuiObscuredDirective, selector: "[tuiObscured]", inputs: ["tuiObscuredEnabled"], outputs: ["tuiObscured"] }, { type: i1__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i4__namespace.PolymorpheusTemplate, selector: "ng-template[polymorpheus]", inputs: ["polymorpheus"], exportAs: ["polymorpheus"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i5__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i1.tuiDefaultProp()
    ], TuiHostedDropdownComponent.prototype, "sided", void 0);
    __decorate([
        i1.tuiDefaultProp()
    ], TuiHostedDropdownComponent.prototype, "canOpen", void 0);
    __decorate([
        i1.tuiDefaultProp()
    ], TuiHostedDropdownComponent.prototype, "open", null);
    __decorate([
        ngEventPlugins.shouldCall(shouldClose)
    ], TuiHostedDropdownComponent.prototype, "onKeyDownEsc", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-hosted-dropdown',
                        templateUrl: './hosted-dropdown.template.html',
                        styleUrls: ['./hosted-dropdown.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i1.tuiAsFocusableItemAccessor(TuiHostedDropdownComponent),
                            {
                                provide: TuiAccessorProxyDirective,
                                deps: [[new i0.Optional(), new i0.Self(), abstract.TuiPositionAccessor]],
                                useFactory: function (position) { return position === null || position === void 0 ? void 0 : position[0]; },
                            },
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i3__namespace.TuiDropdownHoverDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDropdownHoverDirective]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { dropdownHost: [{
                    type: i0.ContentChild,
                    args: [TuiHostedDropdownConnectorDirective, { read: i0.ElementRef }]
                }], wrapper: [{
                    type: i0.ViewChild,
                    args: ['wrapper', { read: i0.ElementRef }]
                }], dropdownDirective: [{
                    type: i0.ViewChild,
                    args: [i3.TuiDropdownDirective]
                }], activeZone: [{
                    type: i0.ViewChild,
                    args: [i1.TuiActiveZoneDirective]
                }], content: [{
                    type: i0.Input
                }], sided: [{
                    type: i0.Input
                }], canOpen: [{
                    type: i0.Input
                }], open$: [{
                    type: i0.Output,
                    args: ['openChange']
                }], focusedChange: [{
                    type: i0.Output
                }], open: [{
                    type: i0.Input
                }], focused: [{
                    type: i0.HostBinding,
                    args: ['class._hosted_dropdown_focused']
                }], onFocusIn: [{
                    type: i0.HostListener,
                    args: ['focusin', ['$event.target']]
                }], onClick: [{
                    type: i0.HostListener,
                    args: ['click', ['$event.target']]
                }], onKeyDownEsc: [{
                    type: i0.HostListener,
                    args: ['document:keydown.silent', ['$event']]
                }], onArrow: [{
                    type: i0.HostListener,
                    args: ['keydown.arrowDown', ['$event', 'true']]
                }, {
                    type: i0.HostListener,
                    args: ['keydown.arrowUp', ['$event', 'false']]
                }] } });

    var TuiHostedDropdownModule = /** @class */ (function () {
        function TuiHostedDropdownModule() {
        }
        return TuiHostedDropdownModule;
    }());
    TuiHostedDropdownModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiHostedDropdownModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownModule, declarations: [TuiAccessorProxyDirective,
            TuiHostedDropdownComponent,
            TuiHostedDropdownConnectorDirective], imports: [i5.CommonModule,
            i4.PolymorpheusModule,
            i1.TuiLetModule,
            i1.TuiObscuredModule,
            i1.TuiActiveZoneModule,
            i3.TuiDropdownModule], exports: [TuiHostedDropdownComponent,
            TuiHostedDropdownConnectorDirective,
            i3.TuiDropdownOptionsDirective] });
    TuiHostedDropdownModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownModule, imports: [[
                i5.CommonModule,
                i4.PolymorpheusModule,
                i1.TuiLetModule,
                i1.TuiObscuredModule,
                i1.TuiActiveZoneModule,
                i3.TuiDropdownModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHostedDropdownModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i5.CommonModule,
                            i4.PolymorpheusModule,
                            i1.TuiLetModule,
                            i1.TuiObscuredModule,
                            i1.TuiActiveZoneModule,
                            i3.TuiDropdownModule,
                        ],
                        declarations: [
                            TuiAccessorProxyDirective,
                            TuiHostedDropdownComponent,
                            TuiHostedDropdownConnectorDirective,
                        ],
                        exports: [
                            TuiHostedDropdownComponent,
                            TuiHostedDropdownConnectorDirective,
                            i3.TuiDropdownOptionsDirective,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiAccessorProxyDirective = TuiAccessorProxyDirective;
    exports.TuiHostedDropdownComponent = TuiHostedDropdownComponent;
    exports.TuiHostedDropdownConnectorDirective = TuiHostedDropdownConnectorDirective;
    exports.TuiHostedDropdownModule = TuiHostedDropdownModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
