// @taiga-ui/core/bundles/taiga-ui-core-pipes-flag.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core/components/svg'), require('@taiga-ui/core/constants'), require('@taiga-ui/i18n')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/pipes/flag', ['exports', '@angular/core', '@taiga-ui/core/components/svg', '@taiga-ui/core/constants', '@taiga-ui/i18n'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.pipes = global["taiga-ui"].core.pipes || {}, global["taiga-ui"].core.pipes.flag = {}), global.ng.core, global["taiga-ui"].core.components.svg, global["taiga-ui"].core.constants, global.i18n));
})(this, (function (exports, i0, svg, constants, i18n) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiFlagPipe = /** @class */ (function () {
        function TuiFlagPipe(svgOptions) {
            this.svgOptions = svgOptions;
            this.staticPath = this.svgOptions
                .path("tuiIcon")
                .replace("tuiIcon.svg#tuiIcon", "")
                .replace("tuiIcon.svg" + constants.TUI_CACHE_BUSTING_PAYLOAD + "#tuiIcon", "");
        }
        TuiFlagPipe.prototype.transform = function (countryIsoCode) {
            if (!countryIsoCode) {
                return null;
            }
            switch (countryIsoCode) {
                case i18n.TuiCountryIsoCode.BL:
                case i18n.TuiCountryIsoCode.BQ:
                case i18n.TuiCountryIsoCode.CW:
                case i18n.TuiCountryIsoCode.GF:
                case i18n.TuiCountryIsoCode.GP:
                case i18n.TuiCountryIsoCode.MF:
                case i18n.TuiCountryIsoCode.MQ:
                case i18n.TuiCountryIsoCode.NC:
                case i18n.TuiCountryIsoCode.RE:
                case i18n.TuiCountryIsoCode.YT:
                    return "" + this.staticPath + i18n.TuiCountryIsoCode.FR + ".png";
                case i18n.TuiCountryIsoCode.SX:
                    return "" + this.staticPath + i18n.TuiCountryIsoCode.NL + ".png";
                default:
                    return "" + this.staticPath + countryIsoCode + ".png";
            }
        };
        return TuiFlagPipe;
    }());
    TuiFlagPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipe, deps: [{ token: svg.TUI_SVG_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiFlagPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipe, name: "tuiFlag" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: "tuiFlag" }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [svg.TUI_SVG_OPTIONS]
                        }] }];
        } });

    var TuiFlagPipeModule = /** @class */ (function () {
        function TuiFlagPipeModule() {
        }
        return TuiFlagPipeModule;
    }());
    TuiFlagPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFlagPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipeModule, declarations: [TuiFlagPipe], exports: [TuiFlagPipe] });
    TuiFlagPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFlagPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFlagPipe],
                        exports: [TuiFlagPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiFlagPipe = TuiFlagPipe;
    exports.TuiFlagPipeModule = TuiFlagPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
