// @taiga-ui/core/bundles/taiga-ui-core-providers.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/directives/mode'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/cdk')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/providers', ['exports', '@angular/core', '@taiga-ui/core/tokens', '@taiga-ui/core/directives/mode', 'rxjs', 'rxjs/operators', '@taiga-ui/cdk'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.providers = {}), global.ng.core, global["taiga-ui"].core.tokens, global["taiga-ui"].core.directives.mode, global.rxjs, global.rxjs.operators, global.cdk));
})(this, (function (exports, core, tokens, mode, rxjs, operators, cdk) { 'use strict';

    var TUI_IS_MOBILE_RES_PROVIDER = {
        provide: tokens.TUI_IS_MOBILE_RES,
        deps: [[new core.SkipSelf(), tokens.TUI_IS_MOBILE_RES], core.ElementRef],
        useFactory: function (mobile$, _a) {
            var nativeElement = _a.nativeElement;
            nativeElement["$.class._mobile"] = mobile$;
            return mobile$;
        },
    };

    var MODE_PROVIDER = {
        provide: tokens.TUI_MODE,
        deps: [[new core.Optional(), mode.TuiModeDirective], core.ElementRef],
        useFactory: function (mode, _a) {
            var nativeElement = _a.nativeElement;
            var mode$ = mode
                ? mode.change$.pipe(operators.startWith(null), operators.map(function () { return mode.mode; }))
                : rxjs.of(null);
            nativeElement["$.data-mode.attr"] = mode$;
            return mode$;
        },
    };

    function tuiWatchedControllerFactory(controller, cdr, destroy$) {
        controller.change$.pipe(cdk.tuiWatch(cdr), operators.takeUntil(destroy$)).subscribe();
        return controller;
    }

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MODE_PROVIDER = MODE_PROVIDER;
    exports.TUI_IS_MOBILE_RES_PROVIDER = TUI_IS_MOBILE_RES_PROVIDER;
    exports.tuiWatchedControllerFactory = tuiWatchedControllerFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
