// @taiga-ui/core/bundles/taiga-ui-core-components-data-list.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/directives'), require('@taiga-ui/core/tokens'), require('rxjs/operators'), require('@taiga-ui/core/directives/dropdown'), require('@tinkoff/ng-event-plugins'), require('@taiga-ui/core/components/svg'), require('@tinkoff/ng-polymorpheus'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/data-list', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/directives', '@taiga-ui/core/tokens', 'rxjs/operators', '@taiga-ui/core/directives/dropdown', '@tinkoff/ng-event-plugins', '@taiga-ui/core/components/svg', '@tinkoff/ng-polymorpheus', '@angular/common', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components["data-list"] = {}), global.ng.core, global.cdk, global["taiga-ui"].core.directives, global["taiga-ui"].core.tokens, global.rxjs.operators, global["taiga-ui"].core.directives.dropdown, global.ngEventPlugins, global["taiga-ui"].core.components.svg, global.i2, global.ng.common, global.rxjs));
})(this, (function (exports, i0, cdk, i3, tokens, operators, i5, ngEventPlugins, i1, i2, i1$1, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function shouldFocus(_a) {
        var currentTarget = _a.currentTarget;
        return !cdk.tuiIsNativeFocused(currentTarget);
    }
    // TODO: Consider all use cases for aria roles
    var TuiOptionComponent = /** @class */ (function () {
        function TuiOptionComponent(content, dataList, el, host, dropdown) {
            this.content = content;
            this.dataList = dataList;
            this.el = el;
            this.host = host;
            this.dropdown = dropdown;
            /** @deprecated use size on {@link TuiDataListComponent} instead */
            this.size = null;
            this.role = 'option';
            this.disabled = false;
        }
        Object.defineProperty(TuiOptionComponent.prototype, "active", {
            get: function () {
                return !!this.dropdown && !!this.dropdown.dropdownBoxRef;
            },
            enumerable: false,
            configurable: true
        });
        TuiOptionComponent.prototype.onClick = function () {
            if (this.host && this.value !== undefined) {
                this.host.handleOption(this.value);
            }
        };
        // @bad TODO: Consider aria-activedescendant for proper accessibility implementation
        TuiOptionComponent.prototype.onMouseMove = function (_a) {
            var currentTarget = _a.currentTarget;
            currentTarget.focus({ preventScroll: true });
        };
        // Preventing focus loss upon focused option removal
        TuiOptionComponent.prototype.ngOnDestroy = function () {
            this.dataList.handleFocusLossIfNecessary(this.el.nativeElement);
        };
        return TuiOptionComponent;
    }());
    TuiOptionComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiOptionComponent, deps: [{ token: tokens.TUI_OPTION_CONTENT, optional: true }, { token: i0.forwardRef(function () { return TuiDataListComponent; }) }, { token: i0.ElementRef }, { token: tokens.TUI_DATA_LIST_HOST, optional: true }, { token: i5.TuiDropdownDirective, optional: true, self: true }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiOptionComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiOptionComponent, selector: "button[tuiOption], a[tuiOption]", inputs: { size: "size", role: "role", disabled: "disabled", value: "value" }, host: { attributes: { "tabIndex": "-1", "type": "button" }, listeners: { "click": "onClick()", "mousemove.silent": "onMouseMove($event)" }, properties: { "attr.disabled": "disabled || null", "attr.data-size": "this.size", "attr.role": "this.role", "class._with-dropdown": "this.active" } }, ngImport: i0__namespace, template: "<ng-container *polymorpheusOutlet=\"content || base as text; context: {$implicit: base}\">\n    {{ text }}\n</ng-container>\n<ng-template #base>\n    <ng-content></ng-content>\n    <tui-svg\n        *ngIf=\"dropdown\"\n        src=\"tuiIconChevronRight\"\n        class=\"t-arrow\"\n    ></tui-svg>\n</ng-template>\n", styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;border:0;background:none;font-size:inherit;line-height:inherit;transition-property:background;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;display:flex;align-items:center;justify-content:space-between;text-align:left;box-sizing:border-box;color:var(--tui-text-01);border-radius:var(--tui-radius-s);outline:none;text-decoration:none;cursor:pointer;background-clip:padding-box;font:var(--tui-font-text-s);min-height:2.5rem;padding:.375rem .5rem;margin:.125rem 0}:host:disabled{opacity:var(--tui-disabled-opacity);cursor:default}:host:focus,:host._with-dropdown{background-color:var(--tui-clear)}:host :host-context([data-list-size=\"s\"]),:host :host-context([data-list-size=\"xs\"]),:host[data-size=s][data-size=s],:host[data-size=xs][data-size=xs]{font:var(--tui-font-text-s);min-height:2rem;padding:.3125rem .5rem;margin:var(--tui-data-list-margin) 0}:host :host-context([data-list-size=\"m\"]),:host[data-size=m][data-size=m]{font:var(--tui-font-text-s);min-height:2.5rem;padding:.375rem .5rem;margin:var(--tui-data-list-margin) 0}:host :host-context([data-list-size=\"l\"]),:host[data-size=l][data-size=l]{font:var(--tui-font-text-m);min-height:2.75rem;padding:.375rem .625rem;margin:var(--tui-data-list-margin) 0}.t-arrow{margin:0 -.5rem 0 .75rem}\n"], components: [{ type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1__namespace$1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiOptionComponent.prototype, "size", void 0);
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiOptionComponent.prototype, "role", void 0);
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiOptionComponent.prototype, "disabled", void 0);
    __decorate([
        ngEventPlugins.shouldCall(shouldFocus)
    ], TuiOptionComponent.prototype, "onMouseMove", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiOptionComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'button[tuiOption], a[tuiOption]',
                        templateUrl: './option.template.html',
                        styleUrls: ['./option.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        host: {
                            tabIndex: '-1',
                            type: 'button',
                            '[attr.disabled]': 'disabled || null',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_OPTION_CONTENT]
                        }] }, { type: TuiDataListComponent, decorators: [{
                            type: i0.Inject,
                            args: [i0.forwardRef(function () { return TuiDataListComponent; })]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_DATA_LIST_HOST]
                        }] }, { type: i5__namespace.TuiDropdownDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i5.TuiDropdownDirective]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], role: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.role']
                }], disabled: [{
                    type: i0.Input
                }], value: [{
                    type: i0.Input
                }], active: [{
                    type: i0.HostBinding,
                    args: ['class._with-dropdown']
                }], onClick: [{
                    type: i0.HostListener,
                    args: ['click']
                }], onMouseMove: [{
                    type: i0.HostListener,
                    args: ['mousemove.silent', ['$event']]
                }] } });

    // TODO: Consider aria-activedescendant for proper accessibility implementation
    var TuiDataListComponent = /** @class */ (function () {
        function TuiDataListComponent(controller, el, defaultEmptyContent$) {
            var _a;
            this.controller = controller;
            this.el = el;
            this.defaultEmptyContent$ = defaultEmptyContent$;
            this.options = cdk.EMPTY_QUERY;
            this.role = 'listbox';
            this.size = ((_a = this.controller) === null || _a === void 0 ? void 0 : _a.size) || 'm';
        }
        Object.defineProperty(TuiDataListComponent.prototype, "empty$", {
            get: function () {
                return cdk.tuiQueryListChanges(this.options).pipe(operators.map(function (_b) {
                    var length = _b.length;
                    return !length;
                }));
            },
            enumerable: false,
            configurable: true
        });
        TuiDataListComponent.prototype.onFocusIn = function (relatedTarget, currentTarget) {
            if (!currentTarget.contains(relatedTarget) && !this.origin) {
                this.origin = relatedTarget;
            }
        };
        TuiDataListComponent.prototype.noop = function () { };
        TuiDataListComponent.prototype.onKeyDownArrow = function (current, step) {
            var elements = this.elements;
            cdk.tuiMoveFocus(elements.indexOf(current), elements, step);
        };
        // TODO: Consider aria-activedescendant for proper accessibility implementation
        TuiDataListComponent.prototype.handleFocusLossIfNecessary = function (element) {
            if (element === void 0) { element = this.el.nativeElement; }
            if (this.origin && cdk.tuiIsNativeFocusedIn(element)) {
                cdk.tuiSetNativeMouseFocused(this.origin, true, true);
            }
        };
        TuiDataListComponent.prototype.getOptions = function (includeDisabled) {
            if (includeDisabled === void 0) { includeDisabled = false; }
            return this.options
                .filter(function (_b) {
                var disabled = _b.disabled;
                return includeDisabled || !disabled;
            })
                .map(function (_b) {
                var value = _b.value;
                return value;
            })
                .filter(cdk.tuiIsPresent);
        };
        TuiDataListComponent.prototype.onFocus = function (_b, top) {
            var target = _b.target;
            if (!cdk.tuiIsElement(target)) {
                return;
            }
            var elements = this.elements;
            cdk.tuiMoveFocus(top ? -1 : elements.length, elements, top ? 1 : -1);
            this.handleFocusLossIfNecessary(target);
        };
        Object.defineProperty(TuiDataListComponent.prototype, "elements", {
            get: function () {
                return Array.from(this.el.nativeElement.querySelectorAll('[tuiOption]'));
            },
            enumerable: false,
            configurable: true
        });
        return TuiDataListComponent;
    }());
    TuiDataListComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListComponent, deps: [{ token: i3.TUI_TEXTFIELD_WATCHED_CONTROLLER, optional: true }, { token: i0.ElementRef }, { token: tokens.TUI_NOTHING_FOUND_MESSAGE }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiDataListComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiDataListComponent, selector: "tui-data-list", inputs: { role: "role", emptyContent: "emptyContent", size: "size" }, host: { listeners: { "focusin": "onFocusIn($event.relatedTarget,$event.currentTarget)", "mousedown.prevent": "noop()", "keydown.arrowDown.prevent": "onKeyDownArrow($event.target,1)", "keydown.arrowUp.prevent": "onKeyDownArrow($event.target,-1)", "wheel.silent.passive": "handleFocusLossIfNecessary()", "mouseleave": "handleFocusLossIfNecessary($event.target)" }, properties: { "attr.role": "this.role", "attr.data-list-size": "this.size" } }, providers: [
            tokens.tuiAsDataListAccessor(TuiDataListComponent),
            i3.TEXTFIELD_CONTROLLER_PROVIDER,
        ], queries: [{ propertyName: "options", predicate: i0__namespace.forwardRef(function () { return TuiOptionComponent; }), descendants: true }], ngImport: i0__namespace, template: "<div\n    tabindex=\"0\"\n    class=\"t-trap\"\n    (focusin)=\"onFocus($event, true)\"\n></div>\n<ng-content></ng-content>\n<div\n    *ngIf=\"empty$ | async\"\n    class=\"t-empty\"\n>\n    <ng-container *polymorpheusOutlet=\"emptyContent || (defaultEmptyContent$ | async) as text\">\n        {{ text }}\n    </ng-container>\n</div>\n<div\n    tabindex=\"0\"\n    class=\"t-trap\"\n    (focusin)=\"onFocus($event, false)\"\n></div>\n", styles: ["tui-data-list{--tui-data-list-padding: .25rem;--tui-data-list-margin: .0625rem;display:flex;font:var(--tui-font-text-m);flex-direction:column;padding:calc(var(--tui-data-list-padding) - var(--tui-data-list-margin)) var(--tui-data-list-padding);color:var(--tui-text-03)}tui-data-list:focus-within .t-trap{display:none}tui-data-list:focus-within [tuiOption]._with-dropdown:not(:focus){background-color:transparent}tui-data-list[data-list-size=s]{--tui-data-list-margin: 0rem}tui-data-list[data-list-size=l]{--tui-data-list-padding: .375rem;--tui-data-list-margin: .125rem}tui-data-list>.t-empty{margin:.75rem 1rem}tui-opt-group{position:relative;display:flex;font:var(--tui-font-text-xs);color:var(--tui-text-02);flex-direction:column;line-height:1rem}tui-data-list[data-list-size=l] tui-opt-group{font:var(--tui-font-text-s);line-height:1.25rem}tui-data-list[data-list-size=l] tui-opt-group:before{padding-left:.625rem;padding-right:.625rem}tui-data-list[data-list-size=l] tui-opt-group:after{left:.625rem;right:.625rem}tui-opt-group:empty:before,tui-opt-group:empty:after{display:none}tui-opt-group:before{content:attr(data-label);padding:var(--tui-data-list-padding) .5rem var(--tui-data-list-padding);margin:var(--tui-data-list-margin) 0;white-space:normal;word-break:break-word}tui-opt-group:after{position:absolute;left:.5rem;right:.5rem;top:var(--tui-data-list-padding);height:1px;background:var(--tui-base-03)}tui-opt-group:not(:empty)~tui-opt-group:before{padding-top:calc(.75rem + var(--tui-data-list-padding))}tui-opt-group:not(:empty)~tui-opt-group[data-label=\"\"]:before{padding:var(--tui-data-list-padding) 0}tui-opt-group:not(:empty)~tui-opt-group:after{content:\"\"}tui-opt-group[data-label=\"\"]:before{padding:0 .5rem}\n"], directives: [{ type: i1__namespace$1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i1__namespace$1.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiDataListComponent.prototype, "role", void 0);
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiDataListComponent.prototype, "size", void 0);
    __decorate([
        cdk.tuiPure
    ], TuiDataListComponent.prototype, "empty$", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-data-list',
                        templateUrl: './data-list.template.html',
                        styleUrls: ['./data-list.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        encapsulation: i0.ViewEncapsulation.None,
                        providers: [
                            tokens.tuiAsDataListAccessor(TuiDataListComponent),
                            i3.TEXTFIELD_CONTROLLER_PROVIDER,
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i3__namespace.TuiTextfieldController, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TUI_TEXTFIELD_WATCHED_CONTROLLER]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i4__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_NOTHING_FOUND_MESSAGE]
                        }] }];
        }, propDecorators: { options: [{
                    type: i0.ContentChildren,
                    args: [i0.forwardRef(function () { return TuiOptionComponent; }), { descendants: true }]
                }], role: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.role']
                }], emptyContent: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-list-size']
                }], empty$: [], onFocusIn: [{
                    type: i0.HostListener,
                    args: ['focusin', ['$event.relatedTarget', '$event.currentTarget']]
                }], noop: [{
                    type: i0.HostListener,
                    args: ['mousedown.prevent']
                }], onKeyDownArrow: [{
                    type: i0.HostListener,
                    args: ['keydown.arrowDown.prevent', ['$event.target', '1']]
                }, {
                    type: i0.HostListener,
                    args: ['keydown.arrowUp.prevent', ['$event.target', '-1']]
                }], handleFocusLossIfNecessary: [{
                    type: i0.HostListener,
                    args: ['wheel.silent.passive']
                }, {
                    type: i0.HostListener,
                    args: ['mouseleave', ['$event.target']]
                }] } });

    var TuiDataListDirective = /** @class */ (function () {
        function TuiDataListDirective() {
        }
        return TuiDataListDirective;
    }());
    TuiDataListDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiDataListDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDataListDirective, selector: "ng-template[tuiDataList]", ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'ng-template[tuiDataList]',
                    }]
            }] });
    function tuiAsDataList(useExisting) {
        return {
            provide: TuiDataListDirective,
            useExisting: useExisting,
        };
    }

    var TuiOptGroupDirective = /** @class */ (function () {
        function TuiOptGroupDirective() {
            this.label = '';
        }
        return TuiOptGroupDirective;
    }());
    TuiOptGroupDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiOptGroupDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiOptGroupDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiOptGroupDirective, selector: "tui-opt-group", inputs: { label: "label" }, host: { attributes: { "role": "group" }, properties: { "attr.data-label": "this.label" } }, ngImport: i0__namespace });
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiOptGroupDirective.prototype, "label", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiOptGroupDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-opt-group',
                        host: {
                            role: 'group',
                        },
                    }]
            }], propDecorators: { label: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-label']
                }] } });

    var TuiDataListModule = /** @class */ (function () {
        function TuiDataListModule() {
        }
        return TuiDataListModule;
    }());
    TuiDataListModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiDataListModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListModule, declarations: [TuiDataListComponent,
            TuiOptionComponent,
            TuiOptGroupDirective,
            TuiDataListDirective], imports: [i1$1.CommonModule, i1.TuiSvgModule, i2.PolymorpheusModule], exports: [TuiDataListComponent,
            TuiOptionComponent,
            TuiOptGroupDirective,
            TuiDataListDirective] });
    TuiDataListModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListModule, imports: [[i1$1.CommonModule, i1.TuiSvgModule, i2.PolymorpheusModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDataListModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1$1.CommonModule, i1.TuiSvgModule, i2.PolymorpheusModule],
                        declarations: [
                            TuiDataListComponent,
                            TuiOptionComponent,
                            TuiOptGroupDirective,
                            TuiDataListDirective,
                        ],
                        exports: [
                            TuiDataListComponent,
                            TuiOptionComponent,
                            TuiOptGroupDirective,
                            TuiDataListDirective,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiDataListComponent = TuiDataListComponent;
    exports.TuiDataListDirective = TuiDataListDirective;
    exports.TuiDataListModule = TuiDataListModule;
    exports.TuiOptGroupDirective = TuiOptGroupDirective;
    exports.TuiOptionComponent = TuiOptionComponent;
    exports.tuiAsDataList = tuiAsDataList;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
