// @taiga-ui/core/bundles/taiga-ui-core-components-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/animations'), require('@taiga-ui/core/tokens'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@ng-web-apis/common'), require('@taiga-ui/core/utils'), require('@taiga-ui/core/components/button')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/dialog', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/animations', '@taiga-ui/core/tokens', '@tinkoff/ng-polymorpheus', 'rxjs', 'rxjs/operators', '@angular/common', '@ng-web-apis/common', '@taiga-ui/core/utils', '@taiga-ui/core/components/button'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components.dialog = {}), global.ng.core, global.i4, global["taiga-ui"].core.animations, global["taiga-ui"].core.tokens, global.i3, global.rxjs, global.rxjs.operators, global.ng.common, global.common, global["taiga-ui"].core.utils, global["taiga-ui"].core.components.button));
})(this, (function (exports, i0, i4, animations, tokens, i3, i5, operators, i2, common, utils, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TUI_DIALOG_DEFAULT_OPTIONS = {
        size: "m",
        required: false,
        closeable: true,
        dismissible: true,
        label: "",
        header: "",
    };
    /**
     * A stream to close dialogs
     */
    var TUI_DIALOGS_CLOSE = new i0.InjectionToken("[TUI_DIALOGS_CLOSE]", {
        factory: function () { return i5.EMPTY; },
    });
    /**
     * Default parameters for dialog component
     */
    var TUI_DIALOG_OPTIONS = new i0.InjectionToken("[TUI_DIALOG_OPTIONS]", {
        factory: function () { return TUI_DIALOG_DEFAULT_OPTIONS; },
    });
    function tuiDialogOptionsProvider(options) {
        return {
            provide: TUI_DIALOG_OPTIONS,
            useValue: Object.assign(Object.assign({}, TUI_DIALOG_DEFAULT_OPTIONS), options),
        };
    }

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var SCROLLBAR_PLACEHOLDER = 17;
    var TuiDialogCloseService = /** @class */ (function (_super) {
        __extends(TuiDialogCloseService, _super);
        function TuiDialogCloseService(win, doc, el) {
            var _this = _super.call(this, function (subscriber) { return i5.merge(_this.click$, _this.esc$, _this.mousedown$).subscribe(subscriber); }) || this;
            _this.win = win;
            _this.doc = doc;
            _this.el = el;
            _this.click$ = i4.tuiTypedFromEvent(_this.element, "click").pipe(operators.filter(i4.tuiIsCurrentTarget));
            _this.esc$ = i4.tuiTypedFromEvent(_this.doc, "keydown").pipe(operators.filter(function (event) {
                var key = event.key;
                var target = i4.tuiGetActualTarget(event);
                return (key === "Escape" &&
                    i4.tuiIsElement(target) &&
                    (_this.element.contains(target) ||
                        !i4.tuiContainsOrAfter(_this.element, target)));
            }));
            _this.mousedown$ = i4.tuiTypedFromEvent(_this.doc, "mousedown").pipe(operators.filter(function (event) {
                var target = i4.tuiGetActualTarget(event);
                var clientX = event.clientX;
                return (i4.tuiIsElement(target) &&
                    utils.tuiGetViewportWidth(_this.win) - clientX > SCROLLBAR_PLACEHOLDER &&
                    !i4.tuiContainsOrAfter(_this.element, target));
            }), operators.switchMap(function () { return i4.tuiTypedFromEvent(_this.doc, "mouseup").pipe(operators.take(1), operators.map(i4.tuiGetActualTarget), operators.filter(function (target) { return i4.tuiIsElement(target) && !i4.tuiContainsOrAfter(_this.element, target); })); }));
            return _this;
        }
        Object.defineProperty(TuiDialogCloseService.prototype, "element", {
            get: function () {
                return this.el.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        return TuiDialogCloseService;
    }(i5.Observable));
    TuiDialogCloseService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogCloseService, deps: [{ token: common.WINDOW }, { token: i2.DOCUMENT }, { token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiDialogCloseService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogCloseService });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogCloseService, decorators: [{
                type: i0.Injectable
            }], ctorParameters: function () {
            return [{ type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }, { type: Document, decorators: [{
                            type: i0.Inject,
                            args: [i2.DOCUMENT]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        } });

    var REQUIRED_ERROR = new Error('Required dialog was dismissed');
    var TuiDialogComponent = /** @class */ (function () {
        function TuiDialogComponent(duration, isMobile, context, destroy$, dialogClose$, close$, closeWord$) {
            var _this = this;
            this.duration = duration;
            this.isMobile = isMobile;
            this.context = context;
            this.closeWord$ = closeWord$;
            this.animation = {
                value: '',
                params: {
                    start: '40px',
                    duration: this.duration,
                },
            };
            this.fullscreenAnimation = {
                value: '',
                params: {
                    start: '100vh',
                    duration: this.duration,
                },
            };
            this.close$ = new i5.Subject();
            i5.merge(this.close$.pipe(operators.switchMap(function () { return toObservable(context.closeable); })), dialogClose$.pipe(operators.switchMap(function () { return toObservable(context.dismissible); })), close$.pipe(operators.map(i4.ALWAYS_TRUE_HANDLER)))
                .pipe(operators.filter(Boolean), operators.takeUntil(destroy$))
                .subscribe(function () {
                _this.close();
            });
        }
        Object.defineProperty(TuiDialogComponent.prototype, "size", {
            get: function () {
                return this.context.size;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "header", {
            get: function () {
                return this.context.header;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "slideInTop", {
            get: function () {
                return this.fullscreen || this.isMobile
                    ? this.fullscreenAnimation
                    : this.animation;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "fullscreen", {
            get: function () {
                return !this.isMobile && (this.size === 'fullscreen' || this.size === 'page');
            },
            enumerable: false,
            configurable: true
        });
        TuiDialogComponent.prototype.close = function () {
            if (this.context.required) {
                this.context.$implicit.error(REQUIRED_ERROR);
            }
            else {
                this.context.$implicit.complete();
            }
        };
        return TuiDialogComponent;
    }());
    TuiDialogComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogComponent, deps: [{ token: tokens.TUI_ANIMATIONS_DURATION }, { token: i4.TUI_IS_MOBILE }, { token: i3.POLYMORPHEUS_CONTEXT }, { token: i4.TuiDestroyService, self: true }, { token: TuiDialogCloseService }, { token: TUI_DIALOGS_CLOSE }, { token: tokens.TUI_CLOSE_WORD }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiDialogComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiDialogComponent, selector: "tui-dialog", host: { properties: { "attr.data-size": "this.size", "class._centered": "this.header", "@tuiSlideInTop": "this.slideInTop", "@tuiFadeIn": "this.slideInTop" } }, providers: [i4.TuiDestroyService, TuiDialogCloseService], ngImport: i0__namespace, template: "<header\n    *ngIf=\"header\"\n    class=\"t-header\"\n>\n    <ng-container *polymorpheusOutlet=\"header as text; context: context\">\n        {{ text }}\n    </ng-container>\n</header>\n<div class=\"t-content\">\n    <h2\n        class=\"t-heading\"\n        [id]=\"context.id\"\n        [textContent]=\"context.label\"\n    ></h2>\n    <section>\n        <!-- TODO: Polymorpheus fix type -->\n        <ng-container *polymorpheusOutlet=\"$any(context.content) as text; context: context\">\n            <div [innerHTML]=\"text\"></div>\n            <div class=\"t-buttons\">\n                <button\n                    type=\"button\"\n                    tuiButton\n                    size=\"m\"\n                    (click)=\"context.$implicit.complete()\"\n                >\n                    {{ $any(context.data)?.button || 'OK' }}\n                </button>\n            </div>\n        </ng-container>\n    </section>\n</div>\n<div class=\"t-filler\"></div>\n<div\n    *ngIf=\"context.closeable\"\n    class=\"t-wrapper\"\n>\n    <button\n        type=\"button\"\n        tuiIconButton\n        tuiPreventDefault=\"mousedown\"\n        automation-id=\"tui-dialog__close\"\n        size=\"s\"\n        icon=\"tuiIconCloseLarge\"\n        shape=\"rounded\"\n        class=\"t-close\"\n        [appearance]=\"fullscreen ? 'secondary' : ''\"\n        [title]=\"closeWord$ | async\"\n        (click)=\"close$.next()\"\n    ></button>\n</div>\n", styles: [":host{position:relative;display:flex;font:var(--tui-font-text-m);flex-direction:column;box-sizing:border-box;margin:auto;border-radius:1rem;border:2.5rem solid transparent}:host:after{position:absolute;top:0;left:0;width:100%;height:100%;box-shadow:0 1.125rem 1.875rem #0000007a;content:\"\";border-radius:inherit;pointer-events:none}:host[data-size=auto]{width:auto}:host[data-size=s]{width:30rem}:host[data-size=s] .t-content{padding:1.5rem}:host[data-size=s] .t-heading{margin-bottom:.5rem;font:var(--tui-font-heading-5)}:host[data-size=m]{width:42.5rem}:host[data-size=m] .t-heading{margin-bottom:.75rem;font:var(--tui-font-heading-4)}:host[data-size=l]{width:55rem}:host[data-size=l] .t-heading{margin-bottom:1rem}:host[data-size=fullscreen],:host[data-size=page]{width:100%;min-height:100%;border-radius:0;border:none;background:var(--tui-elevation-01);box-shadow:0 4rem var(--tui-elevation-01)}:host[data-size=fullscreen] .t-content,:host[data-size=page] .t-content{padding:3rem calc(50vw - (45rem / 2))}:host[data-size=fullscreen] .t-heading,:host[data-size=page] .t-heading{margin-bottom:1rem}:host[data-size=page] .t-content{padding:0}:host._centered{text-align:center}:host :host-context(tui-root._mobile)[data-size]{min-width:100%;width:100%;max-width:100%;border-radius:0;border:none;margin:auto 0 0}:host :host-context(tui-root._mobile)[data-size] .t-content{padding:1.5rem}:host :host-context(tui-root._mobile)[data-size] .t-heading{margin-bottom:.5rem;font:var(--tui-font-heading-5)}.t-heading{margin:0;word-wrap:break-word;font:var(--tui-font-heading-3)}.t-heading:empty{display:none}.t-header{display:flex;border-top-left-radius:inherit;border-top-right-radius:inherit;overflow:hidden}:host[data-size=fullscreen] :host-context(tui-root._mobile) .t-header{flex:1}.t-content{border-radius:inherit;padding:2rem;background:var(--tui-elevation-01)}.t-content:not(:first-child){border-top-left-radius:0;border-top-right-radius:0}.t-wrapper{position:-webkit-sticky;position:sticky;top:0;order:-1}.t-filler{flex-grow:1}.t-close{transition-property:background;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:fixed;top:1.5rem;right:1.5rem;color:var(--tui-base-01)}:host:not([data-size=\"fullscreen\"]):not([data-size=\"page\"]) .t-close{animation:tuiFadeIn var(--tui-duration);background:rgba(104,104,104,.96)}:host:not([data-size=\"fullscreen\"]):not([data-size=\"page\"]) .t-close:hover{background:rgba(159,159,159,.86)}:host:not([data-size=\"fullscreen\"]).ng-animating .t-close{display:none}:host-context(tui-root._mobile) .t-close{position:absolute;top:0;right:0;background:transparent!important;color:var(--tui-base-06)}:host-context(tui-root._mobile) .t-close:hover{color:var(--tui-base-07)}.t-buttons{margin-top:1.25rem;text-align:right}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i4__namespace.TuiPreventDefaultDirective, selector: "[tuiPreventDefault]", inputs: ["tuiPreventDefault"] }], pipes: { "async": i2__namespace.AsyncPipe }, animations: [animations.tuiSlideInTop, animations.tuiFadeIn], changeDetection: i0__namespace.ChangeDetectionStrategy.Default });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-dialog',
                        templateUrl: './dialog.template.html',
                        styleUrls: ['./dialog.style.less'],
                        // So we don't force OnPush on dialog content
                        // eslint-disable-next-line @angular-eslint/prefer-on-push-component-change-detection
                        changeDetection: i0.ChangeDetectionStrategy.Default,
                        providers: [i4.TuiDestroyService, TuiDialogCloseService],
                        animations: [animations.tuiSlideInTop, animations.tuiFadeIn],
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ANIMATIONS_DURATION]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i4.TUI_IS_MOBILE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.POLYMORPHEUS_CONTEXT]
                        }] }, { type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i4.TuiDestroyService]
                        }, {
                            type: i0.Self
                        }] }, { type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TuiDialogCloseService]
                        }] }, { type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_DIALOGS_CLOSE]
                        }] }, { type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_CLOSE_WORD]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], header: [{
                    type: i0.HostBinding,
                    args: ['class._centered']
                }], slideInTop: [{
                    type: i0.HostBinding,
                    args: ['@tuiSlideInTop']
                }, {
                    type: i0.HostBinding,
                    args: ['@tuiFadeIn']
                }] } });
    function toObservable(valueOrStream) {
        return i5.isObservable(valueOrStream) ? valueOrStream : i5.of(valueOrStream);
    }

    var DIALOG = new i3.PolymorpheusComponent(TuiDialogComponent);
    var TuiDialogService = /** @class */ (function (_super) {
        __extends(TuiDialogService, _super);
        function TuiDialogService() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.component = DIALOG;
            _this.defaultOptions = Object.assign(Object.assign({}, i0.inject(TUI_DIALOG_OPTIONS)), { data: undefined });
            return _this;
        }
        return TuiDialogService;
    }(i4.AbstractTuiDialogService));
    TuiDialogService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogService, deps: null, target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiDialogService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: "root",
                    }]
            }] });

    var TuiDialogDirective = /** @class */ (function (_super) {
        __extends(TuiDialogDirective, _super);
        function TuiDialogDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return TuiDialogDirective;
    }(i4.AbstractTuiDialogDirective));
    TuiDialogDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiDialogDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDialogDirective, selector: "ng-template[tuiDialog]", inputs: { options: ["tuiDialogOptions", "options"], open: ["tuiDialog", "open"] }, outputs: { openChange: "tuiDialogChange" }, providers: [
            {
                provide: i4.AbstractTuiDialogService,
                useExisting: TuiDialogService,
            },
        ], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'ng-template[tuiDialog]',
                        providers: [
                            {
                                provide: i4.AbstractTuiDialogService,
                                useExisting: TuiDialogService,
                            },
                        ],
                        inputs: ['options: tuiDialogOptions', 'open: tuiDialog'],
                        outputs: ['openChange: tuiDialogChange'],
                    }]
            }] });

    var TuiDialogModule = /** @class */ (function () {
        function TuiDialogModule() {
        }
        return TuiDialogModule;
    }());
    TuiDialogModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiDialogModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogModule, declarations: [TuiDialogComponent, TuiDialogDirective], imports: [i3.PolymorpheusModule, i1.TuiButtonModule, i2.CommonModule, i4.TuiPreventDefaultModule], exports: [TuiDialogComponent, TuiDialogDirective] });
    TuiDialogModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogModule, providers: [i4.tuiAsDialog(TuiDialogService)], imports: [[i3.PolymorpheusModule, i1.TuiButtonModule, i2.CommonModule, i4.TuiPreventDefaultModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiDialogModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i3.PolymorpheusModule, i1.TuiButtonModule, i2.CommonModule, i4.TuiPreventDefaultModule],
                        declarations: [TuiDialogComponent, TuiDialogDirective],
                        exports: [TuiDialogComponent, TuiDialogDirective],
                        providers: [i4.tuiAsDialog(TuiDialogService)],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_DIALOGS_CLOSE = TUI_DIALOGS_CLOSE;
    exports.TUI_DIALOG_DEFAULT_OPTIONS = TUI_DIALOG_DEFAULT_OPTIONS;
    exports.TUI_DIALOG_OPTIONS = TUI_DIALOG_OPTIONS;
    exports.TuiDialogCloseService = TuiDialogCloseService;
    exports.TuiDialogComponent = TuiDialogComponent;
    exports.TuiDialogDirective = TuiDialogDirective;
    exports.TuiDialogModule = TuiDialogModule;
    exports.TuiDialogService = TuiDialogService;
    exports.tuiDialogOptionsProvider = tuiDialogOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
