// @taiga-ui/core/bundles/taiga-ui-core-pipes-format-date.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core/services')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/pipes/format-date', ['exports', '@angular/core', '@taiga-ui/core/services'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.pipes = global["taiga-ui"].core.pipes || {}, global["taiga-ui"].core.pipes["format-date"] = {}), global.ng.core, global["taiga-ui"].core.services));
})(this, (function (exports, i0, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiFormatDatePipe = /** @class */ (function () {
        function TuiFormatDatePipe(service) {
            this.service = service;
        }
        TuiFormatDatePipe.prototype.transform = function (timestampOrDate) {
            return this.service.format(timestampOrDate.valueOf());
        };
        return TuiFormatDatePipe;
    }());
    TuiFormatDatePipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipe, deps: [{ token: i1.TuiFormatDateService }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiFormatDatePipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipe, name: "tuiFormatDate" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: "tuiFormatDate",
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.TuiFormatDateService, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiFormatDateService]
                        }] }];
        } });

    var TuiFormatDatePipeModule = /** @class */ (function () {
        function TuiFormatDatePipeModule() {
        }
        return TuiFormatDatePipeModule;
    }());
    TuiFormatDatePipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFormatDatePipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipeModule, declarations: [TuiFormatDatePipe], exports: [TuiFormatDatePipe] });
    TuiFormatDatePipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatDatePipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFormatDatePipe],
                        exports: [TuiFormatDatePipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiFormatDatePipe = TuiFormatDatePipe;
    exports.TuiFormatDatePipeModule = TuiFormatDatePipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
