// @taiga-ui/core/bundles/taiga-ui-core-components-svg.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@angular/platform-browser'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('@taiga-ui/core/constants'), require('@taiga-ui/core/services'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/utils/miscellaneous'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/svg', ['exports', '@angular/common', '@angular/core', '@angular/platform-browser', '@ng-web-apis/common', '@taiga-ui/cdk', '@taiga-ui/core/constants', '@taiga-ui/core/services', '@taiga-ui/core/tokens', '@taiga-ui/core/utils/miscellaneous', 'rxjs', 'rxjs/operators', '@taiga-ui/core/utils'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components.svg = {}), global.ng.common, global.ng.core, global.ng.platformBrowser, global.common, global.i1, global["taiga-ui"].core.constants, global["taiga-ui"].core.services, global["taiga-ui"].core.tokens, global["taiga-ui"].core.utils.miscellaneous, global.rxjs, global.rxjs.operators, global["taiga-ui"].core.utils));
})(this, (function (exports, i2, i0, i4, common, i1, constants, i3, tokens, miscellaneous, rxjs, operators, utils) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    var TUI_DEPRECATED_ICONS = {
        tuiIconAttach: "tuiIconPaperclip",
        tuiIconAttention: "tuiIconAlertCircle",
        tuiIconBackCircle: "tuiIconArrowLeftCircle",
        tuiIconCall: "tuiIconPhone",
        tuiIconCallIn: "tuiIconPhoneIncoming",
        tuiIconCallOut: "tuiIconPhoneOutgoing",
        tuiIconCallTransfer: "tuiIconPhoneForwarded",
        tuiIconCancel: "tuiIconXCircle",
        tuiIconCancelCircle: "tuiIconSlash",
        tuiIconCard: "tuiIconCreditCard",
        tuiIconCards: "tuiIconCreditCard",
        tuiIconChartBar: "tuiIconBarChart",
        tuiIconChart: "tuiIconBarChart",
        tuiIconCheckList: "tuiIconClipboard",
        tuiIconCloseCircle: "tuiIconXCircle",
        tuiIconCollapse: "tuiIconMinimize",
        tuiIconComment: "tuiIconMessageSquare",
        tuiIconCommentFw: "tuiIconMessageForward",
        tuiIconCompany: "tuiIconBriefcase",
        tuiIconDefaultDoc: "tuiIconFile",
        tuiIconDesktop: "tuiIconMonitor",
        tuiIconDesktopOff: "tuiIconMonitorOff",
        tuiIconDislike: "tuiIconThumbsDown",
        tuiIconDone: "tuiIconCheckCircle",
        tuiIconDraft: "tuiIconEdit2",
        tuiIconEyeOpen: "tuiIconEye",
        tuiIconEyeClosed: "tuiIconEyeOff",
        tuiIconExpand: "tuiIconMaximize",
        tuiIconExternal: "tuiIconExternalLink",
        tuiIconFrame: "tuiIconLayout",
        tuiIconGeo: "tuiIconMapPin",
        tuiIconHeartFilled: "tuiIconHeart",
        tuiIconHide: "tuiIconEyeOff",
        tuiIconImg: "tuiIconImage",
        tuiIconInfoCircle: "tuiIconInfo",
        tuiIconLike: "tuiIconThumbsUp",
        tuiIconLockOpen: "tuiIconUnlock",
        tuiIconMobile: "tuiIconSmartphone",
        tuiIconMoreHor: "tuiIconMoreHorizontal",
        tuiIconMoreVer: "tuiIconMoreVertical",
        tuiIconMoreVert: "tuiIconMoreVertical",
        tuiIconPicture: "tuiIconImage",
        tuiIconPin: "tuiIconMapPin",
        tuiIconPrint: "tuiIconPrinter",
        tuiIconRefresh: "tuiIconRefreshCw",
        tuiIconRemove: "tuiIconX",
        tuiIconShow: "tuiIconEye",
        tuiIconSound: "tuiIconVolume",
        tuiIconSoundOff: "tuiIconVolumeX",
        tuiIconStarFilled: "tuiIconStar",
        tuiIconStop: "tuiIconSquare",
        tuiIconStructure: "tuiIconGrid",
        tuiIconTime: "tuiIconClock",
        tuiIconTooltip: "tuiIconHelpCircle",
        tuiIconViewList: "tuiIconList",
        tuiIconWifiOn: "tuiIconWifi",
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_SVG_DEFAULT_OPTIONS = {
        path: utils.tuiIconsPathFactory("assets/taiga-ui/icons"),
        srcProcessor: rxjs.identity,
        contentProcessor: i1.tuiSvgLinearGradientProcessor,
        deprecated: function (src) {
            var oldIcon = src.replace("Large", "").replace("Outline", "");
            var newIcon = TUI_DEPRECATED_ICONS[oldIcon];
            return newIcon
                ? oldIcon + "/(Large|Outline) is deprecated, use " + newIcon + "/(Large|Outline) instead"
                : "";
        },
    };
    /**
     * SVG component options
     */
    var TUI_SVG_OPTIONS = new i0.InjectionToken("[TUI_SVG_OPTIONS]", {
        factory: function () { return ({
            path: i0.inject(tokens.TUI_ICONS_PATH),
            deprecated: TUI_SVG_DEFAULT_OPTIONS.deprecated,
            srcProcessor: i0.inject(tokens.TUI_SVG_SRC_PROCESSOR),
            contentProcessor: i0.inject(tokens.TUI_SVG_CONTENT_PROCESSOR),
        }); },
    });
    var tuiSvgOptionsProvider = function (options) { return ({
        provide: TUI_SVG_OPTIONS,
        deps: [[new i0.SkipSelf(), new i0.Optional(), TUI_SVG_OPTIONS]],
        useFactory: function (fallback) { return (Object.assign(Object.assign(Object.assign({}, (fallback || TUI_SVG_DEFAULT_OPTIONS)), options), { path: i1.tuiIsString(options.path)
                ? utils.tuiIconsPathFactory(options.path)
                : options.path || (fallback === null || fallback === void 0 ? void 0 : fallback.path) || TUI_SVG_DEFAULT_OPTIONS.path })); },
    }); };

    var UNDEFINED_NAMED_ICON = 'Attempted to use undefined named icon';
    var MISSING_EXTERNAL_ICON = 'External icon is missing on the given URL';
    var FAILED_EXTERNAL_ICON = 'Failed to load external SVG';
    // TODO: Consider moving to CDK along with SvgService and SvgDefsHostComponent
    var TuiSvgComponent = /** @class */ (function () {
        function TuiSvgComponent(doc, win, options, tuiSanitizer, svgService, staticRequestService, sanitizer, el) {
            var _this = this;
            this.doc = doc;
            this.win = win;
            this.options = options;
            this.tuiSanitizer = tuiSanitizer;
            this.svgService = svgService;
            this.staticRequestService = staticRequestService;
            this.sanitizer = sanitizer;
            this.el = el;
            this.src$ = new rxjs.ReplaySubject(1);
            this.icon = '';
            this.innerHTML$ = this.src$.pipe(operators.switchMap(function () {
                if (i1.tuiIsString(_this.icon)) {
                    return _this.isExternal
                        ? _this.getExternalIcon(_this.icon)
                        : rxjs.of(_this.getSafeHtml(_this.icon));
                }
                return rxjs.of(_this.icon);
            }), operators.startWith(''));
        }
        Object.defineProperty(TuiSvgComponent.prototype, "src", {
            get: function () {
                return this.icon;
            },
            set: function (src) {
                var deprecated = this.options.deprecated(String(src));
                ngDevMode && i1.tuiAssert.assert(!deprecated, deprecated);
                this.icon = this.options.srcProcessor(src);
                this.src$.next();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "use", {
            get: function () {
                if (i1.tuiIsString(this.icon)) {
                    return this.icon.includes('.svg#')
                        ? this.icon
                        : this.resolveName(this.icon, this.options.path);
                }
                return '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isInnerHTML", {
            get: function () {
                return (!i1.tuiIsString(this.icon) ||
                    this.isSrc ||
                    this.isExternal ||
                    (this.isName && this.isShadowDOM));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isShadowDOM", {
            get: function () {
                return i1.tuiGetDocumentOrShadowRoot(this.el.nativeElement) !== this.doc;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isUse", {
            get: function () {
                return this.use.replace(constants.TUI_CACHE_BUSTING_PAYLOAD, '').includes('.svg#');
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isExternal", {
            get: function () {
                return this.isUrl || this.isCrossDomain;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isUrl", {
            get: function () {
                return i1.tuiIsString(this.icon) && this.icon.endsWith('.svg');
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isSrc", {
            get: function () {
                return i1.tuiIsString(this.icon) && miscellaneous.tuiIsPresumedHTMLString(this.icon);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isName", {
            get: function () {
                return !this.isUrl && !this.isUse && !this.isSrc;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isCrossDomain", {
            get: function () {
                var _a = this, use = _a.use, isUse = _a.isUse, win = _a.win;
                return (isUse && use.startsWith('http') && !!win.origin && !use.startsWith(win.origin));
            },
            enumerable: false,
            configurable: true
        });
        TuiSvgComponent.prototype.onError = function (message) {
            if (message === void 0) { message = MISSING_EXTERNAL_ICON; }
            var icon = this.icon;
            var event = new CustomEvent(constants.TUI_ICON_ERROR, {
                bubbles: true,
                detail: {
                    message: message,
                    icon: icon,
                },
            });
            ngDevMode && i1.tuiAssert.assert(false, message, icon);
            this.el.nativeElement.dispatchEvent(event);
        };
        TuiSvgComponent.prototype.resolveName = function (name, iconsPath) {
            return iconsPath(name);
        };
        TuiSvgComponent.prototype.getSafeHtml = function (src) {
            return this.isSrc ? this.sanitize(src) : this.process(src);
        };
        TuiSvgComponent.prototype.process = function (src) {
            var icon = this.svgService.getOriginal(src);
            if (this.isName && !icon && !!src) {
                this.onError(UNDEFINED_NAMED_ICON);
            }
            // Empty line for innerHTML when icon is shown through USE tag
            return !this.isShadowDOM || !this.isName ? '' : this.sanitize(icon || '');
        };
        TuiSvgComponent.prototype.sanitize = function (src) {
            src = this.options.contentProcessor(src);
            return this.tuiSanitizer && i1.tuiIsString(src)
                ? this.sanitizer.bypassSecurityTrustHtml(this.tuiSanitizer.sanitize(i0.SecurityContext.HTML, src) || '')
                : src;
        };
        TuiSvgComponent.prototype.getExternalIcon = function (src) {
            var _this = this;
            var url = src.includes('.svg') ? src : this.use;
            return this.staticRequestService.request(url).pipe(operators.catchError(function () {
                _this.onError(FAILED_EXTERNAL_ICON);
                return rxjs.of('');
            }), operators.map(function (response) { return _this.sanitize(response.replace('<svg', '<svg focusable="false"')); }));
        };
        return TuiSvgComponent;
    }());
    TuiSvgComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgComponent, deps: [{ token: i2.DOCUMENT }, { token: common.WINDOW }, { token: TUI_SVG_OPTIONS }, { token: tokens.TUI_SANITIZER, optional: true }, { token: i3.TuiSvgService }, { token: i1.TuiStaticRequestService }, { token: i4.DomSanitizer }, { token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSvgComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSvgComponent, selector: "tui-svg", inputs: { src: "src" }, ngImport: i0__namespace, template: "<ng-container *tuiLet=\"innerHTML$ | async as innerHTML\">\n    <div\n        *ngIf=\"isInnerHTML; else useTemplate\"\n        class=\"t-src\"\n        [innerHTML]=\"innerHTML\"\n    ></div>\n    <ng-template #useTemplate>\n        <svg\n            version=\"1.1\"\n            xmlns=\"http://www.w3.org/2000/svg\"\n            xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n            focusable=\"false\"\n            width=\"100%\"\n            height=\"100%\"\n            class=\"t-svg\"\n            (error)=\"onError()\"\n        >\n            <use [attr.xlink:href]=\"use\"></use>\n        </svg>\n    </ng-template>\n</ng-container>\n", styles: [":host{display:inline-flex;vertical-align:middle;flex-shrink:0;align-items:center;justify-content:center;line-height:0;height:1.5rem;width:1.5rem;fill:transparent;stroke:transparent;font-size:1rem}.t-src{display:flex;width:100%;height:100%;align-items:center;justify-content:center}.t-svg{overflow:visible}\n"], directives: [{ type: i1__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "async": i2__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i1.tuiRequiredSetter()
    ], TuiSvgComponent.prototype, "src", null);
    __decorate([
        i1.tuiPure
    ], TuiSvgComponent.prototype, "resolveName", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-svg',
                        templateUrl: './svg.template.html',
                        styleUrls: ['./svg.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: Document, decorators: [{
                            type: i0.Inject,
                            args: [i2.DOCUMENT]
                        }] }, { type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SVG_OPTIONS]
                        }] }, { type: i0__namespace.Sanitizer, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_SANITIZER]
                        }] }, { type: i3__namespace.TuiSvgService, decorators: [{
                            type: i0.Inject,
                            args: [i3.TuiSvgService]
                        }] }, { type: i1__namespace.TuiStaticRequestService, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiStaticRequestService]
                        }] }, { type: i4__namespace.DomSanitizer, decorators: [{
                            type: i0.Inject,
                            args: [i4.DomSanitizer]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { src: [{
                    type: i0.Input
                }], resolveName: [] } });

    var TuiSvgModule = /** @class */ (function () {
        function TuiSvgModule() {
        }
        return TuiSvgModule;
    }());
    TuiSvgModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSvgModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgModule, declarations: [TuiSvgComponent], imports: [i2.CommonModule, i1.TuiLetModule], exports: [TuiSvgComponent] });
    TuiSvgModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgModule, imports: [[i2.CommonModule, i1.TuiLetModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSvgModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i2.CommonModule, i1.TuiLetModule],
                        declarations: [TuiSvgComponent],
                        exports: [TuiSvgComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_DEPRECATED_ICONS = TUI_DEPRECATED_ICONS;
    exports.TUI_SVG_DEFAULT_OPTIONS = TUI_SVG_DEFAULT_OPTIONS;
    exports.TUI_SVG_OPTIONS = TUI_SVG_OPTIONS;
    exports.TuiSvgComponent = TuiSvgComponent;
    exports.TuiSvgModule = TuiSvgModule;
    exports.tuiSvgOptionsProvider = tuiSvgOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
