// @taiga-ui/kit/esm2015/components/badged-content/badged-content.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiBadgeModule } from '@taiga-ui/kit/components/badge';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiBadgedContentComponent } from './badged-content.component';
import * as i0 from "@angular/core";
export class TuiBadgedContentModule {
}
TuiBadgedContentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBadgedContentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, declarations: [TuiBadgedContentComponent], imports: [CommonModule, PolymorpheusModule, TuiSvgModule, TuiBadgeModule], exports: [TuiBadgedContentComponent] });
TuiBadgedContentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, imports: [[CommonModule, PolymorpheusModule, TuiSvgModule, TuiBadgeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiSvgModule, TuiBadgeModule],
                    declarations: [TuiBadgedContentComponent],
                    exports: [TuiBadgedContentComponent],
                }]
        }] });
