// @taiga-ui/kit/esm2015/components/radio-labeled/radio-labeled.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiFocusableModule, TuiFocusedModule } from '@taiga-ui/cdk';
import { TuiRadioModule } from '@taiga-ui/kit/components/radio';
import { TuiRadioLabeledComponent } from './radio-labeled.component';
import * as i0 from "@angular/core";
export class TuiRadioLabeledModule {
}
TuiRadioLabeledModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioLabeledModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRadioLabeledModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioLabeledModule, declarations: [TuiRadioLabeledComponent], imports: [CommonModule,
        FormsModule,
        TuiFocusableModule,
        TuiFocusedModule,
        TuiRadioModule], exports: [TuiRadioLabeledComponent] });
TuiRadioLabeledModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioLabeledModule, imports: [[
            CommonModule,
            FormsModule,
            TuiFocusableModule,
            TuiFocusedModule,
            TuiRadioModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioLabeledModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiFocusableModule,
                        TuiFocusedModule,
                        TuiRadioModule,
                    ],
                    declarations: [TuiRadioLabeledComponent],
                    exports: [TuiRadioLabeledComponent],
                }]
        }] });
