// @taiga-ui/cdk/esm2015/directives/pressed/pressed.directive.js
import { Directive, ElementRef, Inject, Output } from '@angular/core';
import { tuiPressedObservable } from '@taiga-ui/cdk/observables';
import { TUI_TAKE_ONLY_TRUSTED_EVENTS } from '@taiga-ui/cdk/tokens';
import * as i0 from "@angular/core";
export class TuiPressedDirective {
    constructor(el, takeOnlyTrustedEvents) {
        this.el = el;
        this.takeOnlyTrustedEvents = takeOnlyTrustedEvents;
        this.tuiPressedChange = tuiPressedObservable(this.el.nativeElement, {
            onlyTrusted: this.takeOnlyTrustedEvents,
        });
    }
}
TuiPressedDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPressedDirective, deps: [{ token: ElementRef }, { token: TUI_TAKE_ONLY_TRUSTED_EVENTS }], target: i0.ɵɵFactoryTarget.Directive });
TuiPressedDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPressedDirective, selector: "[tuiPressedChange]", outputs: { tuiPressedChange: "tuiPressedChange" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPressedDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPressedChange]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_TAKE_ONLY_TRUSTED_EVENTS]
                }] }]; }, propDecorators: { tuiPressedChange: [{
                type: Output
            }] } });
