// @taiga-ui/core/esm2015/components/expand/expand.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLoaderModule } from '@taiga-ui/core/components/loader';
import { TuiExpandComponent } from './expand.component';
import { TuiExpandContentDirective } from './expand-content.directive';
import * as i0 from "@angular/core";
export class TuiExpandModule {
}
TuiExpandModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiExpandModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiExpandModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiExpandModule, declarations: [TuiExpandComponent, TuiExpandContentDirective], imports: [CommonModule, TuiLoaderModule], exports: [TuiExpandComponent, TuiExpandContentDirective] });
TuiExpandModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiExpandModule, imports: [[CommonModule, TuiLoaderModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiExpandModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiLoaderModule],
                    declarations: [TuiExpandComponent, TuiExpandContentDirective],
                    exports: [TuiExpandComponent, TuiExpandContentDirective],
                }]
        }] });
