// @taiga-ui/core/esm2015/directives/mask-accessor/mask-accessor.directive.js
import { Directive, Optional } from '@angular/core';
import { DefaultValueAccessor, NG_VALUE_ACCESSOR, } from '@angular/forms';
import { TUI_VALUE_ACCESSOR } from '@taiga-ui/core/tokens';
import * as i0 from "@angular/core";
export class TuiMaskAccessorDirective {
}
TuiMaskAccessorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMaskAccessorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiMaskAccessorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiMaskAccessorDirective, selector: "input[tuiMaskAccessor]", providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            deps: [[new Optional(), TUI_VALUE_ACCESSOR], DefaultValueAccessor],
            multi: true,
            useFactory: (accessor, fallback) => (accessor ? accessor[0] : fallback),
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMaskAccessorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[tuiMaskAccessor]',
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            deps: [[new Optional(), TUI_VALUE_ACCESSOR], DefaultValueAccessor],
                            multi: true,
                            useFactory: (accessor, fallback) => (accessor ? accessor[0] : fallback),
                        },
                    ],
                }]
        }] });
