// @taiga-ui/core/esm2015/abstract/abstract-driver.directive.js
import { Directive, Inject, Self } from '@angular/core';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { merge } from 'rxjs';
import { distinctUntilChanged, takeUntil } from 'rxjs/operators';
import { TuiDriver } from './driver';
import { TuiVehicle } from './vehicle';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class AbstractTuiDriverDirective {
    constructor(destroy$, drivers, vehicles) {
        this.destroy$ = destroy$;
        this.drivers = drivers;
        this.vehicles = vehicles;
    }
    ngOnInit() {
        const vehicle = this.vehicles.find(({ type }) => type === this.type);
        merge(...this.drivers.filter(({ type }) => type === this.type))
            .pipe(distinctUntilChanged(), takeUntil(this.destroy$))
            .subscribe(value => {
            vehicle === null || vehicle === void 0 ? void 0 : vehicle.toggle(value);
        });
    }
}
AbstractTuiDriverDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiDriverDirective, deps: [{ token: TuiDestroyService, self: true }, { token: TuiDriver }, { token: TuiVehicle }], target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiDriverDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiDriverDirective, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiDriverDirective, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TuiDriver]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TuiVehicle]
                }] }]; } });
