// @taiga-ui/core/esm2015/abstract/abstract-textfield-host.js
import { Directive, Inject } from '@angular/core';
import { AbstractTuiControl } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class AbstractTuiTextfieldHost {
    constructor(host) {
        this.host = host;
    }
    get readOnly() {
        return this.host.readOnly;
    }
    get disabled() {
        return this.host.computedDisabled;
    }
    get invalid() {
        return this.host.computedInvalid;
    }
    get focusable() {
        return this.host.computedFocusable;
    }
    get inputMode() {
        return `text`;
    }
    get value() {
        var _a;
        return ((_a = this.host.value) === null || _a === void 0 ? void 0 : _a.toString()) || ``;
    }
    process(_input) { }
}
AbstractTuiTextfieldHost.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiTextfieldHost, deps: [{ token: AbstractTuiControl }], target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiTextfieldHost.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiTextfieldHost, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiTextfieldHost, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [AbstractTuiControl]
                }] }]; } });
