// @taiga-ui/kit/esm2015/components/calendar-range/calendar-range.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiMapperPipeModule, TuiPreventDefaultModule } from '@taiga-ui/cdk';
import { TuiCalendarModule, TuiDataListModule, TuiSvgModule } from '@taiga-ui/core';
import { TuiPrimitiveCalendarRangeModule } from '@taiga-ui/kit/internal/primitive-calendar-range';
import { TuiCalendarRangeComponent } from './calendar-range.component';
import * as i0 from "@angular/core";
export class TuiCalendarRangeModule {
}
TuiCalendarRangeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCalendarRangeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarRangeModule, declarations: [TuiCalendarRangeComponent], imports: [CommonModule,
        TuiMapperPipeModule,
        TuiPreventDefaultModule,
        TuiCalendarModule,
        TuiCalendarModule,
        TuiSvgModule,
        TuiDataListModule,
        TuiPrimitiveCalendarRangeModule], exports: [TuiCalendarRangeComponent] });
TuiCalendarRangeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarRangeModule, imports: [[
            CommonModule,
            TuiMapperPipeModule,
            TuiPreventDefaultModule,
            TuiCalendarModule,
            TuiCalendarModule,
            TuiSvgModule,
            TuiDataListModule,
            TuiPrimitiveCalendarRangeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiMapperPipeModule,
                        TuiPreventDefaultModule,
                        TuiCalendarModule,
                        TuiCalendarModule,
                        TuiSvgModule,
                        TuiDataListModule,
                        TuiPrimitiveCalendarRangeModule,
                    ],
                    declarations: [TuiCalendarRangeComponent],
                    exports: [TuiCalendarRangeComponent],
                }]
        }] });
