// @taiga-ui/kit/bundles/taiga-ui-kit-directives-value-accessor.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/kit/providers')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/kit/directives/value-accessor', ['exports', '@angular/core', '@taiga-ui/kit/providers'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.directives = global["taiga-ui"].kit.directives || {}, global["taiga-ui"].kit.directives["value-accessor"] = {}), global.ng.core, global["taiga-ui"].kit.providers));
})(this, (function (exports, i0, providers) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

        var TuiValueAccessorDirective = /** @class */ (function () {
            function TuiValueAccessorDirective() {
            }
            return TuiValueAccessorDirective;
        }());
        TuiValueAccessorDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValueAccessorDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiValueAccessorDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiValueAccessorDirective, selector: "[tuiValueAccessor]", providers: [providers.TUI_VALUE_ACCESSOR_PROVIDER], ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValueAccessorDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiValueAccessor]',
                            providers: [providers.TUI_VALUE_ACCESSOR_PROVIDER],
                        }]
                }] });

        var TuiValueAccessorModule = /** @class */ (function () {
            function TuiValueAccessorModule() {
            }
            return TuiValueAccessorModule;
        }());
        TuiValueAccessorModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValueAccessorModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiValueAccessorModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValueAccessorModule, declarations: [TuiValueAccessorDirective], exports: [TuiValueAccessorDirective] });
        TuiValueAccessorModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValueAccessorModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValueAccessorModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiValueAccessorDirective],
                            exports: [TuiValueAccessorDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiValueAccessorDirective = TuiValueAccessorDirective;
        exports.TuiValueAccessorModule = TuiValueAccessorModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
