// @taiga-ui/kit/bundles/taiga-ui-kit-components-slider.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/utils'), require('@ng-web-apis/common'), require('rxjs/operators'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/slider', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/kit/utils', '@ng-web-apis/common', 'rxjs/operators', '@angular/common', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.slider = {}), global.ng.core, global.ng.forms, global.cdk, global["taiga-ui"].kit.utils, global.common, global.rxjs.operators, global.ng.common, global.rxjs));
})(this, (function (exports, i0, i1, cdk, utils, common, operators, i1$2, i1$1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i1__namespace$2 = /*#__PURE__*/_interopNamespace(i1$2);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_SLIDER_DEFAULT_OPTIONS = {
        size: "m",
        trackColor: "var(--tui-base-03)",
    };
    /**
     * Default parameters for Slider component
     */
    var TUI_SLIDER_OPTIONS = new i0.InjectionToken("[TUI_SLIDER_OPTIONS]", { factory: function () { return TUI_SLIDER_DEFAULT_OPTIONS; } });
    function tuiSliderOptionsProvider(options) {
        return {
            provide: TUI_SLIDER_OPTIONS,
            useValue: Object.assign(Object.assign({}, TUI_SLIDER_DEFAULT_OPTIONS), options),
        };
    }

    var TuiSliderComponent = /** @class */ (function () {
        function TuiSliderComponent(control, cdr, options, el, userAgent, injector) {
            var _a;
            this.control = control;
            this.options = options;
            this.el = el;
            this.userAgent = userAgent;
            this.injector = injector;
            this.size = this.options.size;
            this.segments = 1;
            if (control instanceof i1.NgModel) {
                /**
                 * The ValueAccessor.writeValue method is called twice on any value accessor during component initialization,
                 * when a control is bound using [(ngModel)], first time with a phantom null value.
                 * With `changeDetection: ChangeDetectionStrategy.OnPush` the second call of writeValue with real value don't re-render the view.
                 * ___
                 * See this {@link https://github.com/angular/angular/issues/14988 issue}
                 */
                (_a = control.valueChanges) === null || _a === void 0 ? void 0 : _a.pipe(cdk.tuiWatch(cdr), operators.take(1)).subscribe();
            }
        }
        Object.defineProperty(TuiSliderComponent.prototype, "min", {
            get: function () {
                return Number(this.el.nativeElement.min);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "max", {
            get: function () {
                return Number(this.el.nativeElement.max || 100);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "step", {
            get: function () {
                return Number(this.el.nativeElement.step) || 1;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "value", {
            get: function () {
                var _b = this, el = _b.el, control = _b.control, hasKeySteps = _b.hasKeySteps;
                if (!hasKeySteps && control instanceof i1.NgModel) {
                    /**
                     * If developer uses `[(ngModel)]` and programmatically change value,
                     * the `el.nativeElement.value` is equal to the previous value at this moment.
                     */
                    return control.viewModel;
                }
                return Number(el.nativeElement.value) || 0;
            },
            set: function (newValue) {
                this.el.nativeElement.value = "" + newValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "valuePercentage", {
            get: function () {
                return (100 * (this.value - this.min)) / (this.max - this.min) || 0;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "segmentWidth", {
            get: function () {
                return 100 / Math.max(1, this.segments);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "isOldEdge", {
            // TODO: drop support of legacy Edge (EdgeHTML) in v4.x
            get: function () {
                return cdk.tuiIsEdgeOlderThan(cdk.CHROMIUM_EDGE_START_VERSION, this.userAgent);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "hasKeySteps", {
            get: function () {
                return Boolean(this.injector.get(TuiSliderKeyStepsDirective, null));
            },
            enumerable: false,
            configurable: true
        });
        return TuiSliderComponent;
    }());
    TuiSliderComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderComponent, deps: [{ token: i1.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: TUI_SLIDER_OPTIONS }, { token: i0.ElementRef }, { token: common.USER_AGENT }, { token: i0.Injector }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSliderComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSliderComponent, selector: "input[type=range][tuiSlider]", inputs: { size: "size", segments: "segments" }, host: { listeners: { "input": "0" }, properties: { "style.--tui-slider-track-color": "options.trackColor", "attr.data-size": "this.size", "style.--tui-slider-fill-percentage.%": "this.valuePercentage", "style.--tui-slider-segment-width.%": "this.segmentWidth", "class._old-edge": "this.isOldEdge" } }, ngImport: i0__namespace, template: '', isInline: true, styles: [":host{display:block;width:100%;color:var(--tui-primary);cursor:pointer}:host:active{cursor:ew-resize}:host:disabled{opacity:var(--tui-disabled-opacity);cursor:auto}:host:not(._old-edge){-webkit-appearance:none;-moz-appearance:none;appearance:none;height:.125rem;padding:.4375rem 0;background-color:transparent;background-clip:content-box;outline:none;border-radius:var(--tui-radius-m)}:host:not(._old-edge)::-webkit-slider-container{border-radius:inherit}:host:not(._old-edge)[data-size=m]::-webkit-slider-runnable-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0px .625rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width)),linear-gradient(to right,currentColor var(--tui-slider-fill-percentage),transparent var(--tui-slider-fill-percentage));background-position-x:0,.375rem,0;background-size:calc(100% - 1rem),calc(100% - 1rem),auto}:host:not(._old-edge)[data-size=s]::-webkit-slider-runnable-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0px .375rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width)),linear-gradient(to right,currentColor var(--tui-slider-fill-percentage),transparent var(--tui-slider-fill-percentage));background-position-x:0,.125rem,0;background-size:calc(100% - .5rem),calc(100% - .5rem),auto}:host:not(._old-edge)[data-size=m]::-moz-range-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0px .625rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width));background-position-x:0,.375rem;background-size:calc(100% - 1rem)}:host:not(._old-edge)[data-size=s]::-moz-range-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0px .375rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width));background-position-x:0,.125rem;background-size:calc(100% - .5rem)}:host:not(._old-edge)[data-size=m]::-webkit-slider-thumb{-webkit-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:1rem;width:1rem;margin-top:-.4375rem}:not(:disabled):host:not(._old-edge)[data-size=m]::-webkit-slider-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size=m]::-webkit-slider-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size=m]::-webkit-slider-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size=m]::-webkit-slider-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)[data-size=s]::-webkit-slider-thumb{-webkit-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:.5rem;width:.5rem;margin-top:-.1875rem}:not(:disabled):host:not(._old-edge)[data-size=s]::-webkit-slider-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size=s]::-webkit-slider-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size=s]::-webkit-slider-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size=s]::-webkit-slider-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)[data-size=m]::-moz-range-thumb{-moz-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:1rem;width:1rem}:not(:disabled):host:not(._old-edge)[data-size=m]::-moz-range-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size=m]::-moz-range-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size=m]::-moz-range-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size=m]::-moz-range-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)[data-size=s]::-moz-range-thumb{-moz-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:.5rem;width:.5rem}:not(:disabled):host:not(._old-edge)[data-size=s]::-moz-range-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size=s]::-moz-range-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size=s]::-moz-range-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size=s]::-moz-range-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)::-moz-range-progress{border-radius:inherit}:host:not(._old-edge)::-moz-range-progress{height:.125rem;background:currentColor;border-top-right-radius:0;border-bottom-right-radius:0}:host._old-edge::-ms-thumb{background:currentColor;border-radius:50%}:host._old-edge::-ms-fill-lower{background:currentColor}:host._old-edge::-ms-track{background:var(--tui-slider-track-color);border:none}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiSliderComponent.prototype, "size", void 0);
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiSliderComponent.prototype, "segments", void 0);
    __decorate([
        cdk.tuiPure
    ], TuiSliderComponent.prototype, "hasKeySteps", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderComponent, decorators: [{
                type: i0.Component,
                args: [{
                        /**
                         * We have to call our component as `<input tuiSlider type="range" ... />`
                         * because otherwise built-in angular
                         * {@link https://github.com/angular/angular/blob/master/packages/forms/src/directives/range_value_accessor.ts#L45 RangeValueAccessor}
                         * cannot be matched by its CSS selector.
                         */
                        selector: 'input[type=range][tuiSlider]',
                        template: '',
                        styleUrls: ['./slider.style.less'],
                        host: {
                            /**
                             * For change detection.
                             * Webkit does not have built-in method for customization of filling progress (as Firefox).
                             * We draw filling of progress by `background: linear-gradient(...)` of the track.
                             * This function triggers change detection (for {@link valuePercentage} function) when we drag thumb of the input.
                             */
                            '(input)': '0',
                            '[style.--tui-slider-track-color]': 'options.trackColor',
                        },
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SLIDER_OPTIONS]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [common.USER_AGENT]
                        }] }, { type: i0__namespace.Injector, decorators: [{
                            type: i0.Inject,
                            args: [i0.Injector]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], segments: [{
                    type: i0.Input
                }], valuePercentage: [{
                    type: i0.HostBinding,
                    args: ['style.--tui-slider-fill-percentage.%']
                }], segmentWidth: [{
                    type: i0.HostBinding,
                    args: ['style.--tui-slider-segment-width.%']
                }], isOldEdge: [{
                    type: i0.HostBinding,
                    args: ['class._old-edge']
                }], hasKeySteps: [] } });

    var TuiSliderKeyStepsDirective = /** @class */ (function (_super) {
        __extends(TuiSliderKeyStepsDirective, _super);
        function TuiSliderKeyStepsDirective(control, cdr, el, slider) {
            var _this = _super.call(this, control, cdr) || this;
            _this.el = el;
            _this.slider = slider;
            return _this;
        }
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "nativeFocusableElement", {
            get: function () {
                return this.computedDisabled ? null : this.el.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "focused", {
            get: function () {
                return cdk.tuiIsNativeFocused(this.nativeFocusableElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "min", {
            get: function () {
                return this.keySteps[0][1];
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "max", {
            get: function () {
                return this.keySteps[this.keySteps.length - 1][1];
            },
            enumerable: false,
            configurable: true
        });
        TuiSliderKeyStepsDirective.prototype.updateControlValue = function () {
            this.value = utils.tuiPercentageToKeyStepValue(this.slider.valuePercentage, this.keySteps);
        };
        TuiSliderKeyStepsDirective.prototype.writeValue = function (controlValue) {
            if (controlValue === null) {
                return;
            }
            var clampedControlValue = cdk.tuiClamp(controlValue, this.min, this.max);
            ngDevMode &&
                cdk.tuiAssert.assert(controlValue === clampedControlValue, '\n[SliderKeySteps]: You cannot programmatically set value which is less/more than min/max');
            this.slider.value = this.transformToNativeValue(clampedControlValue);
        };
        TuiSliderKeyStepsDirective.prototype.getFallbackValue = function () {
            return 0;
        };
        TuiSliderKeyStepsDirective.prototype.transformToNativeValue = function (controlValue) {
            var _a = this.slider, min = _a.min, max = _a.max;
            var newValuePercentage = utils.tuiKeyStepValueToPercentage(controlValue, this.keySteps);
            return (newValuePercentage * (max - min)) / 100 + min;
        };
        return TuiSliderKeyStepsDirective;
    }(cdk.AbstractTuiControl));
    TuiSliderKeyStepsDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderKeyStepsDirective, deps: [{ token: i1.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i0.ElementRef }, { token: i0.forwardRef(function () { return TuiSliderComponent; }) }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSliderKeyStepsDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSliderKeyStepsDirective, selector: "input[tuiSlider][keySteps]", inputs: { keySteps: "keySteps" }, host: { listeners: { "input": "updateControlValue()", "change": "updateControlValue()" }, properties: { "attr.aria-valuenow": "safeCurrentValue", "attr.aria-valuemin": "min", "attr.aria-valuemax": "max" } }, usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderKeyStepsDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'input[tuiSlider][keySteps]',
                        host: {
                            '[attr.aria-valuenow]': 'safeCurrentValue',
                            '[attr.aria-valuemin]': 'min',
                            '[attr.aria-valuemax]': 'max',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: TuiSliderComponent, decorators: [{
                            type: i0.Inject,
                            args: [i0.forwardRef(function () { return TuiSliderComponent; })]
                        }] }];
        }, propDecorators: { keySteps: [{
                    type: i0.Input
                }], updateControlValue: [{
                    type: i0.HostListener,
                    args: ['input']
                }, {
                    type: i0.HostListener,
                    args: ['change']
                }] } });

    var SLIDER_INTERACTION_KEYS = new Set([
        'ArrowLeft',
        'ArrowRight',
        'ArrowUp',
        'ArrowDown',
        'Home',
        'End',
        'PageUp',
        'PageDown',
    ]);
    /**
     * Native <input type='range' readonly> doesn't work.
     * This directive imitates this native behaviour.
     */
    var TuiSliderReadonlyDirective = /** @class */ (function () {
        function TuiSliderReadonlyDirective(el, doc, destroy$) {
            var _this = this;
            this.readonly = true;
            var touchStart$ = cdk.tuiTypedFromEvent(el.nativeElement, 'touchstart', {
                passive: false,
            });
            var touchMove$ = cdk.tuiTypedFromEvent(doc, 'touchmove', {
                passive: false,
            });
            var touchEnd$ = cdk.tuiTypedFromEvent(doc, 'touchend', {
                passive: true,
            });
            var shouldPreventMove$ = i1$1.merge(touchStart$.pipe(operators.tap(function (e) { return _this.preventEvent(e); }), operators.map(cdk.ALWAYS_TRUE_HANDLER)), touchEnd$.pipe(operators.map(cdk.ALWAYS_FALSE_HANDLER)));
            /**
             * @bad TODO think about another solution.
             * Keep in mind that preventing touch event (on slider) inside `@HostListener('touchstart')` doesn't work for mobile chrome.
             */
            i1$1.combineLatest([touchMove$, shouldPreventMove$])
                .pipe(operators.filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], shouldPreventMove = _b[1];
                return shouldPreventMove;
            }), operators.takeUntil(destroy$))
                .subscribe(function (_a) {
                var _b = __read(_a, 1), moveEvent = _b[0];
                return _this.preventEvent(moveEvent);
            });
        }
        TuiSliderReadonlyDirective.prototype.preventEvent = function (event) {
            if (event.cancelable && cdk.tuiCoerceBooleanProperty(this.readonly)) {
                event.preventDefault();
            }
        };
        TuiSliderReadonlyDirective.prototype.preventKeyboardInteraction = function (event) {
            if (SLIDER_INTERACTION_KEYS.has(event.key)) {
                this.preventEvent(event);
            }
        };
        return TuiSliderReadonlyDirective;
    }());
    TuiSliderReadonlyDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderReadonlyDirective, deps: [{ token: i0.ElementRef }, { token: i1$2.DOCUMENT }, { token: cdk.TuiDestroyService, self: true }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSliderReadonlyDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSliderReadonlyDirective, selector: "input[tuiSlider][readonly]", inputs: { readonly: "readonly" }, host: { listeners: { "mousedown": "preventEvent($event)", "keydown": "preventKeyboardInteraction($event)" } }, providers: [cdk.TuiDestroyService], ngImport: i0__namespace });
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiSliderReadonlyDirective.prototype, "readonly", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderReadonlyDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'input[tuiSlider][readonly]',
                        providers: [cdk.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: Document, decorators: [{
                            type: i0.Inject,
                            args: [i1$2.DOCUMENT]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [cdk.TuiDestroyService]
                        }] }];
        }, propDecorators: { readonly: [{
                    type: i0.Input
                }], preventEvent: [{
                    type: i0.HostListener,
                    args: ['mousedown', ['$event']]
                }], preventKeyboardInteraction: [{
                    type: i0.HostListener,
                    args: ['keydown', ['$event']]
                }] } });

    var TuiSliderThumbLabelComponent = /** @class */ (function () {
        function TuiSliderThumbLabelComponent() {
        }
        Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "size", {
            get: function () {
                var _a;
                return ((_a = this.slider) === null || _a === void 0 ? void 0 : _a.size) || 'm';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "ratio", {
            get: function () {
                var _a;
                return (((_a = this.slider) === null || _a === void 0 ? void 0 : _a.valuePercentage) || 0) / 100;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "ghostLeft", {
            get: function () {
                var _a;
                return this.ratio * (((_a = this.slider) === null || _a === void 0 ? void 0 : _a.el.nativeElement.offsetWidth) || 0);
            },
            enumerable: false,
            configurable: true
        });
        TuiSliderThumbLabelComponent.prototype.ngAfterContentInit = function () {
            var _a;
            ngDevMode &&
                cdk.tuiAssert.assert(Boolean((_a = this.control) === null || _a === void 0 ? void 0 : _a.valueChanges), '\n[tuiSliderThumbLabel] expected <input tuiSlider type="range" /> to use Angular Forms.\n' +
                    'Use [(ngModel)] or [formControl] or formControlName for correct work.');
        };
        return TuiSliderThumbLabelComponent;
    }());
    TuiSliderThumbLabelComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderThumbLabelComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSliderThumbLabelComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSliderThumbLabelComponent, selector: "[tuiSliderThumbLabel]", queries: [{ propertyName: "slider", first: true, predicate: TuiSliderComponent, descendants: true }, { propertyName: "control", first: true, predicate: i1.NgControl, descendants: true }], ngImport: i0__namespace, template: "<ng-container *ngIf=\"control?.valueChanges | async\"></ng-container>\n\n<div\n    class=\"t-ghost\"\n    [attr.data-size]=\"size\"\n    [style.left.px]=\"ghostLeft\"\n    [style.--tui-slider-thumb-ratio]=\"ratio\"\n>\n    <ng-content></ng-content>\n</div>\n\n<ng-content select=\"input[type=range]\"></ng-content>\n", styles: [":host{position:relative}.t-ghost{position:absolute;top:0;bottom:0;margin:auto;border-radius:50%;pointer-events:none}.t-ghost[data-size=s]{width:.5rem;height:.5rem;transform:translate(calc(var(--tui-slider-thumb-ratio) * -.5rem))}.t-ghost[data-size=m]{width:1rem;height:1rem;transform:translate(calc(var(--tui-slider-thumb-ratio) * -1rem))}\n"], directives: [{ type: i1__namespace$2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "async": i1__namespace$2.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderThumbLabelComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: '[tuiSliderThumbLabel]',
                        templateUrl: './slider-thumb-label.template.html',
                        styleUrls: ['./slider-thumb-label.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { slider: [{
                    type: i0.ContentChild,
                    args: [TuiSliderComponent]
                }], control: [{
                    type: i0.ContentChild,
                    args: [i1.NgControl]
                }] } });

    var TuiSliderModule = /** @class */ (function () {
        function TuiSliderModule() {
        }
        return TuiSliderModule;
    }());
    TuiSliderModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSliderModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderModule, declarations: [TuiSliderComponent,
            TuiSliderThumbLabelComponent,
            TuiSliderKeyStepsDirective,
            TuiSliderReadonlyDirective], imports: [i1$2.CommonModule], exports: [TuiSliderComponent,
            TuiSliderThumbLabelComponent,
            TuiSliderKeyStepsDirective,
            TuiSliderReadonlyDirective] });
    TuiSliderModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderModule, imports: [[i1$2.CommonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSliderModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1$2.CommonModule],
                        declarations: [
                            TuiSliderComponent,
                            TuiSliderThumbLabelComponent,
                            TuiSliderKeyStepsDirective,
                            TuiSliderReadonlyDirective,
                        ],
                        exports: [
                            TuiSliderComponent,
                            TuiSliderThumbLabelComponent,
                            TuiSliderKeyStepsDirective,
                            TuiSliderReadonlyDirective,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_SLIDER_DEFAULT_OPTIONS = TUI_SLIDER_DEFAULT_OPTIONS;
    exports.TUI_SLIDER_OPTIONS = TUI_SLIDER_OPTIONS;
    exports.TuiSliderComponent = TuiSliderComponent;
    exports.TuiSliderKeyStepsDirective = TuiSliderKeyStepsDirective;
    exports.TuiSliderModule = TuiSliderModule;
    exports.TuiSliderReadonlyDirective = TuiSliderReadonlyDirective;
    exports.TuiSliderThumbLabelComponent = TuiSliderThumbLabelComponent;
    exports.tuiSliderOptionsProvider = tuiSliderOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
