// @taiga-ui/kit/bundles/taiga-ui-kit-components-push.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@angular/common'), require('rxjs'), require('@tinkoff/ng-polymorpheus'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/push', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@angular/common', 'rxjs', '@tinkoff/ng-polymorpheus', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.push = {}), global.ng.core, global.cdk, global.i1, global.ng.common, global.rxjs, global.i5, global.rxjs.operators));
})(this, (function (exports, i0, cdk, i1, i2, i3, i5, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiPushComponent = /** @class */ (function () {
        function TuiPushComponent(closeWord$) {
            this.closeWord$ = closeWord$;
            this.heading = '';
            this.type = '';
            this.timestamp = 0;
            this.close = new i0.EventEmitter();
        }
        Object.defineProperty(TuiPushComponent.prototype, "closeable", {
            get: function () {
                return cdk.tuiIsObserved(this.close);
            },
            enumerable: false,
            configurable: true
        });
        return TuiPushComponent;
    }());
    TuiPushComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushComponent, deps: [{ token: i1.TUI_CLOSE_WORD }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPushComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPushComponent, selector: "tui-push", inputs: { heading: "heading", type: "type", timestamp: "timestamp" }, outputs: { close: "close" }, ngImport: i0__namespace, template: "<div class=\"t-image\">\n    <ng-content select=\"img\"></ng-content>\n</div>\n<button\n    *ngIf=\"closeable\"\n    type=\"button\"\n    tuiIconButton\n    icon=\"tuiIconCloseLarge\"\n    size=\"xs\"\n    appearance=\"\"\n    shape=\"rounded\"\n    class=\"t-close\"\n    [title]=\"closeWord$ | async\"\n    (click)=\"close.emit()\"\n></button>\n<div class=\"t-top\">\n    <span class=\"t-icon\">\n        <ng-content select=\"tui-svg\"></ng-content>\n    </span>\n    {{ type }}\n    <span\n        *ngIf=\"timestamp\"\n        class=\"t-time\"\n        [textContent]=\"timestamp | tuiFormatDate | async\"\n    ></span>\n</div>\n<h3\n    automation-id=\"tui-push__heading\"\n    class=\"t-heading\"\n>\n    {{ heading }}\n</h3>\n<div class=\"t-content\">\n    <ng-content></ng-content>\n</div>\n<div class=\"t-actions\">\n    <span class=\"t-button\">\n        <ng-content select=\"[tuiButton]\"></ng-content>\n    </span>\n    <span class=\"t-link\">\n        <ng-content select=\"[tuiLink]\"></ng-content>\n    </span>\n</div>\n", styles: [":host{box-shadow:0 .25rem 1.5rem #0000001f;position:relative;display:block;width:22.5rem;max-width:100%;box-sizing:border-box;padding:1.25rem 1.25rem .25rem;overflow:hidden;font:var(--tui-font-text-m);border-radius:var(--tui-radius-xl);background:var(--tui-elevation-02)}.t-image{display:flex;flex-direction:column;max-height:10.625rem;overflow:hidden;margin:-1.25rem -1.25rem 1.25rem}.t-close{position:absolute;top:.75rem;right:.75rem;color:#0000008a;background:rgba(242,242,242,.32);-webkit-backdrop-filter:blur(1rem);backdrop-filter:blur(1rem)}.t-top{display:flex;align-items:center;font:var(--tui-font-text-s);color:var(--tui-text-02)}.t-icon{height:1.25rem;margin-right:.5rem;transform:scale(.833);transform-origin:top left;color:var(--tui-link)}.t-time:before{content:\"\\2022\";display:inline-block;width:1rem;text-align:center}.t-heading{line-height:1.25rem;font-weight:bold;margin:.75rem 0 .25rem}.t-heading,.t-content{display:-webkit-box;-webkit-box-orient:vertical;-webkit-line-clamp:2;word-break:break-word;overflow:hidden}.t-actions{display:flex;align-items:center;font:var(--tui-font-text-s)}.t-button:not(:empty){margin:.75rem .75rem .75rem 0}.t-link{margin:.75rem 0}.t-link:empty{margin:.5rem}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "async": i2__namespace.AsyncPipe, "tuiFormatDate": i1__namespace.TuiFormatDatePipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiPushComponent.prototype, "heading", void 0);
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiPushComponent.prototype, "type", void 0);
    __decorate([
        cdk.tuiDefaultProp()
    ], TuiPushComponent.prototype, "timestamp", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-push',
                        templateUrl: './push.template.html',
                        styleUrls: ['./push.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i3__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_CLOSE_WORD]
                        }] }];
        }, propDecorators: { heading: [{
                    type: i0.Input
                }], type: [{
                    type: i0.Input
                }], timestamp: [{
                    type: i0.Input
                }], close: [{
                    type: i0.Output
                }] } });

    var TuiPushDirective = /** @class */ (function (_super) {
        __extends(TuiPushDirective, _super);
        function TuiPushDirective(modeDirective) {
            var _this = this;
            var _a;
            _this = _super.call(this) || this;
            _this.modeDirective = modeDirective;
            _this.size = 's';
            _this.shape = null;
            _this.change$ = ((_a = _this.modeDirective) === null || _a === void 0 ? void 0 : _a.change$) || new i3.Subject();
            return _this;
        }
        Object.defineProperty(TuiPushDirective.prototype, "appearance", {
            get: function () {
                var _a;
                return ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.mode) === 'onDark' ? 'accent' : 'secondary';
            },
            enumerable: false,
            configurable: true
        });
        return TuiPushDirective;
    }(i1.TuiModeDirective));
    TuiPushDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushDirective, deps: [{ token: i1.TuiModeDirective, optional: true, skipSelf: true }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPushDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPushDirective, selector: "tui-push", providers: [
            {
                provide: i1.TuiModeDirective,
                useExisting: TuiPushDirective,
            },
            {
                provide: i1.TUI_BUTTON_OPTIONS,
                useExisting: TuiPushDirective,
            },
        ], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-push',
                        providers: [
                            {
                                provide: i1.TuiModeDirective,
                                useExisting: TuiPushDirective,
                            },
                            {
                                provide: i1.TUI_BUTTON_OPTIONS,
                                useExisting: TuiPushDirective,
                            },
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.TuiModeDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.SkipSelf
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiModeDirective]
                        }] }];
        } });

    var TUI_PUSH_DEFAULT_OPTIONS = {
        heading: "",
        type: "",
        timestamp: 0,
        image: "",
        icon: "",
        iconColor: "",
        buttons: [],
    };
    /**
     * Default parameters for Push component
     */
    var TUI_PUSH_OPTIONS = new i0.InjectionToken("[TUI_PUSH_OPTIONS]", {
        factory: function () { return TUI_PUSH_DEFAULT_OPTIONS; },
    });
    function tuiPushOptionsProvider(options) {
        return {
            provide: TUI_PUSH_OPTIONS,
            useValue: Object.assign(Object.assign({}, TUI_PUSH_DEFAULT_OPTIONS), options),
        };
    }

    var TuiPushAlertDirective = /** @class */ (function (_super) {
        __extends(TuiPushAlertDirective, _super);
        function TuiPushAlertDirective(template, cdr, destroy$, push) {
            var _this = _super.call(this, template, cdr) || this;
            _this.show$ = new i3.Subject();
            _this.show$
                .pipe(cdk.tuiIfMap(function () { return push.open(_this); }), operators.takeUntil(destroy$))
                .subscribe();
            return _this;
        }
        Object.defineProperty(TuiPushAlertDirective.prototype, "tuiPush", {
            set: function (show) {
                this.show$.next(show);
            },
            enumerable: false,
            configurable: true
        });
        return TuiPushAlertDirective;
    }(i5.PolymorpheusTemplate));
    TuiPushAlertDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushAlertDirective, deps: [{ token: i0.TemplateRef }, { token: i0.ChangeDetectorRef }, { token: cdk.TuiDestroyService, self: true }, { token: i0.forwardRef(function () { return TuiPushService; }) }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPushAlertDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPushAlertDirective, selector: "[tuiPush]", inputs: { tuiPush: "tuiPush" }, providers: [cdk.TuiDestroyService], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushAlertDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPush]',
                        providers: [cdk.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.TemplateRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.TemplateRef]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i3__namespace.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [cdk.TuiDestroyService]
                        }] }, { type: TuiPushService, decorators: [{
                            type: i0.Inject,
                            args: [i0.forwardRef(function () { return TuiPushService; })]
                        }] }];
        }, propDecorators: { tuiPush: [{
                    type: i0.Input
                }] } });

    var TuiPushAlertComponent = /** @class */ (function () {
        function TuiPushAlertComponent(options, context) {
            this.options = options;
            this.context = context;
            this.animation = Object.assign({ value: '' }, this.options);
        }
        Object.defineProperty(TuiPushAlertComponent.prototype, "isDirective", {
            get: function () {
                return this.context.content instanceof TuiPushAlertDirective;
            },
            enumerable: false,
            configurable: true
        });
        return TuiPushAlertComponent;
    }());
    TuiPushAlertComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushAlertComponent, deps: [{ token: i1.TUI_ANIMATION_OPTIONS }, { token: i5.POLYMORPHEUS_CONTEXT }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPushAlertComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPushAlertComponent, selector: "tui-push-alert", host: { attributes: { "role": "alert" }, properties: { "@tuiFadeIn": "this.animation", "@tuiSlideInRight": "this.animation", "@tuiHeightCollapse": "this.animation" } }, ngImport: i0__namespace, template: "<tui-push\n    *ngIf=\"!isDirective; else $any(context.content).template\"\n    [heading]=\"context.heading\"\n    [type]=\"context.type\"\n    [timestamp]=\"context.timestamp\"\n    (close)=\"context.$implicit.complete()\"\n>\n    <img\n        *ngIf=\"context.image\"\n        alt=\"\"\n        [src]=\"context.image\"\n    />\n    <tui-svg\n        *ngIf=\"context.icon\"\n        [style.color]=\"context.iconColor\"\n        [src]=\"context.icon\"\n    ></tui-svg>\n    <div\n        *polymorpheusOutlet=\"context.content as text\"\n        [innerHTML]=\"text\"\n    ></div>\n    <button\n        *ngIf=\"context.buttons.length > 1\"\n        type=\"button\"\n        tuiButton\n        (click)=\"context.$implicit.next(context.buttons[0])\"\n    >\n        {{ context.buttons[0] }}\n    </button>\n    <button\n        *ngIf=\"context.buttons.length\"\n        tuiLink\n        type=\"button\"\n        (click)=\"context.$implicit.next(context.buttons[context.buttons.length - 1])\"\n    >\n        {{ context.buttons[context.buttons.length - 1] }}\n    </button>\n</tui-push>\n", styles: [":host{max-width:calc(100% - 2rem);margin:0 1rem 1rem auto}:host:first-child{margin-top:auto}\n"], components: [{ type: TuiPushComponent, selector: "tui-push", inputs: ["heading", "type", "timestamp"], outputs: ["close"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }, { type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }, { type: i1__namespace.TuiLinkComponent, selector: "a[tuiLink], button[tuiLink]", inputs: ["pseudo", "icon", "iconAlign", "iconRotated", "mode"], exportAs: ["tuiLink"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: TuiPushDirective, selector: "tui-push" }, { type: i5__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [i1.tuiFadeIn, i1.tuiSlideInRight, i1.tuiHeightCollapse], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushAlertComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-push-alert',
                        templateUrl: './push-alert.template.html',
                        styleUrls: ['./push-alert.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        animations: [i1.tuiFadeIn, i1.tuiSlideInRight, i1.tuiHeightCollapse],
                        host: { role: 'alert' },
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_ANIMATION_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i5.POLYMORPHEUS_CONTEXT]
                        }] }];
        }, propDecorators: { animation: [{
                    type: i0.HostBinding,
                    args: ['@tuiFadeIn']
                }, {
                    type: i0.HostBinding,
                    args: ['@tuiSlideInRight']
                }, {
                    type: i0.HostBinding,
                    args: ['@tuiHeightCollapse']
                }] } });

    var TuiPushService = /** @class */ (function (_super) {
        __extends(TuiPushService, _super);
        function TuiPushService() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.component = new i5.PolymorpheusComponent(TuiPushAlertComponent);
            _this.defaultOptions = i0.inject(TUI_PUSH_OPTIONS);
            return _this;
        }
        return TuiPushService;
    }(cdk.AbstractTuiDialogService));
    TuiPushService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushService, deps: null, target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiPushService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushService, providedIn: "root" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushService, decorators: [{
                type: i0.Injectable,
                args: [{ providedIn: "root" }]
            }] });

    var TuiPushModule = /** @class */ (function () {
        function TuiPushModule() {
        }
        return TuiPushModule;
    }());
    TuiPushModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPushModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushModule, declarations: [TuiPushComponent,
            TuiPushDirective,
            TuiPushAlertComponent,
            TuiPushAlertDirective], imports: [i2.CommonModule,
            i5.PolymorpheusModule,
            i1.TuiButtonModule,
            i1.TuiLinkModule,
            i1.TuiSvgModule,
            i1.TuiFormatDatePipeModule], exports: [TuiPushComponent,
            TuiPushDirective,
            TuiPushAlertComponent,
            TuiPushAlertDirective] });
    TuiPushModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushModule, providers: [
            {
                provide: cdk.TUI_ALERTS,
                useExisting: TuiPushService,
                multi: true,
            },
        ], imports: [[
                i2.CommonModule,
                i5.PolymorpheusModule,
                i1.TuiButtonModule,
                i1.TuiLinkModule,
                i1.TuiSvgModule,
                i1.TuiFormatDatePipeModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPushModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i2.CommonModule,
                            i5.PolymorpheusModule,
                            i1.TuiButtonModule,
                            i1.TuiLinkModule,
                            i1.TuiSvgModule,
                            i1.TuiFormatDatePipeModule,
                        ],
                        declarations: [
                            TuiPushComponent,
                            TuiPushDirective,
                            TuiPushAlertComponent,
                            TuiPushAlertDirective,
                        ],
                        exports: [
                            TuiPushComponent,
                            TuiPushDirective,
                            TuiPushAlertComponent,
                            TuiPushAlertDirective,
                        ],
                        providers: [
                            {
                                provide: cdk.TUI_ALERTS,
                                useExisting: TuiPushService,
                                multi: true,
                            },
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_PUSH_DEFAULT_OPTIONS = TUI_PUSH_DEFAULT_OPTIONS;
    exports.TUI_PUSH_OPTIONS = TUI_PUSH_OPTIONS;
    exports.TuiPushAlertComponent = TuiPushAlertComponent;
    exports.TuiPushAlertDirective = TuiPushAlertDirective;
    exports.TuiPushComponent = TuiPushComponent;
    exports.TuiPushDirective = TuiPushDirective;
    exports.TuiPushModule = TuiPushModule;
    exports.TuiPushService = TuiPushService;
    exports.tuiPushOptionsProvider = tuiPushOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
