// @taiga-ui/kit/bundles/taiga-ui-kit-components-arrow.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core'), require('@tinkoff/ng-polymorpheus'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/arrow', ['exports', '@angular/core', '@taiga-ui/core', '@tinkoff/ng-polymorpheus', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.arrow = {}), global.ng.core, global.i1, global.i2, global.ng.common));
})(this, (function (exports, i0, i1, i2, i3) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    /** Default values for arrow options */
    var TUI_ARROW_DEFAULT_OPTIONS = {
        iconSmall: "tuiIconChevronDown",
        iconLarge: "tuiIconChevronDownLarge",
    };
    /**
     * Default parameters for arrow component
     */
    var TUI_ARROW_OPTIONS = new i0.InjectionToken("[TUI_ARROW_OPTIONS]", {
        factory: function () { return TUI_ARROW_DEFAULT_OPTIONS; },
    });
    var tuiArrowOptionsProvider = function (options) { return ({
        provide: TUI_ARROW_OPTIONS,
        useValue: Object.assign(Object.assign({}, TUI_ARROW_DEFAULT_OPTIONS), options),
    }); };
    /**
     * Type of icon in dropdowns for interactive or disable mode
     */
    var TUI_ARROW_MODE = new i0.InjectionToken("[TUI_ARROW_MODE]", {
        factory: function () { return ({
            interactive: TUI_ARROW,
            disabled: TUI_ARROW,
        }); },
    });
    var tuiArrowModeProvider = function (options) { return ({
        provide: TUI_ARROW_MODE,
        useValue: Object.assign({ interactive: TUI_ARROW, disabled: TUI_ARROW }, options),
    }); };

    var TuiArrowComponent = /** @class */ (function () {
        function TuiArrowComponent(dropdown, textfieldSize, options) {
            this.dropdown = dropdown;
            this.textfieldSize = textfieldSize;
            this.options = options;
        }
        Object.defineProperty(TuiArrowComponent.prototype, "rotated", {
            get: function () {
                return this.dropdown.open;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiArrowComponent.prototype, "arrowIcon", {
            get: function () {
                return i1.tuiSizeBigger(this.textfieldSize.size)
                    ? this.options.iconLarge
                    : this.options.iconSmall;
            },
            enumerable: false,
            configurable: true
        });
        return TuiArrowComponent;
    }());
    TuiArrowComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiArrowComponent, deps: [{ token: i1.TuiHostedDropdownComponent }, { token: i1.TUI_TEXTFIELD_SIZE }, { token: TUI_ARROW_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiArrowComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiArrowComponent, selector: "tui-arrow", host: { properties: { "class._rotated": "this.rotated" } }, ngImport: i0__namespace, template: "<tui-svg\n    *polymorpheusOutlet=\"arrowIcon as src\"\n    [src]=\"src\"\n></tui-svg>\n<ng-container *ngIf=\"dropdown.openChange | async\"></ng-container>\n", styles: [":host{transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;display:block;cursor:pointer;pointer-events:none}:host._rotated{transform:rotate(180deg)}\n"], components: [{ type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "async": i3__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiArrowComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-arrow',
                        templateUrl: './arrow.template.html',
                        styleUrls: ['./arrow.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.TuiHostedDropdownComponent, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiHostedDropdownComponent]
                        }] }, { type: i1__namespace.TuiTextfieldSizeDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_SIZE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_ARROW_OPTIONS]
                        }] }];
        }, propDecorators: { rotated: [{
                    type: i0.HostBinding,
                    args: ['class._rotated']
                }] } });
    var TUI_ARROW = new i2.PolymorpheusComponent(TuiArrowComponent);

    var TuiArrowModule = /** @class */ (function () {
        function TuiArrowModule() {
        }
        return TuiArrowModule;
    }());
    TuiArrowModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiArrowModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiArrowModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiArrowModule, declarations: [TuiArrowComponent], imports: [i3.CommonModule, i1.TuiSvgModule, i2.PolymorpheusModule], exports: [TuiArrowComponent] });
    TuiArrowModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiArrowModule, imports: [[i3.CommonModule, i1.TuiSvgModule, i2.PolymorpheusModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiArrowModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i3.CommonModule, i1.TuiSvgModule, i2.PolymorpheusModule],
                        declarations: [TuiArrowComponent],
                        exports: [TuiArrowComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_ARROW = TUI_ARROW;
    exports.TUI_ARROW_DEFAULT_OPTIONS = TUI_ARROW_DEFAULT_OPTIONS;
    exports.TUI_ARROW_MODE = TUI_ARROW_MODE;
    exports.TUI_ARROW_OPTIONS = TUI_ARROW_OPTIONS;
    exports.TuiArrowComponent = TuiArrowComponent;
    exports.TuiArrowModule = TuiArrowModule;
    exports.tuiArrowModeProvider = tuiArrowModeProvider;
    exports.tuiArrowOptionsProvider = tuiArrowOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
