// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-number.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/kit/directives'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-number', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@tinkoff/ng-polymorpheus', '@taiga-ui/kit/directives', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-number"] = {}), global.ng.core, global.ng.forms, global.i5, global.i1, global.i4, global["taiga-ui"].kit.directives, global.ng.common));
})(this, (function (exports, i0, i6, i5, i1, i4, i2, i3) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var DEFAULT_MAX_LENGTH = 18;
    var TuiInputNumberComponent = /** @class */ (function (_super) {
        __extends(TuiInputNumberComponent, _super);
        function TuiInputNumberComponent(control, cdr, numberFormat, isIOS) {
            var _this = _super.call(this, control, cdr) || this;
            _this.numberFormat = numberFormat;
            _this.isIOS = isIOS;
            _this.unfinishedValue = '';
            _this.min = Number.MIN_SAFE_INTEGER;
            _this.max = Number.MAX_SAFE_INTEGER;
            _this.decimal = 'not-zero';
            _this.precision = 2;
            /** @deprecated use `tuiTextfieldPrefix` from {@link TuiTextfieldControllerModule} instead */
            _this.prefix = '';
            /** @deprecated use `tuiTextfieldPostfix` from {@link TuiTextfieldControllerModule} instead */
            _this.postfix = '';
            _this.polymorpheusValueContent = i5.EMPTY_QUERY;
            _this.mask = function (allowNegative, decimal, decimalLimit, nativeFocusableElement) { return ({
                mask: i1.tuiCreateNumberMask({
                    allowNegative: allowNegative,
                    decimalLimit: decimalLimit,
                    allowDecimal: decimal !== 'never',
                    requireDecimal: decimal === 'always',
                    decimalSymbol: _this.numberFormat.decimalSeparator,
                    thousandSymbol: _this.numberFormat.thousandSeparator,
                    autoCorrectDecimalSymbol: i1.tuiEnableAutoCorrectDecimalSymbol(_this.numberFormat),
                }),
                pipe: i1.tuiCreateAutoCorrectedNumberPipe(decimal === 'always' ? decimalLimit : 0, _this.numberFormat.decimalSeparator, _this.numberFormat.thousandSeparator, nativeFocusableElement, allowNegative, _this.isIOS),
                guide: false,
            }); };
            return _this;
        }
        Object.defineProperty(TuiInputNumberComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.primitiveTextfield || this.computedDisabled
                    ? null
                    : this.primitiveTextfield.nativeFocusableElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "focused", {
            get: function () {
                return !!this.primitiveTextfield && this.primitiveTextfield.focused;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "isNegativeAllowed", {
            get: function () {
                return this.min < 0;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "inputMode", {
            get: function () {
                if (this.isIOS && this.isNegativeAllowed) {
                    // iPhone does not have minus sign if inputMode is equal to 'numeric' / 'decimal'
                    return 'text';
                }
                return this.decimal === 'never' ? 'numeric' : 'decimal';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "calculatedMaxLength", {
            get: function () {
                var decimalPart = this.decimal !== 'never' &&
                    this.nativeValue.includes(this.numberFormat.decimalSeparator);
                var precision = decimalPart ? Math.min(this.precision + 1, 20) : 0;
                var takeThousand = this.numberFormat.thousandSeparator.repeat(5).length;
                return DEFAULT_MAX_LENGTH + precision + takeThousand;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "formattedValue", {
            get: function () {
                return this.getFormattedValue(this.value || 0);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "computedValue", {
            get: function () {
                if (this.focused) {
                    return this.nativeValue;
                }
                return this.value === null ? '' : this.formattedValue;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputNumberComponent.prototype.onZero = function (event) {
            var decimal = this.nativeValue.split(this.numberFormat.decimalSeparator)[1] || '';
            var nativeFocusableElement = this.nativeFocusableElement;
            if (decimal.length < this.precision ||
                !nativeFocusableElement ||
                !nativeFocusableElement.selectionStart ||
                this.nativeValue[nativeFocusableElement.selectionStart] !== '0') {
                return;
            }
            event.preventDefault();
            nativeFocusableElement.selectionStart++;
        };
        TuiInputNumberComponent.prototype.onValueChange = function (value) {
            if (i1.tuiMaskedMoneyValueIsEmpty(value)) {
                this.value = null;
                return;
            }
            if (this.isNativeValueNotFinished) {
                this.unfinishedValue = value;
                return;
            }
            this.unfinishedValue = null;
            var capped = this.absoluteCapInputValue(value);
            if (capped === null || Number.isNaN(capped)) {
                return;
            }
            this.value = capped;
            if (capped !==
                i1.tuiMaskedNumberStringToNumber(value, this.numberFormat.decimalSeparator, this.numberFormat.thousandSeparator)) {
                this.nativeValue = this.formattedValue;
            }
        };
        TuiInputNumberComponent.prototype.onKeyDown = function (event) {
            if (!i1.TUI_DECIMAL_SYMBOLS.includes(event.key)) {
                return;
            }
            if (this.decimal === 'never') {
                event.preventDefault();
                return;
            }
            if (this.nativeValue.includes(this.numberFormat.decimalSeparator)) {
                event.preventDefault();
                this.setCaretAfterComma();
            }
        };
        TuiInputNumberComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
            if (focused) {
                return;
            }
            var nativeNumberValue = this.unfinishedValue
                ? i1.tuiMaskedNumberStringToNumber(this.unfinishedValue, this.numberFormat.decimalSeparator, this.numberFormat.thousandSeparator)
                : this.nativeNumberValue;
            this.unfinishedValue = null;
            if (Number.isNaN(nativeNumberValue)) {
                this.clear();
                return;
            }
            this.value = Math.min(this.max, Math.max(this.min, nativeNumberValue));
            this.nativeValue = this.formattedValue;
        };
        TuiInputNumberComponent.prototype.getFormattedValue = function (value) {
            var absValue = Math.abs(value);
            var hasFraction = absValue % 1 > 0;
            var decimalLimit = this.decimal === 'always' || (hasFraction && this.decimal !== 'never')
                ? this.precision
                : 0;
            var fraction = hasFraction
                ? i1.tuiGetFractionPartPadded(value, this.precision)
                : '';
            if (this.focused && this.decimal !== 'always') {
                decimalLimit = fraction.length;
            }
            return i1.tuiFormatNumber(value, Object.assign(Object.assign({}, this.numberFormat), { decimalLimit: decimalLimit }));
        };
        Object.defineProperty(TuiInputNumberComponent.prototype, "isNativeValueNotFinished", {
            get: function () {
                var nativeNumberValue = this.nativeNumberValue;
                return nativeNumberValue < 0
                    ? nativeNumberValue > this.max
                    : nativeNumberValue < this.min;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.primitiveTextfield || !this.nativeFocusableElement) {
                    return;
                }
                this.primitiveTextfield.value = value;
                this.nativeFocusableElement.value = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "nativeNumberValue", {
            get: function () {
                return i1.tuiMaskedNumberStringToNumber(this.nativeValue, this.numberFormat.decimalSeparator, this.numberFormat.thousandSeparator);
            },
            enumerable: false,
            configurable: true
        });
        TuiInputNumberComponent.prototype.clear = function () {
            this.nativeValue = '';
            this.value = null;
        };
        TuiInputNumberComponent.prototype.absoluteCapInputValue = function (inputValue) {
            var value = i1.tuiMaskedNumberStringToNumber(inputValue, this.numberFormat.decimalSeparator, this.numberFormat.thousandSeparator);
            var capped = value < 0
                ? Math.max(Math.max(this.min, Number.MIN_SAFE_INTEGER), value)
                : Math.min(value, Math.min(this.max, Number.MAX_SAFE_INTEGER));
            var ineligibleValue = Number.isNaN(capped) || capped < this.min || capped > this.max;
            return ineligibleValue ? null : capped;
        };
        TuiInputNumberComponent.prototype.setCaretAfterComma = function () {
            if (!this.nativeFocusableElement) {
                return;
            }
            var afterCommaPosition = this.nativeValue.length - this.precision;
            this.nativeFocusableElement.setSelectionRange(afterCommaPosition, afterCommaPosition);
        };
        return TuiInputNumberComponent;
    }(i5.AbstractTuiNullableControl));
    TuiInputNumberComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberComponent, deps: [{ token: i6.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i1.TUI_NUMBER_FORMAT }, { token: i5.TUI_IS_IOS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputNumberComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputNumberComponent, selector: "tui-input-number", inputs: { min: "min", max: "max", decimal: "decimal", precision: "precision", prefix: "prefix", postfix: "postfix" }, host: { listeners: { "keydown.0": "onZero($event)" } }, providers: [
            i5.tuiAsFocusableItemAccessor(TuiInputNumberComponent),
            i5.tuiAsControl(TuiInputNumberComponent),
        ], queries: [{ propertyName: "polymorpheusValueContent", predicate: i4.PolymorpheusOutletDirective, descendants: true }], viewQueries: [{ propertyName: "primitiveTextfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-primitive-textfield\n    tuiValueAccessor\n    class=\"t-textfield\"\n    [pseudoHover]=\"pseudoHover\"\n    [pseudoFocus]=\"computedFocused\"\n    [pseudoActive]=\"pseudoActive\"\n    [invalid]=\"computedInvalid\"\n    [readOnly]=\"readOnly\"\n    [disabled]=\"computedDisabled\"\n    [textMask]=\"isNegativeAllowed | tuiMapper : mask : decimal : precision : nativeFocusableElement\"\n    [value]=\"computedValue\"\n    [prefix]=\"prefix\"\n    [postfix]=\"postfix\"\n    [focusable]=\"focusable\"\n    (valueChange)=\"onValueChange($event)\"\n    (focusedChange)=\"onFocused($event)\"\n    (keydown)=\"onKeyDown($event)\"\n>\n    <ng-content></ng-content>\n    <ng-content\n        select=\"input\"\n        ngProjectAs=\"input\"\n    ></ng-content>\n    <div\n        *ngIf=\"polymorpheusValueContent.length\"\n        ngProjectAs=\"tuiContent\"\n        class=\"t-value-content\"\n    >\n        <ng-container *polymorpheusOutlet=\"valueContent as text\">\n            {{ text }}\n        </ng-container>\n    </div>\n</tui-primitive-textfield>\n\n<ng-container *ngIf=\"polymorpheusValueContent.changes | async\"></ng-container>\n\n<ng-template #valueContent>\n    <ng-content select=\"tuiContent\"></ng-content>\n</ng-template>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}.t-textfield{border-radius:inherit;text-align:inherit}.t-value-content{width:100%}\n"], components: [{ type: i1__namespace.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }], directives: [{ type: i1__namespace.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i2__namespace.TuiValueAccessorDirective, selector: "[tuiValueAccessor]" }, { type: i2__namespace.MaskedInputDirective, selector: "[textMask]", inputs: ["textMask"], exportAs: ["textMask"] }, { type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "tuiMapper": i5__namespace.TuiMapperPipe, "async": i3__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i5.tuiDefaultProp()
    ], TuiInputNumberComponent.prototype, "min", void 0);
    __decorate([
        i5.tuiDefaultProp()
    ], TuiInputNumberComponent.prototype, "max", void 0);
    __decorate([
        i5.tuiDefaultProp()
    ], TuiInputNumberComponent.prototype, "decimal", void 0);
    __decorate([
        i5.tuiDefaultProp()
    ], TuiInputNumberComponent.prototype, "precision", void 0);
    __decorate([
        i5.tuiDefaultProp()
    ], TuiInputNumberComponent.prototype, "prefix", void 0);
    __decorate([
        i5.tuiDefaultProp()
    ], TuiInputNumberComponent.prototype, "postfix", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-number',
                        templateUrl: './input-number.template.html',
                        styleUrls: ['./input-number.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i5.tuiAsFocusableItemAccessor(TuiInputNumberComponent),
                            i5.tuiAsControl(TuiInputNumberComponent),
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i6__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i6.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_NUMBER_FORMAT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i5.TUI_IS_IOS]
                        }] }];
        }, propDecorators: { primitiveTextfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], decimal: [{
                    type: i0.Input
                }], precision: [{
                    type: i0.Input
                }], prefix: [{
                    type: i0.Input
                }], postfix: [{
                    type: i0.Input
                }], polymorpheusValueContent: [{
                    type: i0.ContentChildren,
                    args: [i4.PolymorpheusOutletDirective, { descendants: true }]
                }], onZero: [{
                    type: i0.HostListener,
                    args: ['keydown.0', ['$event']]
                }] } });

    var TuiInputNumberDirective = /** @class */ (function (_super) {
        __extends(TuiInputNumberDirective, _super);
        function TuiInputNumberDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(TuiInputNumberDirective.prototype, "value", {
            get: function () {
                return this.host.computedValue;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputNumberDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        TuiInputNumberDirective.prototype.ngDoCheck = function () {
            if (!this.host.nativeFocusableElement) {
                return;
            }
            this.host.nativeFocusableElement.maxLength = this.host.calculatedMaxLength;
            this.host.nativeFocusableElement.inputMode = this.host.inputMode;
        };
        return TuiInputNumberDirective;
    }(i1.AbstractTuiTextfieldHost));
    TuiInputNumberDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiInputNumberDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputNumberDirective, selector: "tui-input-number", providers: [i1.tuiAsTextfieldHost(TuiInputNumberDirective)], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-input-number',
                        providers: [i1.tuiAsTextfieldHost(TuiInputNumberDirective)],
                    }]
            }] });

    var TuiInputNumberModule = /** @class */ (function () {
        function TuiInputNumberModule() {
        }
        return TuiInputNumberModule;
    }());
    TuiInputNumberModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputNumberModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberModule, declarations: [TuiInputNumberComponent, TuiInputNumberDirective], imports: [i3.CommonModule,
            i2.TextMaskModule,
            i5.TuiMapperPipeModule,
            i1.TuiPrimitiveTextfieldModule,
            i1.TuiTextfieldControllerModule,
            i2.TuiValueAccessorModule,
            i4.PolymorpheusModule], exports: [TuiInputNumberComponent, TuiInputNumberDirective, i1.TuiTextfieldComponent] });
    TuiInputNumberModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberModule, imports: [[
                i3.CommonModule,
                i2.TextMaskModule,
                i5.TuiMapperPipeModule,
                i1.TuiPrimitiveTextfieldModule,
                i1.TuiTextfieldControllerModule,
                i2.TuiValueAccessorModule,
                i4.PolymorpheusModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputNumberModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            i2.TextMaskModule,
                            i5.TuiMapperPipeModule,
                            i1.TuiPrimitiveTextfieldModule,
                            i1.TuiTextfieldControllerModule,
                            i2.TuiValueAccessorModule,
                            i4.PolymorpheusModule,
                        ],
                        declarations: [TuiInputNumberComponent, TuiInputNumberDirective],
                        exports: [TuiInputNumberComponent, TuiInputNumberDirective, i1.TuiTextfieldComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiInputNumberComponent = TuiInputNumberComponent;
    exports.TuiInputNumberDirective = TuiInputNumberDirective;
    exports.TuiInputNumberModule = TuiInputNumberModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
