// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-tag.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/providers'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/kit/components/tag'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-tag', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/providers', 'rxjs', 'rxjs/operators', '@taiga-ui/kit/components/tag', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-tag"] = {}), global.ng.core, global.ng.forms, global.i4, global.i1, global["taiga-ui"].kit.providers, global.rxjs, global.rxjs.operators, global["taiga-ui"].kit.components.tag, global.ng.common, global.i5));
})(this, (function (exports, i0, i6, i4, i1, providers, i7, operators, i2, i3, i5) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_INPUT_TAG_DEFAULT_OPTIONS = {
        separator: ",",
        uniqueTags: true,
        tagStatus: "primary",
    };
    /**
     * Default parameters for InputTag component
     */
    var TUI_INPUT_TAG_OPTIONS = new i0.InjectionToken("[TUI_INPUT_TAG_OPTIONS]", {
        factory: function () { return TUI_INPUT_TAG_DEFAULT_OPTIONS; },
    });
    function tuiInputTagOptionsProvider(options) {
        return {
            provide: TUI_INPUT_TAG_OPTIONS,
            useValue: Object.assign(Object.assign({}, TUI_INPUT_TAG_DEFAULT_OPTIONS), options),
        };
    }

    var EVENT_Y_TO_X_COEFFICIENT = 3;
    var TAG_SIZE_REM = {
        s: 1.25,
        m: 1.5,
        l: 2,
    };
    var LINE_HEIGHT_REM = {
        s: 1,
        m: 1.25,
        l: 1.25,
    };
    var TAG_VERTICAL_SPACE_REM = 0.125;
    var TuiInputTagComponent = /** @class */ (function (_super) {
        __extends(TuiInputTagComponent, _super);
        function TuiInputTagComponent(control, cdr, tuiScrollService, el, modeDirective, mode$, hintOptions, controller, options, parentHostedDropdown) {
            var _this = _super.call(this, control, cdr) || this;
            _this.tuiScrollService = tuiScrollService;
            _this.el = el;
            _this.modeDirective = modeDirective;
            _this.mode$ = mode$;
            _this.hintOptions = hintOptions;
            _this.controller = controller;
            _this.options = options;
            _this.parentHostedDropdown = parentHostedDropdown;
            _this.tags = i4.EMPTY_QUERY;
            _this.scrollToStart$ = new i7.Subject();
            _this.scrollToEnd$ = new i7.Subject();
            _this.separator = _this.options.separator;
            _this.search = '';
            _this.editable = true;
            _this.tagValidator = i4.ALWAYS_TRUE_HANDLER;
            // TODO: 4.0 Consider removing and use rows = 1 instead
            _this.expandable = true;
            _this.rows = Infinity;
            _this.inputHidden = false;
            _this.uniqueTags = _this.options.uniqueTags;
            _this.maxLength = null;
            _this.placeholder = '';
            _this.removable = true;
            _this.disabledItemHandler = i4.ALWAYS_FALSE_HANDLER;
            _this.searchChange = new i0.EventEmitter();
            _this.status$ = _this.mode$.pipe(operators.map(function () { return _this.status; }));
            _this.open = false;
            return _this;
        }
        Object.defineProperty(TuiInputTagComponent.prototype, "pseudoFocusedSetter", {
            set: function (value) {
                if (!value && !this.focused) {
                    this.scrollToStart$.next();
                }
                this.pseudoFocus = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "scrollerSetter", {
            set: function (scroller) {
                this.initScrollerSubscription(scroller);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.focusableElement || this.computedDisabled
                    ? null
                    : this.focusableElement.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "focused", {
            get: function () {
                var _a;
                return (i4.tuiIsNativeFocusedIn(this.el.nativeElement) || !!((_a = this.hostedDropdown) === null || _a === void 0 ? void 0 : _a.focused));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "appearance", {
            get: function () {
                return this.controller.appearance;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "size", {
            get: function () {
                return this.controller.size;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "labelOutside", {
            get: function () {
                var _b = this.controller, size = _b.size, labelOutside = _b.labelOutside;
                return size === 's' || labelOutside;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "icon", {
            get: function () {
                return this.controller.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "iconCleaner", {
            get: function () {
                return this.controller.options.iconCleaner;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "iconLeft", {
            get: function () {
                return this.controller.iconLeft;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasCleaner", {
            get: function () {
                return this.controller.cleaner && this.hasValue && this.interactive;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasNativeValue", {
            get: function () {
                return !!this.search;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasValue", {
            get: function () {
                return !!this.value.length || this.hasNativeValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasPlaceholder", {
            get: function () {
                return (!this.labelOutside ||
                    (!this.hasValue && (!this.hasExampleText || this.inputHidden)));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "placeholderRaised", {
            get: function () {
                return (!this.labelOutside &&
                    ((this.computedFocused && !this.readOnly) || this.hasValue));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasExampleText", {
            get: function () {
                var _a;
                return (!!((_a = this.nativeFocusableElement) === null || _a === void 0 ? void 0 : _a.placeholder) &&
                    this.computedFocused &&
                    !this.hasValue &&
                    !this.readOnly);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasRightIcons", {
            get: function () {
                var _a;
                return (this.hasCleaner ||
                    !!this.icon ||
                    (!!((_a = this.hintOptions) === null || _a === void 0 ? void 0 : _a.content) && !this.computedDisabled));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "status", {
            get: function () {
                var _a;
                return ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.mode) ? 'default' : this.options.tagStatus;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "canOpen", {
            get: function () {
                return this.interactive && !!this.datalist;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "computeMaxHeight", {
            get: function () {
                return this.expandable ? this.rows * this.lineHeight : null;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputTagComponent.prototype.getLeftContent = function (tag) {
            return !this.tagValidator(tag) && this.errorIconTemplate
                ? this.errorIconTemplate
                : '';
        };
        TuiInputTagComponent.prototype.onCleanerClick = function () {
            var _a;
            this.updateSearch('');
            this.clear();
            this.focusInput();
            (_a = this.parentHostedDropdown) === null || _a === void 0 ? void 0 : _a.updateOpen(true);
        };
        TuiInputTagComponent.prototype.onActiveZone = function (active) {
            this.open = false;
            this.addTag();
            this.updateFocused(active);
            if (!active && !this.computedFocused) {
                this.scrollToStart$.next();
            }
        };
        TuiInputTagComponent.prototype.onMouseDown = function (event) {
            var _a;
            var actualTarget = i4.tuiGetActualTarget(event);
            if (!this.focusableElement ||
                actualTarget === this.focusableElement.nativeElement ||
                !i4.tuiIsElement(event.target) ||
                ((_a = this.cleanerSvg) === null || _a === void 0 ? void 0 : _a.nativeElement.contains(event.target)) ||
                (this.tagsContainer &&
                    actualTarget !== this.tagsContainer.nativeElement &&
                    this.tagsContainer.nativeElement.contains(actualTarget))) {
                return;
            }
            event.preventDefault();
            this.focusInput();
        };
        TuiInputTagComponent.prototype.onFieldKeyDownBackspace = function (event) {
            if (!this.labelOutside && !this.hasNativeValue && this.value.length) {
                this.deleteLastEnabledItem();
            }
            else {
                this.onFieldKeyDownArrowLeft(event);
            }
        };
        TuiInputTagComponent.prototype.onFieldKeyDownArrowLeft = function (event) {
            if (!this.labelOutside || this.hasNativeValue || !this.value.length) {
                return;
            }
            event.preventDefault();
            this.tags.last.nativeElement.focus();
        };
        TuiInputTagComponent.prototype.onFieldKeyDownEnter = function () {
            this.addTag();
            this.scrollToEnd$.next();
        };
        TuiInputTagComponent.prototype.onTagKeyDownArrowLeft = function (currentIndex) {
            if (currentIndex > 0) {
                this.onScrollKeyDown(currentIndex, -1);
            }
        };
        TuiInputTagComponent.prototype.onTagKeyDownArrowRight = function (currentIndex) {
            if (currentIndex === this.value.length - 1) {
                this.focusInput();
            }
            else {
                this.onScrollKeyDown(currentIndex, 1);
            }
        };
        TuiInputTagComponent.prototype.onTagEdited = function (value, index) {
            var _this = this;
            this.focusInput();
            this.value = this.filterValue(this.value
                .map(function (tag, tagIndex) { return tagIndex !== index
                ? tag
                : value
                    .split(_this.separator)
                    .map(function (tag) { return tag.trim(); })
                    .filter(Boolean); })
                .reduce(function (result, item) { return result.concat(item); }, []));
        };
        TuiInputTagComponent.prototype.handleOption = function (item) {
            this.focusInput();
            this.updateSearch('');
            this.value = this.filterValue(this.value.concat(item));
            this.open = false;
            this.scrollToEnd$.next();
        };
        TuiInputTagComponent.prototype.onInput = function (value) {
            var _this = this;
            var array = value.split(this.separator);
            var tags = array
                .map(function (item) { return _this.clippedValue(item.trim()); })
                .filter(function (item, index, _b) {
                var length = _b.length;
                return item.length > 0 && index !== length - 1;
            });
            var validated = tags.filter(function (tag) { return !_this.disabledItemHandler(tag); });
            if (array.length > 1) {
                this.updateSearch(this.clippedValue(array[array.length - 1].trim()));
                this.value = this.filterValue(__spreadArray(__spreadArray([], __read(this.value)), __read(validated)));
            }
            else {
                this.updateSearch(this.clippedValue(value));
            }
            this.open = this.hasNativeValue;
        };
        TuiInputTagComponent.prototype.onPaste = function (event) {
            var pasted = i4.tuiGetClipboardDataText(event);
            this.onInput(pasted || '');
        };
        TuiInputTagComponent.prototype.onDrop = function (_b) {
            var dataTransfer = _b.dataTransfer;
            if (dataTransfer) {
                this.onInput(dataTransfer.getData('text') || '');
            }
        };
        TuiInputTagComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.open = false;
        };
        TuiInputTagComponent.prototype.trackByFn = function (_, tag) {
            // Actually tag has TuiStringifiableItem type not string
            return tag.toString();
        };
        TuiInputTagComponent.prototype.filterValue = function (value) {
            var _this = this;
            var seen = new Set();
            return value
                .reverse()
                .filter(function (item) { return !_this.uniqueTags || (item && !seen.has(item) && seen.add(item)); })
                .reverse();
        };
        TuiInputTagComponent.prototype.onScrollKeyDown = function (currentIndex, flag) {
            var tag = this.tags.find(function (_item, index) { return index === currentIndex + flag; });
            if (!tag || !this.scrollBar) {
                return;
            }
            tag.nativeElement.focus();
            if (flag * this.scrollBar.nativeElement.clientWidth -
                flag * tag.nativeElement.offsetLeft -
                tag.nativeElement.clientWidth <
                0) {
                this.scrollBar.nativeElement.scrollLeft +=
                    flag * tag.nativeElement.clientWidth;
            }
        };
        TuiInputTagComponent.prototype.initScrollerSubscription = function (scroller) {
            var _this = this;
            if (!(scroller === null || scroller === void 0 ? void 0 : scroller.browserScrollRef)) {
                return;
            }
            var nativeElement = scroller.browserScrollRef.nativeElement;
            var wheel$ = i4.tuiTypedFromEvent(nativeElement, 'wheel', { passive: false }).pipe(operators.filter(function (event) { return event.deltaX === 0 && _this.shouldScroll(nativeElement); }), i4.tuiPreventDefault(), operators.map(function (_b) {
                var deltaY = _b.deltaY;
                return Math.max(nativeElement.scrollLeft + deltaY * EVENT_Y_TO_X_COEFFICIENT, 0);
            }));
            var start$ = this.scrollToStart$.pipe(operators.map(function () { return 0; }));
            var end$ = this.scrollToEnd$.pipe(operators.map(function () { return nativeElement.scrollWidth; }));
            i7.merge(wheel$, start$, end$)
                .pipe(operators.switchMap(function (left) { return _this.tuiScrollService.scroll$(nativeElement, nativeElement.scrollHeight, left); }), operators.takeUntil(this.destroy$))
                .subscribe();
        };
        TuiInputTagComponent.prototype.updateSearch = function (value) {
            if (this.focusableElement) {
                this.focusableElement.nativeElement.value = value;
            }
            this.search = value;
            this.searchChange.emit(value);
        };
        TuiInputTagComponent.prototype.shouldScroll = function (_b) {
            var scrollWidth = _b.scrollWidth, offsetWidth = _b.offsetWidth;
            return scrollWidth > offsetWidth;
        };
        TuiInputTagComponent.prototype.addTag = function () {
            var inputValue = this.search.trim();
            if (!inputValue || this.disabledItemHandler(inputValue)) {
                return;
            }
            this.updateSearch('');
            this.value = this.filterValue(this.value.concat(inputValue));
        };
        TuiInputTagComponent.prototype.deleteLastEnabledItem = function () {
            for (var index = this.value.length - 1; index >= 0; index--) {
                if (!this.disabledItemHandler(this.value[index])) {
                    this.value = i4.tuiArrayRemove(this.value, index);
                    break;
                }
            }
        };
        TuiInputTagComponent.prototype.focusInput = function (preventScroll) {
            if (preventScroll === void 0) { preventScroll = false; }
            if (this.nativeFocusableElement) {
                this.nativeFocusableElement.focus({ preventScroll: preventScroll });
            }
        };
        TuiInputTagComponent.prototype.clippedValue = function (value) {
            return value.slice(0, this.maxLength || value.length);
        };
        Object.defineProperty(TuiInputTagComponent.prototype, "lineHeight", {
            get: function () {
                return this.labelOutside
                    ? TAG_SIZE_REM[this.controller.size] + 2 * TAG_VERTICAL_SPACE_REM
                    : LINE_HEIGHT_REM[this.controller.size];
            },
            enumerable: false,
            configurable: true
        });
        return TuiInputTagComponent;
    }(i4.AbstractTuiMultipleControl));
    TuiInputTagComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTagComponent, deps: [{ token: i6.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i4.TuiScrollService }, { token: i0.ElementRef }, { token: i1.TuiModeDirective, optional: true }, { token: i1.TUI_MODE }, { token: i1.TuiHintOptionsDirective, optional: true }, { token: i1.TUI_TEXTFIELD_WATCHED_CONTROLLER }, { token: TUI_INPUT_TAG_OPTIONS }, { token: i1.TuiHostedDropdownComponent, optional: true }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputTagComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputTagComponent, selector: "tui-input-tag", inputs: { separator: "separator", search: "search", editable: "editable", tagValidator: "tagValidator", expandable: "expandable", rows: "rows", inputHidden: "inputHidden", uniqueTags: "uniqueTags", maxLength: "maxLength", placeholder: "placeholder", removable: "removable", disabledItemHandler: "disabledItemHandler", pseudoFocusedSetter: ["pseudoFocused", "pseudoFocusedSetter"] }, outputs: { searchChange: "searchChange" }, host: { properties: { "class._expandable": "this.expandable", "attr.data-size": "this.size", "class._label-outside": "this.labelOutside" } }, providers: [
            i4.tuiAsFocusableItemAccessor(TuiInputTagComponent),
            i4.tuiAsControl(TuiInputTagComponent),
            i1.tuiAsDataListHost(TuiInputTagComponent),
            i1.TEXTFIELD_CONTROLLER_PROVIDER,
            i1.MODE_PROVIDER,
        ], queries: [{ propertyName: "datalist", first: true, predicate: i1.TuiDataListDirective, descendants: true, read: i0.TemplateRef }], viewQueries: [{ propertyName: "hostedDropdown", first: true, predicate: i1.TuiHostedDropdownComponent, descendants: true }, { propertyName: "focusableElement", first: true, predicate: ["focusableElement"], descendants: true }, { propertyName: "tagsContainer", first: true, predicate: ["tagsContainer"], descendants: true }, { propertyName: "cleanerSvg", first: true, predicate: ["cleaner"], descendants: true, read: i0.ElementRef }, { propertyName: "scrollBar", first: true, predicate: i1.TuiScrollbarComponent, descendants: true, read: i0.ElementRef }, { propertyName: "errorIconTemplate", first: true, predicate: ["errorIcon"], descendants: true }, { propertyName: "scrollerSetter", first: true, predicate: i1.TuiScrollbarComponent, descendants: true }, { propertyName: "tags", predicate: ["tag"], descendants: true, read: i0.ElementRef }], usesInheritance: true, ngImport: i0__namespace, template: "<ng-container *ngIf=\"hintOptions?.change$ | async\"></ng-container>\n<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"canOpen\"\n    [content]=\"datalist || ''\"\n    [(open)]=\"open\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <div\n        tuiWrapper\n        [appearance]=\"appearance\"\n        [readOnly]=\"readOnly\"\n        [focus]=\"computedFocused\"\n        [hover]=\"pseudoHover\"\n        [disabled]=\"computedDisabled\"\n        [invalid]=\"computedInvalid\"\n        (mousedown)=\"onMouseDown($event)\"\n    >\n        <div class=\"t-content\">\n            <div\n                *ngIf=\"iconLeft\"\n                class=\"t-icons t-icons_left t-textfield-icon\"\n            >\n                <tui-svg\n                    *polymorpheusOutlet=\"iconLeft as src; context: {$implicit: size}\"\n                    tuiWrapper\n                    appearance=\"icon\"\n                    class=\"t-icon\"\n                    [src]=\"src\"\n                ></tui-svg>\n            </div>\n            <div class=\"t-wrapper\">\n                <div class=\"t-absolute-wrapper\">\n                    <div\n                        *ngIf=\"hasPlaceholder\"\n                        automation-id=\"tui-input-tag__placeholder\"\n                        class=\"t-placeholder\"\n                        [class.t-placeholder_raised]=\"placeholderRaised\"\n                    >\n                        <ng-content></ng-content>\n                    </div>\n                </div>\n                <tui-scrollbar\n                    class=\"t-scrollbar\"\n                    [hidden]=\"!expandable\"\n                    [style.maxHeight.rem]=\"computeMaxHeight\"\n                >\n                    <div\n                        class=\"t-tags\"\n                        [class.t-tags_expandable]=\"expandable\"\n                    >\n                        <ng-container *ngIf=\"labelOutside; else text\">\n                            <tui-tag\n                                *ngFor=\"let item of value; index as index; trackBy: trackByFn\"\n                                #tag\n                                automation-id=\"tui-input-tag__tag\"\n                                class=\"t-tag\"\n                                [status]=\"(status$ | async)!\"\n                                [leftContent]=\"getLeftContent(item)\"\n                                [tuiFocusable]=\"false\"\n                                [disabled]=\"computedDisabled || disabledItemHandler(item)\"\n                                [editable]=\"editable && !readOnly\"\n                                [hoverable]=\"!readOnly\"\n                                [removable]=\"!readOnly && removable\"\n                                [separator]=\"separator\"\n                                [maxLength]=\"maxLength\"\n                                [size]=\"controller.size\"\n                                [value]=\"item.toString()\"\n                                (edited)=\"onTagEdited($event, index)\"\n                                (keydown.arrowLeft.prevent)=\"onTagKeyDownArrowLeft(index)\"\n                                (keydown.arrowRight.prevent)=\"onTagKeyDownArrowRight(index)\"\n                            ></tui-tag>\n                        </ng-container>\n                        <ng-template #text>\n                            <span\n                                *ngFor=\"let item of value\"\n                                class=\"t-text\"\n                                [class.t-text_comma]=\"computedFocused && !inputHidden\"\n                                [class.t-text_disabled]=\"disabledItemHandler(item)\"\n                                [class.t-text_error]=\"!tagValidator(item)\"\n                                [textContent]=\"item\"\n                            ></span>\n                        </ng-template>\n                        <div\n                            class=\"t-input-wrapper\"\n                            [class.t-input-wrapper_collapsed]=\"computedDisabled || readOnly || inputHidden\"\n                        >\n                            <div class=\"t-ghost\">{{ search }}</div>\n                            <input\n                                #focusableElement\n                                type=\"text\"\n                                automation-id=\"tui-input-tag__native\"\n                                class=\"t-native\"\n                                [placeholder]=\"placeholder\"\n                                [class.t-native_hidden]=\"inputHidden\"\n                                [attr.maxLength]=\"maxLength\"\n                                [id]=\"id\"\n                                [disabled]=\"computedDisabled\"\n                                [readOnly]=\"readOnly || inputHidden\"\n                                [tuiFocusable]=\"computedFocusable\"\n                                [ngModel]=\"search\"\n                                (ngModelChange)=\"onInput($event)\"\n                                (paste.prevent)=\"onPaste($event)\"\n                                (drop.prevent)=\"onDrop($any($event))\"\n                                (keydown.arrowLeft)=\"onFieldKeyDownArrowLeft($event)\"\n                                (keydown.backspace)=\"onFieldKeyDownBackspace($event)\"\n                                (keydown.enter.prevent)=\"onFieldKeyDownEnter()\"\n                            />\n                        </div>\n                    </div>\n                </tui-scrollbar>\n            </div>\n            <ng-content select=\"select\"></ng-content>\n            <div\n                *ngIf=\"hasRightIcons\"\n                class=\"t-icons t-icons_right\"\n            >\n                <ng-container *ngIf=\"hasCleaner\">\n                    <tui-svg\n                        *polymorpheusOutlet=\"iconCleaner as src; context: {$implicit: size}\"\n                        #cleaner\n                        tuiWrapper\n                        appearance=\"icon\"\n                        automation-id=\"tui-input-tag__cleaner\"\n                        class=\"t-cleaner\"\n                        [src]=\"src\"\n                        (click.stop)=\"onCleanerClick()\"\n                    ></tui-svg>\n                </ng-container>\n                <tui-tooltip\n                    *ngIf=\"hintOptions?.content && !computedDisabled\"\n                    automation-id=\"tui-input-tag__tooltip\"\n                    class=\"t-tooltip\"\n                    [content]=\"hintOptions?.content\"\n                ></tui-tooltip>\n                <div\n                    *ngIf=\"icon\"\n                    class=\"t-icon t-textfield-icon\"\n                >\n                    <tui-svg\n                        *polymorpheusOutlet=\"icon as src; context: {$implicit: size}\"\n                        tuiWrapper\n                        appearance=\"icon\"\n                        [src]=\"src\"\n                    ></tui-svg>\n                </div>\n            </div>\n        </div>\n        <ng-template #errorIcon>\n            <tui-svg\n                src=\"tuiIconAlertCircle\"\n                class=\"t-error-icon\"\n            ></tui-svg>\n        </ng-template>\n    </div>\n</tui-hosted-dropdown>\n", styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);position:relative;display:block;border-radius:var(--tui-radius-m);text-align:left}:host[data-size=s]{height:var(--tui-height-s);min-height:var(--tui-height-s);max-height:var(--tui-height-s)}:host[data-size=m]{height:var(--tui-height-m);min-height:var(--tui-height-m);max-height:var(--tui-height-m)}:host[data-size=l]{height:var(--tui-height-l);min-height:var(--tui-height-l);max-height:var(--tui-height-l);font:var(--tui-font-text-m);line-height:1.25rem}.t-input{font:var(--tui-font-text-s);color:var(--tui-text-01);padding:0;margin:0;border:0;border-radius:inherit;background:none;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;outline:none;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;padding:0 var(--tui-padding-m);border:solid transparent;border-width:0 var(--border-end, 0) 0 var(--border-start, 0);border-inline-start-width:var(--border-start, 0);border-inline-end-width:var(--border-end, 0);text-indent:var(--text-indent);text-align:inherit;box-sizing:border-box;white-space:nowrap;overflow:hidden;text-transform:inherit;resize:none}.t-input:-webkit-autofill,.t-input:-webkit-autofill:hover,.t-input:-webkit-autofill:focus{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}:host[data-mode=onDark] .t-input:-webkit-autofill,.t-input :host-context(tui-primitive-textfield[data-mode=\"onDark\"]):-webkit-autofill,.t-input :host-context(tui-text-area[data-mode=\"onDark\"]):-webkit-autofill,:host[data-mode=onDark] .t-input:-webkit-autofill:hover,.t-input :host-context(tui-primitive-textfield[data-mode=\"onDark\"]):-webkit-autofill:hover,.t-input :host-context(tui-text-area[data-mode=\"onDark\"]):-webkit-autofill:hover,:host[data-mode=onDark] .t-input:-webkit-autofill:focus,.t-input :host-context(tui-primitive-textfield[data-mode=\"onDark\"]):-webkit-autofill:focus,.t-input :host-context(tui-text-area[data-mode=\"onDark\"]):-webkit-autofill:focus{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01-night)!important;border-color:var(--tui-autofill-night);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill-night) inset!important}:host[data-size=s] .t-input,.t-input :host-context(tui-primitive-textfield[data-size=\"s\"]):not(tui-primitive-textfield),.t-input :host-context(tui-text-area[data-size=\"s\"]):not(tui-text-area){padding:0 var(--tui-padding-s)}:host[data-size=l] .t-input,.t-input :host-context(tui-primitive-textfield[data-size=\"l\"]):not(tui-primitive-textfield),.t-input :host-context(tui-text-area[data-size=\"l\"]):not(tui-text-area){padding:0 var(--tui-padding-l)}:host._disabled .t-input,.t-input :host-context(tui-primitive-textfield._disabled),.t-input :host-context(tui-text-area._disabled){pointer-events:none}:host[data-size=l]:not(._label-outside) .t-input,.t-input :host-context(tui-primitive-textfield[data-size=\"l\"]:not(._label-outside)):not(tui-primitive-textfield){padding-top:1.25rem}:host[data-size=l]:not(._label-outside) .t-input:-webkit-autofill+.t-content .t-placeholder,.t-input :host-context(tui-primitive-textfield[data-size=\"l\"]:not(._label-outside)):not(tui-primitive-textfield):-webkit-autofill+.t-content .t-placeholder{font-size:.8156rem;transform:translateY(-.625rem)}:host[data-size=m]:not(._label-outside) .t-input,.t-input :host-context(tui-primitive-textfield[data-size=\"m\"]:not(._label-outside)):not(tui-primitive-textfield){padding-top:1.125rem}:host[data-size=m]:not(._label-outside) .t-input:-webkit-autofill+.t-content .t-placeholder,.t-input :host-context(tui-primitive-textfield[data-size=\"m\"]:not(._label-outside)):not(tui-primitive-textfield):-webkit-autofill+.t-content .t-placeholder{font-size:.69rem;transform:translateY(-.5rem)}:host._hidden input.t-input,.t-input :host-context(tui-primitive-textfield._hidden){opacity:0;text-indent:-10em;-webkit-user-select:none}.t-content{display:flex;height:100%;width:100%;padding:0 var(--tui-padding-m);box-sizing:border-box;align-items:center;overflow:hidden}:host[data-size=s] .t-content{padding:0 var(--tui-padding-s)}:host[data-size=l] .t-content{padding:0 var(--tui-padding-l)}.t-content:after{content:\"\";margin-right:-.25rem}:host[data-size=m] .t-content:after{display:none}.t-wrapper{flex:1;min-width:0;padding-right:.25rem;-webkit-padding-end:.25rem;padding-inline-end:.25rem;-webkit-padding-start:0;padding-inline-start:0}.t-wrapper+*{margin:0}.t-placeholder{transition-property:transform,font-size,color,letter-spacing;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;font:var(--tui-font-text-s);color:var(--tui-text-01);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;width:100%;-webkit-user-select:none;-moz-user-select:none;user-select:none;color:var(--tui-text-02);pointer-events:none;will-change:transform;transform:translateY(0)}.t-placeholder_raised{transform:translateY(-.625rem)}:host[data-size=m] .t-placeholder_raised{font:var(--tui-font-text-xs);transform:translateY(-.5rem);letter-spacing:.025rem}:host._invalid:not(._focused) .t-placeholder_raised,:host._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill)}:host[data-mode=onDark]._invalid:not(._focused) .t-placeholder_raised,:host[data-mode=onDark]._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill-night)}:host._focused .t-placeholder,:host[data-size=m]._focused._label-outside .t-placeholder,:host[data-size=l]._focused._label-outside .t-placeholder{color:var(--tui-text-03)}:host[data-size=l] .t-placeholder{font-size:.9375rem}:host[data-size=l] .t-placeholder_raised{font-size:.8156rem}:host[data-size=m]._focused:not(._label-outside) .t-placeholder,:host[data-size=l]._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01)}:host[data-mode=onDark] .t-placeholder{color:var(--tui-text-02-night)}:host[data-size=m][data-mode=onDark]._focused:not(._label-outside) .t-placeholder,:host[data-size=l][data-mode=onDark]._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01-night)}:host[data-mode=onDark]._focused .t-placeholder,:host[data-size=m][data-mode=onDark]._focused._label-outside .t-placeholder,:host[data-size=l][data-mode=onDark]._focused._label-outside .t-placeholder{color:var(--tui-text-02-night)}@supports (-webkit-hyphens: none){.t-placeholder{will-change:unset;transition-property:transform,color,letter-spacing}}.t-cleaner{position:relative;display:flex;width:1.5rem;height:1.5rem;margin:0 0 0 .25rem;-webkit-margin-start:.25rem;margin-inline-start:.25rem;-webkit-margin-end:0;margin-inline-end:0;align-items:center;justify-content:center;box-sizing:border-box;cursor:pointer}:host._readonly .t-cleaner,:host._disabled .t-cleaner{pointer-events:none}.t-icon{position:relative;display:flex;width:1.5rem;height:1.5rem;margin:0 0 0 .25rem;-webkit-margin-start:.25rem;margin-inline-start:.25rem;-webkit-margin-end:0;margin-inline-end:0;align-items:center;justify-content:center;box-sizing:border-box;cursor:pointer;pointer-events:none}.t-icon_left{margin:0 .375rem 0 -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.375rem;margin-inline-end:.375rem}:host[data-size=l] .t-icon_left{margin:0 .5rem 0 -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.5rem;margin-inline-end:.5rem}:host{cursor:text}:host._expandable{height:auto;max-height:none}:host._disabled{pointer-events:none}:host._readonly{cursor:default}.t-hosted{display:block;border-radius:inherit}.t-scrollbar{margin:0 -.25rem;min-width:100%}.t-tags{display:flex;padding-left:.25rem;-webkit-padding-start:.25rem;padding-inline-start:.25rem;padding-right:.5rem;-webkit-padding-end:.5rem;padding-inline-end:.5rem}.t-tags_expandable{overflow:hidden}:host._expandable .t-tags{flex-wrap:wrap;white-space:normal}:host._readonly .t-tags{pointer-events:none}.t-tags :host[data-size]{-webkit-padding-start:0;padding-inline-start:0;-webkit-padding-end:.25rem;padding-inline-end:.25rem}:host[data-size]:not(._label-outside) .t-tags{-webkit-padding-end:1rem;padding-inline-end:1rem}.t-content{align-items:flex-start}.t-icons{display:flex;align-items:center;padding:0 0 0 .75rem;-webkit-padding-start:.75rem;padding-inline-start:.75rem;-webkit-padding-end:0;padding-inline-end:0}.t-icons_left{padding:0;margin:0 .75rem 0 -.5rem;-webkit-margin-start:-.5rem;margin-inline-start:-.5rem;-webkit-margin-end:.75rem;margin-inline-end:.75rem}:host._expandable .t-icons_right{margin:0 0 0 -.625rem;-webkit-margin-start:-.625rem;margin-inline-start:-.625rem;-webkit-margin-end:0;margin-inline-end:0}:host[data-size=s] .t-icons{height:var(--tui-height-s)}:host[data-size=m] .t-icons{height:var(--tui-height-m)}:host[data-size=l] .t-icons{height:var(--tui-height-l)}.t-tooltip{margin:0 0 0 .25rem;-webkit-margin-start:.25rem;margin-inline-start:.25rem;-webkit-margin-end:0;margin-inline-end:0}.t-absolute-wrapper{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;flex-direction:column;justify-content:center}:host[data-size=m] .t-absolute-wrapper{height:var(--tui-height-m)}:host[data-size=l] .t-absolute-wrapper{height:var(--tui-height-l)}.t-wrapper{position:relative;display:flex;align-items:center;padding:0;box-sizing:border-box}:host[data-size=s] .t-wrapper{min-height:var(--tui-height-s);padding:.25rem 0}:host[data-size=m] .t-wrapper{min-height:var(--tui-height-m);padding:.5rem 0}:host[data-size=l] .t-wrapper{padding:.625rem 0}:host[data-size=m]:not(._label-outside) .t-wrapper{padding:1.1875rem 0 0}:host[data-size=l]:not(._label-outside) .t-wrapper{padding:1.6875rem 0 .5625rem}.t-tag{margin:.125rem .5rem .125rem -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.5rem;margin-inline-end:.5rem;max-width:100%;flex-shrink:0}.t-text:after{content:\",\\a0\"}.t-text_disabled{color:var(--tui-text-03)}.t-text_error{color:var(--tui-negative)}.t-text:not(.t-text_comma):last-of-type:after{content:\"\"}:host:not(._expandable) .t-text{white-space:nowrap}.t-input-wrapper{position:relative;flex:1;max-width:100%}.t-input-wrapper_collapsed{flex:0;margin:0 0 0 -.5rem;-webkit-margin-start:-.5rem;margin-inline-start:-.5rem;-webkit-margin-end:0;margin-inline-end:0}:host[data-size=s] .t-input-wrapper{min-height:1.5rem}:host[data-size=m] .t-input-wrapper{min-height:calc(var(--tui-height-xs) + 2 * .125rem)}:host[data-size=l] .t-input-wrapper{min-height:calc(var(--tui-height-s) + 2 * .125rem)}:host:not(._label-outside) .t-input-wrapper{min-height:1.25rem}.t-ghost{visibility:hidden;white-space:pre;text-overflow:clip;min-width:.125rem}:host:not(._expandable) .t-ghost{min-width:2rem}.t-native{margin:0;border:0;border-radius:inherit;background:none;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;outline:none;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;box-sizing:border-box;padding:0;cursor:inherit}.t-native:-webkit-autofill,.t-native:-webkit-autofill:hover,.t-native:-webkit-autofill:focus{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}.t-native_hidden{opacity:0;text-indent:-10em}.t-native::placeholder{color:var(--tui-text-03);opacity:0}:host[data-mode=onDark] .t-native::placeholder{color:var(--tui-text-03-night)}:host._focused .t-native:not(:-moz-read-only)::placeholder{opacity:1}:host._focused .t-native:not(:read-only)::placeholder{opacity:1}.t-error-icon{display:block;color:var(--tui-error-fill);width:1rem;height:1rem}\n"], components: [{ type: i1__namespace.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }, { type: i1__namespace.TuiScrollbarComponent, selector: "tui-scrollbar", inputs: ["hidden"] }, { type: i2__namespace.TuiTagComponent, selector: "tui-tag, a[tuiTag], button[tuiTag]", inputs: ["value", "editable", "separator", "maxLength", "size", "showLoader", "status", "hoverable", "removable", "disabled", "autoColor", "leftContent"], outputs: ["edited"] }, { type: i1__namespace.TuiTooltipComponent, selector: "tui-tooltip", inputs: ["content", "direction", "appearance", "showDelay", "hideDelay", "describeId"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i1__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }, { type: i5__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i3__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i4__namespace.TuiFocusableDirective, selector: "[tuiFocusable]", inputs: ["tuiFocusable"] }, { type: i6__namespace.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { type: i6__namespace.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i6__namespace.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }], pipes: { "async": i3__namespace.AsyncPipe }, viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "separator", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "search", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "editable", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "tagValidator", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "expandable", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "inputHidden", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "uniqueTags", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "maxLength", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "placeholder", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "removable", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputTagComponent.prototype, "disabledItemHandler", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTagComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-tag',
                        templateUrl: './input-tag.template.html',
                        styleUrls: ['./input-tag.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i4.tuiAsFocusableItemAccessor(TuiInputTagComponent),
                            i4.tuiAsControl(TuiInputTagComponent),
                            i1.tuiAsDataListHost(TuiInputTagComponent),
                            i1.TEXTFIELD_CONTROLLER_PROVIDER,
                            i1.MODE_PROVIDER,
                        ],
                        viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER],
                    }]
            }], ctorParameters: function () {
            return [{ type: i6__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i6.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i4__namespace.TuiScrollService, decorators: [{
                            type: i0.Inject,
                            args: [i4.TuiScrollService]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i1__namespace.TuiModeDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiModeDirective]
                        }] }, { type: i7__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_MODE]
                        }] }, { type: i1__namespace.TuiHintOptionsDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiHintOptionsDirective]
                        }] }, { type: i1__namespace.TuiTextfieldController, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_WATCHED_CONTROLLER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_INPUT_TAG_OPTIONS]
                        }] }, { type: i1__namespace.TuiHostedDropdownComponent, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiHostedDropdownComponent]
                        }] }];
        }, propDecorators: { hostedDropdown: [{
                    type: i0.ViewChild,
                    args: [i1.TuiHostedDropdownComponent]
                }], focusableElement: [{
                    type: i0.ViewChild,
                    args: ['focusableElement']
                }], tagsContainer: [{
                    type: i0.ViewChild,
                    args: ['tagsContainer']
                }], tags: [{
                    type: i0.ViewChildren,
                    args: ['tag', { read: i0.ElementRef }]
                }], cleanerSvg: [{
                    type: i0.ViewChild,
                    args: ['cleaner', { read: i0.ElementRef }]
                }], scrollBar: [{
                    type: i0.ViewChild,
                    args: [i1.TuiScrollbarComponent, { read: i0.ElementRef }]
                }], separator: [{
                    type: i0.Input
                }], search: [{
                    type: i0.Input
                }], editable: [{
                    type: i0.Input
                }], tagValidator: [{
                    type: i0.Input
                }], expandable: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._expandable']
                }], rows: [{
                    type: i0.Input
                }], inputHidden: [{
                    type: i0.Input
                }], uniqueTags: [{
                    type: i0.Input
                }], maxLength: [{
                    type: i0.Input
                }], placeholder: [{
                    type: i0.Input
                }], removable: [{
                    type: i0.Input
                }], disabledItemHandler: [{
                    type: i0.Input
                }], pseudoFocusedSetter: [{
                    type: i0.Input,
                    args: ['pseudoFocused']
                }], searchChange: [{
                    type: i0.Output
                }], datalist: [{
                    type: i0.ContentChild,
                    args: [i1.TuiDataListDirective, { read: i0.TemplateRef }]
                }], errorIconTemplate: [{
                    type: i0.ViewChild,
                    args: ['errorIcon']
                }], scrollerSetter: [{
                    type: i0.ViewChild,
                    args: [i1.TuiScrollbarComponent]
                }], size: [{
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], labelOutside: [{
                    type: i0.HostBinding,
                    args: ['class._label-outside']
                }] } });

    var TuiInputTagModule = /** @class */ (function () {
        function TuiInputTagModule() {
        }
        return TuiInputTagModule;
    }());
    TuiInputTagModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTagModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputTagModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTagModule, declarations: [TuiInputTagComponent], imports: [i3.CommonModule,
            i6.FormsModule,
            i5.PolymorpheusModule,
            i4.TuiFocusableModule,
            i4.TuiHoveredModule,
            i4.TuiActiveZoneModule,
            i1.TuiSvgModule,
            i1.TuiScrollbarModule,
            i1.TuiTooltipModule,
            i1.TuiHostedDropdownModule,
            i2.TuiTagModule,
            i1.TuiWrapperModule], exports: [TuiInputTagComponent] });
    TuiInputTagModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTagModule, providers: [i4.TuiScrollService], imports: [[
                i3.CommonModule,
                i6.FormsModule,
                i5.PolymorpheusModule,
                i4.TuiFocusableModule,
                i4.TuiHoveredModule,
                i4.TuiActiveZoneModule,
                i1.TuiSvgModule,
                i1.TuiScrollbarModule,
                i1.TuiTooltipModule,
                i1.TuiHostedDropdownModule,
                i2.TuiTagModule,
                i1.TuiWrapperModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTagModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            i6.FormsModule,
                            i5.PolymorpheusModule,
                            i4.TuiFocusableModule,
                            i4.TuiHoveredModule,
                            i4.TuiActiveZoneModule,
                            i1.TuiSvgModule,
                            i1.TuiScrollbarModule,
                            i1.TuiTooltipModule,
                            i1.TuiHostedDropdownModule,
                            i2.TuiTagModule,
                            i1.TuiWrapperModule,
                        ],
                        providers: [i4.TuiScrollService],
                        declarations: [TuiInputTagComponent],
                        exports: [TuiInputTagComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_INPUT_TAG_DEFAULT_OPTIONS = TUI_INPUT_TAG_DEFAULT_OPTIONS;
    exports.TUI_INPUT_TAG_OPTIONS = TUI_INPUT_TAG_OPTIONS;
    exports.TuiInputTagComponent = TuiInputTagComponent;
    exports.TuiInputTagModule = TuiInputTagModule;
    exports.tuiInputTagOptionsProvider = tuiInputTagOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
