// @taiga-ui/kit/bundles/taiga-ui-kit-directives-present.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/directives/present', ['exports', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.directives = global["taiga-ui"].kit.directives || {}, global["taiga-ui"].kit.directives.present = {}), global.ng.core, global.common, global.cdk, global.rxjs, global.rxjs.operators));
})(this, (function (exports, i0, common, cdk, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiPresentDirective = /** @class */ (function () {
        function TuiPresentDirective(_a, userAgent) {
            var _this = this;
            var nativeElement = _a.nativeElement;
            this.visibility$ = new rxjs.BehaviorSubject(false);
            this.tuiPresentChange = this.visibility$.pipe(operators.distinctUntilChanged(), operators.skip(1));
            if (cdk.tuiIsFirefox(userAgent)) {
                return;
            }
            var observer = new MutationObserver(function () {
                if (!nativeElement.offsetParent &&
                    nativeElement.offsetWidth === 0 &&
                    nativeElement.offsetHeight === 0) {
                    _this.visibility$.next(false);
                }
            });
            observer.observe(nativeElement, {
                attributes: true,
                attributeFilter: ['style', 'class'],
            });
        }
        /**
         * Someday animationcancel would work and mutation observer would not be needed:
         * https://drafts.csswg.org/css-animations/#eventdef-animationevent-animationcancel
         * It would also trigger on CSS like display: none on parent nodes which is awesome
         * but currently only works in Firefox
         * ___
         * TODO: remove MutationObserver when we bump versions of supported browsers:
         *** Safari 12+
         *** Chrome 83+
         * See: {@link https://caniuse.com/mdn-api_window_animationcancel_event}
         */
        TuiPresentDirective.prototype.onAnimation = function (visibility) {
            this.visibility$.next(visibility);
        };
        TuiPresentDirective.prototype.ngOnDestroy = function () {
            this.visibility$.next(false);
        };
        return TuiPresentDirective;
    }());
    TuiPresentDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPresentDirective, deps: [{ token: i0.ElementRef }, { token: common.USER_AGENT }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPresentDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPresentDirective, selector: "[tuiPresentChange]", outputs: { tuiPresentChange: "tuiPresentChange" }, host: { listeners: { "animationcancel.self": "onAnimation(false)", "animationstart.self": "onAnimation(true)" }, properties: { "style.animation": "\"tuiPresent 1s infinite\"" } }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPresentDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPresentChange]',
                        host: {
                            '[style.animation]': '"tuiPresent 1s infinite"',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [common.USER_AGENT]
                        }] }];
        }, propDecorators: { tuiPresentChange: [{
                    type: i0.Output
                }], onAnimation: [{
                    type: i0.HostListener,
                    args: ['animationcancel.self', ['false']]
                }, {
                    type: i0.HostListener,
                    args: ['animationstart.self', ['true']]
                }] } });

    var TuiPresentModule = /** @class */ (function () {
        function TuiPresentModule() {
        }
        return TuiPresentModule;
    }());
    TuiPresentModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPresentModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPresentModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPresentModule, declarations: [TuiPresentDirective], exports: [TuiPresentDirective] });
    TuiPresentModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPresentModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPresentModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiPresentDirective],
                        exports: [TuiPresentDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPresentDirective = TuiPresentDirective;
    exports.TuiPresentModule = TuiPresentModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
