// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-phone-international.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/i18n'), require('@taiga-ui/kit/components/arrow'), require('@taiga-ui/kit/components/input-phone'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/tokens'), require('@angular/common'), require('@tinkoff/ng-polymorpheus'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-phone-international', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/i18n', '@taiga-ui/kit/components/arrow', '@taiga-ui/kit/components/input-phone', '@taiga-ui/kit/providers', '@taiga-ui/kit/tokens', '@angular/common', '@tinkoff/ng-polymorpheus', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-phone-international"] = {}), global.ng.core, global.ng.forms, global.i4, global.i1, global.i18n, global["taiga-ui"].kit.components.arrow, global["taiga-ui"].kit.components["input-phone"], global["taiga-ui"].kit.providers, global["taiga-ui"].kit.tokens, global.ng.common, global.i6, global.rxjs));
})(this, (function (exports, i0, i5, i4, i1, i18n, arrow, i2, providers, tokens, i3, i6, i7) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);

    var MASK_AFTER_CODE_REGEXP = /\([#]+\)|[#\- ]/g;

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS = {
        countries: [],
        countryIsoCode: i18n.TuiCountryIsoCode.RU,
    };
    /**
     * Default parameters for input phone international component
     */
    var TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS = new i0.InjectionToken("[TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS]", {
        factory: function () { return TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS; },
    });
    var tuiInputPhoneInternationalOptionsProvider = function (options) { return ({
        provide: TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS,
        useValue: Object.assign(Object.assign({}, TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS), options),
    }); };

    var TUI_COUNTRIES_MASKS = new i0.InjectionToken("[TUI_COUNTRIES_MASKS]", {
        factory: function () {
            var _a;
            return (_a = {},
                _a[i18n.TuiCountryIsoCode.AD] = "+376###-###",
                _a[i18n.TuiCountryIsoCode.AE] = "+971-##-###-####",
                _a[i18n.TuiCountryIsoCode.AF] = "+93##-###-####",
                _a[i18n.TuiCountryIsoCode.AG] = "+1(268) ###-####",
                _a[i18n.TuiCountryIsoCode.AI] = "+1(264) ###-####",
                _a[i18n.TuiCountryIsoCode.AL] = "+355(###) ###-###",
                _a[i18n.TuiCountryIsoCode.AM] = "+374##-###-###",
                _a[i18n.TuiCountryIsoCode.AO] = "+244(###) ###-###",
                _a[i18n.TuiCountryIsoCode.AR] = "+54(###) ###-####",
                _a[i18n.TuiCountryIsoCode.AT] = "+43(###) ###-####",
                _a[i18n.TuiCountryIsoCode.AU] = "+61#-####-####",
                _a[i18n.TuiCountryIsoCode.AW] = "+297###-####",
                _a[i18n.TuiCountryIsoCode.AZ] = "+994##-###-##-##",
                _a[i18n.TuiCountryIsoCode.BA] = "+387 ###-####-##",
                _a[i18n.TuiCountryIsoCode.BB] = "+1(246) ###-####",
                _a[i18n.TuiCountryIsoCode.BD] = "+880##-###-###",
                _a[i18n.TuiCountryIsoCode.BE] = "+32(###) ###-###",
                _a[i18n.TuiCountryIsoCode.BF] = "+226##-##-####",
                _a[i18n.TuiCountryIsoCode.BG] = "+359(###) ###-###",
                _a[i18n.TuiCountryIsoCode.BH] = "+973####-####",
                _a[i18n.TuiCountryIsoCode.BI] = "+257##-##-####",
                _a[i18n.TuiCountryIsoCode.BJ] = "+229##-##-####",
                _a[i18n.TuiCountryIsoCode.BL] = "+590 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.BM] = "+1(441) ###-####",
                _a[i18n.TuiCountryIsoCode.BN] = "+673###-####",
                _a[i18n.TuiCountryIsoCode.BO] = "+591#-###-####",
                _a[i18n.TuiCountryIsoCode.BQ] = "+599-###-####",
                _a[i18n.TuiCountryIsoCode.BR] = "+55(##) ####-####",
                _a[i18n.TuiCountryIsoCode.BS] = "+1(242) ###-####",
                _a[i18n.TuiCountryIsoCode.BT] = "+975#-###-###",
                _a[i18n.TuiCountryIsoCode.BW] = "+267##-###-###",
                _a[i18n.TuiCountryIsoCode.BY] = "+375(##) ###-##-##",
                _a[i18n.TuiCountryIsoCode.BZ] = "+501###-####",
                _a[i18n.TuiCountryIsoCode.CA] = "+1(###) ###-####",
                _a[i18n.TuiCountryIsoCode.CD] = "+243(###) ###-###",
                _a[i18n.TuiCountryIsoCode.CF] = "+236-##-##-####",
                _a[i18n.TuiCountryIsoCode.CG] = "+242##-###-####",
                _a[i18n.TuiCountryIsoCode.CH] = "+41##-###-####",
                _a[i18n.TuiCountryIsoCode.CI] = "+225##-##-###-###",
                _a[i18n.TuiCountryIsoCode.CL] = "+56#-####-####",
                _a[i18n.TuiCountryIsoCode.CM] = "+237#####-####",
                _a[i18n.TuiCountryIsoCode.CN] = "+86(###) ####-####",
                _a[i18n.TuiCountryIsoCode.CO] = "+57(###) ###-####",
                _a[i18n.TuiCountryIsoCode.CR] = "+506####-####",
                _a[i18n.TuiCountryIsoCode.CU] = "+53#-###-####",
                _a[i18n.TuiCountryIsoCode.CV] = "+238(###) ##-##",
                _a[i18n.TuiCountryIsoCode.CW] = "+5999-###-####",
                _a[i18n.TuiCountryIsoCode.CY] = "+357##-###-###",
                _a[i18n.TuiCountryIsoCode.CZ] = "+420(###) ###-###",
                _a[i18n.TuiCountryIsoCode.DE] = "+49(###) ###-###-##",
                _a[i18n.TuiCountryIsoCode.DJ] = "+253##-##-##-##",
                _a[i18n.TuiCountryIsoCode.DK] = "+45##-##-##-##",
                _a[i18n.TuiCountryIsoCode.DM] = "+1(767) ###-####",
                _a[i18n.TuiCountryIsoCode.DO] = "+1(###) ###-####",
                _a[i18n.TuiCountryIsoCode.DZ] = "+213##-###-####",
                _a[i18n.TuiCountryIsoCode.EC] = "+593##-###-####",
                _a[i18n.TuiCountryIsoCode.EE] = "+372####-####",
                _a[i18n.TuiCountryIsoCode.EG] = "+20(###) ###-####",
                _a[i18n.TuiCountryIsoCode.ER] = "+291#-###-###",
                _a[i18n.TuiCountryIsoCode.ES] = "+34(###) ###-###",
                _a[i18n.TuiCountryIsoCode.ET] = "+251##-###-####",
                _a[i18n.TuiCountryIsoCode.FI] = "+358(###) ###-##-##",
                _a[i18n.TuiCountryIsoCode.FJ] = "+679##-#####",
                _a[i18n.TuiCountryIsoCode.FK] = "+500#####",
                _a[i18n.TuiCountryIsoCode.FM] = "+691###-####",
                _a[i18n.TuiCountryIsoCode.FR] = "+33 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.GA] = "+241##-##-##-##",
                _a[i18n.TuiCountryIsoCode.GB] = "+44##-####-####",
                _a[i18n.TuiCountryIsoCode.GD] = "+1(473) ###-####",
                _a[i18n.TuiCountryIsoCode.GE] = "+995(###) ###-###",
                _a[i18n.TuiCountryIsoCode.GF] = "+594 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.GH] = "+233(###) ###-###",
                _a[i18n.TuiCountryIsoCode.GI] = "+350###-#####",
                _a[i18n.TuiCountryIsoCode.GL] = "+299##-##-##",
                _a[i18n.TuiCountryIsoCode.GM] = "+220(###) ##-##",
                _a[i18n.TuiCountryIsoCode.GN] = "+224##-###-###",
                _a[i18n.TuiCountryIsoCode.GP] = "+590 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.GQ] = "+240##-###-####",
                _a[i18n.TuiCountryIsoCode.GR] = "+30(###) ###-####",
                _a[i18n.TuiCountryIsoCode.GT] = "+502#-###-####",
                _a[i18n.TuiCountryIsoCode.GW] = "+245#-######",
                _a[i18n.TuiCountryIsoCode.GY] = "+592###-####",
                _a[i18n.TuiCountryIsoCode.HK] = "+852####-####",
                _a[i18n.TuiCountryIsoCode.HN] = "+504####-####",
                _a[i18n.TuiCountryIsoCode.HR] = "+385##-###-####",
                _a[i18n.TuiCountryIsoCode.HT] = "+509##-##-####",
                _a[i18n.TuiCountryIsoCode.HU] = "+36(###) ###-###",
                _a[i18n.TuiCountryIsoCode.ID] = "+62(###) ###-##-###",
                _a[i18n.TuiCountryIsoCode.IE] = "+353(###) ###-###",
                _a[i18n.TuiCountryIsoCode.IL] = "+972##-###-####",
                _a[i18n.TuiCountryIsoCode.IN] = "+91(####) ###-###",
                _a[i18n.TuiCountryIsoCode.IQ] = "+964(###) ###-####",
                _a[i18n.TuiCountryIsoCode.IR] = "+98(###) ###-####",
                _a[i18n.TuiCountryIsoCode.IS] = "+354###-####",
                _a[i18n.TuiCountryIsoCode.IT] = "+39(###) ####-###",
                _a[i18n.TuiCountryIsoCode.JM] = "+1(876) ###-####",
                _a[i18n.TuiCountryIsoCode.JO] = "+962#-####-####",
                _a[i18n.TuiCountryIsoCode.JP] = "+81-##-####-####",
                _a[i18n.TuiCountryIsoCode.KE] = "+254###-######",
                _a[i18n.TuiCountryIsoCode.KG] = "+996(###) ###-###",
                _a[i18n.TuiCountryIsoCode.KH] = "+855##-###-###",
                _a[i18n.TuiCountryIsoCode.KM] = "+269##-#####",
                _a[i18n.TuiCountryIsoCode.KN] = "+1(869) ###-####",
                _a[i18n.TuiCountryIsoCode.KP] = "+850####-#############",
                _a[i18n.TuiCountryIsoCode.KR] = "+82##-###-####",
                _a[i18n.TuiCountryIsoCode.KW] = "+965####-####",
                _a[i18n.TuiCountryIsoCode.KY] = "+1(345) ###-####",
                _a[i18n.TuiCountryIsoCode.KZ] = "+7(###) ###-##-##",
                _a[i18n.TuiCountryIsoCode.LA] = "+856##-##-###-###",
                _a[i18n.TuiCountryIsoCode.LB] = "+961##-###-###",
                _a[i18n.TuiCountryIsoCode.LC] = "+1(758) ###-####",
                _a[i18n.TuiCountryIsoCode.LI] = "+423(###) ###-####",
                _a[i18n.TuiCountryIsoCode.LK] = "+94##-###-####",
                _a[i18n.TuiCountryIsoCode.LR] = "+231##-###-###",
                _a[i18n.TuiCountryIsoCode.LS] = "+266#-###-####",
                _a[i18n.TuiCountryIsoCode.LT] = "+370(###) ##-###",
                _a[i18n.TuiCountryIsoCode.LU] = "+352(###) ###-###",
                _a[i18n.TuiCountryIsoCode.LV] = "+371##-###-###",
                _a[i18n.TuiCountryIsoCode.LY] = "+218##-###-####",
                _a[i18n.TuiCountryIsoCode.MA] = "+212##-####-###",
                _a[i18n.TuiCountryIsoCode.MC] = "+377###-###-###",
                _a[i18n.TuiCountryIsoCode.MD] = "+373####-####",
                _a[i18n.TuiCountryIsoCode.ME] = "+382##-###-###",
                _a[i18n.TuiCountryIsoCode.MF] = "+590 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.MG] = "+261##-##-#####",
                _a[i18n.TuiCountryIsoCode.MK] = "+389##-###-###",
                _a[i18n.TuiCountryIsoCode.ML] = "+223##-##-####",
                _a[i18n.TuiCountryIsoCode.MM] = "+95##-###-###",
                _a[i18n.TuiCountryIsoCode.MN] = "+976##-##-####",
                _a[i18n.TuiCountryIsoCode.MO] = "+853####-####",
                _a[i18n.TuiCountryIsoCode.MQ] = "+596 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.MR] = "+222##-##-####",
                _a[i18n.TuiCountryIsoCode.MS] = "+1(664) ###-####",
                _a[i18n.TuiCountryIsoCode.MT] = "+356####-####",
                _a[i18n.TuiCountryIsoCode.MU] = "+230####-####",
                _a[i18n.TuiCountryIsoCode.MV] = "+960###-####",
                _a[i18n.TuiCountryIsoCode.MW] = "+265#-####-####",
                _a[i18n.TuiCountryIsoCode.MX] = "+52(###) ###-####",
                _a[i18n.TuiCountryIsoCode.MY] = "+60(###) ###-###",
                _a[i18n.TuiCountryIsoCode.MZ] = "+258##-###-###",
                _a[i18n.TuiCountryIsoCode.NA] = "+264##-###-####",
                _a[i18n.TuiCountryIsoCode.NC] = "+687 ### ###",
                _a[i18n.TuiCountryIsoCode.NE] = "+227##-##-####",
                _a[i18n.TuiCountryIsoCode.NG] = "+234(###) ###-####",
                _a[i18n.TuiCountryIsoCode.NI] = "+505####-####",
                _a[i18n.TuiCountryIsoCode.NL] = "+31##-###-####",
                _a[i18n.TuiCountryIsoCode.NO] = "+47(###) ##-###",
                _a[i18n.TuiCountryIsoCode.NP] = "+977##-###-###",
                _a[i18n.TuiCountryIsoCode.NZ] = "+64(###) ###-####",
                _a[i18n.TuiCountryIsoCode.OM] = "+968##-###-###",
                _a[i18n.TuiCountryIsoCode.PA] = "+507###-####",
                _a[i18n.TuiCountryIsoCode.PE] = "+51(###) ###-###",
                _a[i18n.TuiCountryIsoCode.PF] = "+689 ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.PG] = "+675(###) ##-###",
                _a[i18n.TuiCountryIsoCode.PH] = "+63(###) ###-####",
                _a[i18n.TuiCountryIsoCode.PK] = "+92(###) ###-####",
                _a[i18n.TuiCountryIsoCode.PL] = "+48(###) ###-###",
                _a[i18n.TuiCountryIsoCode.PT] = "+351##-###-####",
                _a[i18n.TuiCountryIsoCode.PW] = "+680###-####",
                _a[i18n.TuiCountryIsoCode.PY] = "+595(###) ###-###",
                _a[i18n.TuiCountryIsoCode.QA] = "+974####-####",
                _a[i18n.TuiCountryIsoCode.RE] = "+262 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.RO] = "+40##-###-####",
                _a[i18n.TuiCountryIsoCode.RS] = "+381##-###-####",
                _a[i18n.TuiCountryIsoCode.RU] = "+7### ###-##-##",
                _a[i18n.TuiCountryIsoCode.RW] = "+250(###) ###-###",
                _a[i18n.TuiCountryIsoCode.SA] = "+966#-####-####",
                _a[i18n.TuiCountryIsoCode.SB] = "+677###-####",
                _a[i18n.TuiCountryIsoCode.SC] = "+248#-###-###",
                _a[i18n.TuiCountryIsoCode.SD] = "+249##-###-####",
                _a[i18n.TuiCountryIsoCode.SE] = "+46##-###-####",
                _a[i18n.TuiCountryIsoCode.SG] = "+65####-####",
                _a[i18n.TuiCountryIsoCode.SH] = "+290####",
                _a[i18n.TuiCountryIsoCode.SI] = "+386##-###-###",
                _a[i18n.TuiCountryIsoCode.SK] = "+421(###) ###-###",
                _a[i18n.TuiCountryIsoCode.SL] = "+232##-######",
                _a[i18n.TuiCountryIsoCode.SM] = "+378####-######",
                _a[i18n.TuiCountryIsoCode.SN] = "+221##-###-####",
                _a[i18n.TuiCountryIsoCode.SO] = "+252##-###-###",
                _a[i18n.TuiCountryIsoCode.SR] = "+597###-####",
                _a[i18n.TuiCountryIsoCode.ST] = "+239##-#####",
                _a[i18n.TuiCountryIsoCode.SV] = "+503##-##-####",
                _a[i18n.TuiCountryIsoCode.SX] = "+1(721) ###-####",
                _a[i18n.TuiCountryIsoCode.SY] = "+963##-####-###",
                _a[i18n.TuiCountryIsoCode.SZ] = "+268##-##-####",
                _a[i18n.TuiCountryIsoCode.TC] = "+1(649) ###-####",
                _a[i18n.TuiCountryIsoCode.TD] = "+235##-##-##-##",
                _a[i18n.TuiCountryIsoCode.TG] = "+228##-###-###",
                _a[i18n.TuiCountryIsoCode.TH] = "+66##-###-####",
                _a[i18n.TuiCountryIsoCode.TJ] = "+992##-###-####",
                _a[i18n.TuiCountryIsoCode.TL] = "+670###-#####",
                _a[i18n.TuiCountryIsoCode.TM] = "+993#-###-####",
                _a[i18n.TuiCountryIsoCode.TN] = "+216##-###-###",
                _a[i18n.TuiCountryIsoCode.TO] = "+676#####",
                _a[i18n.TuiCountryIsoCode.TR] = "+90(###) ###-####",
                _a[i18n.TuiCountryIsoCode.TT] = "+1(868) ###-####",
                _a[i18n.TuiCountryIsoCode.TW] = "+886#-####-####",
                _a[i18n.TuiCountryIsoCode.TZ] = "+255##-###-####",
                _a[i18n.TuiCountryIsoCode.UA] = "+380(##) ###-##-##",
                _a[i18n.TuiCountryIsoCode.UG] = "+256(###) ###-###",
                _a[i18n.TuiCountryIsoCode.US] = "+1(###) ###-####",
                _a[i18n.TuiCountryIsoCode.UY] = "+598#-###-##-##",
                _a[i18n.TuiCountryIsoCode.UZ] = "+998##-###-####",
                _a[i18n.TuiCountryIsoCode.VC] = "+1(784) ###-####",
                _a[i18n.TuiCountryIsoCode.VE] = "+58(###) ###-####",
                _a[i18n.TuiCountryIsoCode.VG] = "+1(284)###-####",
                _a[i18n.TuiCountryIsoCode.VN] = "+84(###) ####-###",
                _a[i18n.TuiCountryIsoCode.VU] = "+678##-#####",
                _a[i18n.TuiCountryIsoCode.WS] = "+685##-####",
                _a[i18n.TuiCountryIsoCode.XK] = "+383##-###-###",
                _a[i18n.TuiCountryIsoCode.YE] = "+967###-###-###",
                _a[i18n.TuiCountryIsoCode.YT] = "+262 ## ## ## ## ##",
                _a[i18n.TuiCountryIsoCode.ZA] = "+27##-###-####",
                _a[i18n.TuiCountryIsoCode.ZM] = "+260##-###-####",
                _a[i18n.TuiCountryIsoCode.ZW] = "+263#-######",
                _a);
        },
    });

    function tuiExtractValueFromEvent(event) {
        var _a;
        return "dataTransfer" in event
            ? ((_a = event.dataTransfer) === null || _a === void 0 ? void 0 : _a.getData("text/plain")) || ""
            : i4.tuiGetClipboardDataText(event);
    }

    function tuiNotKzRegion(value) {
        var region = Number(value.slice(1, 4));
        return region < 600 || region > 799;
    }

    var TuiInputPhoneInternationalComponent = /** @class */ (function (_super) {
        __extends(TuiInputPhoneInternationalComponent, _super);
        function TuiInputPhoneInternationalComponent(control, cdr, countriesNames$, countriesMasks, options, flagPipe) {
            var _this = _super.call(this, control, cdr) || this;
            _this.countriesNames$ = countriesNames$;
            _this.countriesMasks = countriesMasks;
            _this.options = options;
            _this.flagPipe = flagPipe;
            _this.countries = _this.options.countries;
            _this.countryIsoCodeChange = new i0.EventEmitter();
            _this.countryIsoCode = _this.options.countryIsoCode;
            _this.open = false;
            _this.arrow = arrow.TUI_ARROW;
            _this.isoToCountryCodeMapper = function (item) { return _this.isoToCountryCode(item); };
            return _this;
        }
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "isoCode", {
            set: function (code) {
                var _a;
                if (this.countryIsoCode === code) {
                    return;
                }
                (_a = this.inputPhoneComponent) === null || _a === void 0 ? void 0 : _a.writeValue(this.value);
                this.countryIsoCode = code;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.inputPhoneComponent && !this.computedDisabled
                    ? this.inputPhoneComponent.nativeFocusableElement
                    : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "focused", {
            get: function () {
                return ((!!this.primitiveTextfield && this.primitiveTextfield.focused) ||
                    (!!this.inputPhoneComponent && this.inputPhoneComponent.focused));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "inputPhoneCountryCode", {
            get: function () {
                return this.isoToCountryCode(this.countryIsoCode);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "phoneMaskAfterCountryCode", {
            get: function () {
                var countryCode = this.isoToCountryCode(this.countryIsoCode);
                return this.calculateMaskAfterCountryCode(this.countriesMasks[this.countryIsoCode], countryCode);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "countryFlagPath", {
            /**
             * @deprecated use `<img [src]="countryIsoCode | tuiFlagPipe" />`
             * TODO drop in v4.0
             */
            get: function () {
                return this.getFlagPath(this.countryIsoCode);
            },
            enumerable: false,
            configurable: true
        });
        TuiInputPhoneInternationalComponent.prototype.onPaste = function (event) {
            var value = tuiExtractValueFromEvent(event).replace(i1.TUI_NON_DIGITS_REGEXP, '');
            var countryIsoCode = this.extractCountryCode(value);
            if (!countryIsoCode) {
                this.value = ("" + this.inputPhoneCountryCode + value)
                    .replace(i1.TUI_MASK_SYMBOLS_REGEXP, '')
                    .slice(0, this.getMaxAllowedLength(this.countryIsoCode));
                return;
            }
            if (countryIsoCode === i18n.TuiCountryIsoCode.RU) {
                value = value.replace(/^8/, '7');
            }
            this.updateCountryIsoCode(countryIsoCode);
            this.value = "" + i4.CHAR_PLUS + value;
        };
        /**
         * @deprecated use `<img [src]="countryIsoCode | tuiFlagPipe" />`
         * TODO drop in v4.0
         */
        TuiInputPhoneInternationalComponent.prototype.getFlagPath = function (code) {
            return this.flagPipe.transform(code);
        };
        TuiInputPhoneInternationalComponent.prototype.onItemClick = function (isoCode) {
            this.open = false;
            this.updateCountryIsoCode(isoCode);
            // recalculates mask inside inputPhone to prevent isoCode conflict
            this.cdr.detectChanges();
            var maxLength = this.getMaxAllowedLength(isoCode);
            if (this.value.length > maxLength) {
                this.value = this.value.slice(0, maxLength);
            }
            if (this.nativeFocusableElement) {
                this.nativeFocusableElement.focus();
            }
        };
        TuiInputPhoneInternationalComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.close();
        };
        TuiInputPhoneInternationalComponent.prototype.isoToCountryCode = function (isoCode) {
            return this.countriesMasks[isoCode].replace(MASK_AFTER_CODE_REGEXP, '');
        };
        /** @deprecated use 'value' setter */
        TuiInputPhoneInternationalComponent.prototype.onModelChange = function (value) {
            this.value = value;
        };
        TuiInputPhoneInternationalComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
        };
        TuiInputPhoneInternationalComponent.prototype.getFallbackValue = function () {
            return '';
        };
        TuiInputPhoneInternationalComponent.prototype.calculateMaskAfterCountryCode = function (mask, countryCode) {
            return mask.replace(countryCode, '').trim();
        };
        TuiInputPhoneInternationalComponent.prototype.close = function () {
            this.open = false;
        };
        TuiInputPhoneInternationalComponent.prototype.getMaxAllowedLength = function (isoCode) {
            return this.countriesMasks[isoCode].replace(/[()\- ]/g, '').length;
        };
        TuiInputPhoneInternationalComponent.prototype.updateCountryIsoCode = function (code) {
            this.countryIsoCode = code;
            this.countryIsoCodeChange.emit(code);
        };
        TuiInputPhoneInternationalComponent.prototype.extractCountryCode = function (value) {
            var _this = this;
            return this.countries.find(function (countryIsoCode) {
                var ruCodeTest = countryIsoCode === i18n.TuiCountryIsoCode.RU &&
                    /^[7 | 8]/.test(value) &&
                    /^(?!880[1-9 ])/.test(value) &&
                    value.length + 1 === _this.getMaxAllowedLength(i18n.TuiCountryIsoCode.RU);
                var matched = ruCodeTest ||
                    (value.startsWith(_this.isoToCountryCode(countryIsoCode).replace(i4.CHAR_PLUS, '')) &&
                        value.length + 1 === _this.getMaxAllowedLength(countryIsoCode));
                return matched && countryIsoCode === i18n.TuiCountryIsoCode.RU
                    ? tuiNotKzRegion(value)
                    : matched;
            });
        };
        return TuiInputPhoneInternationalComponent;
    }(i4.AbstractTuiControl));
    TuiInputPhoneInternationalComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputPhoneInternationalComponent, deps: [{ token: i5.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: tokens.TUI_COUNTRIES }, { token: TUI_COUNTRIES_MASKS }, { token: TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS }, { token: i1.TuiFlagPipe }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputPhoneInternationalComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputPhoneInternationalComponent, selector: "tui-input-phone-international", inputs: { isoCode: ["countryIsoCode", "isoCode"], countries: "countries" }, outputs: { countryIsoCodeChange: "countryIsoCodeChange" }, host: { listeners: { "paste.capture.prevent.stop": "onPaste($event)", "drop.capture.prevent.stop": "onPaste($event)" } }, providers: [
            i4.tuiAsFocusableItemAccessor(TuiInputPhoneInternationalComponent),
            i4.tuiAsControl(TuiInputPhoneInternationalComponent),
            // TODO: for backward compatibility only. Drop in v4.0
            i1.TuiFlagPipe,
        ], viewQueries: [{ propertyName: "inputPhoneComponent", first: true, predicate: i2.TuiInputPhoneComponent, descendants: true }, { propertyName: "primitiveTextfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-hosted-dropdown\n    *ngIf=\"countriesNames$ | async as countriesNames\"\n    class=\"t-hosted-dropdown\"\n    [content]=\"dropdown\"\n    [canOpen]=\"!readOnly\"\n    [(open)]=\"open\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <div tuiGroup>\n        <tui-primitive-textfield\n            tuiHintContent=\"\"\n            tuiTextfieldPrefix=\"\"\n            tuiTextfieldPostfix=\"\"\n            class=\"t-country-select tui-group__auto-width-item\"\n            [tuiTextfieldCustomContent]=\"countryValueContent\"\n            [tuiTextfieldLabelOutside]=\"true\"\n            [tuiTextfieldIcon]=\"icon\"\n            [disabled]=\"disabled\"\n            [focusable]=\"focusable\"\n            [editable]=\"false\"\n            [pseudoFocus]=\"open || null\"\n            [readOnly]=\"readOnly\"\n        ></tui-primitive-textfield>\n        <tui-input-phone\n            class=\"t-input-phone tui-group__auto-width-item\"\n            [disabled]=\"disabled\"\n            [focusable]=\"focusable\"\n            [countryCode]=\"inputPhoneCountryCode\"\n            [phoneMaskAfterCountryCode]=\"phoneMaskAfterCountryCode\"\n            [readOnly]=\"readOnly\"\n            [pseudoInvalid]=\"computedInvalid\"\n            [pseudoFocus]=\"pseudoFocus\"\n            [pseudoHover]=\"pseudoHover\"\n            [(ngModel)]=\"value\"\n        >\n            <ng-content></ng-content>\n            <input\n                tuiTextfield\n                autocomplete=\"new-password\"\n            />\n        </tui-input-phone>\n    </div>\n\n    <ng-template #dropdown>\n        <tui-data-list>\n            <button\n                *ngFor=\"let item of countries\"\n                tuiOption\n                (click)=\"onItemClick(item)\"\n            >\n                <img\n                    alt=\"\"\n                    class=\"t-country-item-flag\"\n                    [src]=\"item | tuiFlag\"\n                />\n                <span class=\"t-country-item-name\">\n                    {{ countriesNames[item] }}\n                </span>\n                <span class=\"t-country-item-code\">\n                    {{ item | tuiMapper : isoToCountryCodeMapper }}\n                </span>\n            </button>\n        </tui-data-list>\n    </ng-template>\n\n    <ng-template #countryValueContent>\n        <!-- eslint-disable @html-eslint/require-img-alt -->\n        <img\n            class=\"t-flag\"\n            [alt]=\"countriesNames[countryIsoCode]\"\n            [src]=\"countryIsoCode | tuiFlag\"\n        />\n    </ng-template>\n\n    <ng-template #icon>\n        <div\n            tuiWrapper\n            appearance=\"icon\"\n        >\n            <ng-container *polymorpheusOutlet=\"arrow\"></ng-container>\n        </div>\n    </ng-template>\n</tui-hosted-dropdown>\n", styles: [":host{display:block}:host._disabled{pointer-events:none}.t-hosted-dropdown{display:block}.t-country-select{width:5.625rem}.t-country-select:not(._readonly) ::ng-deep input:not(:disabled){cursor:pointer}.t-country-select._readonly ::ng-deep input{cursor:default}.t-country-select[data-size=m]{width:5.5rem}.t-country-select[data-size=s]{width:2rem}.t-country-select[data-size=s] .t-flag{margin-left:-1rem}.t-arrow-icon{position:relative;display:flex;width:1.5rem;height:1.5rem;margin:0 0 0 .25rem;-webkit-margin-start:.25rem;margin-inline-start:.25rem;-webkit-margin-end:0;margin-inline-end:0;align-items:center;justify-content:center;box-sizing:border-box;cursor:pointer}:host._readonly .t-arrow-icon,:host._disabled .t-arrow-icon{pointer-events:none}.t-arrow-icon_open{transform:rotate(180deg)}.t-input-phone{flex:1}.t-flag{width:1.75rem;height:1.25rem;margin-left:-.5rem}.t-country-item-flag{width:1.75rem;height:1.25rem}.t-country-item-name{margin-left:.75rem;margin-right:auto}.t-country-item-code{color:var(--tui-text-02);margin-right:.25rem}\n"], components: [{ type: i1__namespace.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i1__namespace.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }, { type: i2__namespace.TuiInputPhoneComponent, selector: "tui-input-phone", inputs: ["countryCode", "phoneMaskAfterCountryCode", "allowText", "search"], outputs: ["searchChange"] }, { type: i1__namespace.TuiTextfieldComponent, selector: "input[tuiTextfield], textarea[tuiTextfield]" }, { type: i1__namespace.TuiDataListComponent, selector: "tui-data-list", inputs: ["role", "emptyContent", "size"] }, { type: i1__namespace.TuiOptionComponent, selector: "button[tuiOption], a[tuiOption]", inputs: ["size", "role", "disabled", "value"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i1__namespace.TuiGroupDirective, selector: "[tuiGroup]:not(ng-container)", inputs: ["orientation", "adaptive", "collapsed", "rounded", "size"] }, { type: i1__namespace.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i1__namespace.TuiHintOptionsDirective, selector: "[tuiHintContent]", inputs: ["tuiHintContent", "tuiHintDirection", "tuiHintAppearance", "tuiHintShowDelay", "tuiHintHideDelay"] }, { type: i1__namespace.TuiTextfieldPrefixDirective, selector: "[tuiTextfieldPrefix]", inputs: ["tuiTextfieldPrefix"] }, { type: i1__namespace.TuiTextfieldPostfixDirective, selector: "[tuiTextfieldPostfix]", inputs: ["tuiTextfieldPostfix"] }, { type: i1__namespace.TuiTextfieldCustomContentDirective, selector: "[tuiTextfieldCustomContent]", inputs: ["tuiTextfieldCustomContent"] }, { type: i1__namespace.TuiTextfieldLabelOutsideDirective, selector: "[tuiTextfieldLabelOutside]", inputs: ["tuiTextfieldLabelOutside"] }, { type: i1__namespace.TuiTextfieldIconDirective, selector: "[tuiTextfieldIcon]", inputs: ["tuiTextfieldIcon"] }, { type: i2__namespace.TuiInputPhoneDirective, selector: "tui-input-phone" }, { type: i5__namespace.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i5__namespace.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { type: i3__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i1__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }, { type: i6__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i3__namespace.AsyncPipe, "tuiFlag": i1__namespace.TuiFlagPipe, "tuiMapper": i4__namespace.TuiMapperPipe }, viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputPhoneInternationalComponent.prototype, "isoCode", null);
    __decorate([
        i4.tuiPure
    ], TuiInputPhoneInternationalComponent.prototype, "calculateMaskAfterCountryCode", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputPhoneInternationalComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-phone-international',
                        templateUrl: './input-phone-international.template.html',
                        styleUrls: ['./input-phone-international.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i4.tuiAsFocusableItemAccessor(TuiInputPhoneInternationalComponent),
                            i4.tuiAsControl(TuiInputPhoneInternationalComponent),
                            // TODO: for backward compatibility only. Drop in v4.0
                            i1.TuiFlagPipe,
                        ],
                        viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER],
                    }]
            }], ctorParameters: function () {
            return [{ type: i5__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i5.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i7__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_COUNTRIES]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_COUNTRIES_MASKS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS]
                        }] }, { type: i1__namespace.TuiFlagPipe, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiFlagPipe]
                        }] }];
        }, propDecorators: { inputPhoneComponent: [{
                    type: i0.ViewChild,
                    args: [i2.TuiInputPhoneComponent]
                }], primitiveTextfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], isoCode: [{
                    type: i0.Input,
                    args: ['countryIsoCode']
                }], countries: [{
                    type: i0.Input
                }], countryIsoCodeChange: [{
                    type: i0.Output
                }], onPaste: [{
                    type: i0.HostListener,
                    args: ['paste.capture.prevent.stop', ['$event']]
                }, {
                    type: i0.HostListener,
                    args: ['drop.capture.prevent.stop', ['$event']]
                }], calculateMaskAfterCountryCode: [] } });

    var TuiInputPhoneInternationalModule = /** @class */ (function () {
        function TuiInputPhoneInternationalModule() {
        }
        return TuiInputPhoneInternationalModule;
    }());
    TuiInputPhoneInternationalModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputPhoneInternationalModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputPhoneInternationalModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputPhoneInternationalModule, declarations: [TuiInputPhoneInternationalComponent], imports: [i3.CommonModule,
            i5.FormsModule,
            i6.PolymorpheusModule,
            i1.TuiSvgModule,
            i2.TuiInputPhoneModule,
            i1.TuiGroupModule,
            i1.TuiPrimitiveTextfieldModule,
            i1.TuiTextfieldControllerModule,
            i1.TuiHintModule,
            i1.TuiHostedDropdownModule,
            i1.TuiDataListModule,
            arrow.TuiArrowModule,
            i1.TuiWrapperModule,
            i4.TuiLetModule,
            i4.TuiMapperPipeModule,
            i4.TuiActiveZoneModule,
            i1.TuiFlagPipeModule], exports: [TuiInputPhoneInternationalComponent] });
    TuiInputPhoneInternationalModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputPhoneInternationalModule, imports: [[
                i3.CommonModule,
                i5.FormsModule,
                i6.PolymorpheusModule,
                i1.TuiSvgModule,
                i2.TuiInputPhoneModule,
                i1.TuiGroupModule,
                i1.TuiPrimitiveTextfieldModule,
                i1.TuiTextfieldControllerModule,
                i1.TuiHintModule,
                i1.TuiHostedDropdownModule,
                i1.TuiDataListModule,
                arrow.TuiArrowModule,
                i1.TuiWrapperModule,
                i4.TuiLetModule,
                i4.TuiMapperPipeModule,
                i4.TuiActiveZoneModule,
                i1.TuiFlagPipeModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputPhoneInternationalModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            i5.FormsModule,
                            i6.PolymorpheusModule,
                            i1.TuiSvgModule,
                            i2.TuiInputPhoneModule,
                            i1.TuiGroupModule,
                            i1.TuiPrimitiveTextfieldModule,
                            i1.TuiTextfieldControllerModule,
                            i1.TuiHintModule,
                            i1.TuiHostedDropdownModule,
                            i1.TuiDataListModule,
                            arrow.TuiArrowModule,
                            i1.TuiWrapperModule,
                            i4.TuiLetModule,
                            i4.TuiMapperPipeModule,
                            i4.TuiActiveZoneModule,
                            i1.TuiFlagPipeModule,
                        ],
                        declarations: [TuiInputPhoneInternationalComponent],
                        exports: [TuiInputPhoneInternationalComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MASK_AFTER_CODE_REGEXP = MASK_AFTER_CODE_REGEXP;
    exports.TUI_COUNTRIES_MASKS = TUI_COUNTRIES_MASKS;
    exports.TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS = TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS;
    exports.TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS = TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS;
    exports.TuiInputPhoneInternationalComponent = TuiInputPhoneInternationalComponent;
    exports.TuiInputPhoneInternationalModule = TuiInputPhoneInternationalModule;
    exports.tuiExtractValueFromEvent = tuiExtractValueFromEvent;
    exports.tuiInputPhoneInternationalOptionsProvider = tuiInputPhoneInternationalOptionsProvider;
    exports.tuiNotKzRegion = tuiNotKzRegion;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
