// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-time.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/components/select-option'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/mask'), require('rxjs/operators'), require('@taiga-ui/kit/constants'), require('@angular/common'), require('@taiga-ui/kit/directives'), require('@tinkoff/ng-polymorpheus'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-time', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/components/select-option', '@taiga-ui/kit/providers', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/mask', 'rxjs/operators', '@taiga-ui/kit/constants', '@angular/common', '@taiga-ui/kit/directives', '@tinkoff/ng-polymorpheus', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-time"] = {}), global.ng.core, global.ng.forms, global.i2, global.i1, global["taiga-ui"].kit.components["select-option"], global["taiga-ui"].kit.providers, global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils.mask, global.rxjs.operators, global["taiga-ui"].kit.constants, global.ng.common, global["taiga-ui"].kit.directives, global.i5, global.rxjs));
})(this, (function (exports, i0, i6, i2, i1, selectOption, providers, tokens, mask, operators, constants, i4, i3, i5, i7) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_INPUT_TIME_DEFAULT_OPTIONS = {
        icon: function (_a) {
            var $implicit = _a.$implicit;
            return ($implicit === "s" ? "tuiIconClock" : "tuiIconClockLarge");
        },
        mode: "HH:MM",
        postfix: "",
        maxValues: constants.MAX_TIME_VALUES,
        itemSize: "m",
        nativePicker: false,
    };
    /**
     * Default parameters for InputTime component
     */
    var TUI_INPUT_TIME_OPTIONS = new i0.InjectionToken("[TUI_INPUT_TIME_OPTIONS]", {
        factory: function () { return TUI_INPUT_TIME_DEFAULT_OPTIONS; },
    });
    var tuiInputTimeOptionsProvider = function (options) { return ({
        provide: TUI_INPUT_TIME_OPTIONS,
        useValue: Object.assign(Object.assign({}, TUI_INPUT_TIME_DEFAULT_OPTIONS), options),
    }); };

    var TuiNativeTimeComponent = /** @class */ (function () {
        function TuiNativeTimeComponent(host, idService) {
            this.host = host;
            this.autoIdString = idService.generate();
        }
        Object.defineProperty(TuiNativeTimeComponent.prototype, "items", {
            get: function () {
                var _this = this;
                return this.host.items.map(function (item) { return item.toString(_this.host.mode); });
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiNativeTimeComponent.prototype, "value", {
            get: function () {
                return this.host.value.length === this.host.mode.length ? this.host.value : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiNativeTimeComponent.prototype, "step", {
            get: function () {
                switch (this.host.mode) {
                    case 'HH:MM:SS':
                        return 1;
                    case 'HH:MM:SS.MSS':
                        return 0.001;
                    default:
                        return 60;
                }
            },
            enumerable: false,
            configurable: true
        });
        TuiNativeTimeComponent.prototype.onChange = function (value) {
            this.host.onValueChange(value);
        };
        return TuiNativeTimeComponent;
    }());
    TuiNativeTimeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeTimeComponent, deps: [{ token: i1.TUI_TEXTFIELD_HOST }, { token: i2.TuiIdService }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiNativeTimeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiNativeTimeComponent, selector: "input[tuiTime]", host: { attributes: { "type": "time" }, listeners: { "change.stop": "onChange($event.target.value)", "click.stop.silent": "0", "mousedown.stop.silent": "0" }, properties: { "attr.list": "autoIdString", "tabIndex": "-1", "value": "value", "step": "step" } }, ngImport: i0__namespace, template: "\n        <datalist\n            *ngIf=\"items.length\"\n            [id]=\"autoIdString\"\n        >\n            <option\n                *ngFor=\"let item of items\"\n                value=\"{{ item }}\"\n            ></option>\n        </datalist>\n    ", isInline: true, styles: [":host{position:absolute;top:0;left:0;width:100%;height:100%;opacity:0;pointer-events:auto;font-size:2rem}\n"], directives: [{ type: i4__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeTimeComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'input[tuiTime]',
                        template: "\n        <datalist\n            *ngIf=\"items.length\"\n            [id]=\"autoIdString\"\n        >\n            <option\n                *ngFor=\"let item of items\"\n                value=\"{{ item }}\"\n            ></option>\n        </datalist>\n    ",
                        host: {
                            type: 'time',
                            '[attr.list]': 'autoIdString',
                            '[tabIndex]': '-1',
                            '[value]': 'value',
                            '[step]': 'step',
                            '(change.stop)': 'onChange($event.target.value)',
                            '(click.stop.silent)': '0',
                            '(mousedown.stop.silent)': '0',
                        },
                        styleUrls: ['./native-time.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_HOST]
                        }] }, { type: i2__namespace.TuiIdService, decorators: [{
                            type: i0.Inject,
                            args: [i2.TuiIdService]
                        }] }];
        } });

    var TuiInputTimeComponent = /** @class */ (function (_super) {
        __extends(TuiInputTimeComponent, _super);
        function TuiInputTimeComponent(control, cdr, timeTexts$, options, isMobile, textfieldSize, isIos) {
            var _this = _super.call(this, control, cdr) || this;
            _this.timeTexts$ = timeTexts$;
            _this.options = options;
            _this.isMobile = isMobile;
            _this.textfieldSize = textfieldSize;
            _this.isIos = isIos;
            _this.disabledItemHandler = i2.ALWAYS_FALSE_HANDLER;
            _this.items = [];
            _this.itemSize = _this.options.itemSize;
            _this.strict = false;
            _this.mode = _this.options.mode;
            /** @deprecated use `tuiTextfieldPostfix` from {@link TuiTextfieldControllerModule} instead */
            _this.postfix = _this.options.postfix;
            _this.open = false;
            return _this;
        }
        Object.defineProperty(TuiInputTimeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "focused", {
            get: function () {
                return i2.tuiIsNativeFocused(this.nativeFocusableElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "canOpen", {
            get: function () {
                return this.interactive && !!this.filtered.length;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "filtered", {
            get: function () {
                return this.filter(this.items, this.mode, this.computedSearch);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "showNativePicker", {
            get: function () {
                return (this.nativePicker &&
                    (!this.isIos || (this.mode === 'HH:MM' && !this.items.length)));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "nativeDatalist", {
            get: function () {
                return this.nativePicker && !this.isIos;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "size", {
            get: function () {
                return this.textfieldSize.size;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "textMaskOptions", {
            get: function () {
                return this.calculateMask(this.mode);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "computedValue", {
            get: function () {
                return this.value ? this.value.toString(this.mode) : this.nativeValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "computedSearch", {
            get: function () {
                return this.computedValue.length !== this.mode.length ? this.computedValue : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "innerPseudoFocused", {
            get: function () {
                if (this.pseudoFocus === false) {
                    return false;
                }
                if ((this.open && this.canOpen) || this.computedFocused) {
                    return true;
                }
                return null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "icon", {
            get: function () {
                return this.options.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.nativeFocusableElement) {
                    return;
                }
                this.nativeFocusableElement.value = value;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputTimeComponent.prototype.getFiller$ = function (mode) {
            return this.timeTexts$.pipe(operators.map(function (texts) { return texts[mode]; }));
        };
        TuiInputTimeComponent.prototype.onClick = function () {
            this.open = !this.open;
        };
        TuiInputTimeComponent.prototype.onValueChange = function (value) {
            this.open = !!this.items.length;
            if (this.control) {
                this.control.updateValueAndValidity({ emitEvent: false });
            }
            var match = this.getMatch(value);
            if (match !== undefined) {
                this.value = match;
                return;
            }
            if (value.length !== this.mode.length) {
                this.value = null;
                return;
            }
            var time = i2.TuiTime.fromString(value);
            this.value = this.strict ? this.findNearestTimeFromItems(time) : time;
        };
        TuiInputTimeComponent.prototype.onFocused = function (focused) {
            var _this = this;
            this.updateFocused(focused);
            if (focused ||
                this.value !== null ||
                this.nativeValue === '' ||
                this.mode === 'HH:MM') {
                return;
            }
            this.value = i2.TuiTime.fromString(this.nativeValue);
            setTimeout(function () {
                if (_this.nativeValue.endsWith('.') || _this.nativeValue.endsWith(':')) {
                    _this.nativeValue = _this.nativeValue.slice(0, -1);
                }
            });
        };
        TuiInputTimeComponent.prototype.onArrowUp = function (event) {
            if (this.items.length) {
                return;
            }
            this.processArrow(event, 1);
        };
        TuiInputTimeComponent.prototype.onArrowDown = function (event) {
            if (this.items.length) {
                return;
            }
            this.processArrow(event, -1);
        };
        TuiInputTimeComponent.prototype.handleOption = function (item) {
            this.focusInput();
            this.value = item;
        };
        TuiInputTimeComponent.prototype.onOpen = function (open) {
            this.open = open;
        };
        TuiInputTimeComponent.prototype.writeValue = function (value) {
            _super.prototype.writeValue.call(this, value);
            this.nativeValue = value ? this.computedValue : '';
        };
        Object.defineProperty(TuiInputTimeComponent.prototype, "nativePicker", {
            get: function () {
                return !!this.options.nativePicker && this.isMobile;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputTimeComponent.prototype.calculateMask = function (mode) {
            return {
                mask: mask.tuiCreateTimeMask(mode, this.options.maxValues),
                pipe: mask.tuiCreateAutoCorrectedTimePipe(mode, this.options.maxValues),
                guide: false,
            };
        };
        TuiInputTimeComponent.prototype.filter = function (items, mode, search) {
            return items.filter(function (item) { return item.toString(mode).includes(search); });
        };
        TuiInputTimeComponent.prototype.findNearestTimeFromItems = function (value) {
            return this.items.reduce(function (previous, current) { return Math.abs(current.toAbsoluteMilliseconds() - value.toAbsoluteMilliseconds()) <
                Math.abs(previous.toAbsoluteMilliseconds() - value.toAbsoluteMilliseconds())
                ? current
                : previous; });
        };
        TuiInputTimeComponent.prototype.getMatch = function (value) {
            return this.items.find(function (item) { return i2.TUI_STRICT_MATCHER(item, value); });
        };
        TuiInputTimeComponent.prototype.close = function () {
            this.open = false;
        };
        TuiInputTimeComponent.prototype.processArrow = function (event, shift) {
            var target = event.target;
            if (this.readOnly || !i2.tuiIsElement(target) || !i2.tuiIsInput(target)) {
                return;
            }
            var selectionStart = target.selectionStart || 0;
            this.shiftTime(this.calculateShift(selectionStart, shift));
            target.setSelectionRange(selectionStart, selectionStart);
            event.preventDefault();
        };
        TuiInputTimeComponent.prototype.calculateShift = function (selectionStart, shift) {
            if (selectionStart <= 2) {
                return { hours: shift };
            }
            if (selectionStart <= 5) {
                return { minutes: shift };
            }
            if (selectionStart <= 8) {
                return { seconds: shift };
            }
            return { ms: shift };
        };
        TuiInputTimeComponent.prototype.shiftTime = function (shift) {
            if (this.value === null) {
                return;
            }
            var increasedTime = this.value.shift(shift);
            // Manual update so we can set caret position properly
            this.nativeValue = increasedTime.toString(this.mode);
            this.value = increasedTime;
        };
        TuiInputTimeComponent.prototype.focusInput = function (preventScroll) {
            if (preventScroll === void 0) { preventScroll = false; }
            if (this.nativeFocusableElement) {
                this.nativeFocusableElement.focus({ preventScroll: preventScroll });
                this.close();
            }
        };
        return TuiInputTimeComponent;
    }(i2.AbstractTuiNullableControl));
    TuiInputTimeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeComponent, deps: [{ token: i6.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: tokens.TUI_TIME_TEXTS }, { token: TUI_INPUT_TIME_OPTIONS }, { token: i2.TUI_IS_MOBILE }, { token: i1.TUI_TEXTFIELD_SIZE }, { token: i2.TUI_IS_IOS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputTimeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputTimeComponent, selector: "tui-input-time", inputs: { disabledItemHandler: "disabledItemHandler", items: "items", itemSize: "itemSize", strict: "strict", mode: "mode", postfix: "postfix" }, host: { listeners: { "click": "onClick()" } }, providers: [
            i2.tuiAsFocusableItemAccessor(TuiInputTimeComponent),
            i2.tuiAsControl(TuiInputTimeComponent),
            i1.tuiAsDataListHost(TuiInputTimeComponent),
            i1.tuiAsOptionContent(selectOption.TUI_SELECT_OPTION),
        ], viewQueries: [{ propertyName: "textfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-hosted-dropdown\n    class=\"t-wrapper\"\n    [canOpen]=\"canOpen && !nativeDatalist\"\n    [content]=\"dropdownContent\"\n    [open]=\"canOpen && open\"\n    (openChange)=\"onOpen($event)\"\n    (focusedChange)=\"onFocused($event)\"\n>\n    <tui-primitive-textfield\n        tuiValueAccessor\n        class=\"t-textfield\"\n        [tuiTextfieldFiller]=\"(getFiller$(mode) | async) || ''\"\n        [nativeId]=\"nativeId\"\n        [pseudoFocus]=\"innerPseudoFocused\"\n        [pseudoHover]=\"pseudoHover\"\n        [pseudoActive]=\"pseudoActive\"\n        [invalid]=\"computedInvalid\"\n        [focusable]=\"focusable\"\n        [disabled]=\"disabled\"\n        [readOnly]=\"readOnly\"\n        [textMask]=\"textMaskOptions\"\n        [tuiTextfieldIcon]=\"iconContent\"\n        [value]=\"computedValue\"\n        [postfix]=\"postfix\"\n        (valueChange)=\"onValueChange($event)\"\n        (keydown.arrowUp)=\"onArrowUp($event)\"\n        (keydown.arrowDown)=\"onArrowDown($event)\"\n    >\n        <ng-content></ng-content>\n        <input\n            tuiTextfield\n            inputmode=\"numeric\"\n        />\n    </tui-primitive-textfield>\n</tui-hosted-dropdown>\n<ng-template #dropdownContent>\n    <tui-data-list\n        automation-id=\"tui-input-time__dropdown\"\n        [size]=\"itemSize\"\n    >\n        <button\n            *ngFor=\"let item of filtered\"\n            tuiOption\n            automation-id=\"tui-input-time__item\"\n            [disabled]=\"disabledItemHandler(item)\"\n            [value]=\"item\"\n        >\n            {{ item }}\n        </button>\n    </tui-data-list>\n</ng-template>\n\n<ng-template #iconContent>\n    <tui-svg\n        *polymorpheusOutlet=\"icon as src; context: {$implicit: size}\"\n        tuiWrapper\n        appearance=\"icon\"\n        [src]=\"src\"\n    ></tui-svg>\n    <input\n        *ngIf=\"showNativePicker\"\n        tuiTime\n    />\n</ng-template>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}:host._disabled{pointer-events:none}.t-wrapper{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}\n"], components: [{ type: i1__namespace.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i1__namespace.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }, { type: i1__namespace.TuiTextfieldComponent, selector: "input[tuiTextfield], textarea[tuiTextfield]" }, { type: i1__namespace.TuiDataListComponent, selector: "tui-data-list", inputs: ["role", "emptyContent", "size"] }, { type: i1__namespace.TuiOptionComponent, selector: "button[tuiOption], a[tuiOption]", inputs: ["size", "role", "disabled", "value"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }, { type: TuiNativeTimeComponent, selector: "input[tuiTime]" }], directives: [{ type: i1__namespace.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i3__namespace.TuiValueAccessorDirective, selector: "[tuiValueAccessor]" }, { type: i1__namespace.TuiTextfieldFillerDirective, selector: "[tuiTextfieldFiller]", inputs: ["tuiTextfieldFiller"] }, { type: i3__namespace.MaskedInputDirective, selector: "[textMask]", inputs: ["textMask"], exportAs: ["textMask"] }, { type: i1__namespace.TuiTextfieldIconDirective, selector: "[tuiTextfieldIcon]", inputs: ["tuiTextfieldIcon"] }, { type: i4__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i5__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }, { type: i4__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "async": i4__namespace.AsyncPipe }, viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i2.tuiDefaultProp()
    ], TuiInputTimeComponent.prototype, "disabledItemHandler", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiInputTimeComponent.prototype, "items", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiInputTimeComponent.prototype, "itemSize", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiInputTimeComponent.prototype, "strict", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiInputTimeComponent.prototype, "mode", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiInputTimeComponent.prototype, "postfix", void 0);
    __decorate([
        i2.tuiPure
    ], TuiInputTimeComponent.prototype, "getFiller$", null);
    __decorate([
        i2.tuiPure
    ], TuiInputTimeComponent.prototype, "calculateMask", null);
    __decorate([
        i2.tuiPure
    ], TuiInputTimeComponent.prototype, "filter", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-time',
                        templateUrl: './input-time.template.html',
                        styleUrls: ['./input-time.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i2.tuiAsFocusableItemAccessor(TuiInputTimeComponent),
                            i2.tuiAsControl(TuiInputTimeComponent),
                            i1.tuiAsDataListHost(TuiInputTimeComponent),
                            i1.tuiAsOptionContent(selectOption.TUI_SELECT_OPTION),
                        ],
                        viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER],
                    }]
            }], ctorParameters: function () {
            return [{ type: i6__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i6.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i7__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_TIME_TEXTS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_INPUT_TIME_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i2.TUI_IS_MOBILE]
                        }] }, { type: i1__namespace.TuiTextfieldSizeDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_SIZE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i2.TUI_IS_IOS]
                        }] }];
        }, propDecorators: { textfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], disabledItemHandler: [{
                    type: i0.Input
                }], items: [{
                    type: i0.Input
                }], itemSize: [{
                    type: i0.Input
                }], strict: [{
                    type: i0.Input
                }], mode: [{
                    type: i0.Input
                }], postfix: [{
                    type: i0.Input
                }], getFiller$: [], onClick: [{
                    type: i0.HostListener,
                    args: ['click']
                }], calculateMask: [], filter: [] } });

    var TuiInputTimeDirective = /** @class */ (function (_super) {
        __extends(TuiInputTimeDirective, _super);
        function TuiInputTimeDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(TuiInputTimeDirective.prototype, "value", {
            get: function () {
                return this.host.computedValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeDirective.prototype, "mode", {
            get: function () {
                return this.host.mode;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeDirective.prototype, "items", {
            get: function () {
                return this.host.items;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputTimeDirective.prototype.onValueChange = function (value) {
            if (!value) {
                this.host.nativeValue = '';
            }
            this.host.onValueChange(value);
        };
        TuiInputTimeDirective.prototype.process = function (input) {
            input.inputMode = 'numeric';
        };
        return TuiInputTimeDirective;
    }(i1.AbstractTuiTextfieldHost));
    TuiInputTimeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiInputTimeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputTimeDirective, selector: "tui-input-time", providers: [i1.tuiAsTextfieldHost(TuiInputTimeDirective)], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-input-time',
                        providers: [i1.tuiAsTextfieldHost(TuiInputTimeDirective)],
                    }]
            }] });

    var TuiInputTimeModule = /** @class */ (function () {
        function TuiInputTimeModule() {
        }
        return TuiInputTimeModule;
    }());
    TuiInputTimeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputTimeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeModule, declarations: [TuiInputTimeComponent, TuiNativeTimeComponent, TuiInputTimeDirective], imports: [i4.CommonModule,
            i3.TextMaskModule,
            selectOption.TuiSelectOptionModule,
            i1.TuiDataListModule,
            i1.TuiWrapperModule,
            i1.TuiHostedDropdownModule,
            i1.TuiPrimitiveTextfieldModule,
            i3.TuiValueAccessorModule,
            i1.TuiTextfieldControllerModule,
            i1.TuiSvgModule,
            i5.PolymorpheusModule], exports: [TuiInputTimeComponent, TuiInputTimeDirective] });
    TuiInputTimeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeModule, imports: [[
                i4.CommonModule,
                i3.TextMaskModule,
                selectOption.TuiSelectOptionModule,
                i1.TuiDataListModule,
                i1.TuiWrapperModule,
                i1.TuiHostedDropdownModule,
                i1.TuiPrimitiveTextfieldModule,
                i3.TuiValueAccessorModule,
                i1.TuiTextfieldControllerModule,
                i1.TuiSvgModule,
                i5.PolymorpheusModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputTimeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i4.CommonModule,
                            i3.TextMaskModule,
                            selectOption.TuiSelectOptionModule,
                            i1.TuiDataListModule,
                            i1.TuiWrapperModule,
                            i1.TuiHostedDropdownModule,
                            i1.TuiPrimitiveTextfieldModule,
                            i3.TuiValueAccessorModule,
                            i1.TuiTextfieldControllerModule,
                            i1.TuiSvgModule,
                            i5.PolymorpheusModule,
                        ],
                        declarations: [TuiInputTimeComponent, TuiNativeTimeComponent, TuiInputTimeDirective],
                        exports: [TuiInputTimeComponent, TuiInputTimeDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_INPUT_TIME_DEFAULT_OPTIONS = TUI_INPUT_TIME_DEFAULT_OPTIONS;
    exports.TUI_INPUT_TIME_OPTIONS = TUI_INPUT_TIME_OPTIONS;
    exports.TuiInputTimeComponent = TuiInputTimeComponent;
    exports.TuiInputTimeDirective = TuiInputTimeDirective;
    exports.TuiInputTimeModule = TuiInputTimeModule;
    exports.tuiInputTimeOptionsProvider = tuiInputTimeOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
