// @taiga-ui/kit/bundles/taiga-ui-kit-utils-mask.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@taiga-ui/cdk'), require('@taiga-ui/kit/constants'), require('@taiga-ui/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/utils/mask', ['exports', '@taiga-ui/cdk', '@taiga-ui/kit/constants', '@taiga-ui/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.utils = global["taiga-ui"].kit.utils || {}, global["taiga-ui"].kit.utils.mask = {}), global.cdk, global["taiga-ui"].kit.constants, global.core));
})(this, (function (exports, cdk, constants, core) { 'use strict';

    function tuiNormalizeDateValue(dateValue, _a) {
        var value = _a.value, min = _a.min, max = _a.max, dateFormat = _a.dateFormat, dateSeparator = _a.dateSeparator;
        return value && value.toString(dateFormat, dateSeparator) === dateValue
            ? dateValue
            : cdk.TuiDay.normalizeParse(dateValue, dateFormat)
                .dayLimit(min, max)
                .toString(dateFormat, dateSeparator);
    }
    function tuiCreateAutoCorrectedDatePipe(config) {
        return function (value) {
            if (value.length !== cdk.DATE_FILLER_LENGTH) {
                return { value: value };
            }
            return {
                value: tuiNormalizeDateValue(value, config),
            };
        };
    }

    function parseWithLimit(value, config) {
        return cdk.TuiDay.normalizeParse(value.slice(0, cdk.DATE_FILLER_LENGTH), config.dateFormat).dayLimit(config.min, config.max);
    }
    function processRawValue(value, config) {
        var dateFormat = config.dateFormat, dateSeparator = config.dateSeparator;
        switch (value.length) {
            case cdk.DATE_FILLER_LENGTH:
                return parseWithLimit(value, config).toString(dateFormat, dateSeparator);
            case cdk.DATE_FILLER_LENGTH + cdk.RANGE_SEPARATOR_CHAR.length:
                return (parseWithLimit(value, config).toString(dateFormat, dateSeparator) +
                    cdk.RANGE_SEPARATOR_CHAR);
            case cdk.DATE_RANGE_FILLER_LENGTH:
                return config.value &&
                    config.value.toString(dateFormat, dateSeparator) === value
                    ? value
                    : cdk.TuiDayRange.sort(parseWithLimit(value.slice(0, cdk.DATE_FILLER_LENGTH), config), parseWithLimit(value.slice(cdk.DATE_FILLER_LENGTH + cdk.RANGE_SEPARATOR_CHAR.length), config)).toString(dateFormat, dateSeparator);
            default:
                return value;
        }
    }
    /**
     * Normalizes date in formatted string
     *
     * Normalizes when:
     *
     * 1. It is a single date
     * 2. It is a single date and a separator
     * 3. It is two dates and a separator between them
     *
     * In **other** cases, the value does not change.
     *
     * @param config with min and max date
     * @return mask pipe handler that handles `min` and `max`
     */
    function tuiCreateAutoCorrectedDateRangePipe(config) {
        return function (value) { return ({ value: processRawValue(value, config) }); };
    }

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    /**
     * Adjusts the entered time by omitting only suitable values for hours and minutes
     * @returns time as a string
     */
    function tuiCreateAutoCorrectedTimePipe(timeMode, maxValues) {
        if (timeMode === void 0) { timeMode = "HH:MM"; }
        if (maxValues === void 0) { maxValues = {}; }
        var timeFormatArray = ["HH", "MM", "SS", "MS"];
        var safeValues = Object.assign(Object.assign({}, constants.MAX_TIME_VALUES), maxValues);
        return function (conformedValue) {
            var indexesOfPipedChars = [];
            var conformedValueArr = conformedValue.split("");
            timeFormatArray.forEach(function (format) {
                var position = timeMode.indexOf(format);
                var maxFirstDigit = parseInt(String(safeValues[format]).slice(0, 1), 10);
                if (parseInt(conformedValueArr[position], 10) > maxFirstDigit) {
                    conformedValueArr[position + 1] = conformedValueArr[position];
                    conformedValueArr[position] = "0";
                    indexesOfPipedChars.push(position);
                }
            });
            var isInvalid = timeFormatArray.some(function (format) { return parseInt(conformedValue.slice(timeMode.indexOf(format), 2), 10) >
                safeValues[format]; });
            return isInvalid
                ? false
                : {
                    value: conformedValueArr.join(""),
                    indexesOfPipedChars: indexesOfPipedChars,
                };
        };
    }

    function tuiCreateAutoCorrectedDateTimePipe(configs) {
        var timePipe = tuiCreateAutoCorrectedTimePipe(configs.timeMode);
        return function (value) {
            if (value.length < cdk.DATE_FILLER_LENGTH) {
                return { value: value };
            }
            var _a = __read(value.split(constants.DATE_TIME_SEPARATOR), 2), date = _a[0], time = _a[1];
            var formattedDate = tuiNormalizeDateValue(date, configs);
            if (!time) {
                return { value: formattedDate };
            }
            var pipedTime = timePipe(time, {});
            if (!pipedTime || cdk.tuiIsString(pipedTime)) {
                return false;
            }
            return {
                value: "" + formattedDate + constants.DATE_TIME_SEPARATOR + pipedTime.value,
                indexesOfPipedChars: pipedTime.indexesOfPipedChars
                    ? pipedTime.indexesOfPipedChars.map(function (i) { return i + date.length + 2; })
                    : undefined,
            };
        };
    }

    var TWO_DIGITS = new Array(2).fill(core.TUI_DIGIT_REGEXP);
    var FOUR_DIGITS = new Array(4).fill(core.TUI_DIGIT_REGEXP);
    function tuiCreateDateMask(mode, separator) {
        ngDevMode &&
            cdk.tuiAssert.assert(separator.length === 1, "Separator should consist of only 1 symbol");
        switch (mode) {
            case "YMD":
                return __spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray([], __read(FOUR_DIGITS)), [separator]), __read(TWO_DIGITS)), [separator]), __read(TWO_DIGITS));
            case "MDY":
            case "DMY":
            default:
                return __spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray([], __read(TWO_DIGITS)), [separator]), __read(TWO_DIGITS)), [separator]), __read(FOUR_DIGITS));
        }
    }

    function tuiCreateDateRangeMask(dateMode, dateSeparator) {
        ngDevMode &&
            cdk.tuiAssert.assert(dateSeparator.length === 1, "Separator should consist of only 1 symbol");
        var dateMask = tuiCreateDateMask(dateMode, dateSeparator);
        return __spreadArray(__spreadArray(__spreadArray([], __read(dateMask)), [
            cdk.CHAR_NO_BREAK_SPACE,
            cdk.CHAR_EN_DASH,
            cdk.CHAR_NO_BREAK_SPACE
        ]), __read(dateMask));
    }

    function tuiCreateTimePartMask(maxPartValue, prefix) {
        var length = String(maxPartValue).length;
        var regExp = new Array(length).fill(core.TUI_DIGIT_REGEXP);
        if (prefix) {
            regExp.unshift(prefix);
        }
        return regExp;
    }
    function tuiCreateTimeMask(mode, maxValues) {
        if (maxValues === void 0) { maxValues = {}; }
        var _a = Object.assign(Object.assign({}, constants.MAX_TIME_VALUES), maxValues), HH = _a.HH, MM = _a.MM, SS = _a.SS, MS = _a.MS;
        return __spreadArray(__spreadArray(__spreadArray(__spreadArray([], __read(tuiCreateTimePartMask(HH))), __read(tuiCreateTimePartMask(MM, ":"))), __read((mode.includes("HH:MM:SS") ? tuiCreateTimePartMask(SS, ":") : []))), __read((mode === "HH:MM:SS.MSS" ? tuiCreateTimePartMask(MS, ".") : [])));
    }

    /**
     * Generated bundle index. Do not edit.
     */

    exports.tuiCreateAutoCorrectedDatePipe = tuiCreateAutoCorrectedDatePipe;
    exports.tuiCreateAutoCorrectedDateRangePipe = tuiCreateAutoCorrectedDateRangePipe;
    exports.tuiCreateAutoCorrectedDateTimePipe = tuiCreateAutoCorrectedDateTimePipe;
    exports.tuiCreateAutoCorrectedTimePipe = tuiCreateAutoCorrectedTimePipe;
    exports.tuiCreateDateMask = tuiCreateDateMask;
    exports.tuiCreateDateRangeMask = tuiCreateDateRangeMask;
    exports.tuiCreateTimeMask = tuiCreateTimeMask;
    exports.tuiNormalizeDateValue = tuiNormalizeDateValue;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
