// @taiga-ui/kit/bundles/taiga-ui-kit-components-prompt.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/tokens'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/core'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/prompt', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/kit/tokens', '@tinkoff/ng-polymorpheus', '@taiga-ui/core', '@angular/common', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.prompt = {}), global.ng.core, global.i4, global["taiga-ui"].kit.tokens, global.i2, global.i1, global.ng.common, global.rxjs));
})(this, (function (exports, i0, i4, tokens, i2, i1, i3, i5) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    var TuiPromptComponent = /** @class */ (function () {
        function TuiPromptComponent(words$, context, isMobile) {
            this.words$ = words$;
            this.context = context;
            this.isMobile = isMobile;
        }
        Object.defineProperty(TuiPromptComponent.prototype, "appearance", {
            get: function () {
                return this.isMobile ? "secondary" /* Secondary */ : "flat" /* Flat */;
            },
            enumerable: false,
            configurable: true
        });
        return TuiPromptComponent;
    }());
    TuiPromptComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPromptComponent, deps: [{ token: tokens.TUI_PROMPT_WORDS }, { token: i2.POLYMORPHEUS_CONTEXT }, { token: i4.TUI_IS_MOBILE }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPromptComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPromptComponent, selector: "tui-prompt", ngImport: i0__namespace, template: "<div\n    *polymorpheusOutlet=\"context.data?.content as text\"\n    class=\"t-content\"\n    [innerHTML]=\"text\"\n></div>\n<div\n    *ngIf=\"words$ | async as words\"\n    class=\"t-buttons\"\n>\n    <button\n        tuiButton\n        type=\"button\"\n        size=\"m\"\n        class=\"t-button\"\n        [appearance]=\"appearance\"\n        (click)=\"context.completeWith(false)\"\n    >\n        {{ context.data?.no || words.no }}\n    </button>\n    <button\n        tuiButton\n        tuiAutoFocus\n        type=\"button\"\n        size=\"m\"\n        class=\"t-button\"\n        (click)=\"context.completeWith(true)\"\n    >\n        {{ context.data?.yes || words.yes }}\n    </button>\n</div>\n", styles: [".t-content:not(:empty){margin-bottom:1.25rem}.t-buttons{display:flex;justify-content:flex-end;margin-top:.75rem;width:100%}.t-button:last-child{margin-left:.75rem}:host-context(tui-root._mobile) .t-button{flex:1}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.TuiAutoFocusDirective, selector: "[tuiAutoFocus]", inputs: ["tuiAutoFocus"] }], pipes: { "async": i3__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPromptComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-prompt',
                        templateUrl: './prompt.template.html',
                        styleUrls: ['./prompt.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PROMPT_WORDS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i2.POLYMORPHEUS_CONTEXT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i4.TUI_IS_MOBILE]
                        }] }];
        } });
    var TUI_PROMPT = new i2.PolymorpheusComponent(TuiPromptComponent);

    var TuiPromptModule = /** @class */ (function () {
        function TuiPromptModule() {
        }
        return TuiPromptModule;
    }());
    TuiPromptModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPromptModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPromptModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPromptModule, declarations: [TuiPromptComponent], imports: [i3.CommonModule, i2.PolymorpheusModule, i1.TuiButtonModule, i4.TuiAutoFocusModule], exports: [TuiPromptComponent] });
    TuiPromptModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPromptModule, imports: [[i3.CommonModule, i2.PolymorpheusModule, i1.TuiButtonModule, i4.TuiAutoFocusModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPromptModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i3.CommonModule, i2.PolymorpheusModule, i1.TuiButtonModule, i4.TuiAutoFocusModule],
                        declarations: [TuiPromptComponent],
                        exports: [TuiPromptComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_PROMPT = TUI_PROMPT;
    exports.TuiPromptComponent = TuiPromptComponent;
    exports.TuiPromptModule = TuiPromptModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
