// @taiga-ui/kit/bundles/taiga-ui-kit-pipes-stringify-content.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/pipes/stringify-content', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.pipes = global["taiga-ui"].kit.pipes || {}, global["taiga-ui"].kit.pipes["stringify-content"] = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiStringifyContentPipe = /** @class */ (function () {
        function TuiStringifyContentPipe() {
        }
        TuiStringifyContentPipe.prototype.transform = function (stringify) {
            return function (_a) {
                var $implicit = _a.$implicit;
                return stringify($implicit);
            };
        };
        return TuiStringifyContentPipe;
    }());
    TuiStringifyContentPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiStringifyContentPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipe, name: "tuiStringifyContent" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: "tuiStringifyContent",
                    }]
            }] });

    var TuiStringifyContentPipeModule = /** @class */ (function () {
        function TuiStringifyContentPipeModule() {
        }
        return TuiStringifyContentPipeModule;
    }());
    TuiStringifyContentPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiStringifyContentPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipeModule, declarations: [TuiStringifyContentPipe], exports: [TuiStringifyContentPipe] });
    TuiStringifyContentPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyContentPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiStringifyContentPipe],
                        exports: [TuiStringifyContentPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiStringifyContentPipe = TuiStringifyContentPipe;
    exports.TuiStringifyContentPipeModule = TuiStringifyContentPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
