// @taiga-ui/kit/bundles/taiga-ui-kit-components-pagination.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/math'), require('@angular/common'), require('@tinkoff/ng-polymorpheus'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/pagination', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/math', '@angular/common', '@tinkoff/ng-polymorpheus', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.pagination = {}), global.ng.core, global.i2, global.i1, global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils.math, global.ng.common, global.i4, global.rxjs));
})(this, (function (exports, i0, i2, i1, tokens, math, i3, i4, i5) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var DOTS_LENGTH = 1;
    var ACTIVE_ITEM_LENGTH = 1;
    var TuiPaginationComponent = /** @class */ (function (_super) {
        __extends(TuiPaginationComponent, _super);
        function TuiPaginationComponent(el, modeDirective, texts$) {
            var _this = _super.call(this) || this;
            _this.el = el;
            _this.modeDirective = modeDirective;
            _this.texts$ = texts$;
            _this.els = i2.EMPTY_QUERY;
            _this.length = 1;
            _this.size = 'm';
            _this.disabled = false;
            /**
             * Amount of visible pages around active page
             */
            _this.activePadding = 1;
            /**
             * Amount of visible pages at the edges
             */
            _this.sidePadding = 1;
            /**
             * Active page index
             */
            _this.index = 0;
            _this.indexChange = new i0.EventEmitter();
            return _this;
        }
        Object.defineProperty(TuiPaginationComponent.prototype, "nativeFocusableElement", {
            get: function () {
                if (this.disabled) {
                    return null;
                }
                var activeElementIndex = 0;
                var elementsLength = this.elementsLength;
                for (var i = 0; i < elementsLength; i++) {
                    var itemIndex = this.getItemIndexByElementIndex(i);
                    if (itemIndex) {
                        activeElementIndex++;
                    }
                    if (itemIndex === this.index) {
                        break;
                    }
                }
                var activeElement = this.els.find(function (_, index) { return index === activeElementIndex; });
                return activeElement ? activeElement.nativeFocusableElement : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "focused", {
            get: function () {
                return i2.tuiIsNativeFocusedIn(this.el.nativeElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "elementsLength", {
            /**
             * Number of items in a container.
             */
            get: function () {
                return this.itemsFit ? this.length : this.maxElementsLength;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "sizeM", {
            get: function () {
                return this.size === 'm';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "mode", {
            get: function () {
                return this.modeDirective ? this.modeDirective.mode : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "arrowIsDisabledLeft", {
            get: function () {
                return this.index === 0;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "arrowIsDisabledRight", {
            get: function () {
                return this.reverseIndex === 0;
            },
            enumerable: false,
            configurable: true
        });
        TuiPaginationComponent.prototype.elementIsFocusable = function (index) {
            return this.index === index && !this.focused;
        };
        /**
         * Get index by element index
         * @param elementIndex
         * @returns index or null (for '…')
         */
        TuiPaginationComponent.prototype.getItemIndexByElementIndex = function (elementIndex) {
            if (!this.sizeM) {
                return elementIndex;
            }
            if (elementIndex < this.sidePadding) {
                return elementIndex;
            }
            if (elementIndex === this.sidePadding && this.hasCollapsedItems(this.index)) {
                return null;
            }
            var reverseElementIndex = this.lastElementIndex - elementIndex;
            if (reverseElementIndex === this.sidePadding &&
                this.hasCollapsedItems(this.reverseIndex)) {
                return null;
            }
            if (reverseElementIndex < this.sidePadding) {
                return this.lastIndex - reverseElementIndex;
            }
            var computedIndex = this.index - this.maxHalfLength + elementIndex;
            return i2.tuiClamp(computedIndex, elementIndex, this.lastIndex - reverseElementIndex);
        };
        TuiPaginationComponent.prototype.getElementMode = function (index) {
            return this.index === index ? "primary" /* Primary */ : "flat" /* Flat */;
        };
        TuiPaginationComponent.prototype.getSmallElementMode = function (index, mode) {
            return this.index === index && mode !== 'onLight'
                ? "primary" /* Primary */
                : "secondary" /* Secondary */;
        };
        TuiPaginationComponent.prototype.onElementClick = function (index) {
            this.updateIndex(index);
        };
        TuiPaginationComponent.prototype.onElementKeyDownArrowLeft = function (element) {
            if (element === this.els.first) {
                return;
            }
            var previous = this.els.find(function (_, index, array) { return array[index + 1] === element; });
            if (previous === null || previous === void 0 ? void 0 : previous.nativeFocusableElement) {
                previous.nativeFocusableElement.focus();
            }
        };
        TuiPaginationComponent.prototype.onElementKeyDownArrowRight = function (element) {
            if (element === this.els.last) {
                return;
            }
            var next = this.els.find(function (_, index, array) { return array[index - 1] === element; });
            if (next === null || next === void 0 ? void 0 : next.nativeFocusableElement) {
                next.nativeFocusableElement.focus();
            }
        };
        TuiPaginationComponent.prototype.onArrowClick = function (direction) {
            this.tryChangeTo(direction);
            this.focusActive();
        };
        TuiPaginationComponent.prototype.onActiveZone = function (focused) {
            this.updateFocused(focused);
        };
        Object.defineProperty(TuiPaginationComponent.prototype, "reverseIndex", {
            /**
             * Active index from the end
             */
            get: function () {
                return this.lastIndex - this.index;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "maxHalfLength", {
            /**
             * Max number of elements in half (not counting the middle one).
             */
            get: function () {
                return this.sidePadding + DOTS_LENGTH + this.activePadding;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "itemsFit", {
            /**
             * Is there '...' anywhere
             */
            get: function () {
                return this.length <= this.maxElementsLength;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "maxElementsLength", {
            /**
             * Max number of elements
             */
            get: function () {
                return this.maxHalfLength * 2 + ACTIVE_ITEM_LENGTH;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "lastIndex", {
            get: function () {
                return this.length - 1;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "lastElementIndex", {
            get: function () {
                return this.elementsLength - 1;
            },
            enumerable: false,
            configurable: true
        });
        /**
         * Are there collapsed items at that index
         * @param index
         * @returns there are collapsed items
         */
        TuiPaginationComponent.prototype.hasCollapsedItems = function (index) {
            return !this.itemsFit && index > this.maxHalfLength;
        };
        TuiPaginationComponent.prototype.tryChangeTo = function (direction) {
            this.updateIndex(i2.tuiClamp(this.index + math.tuiHorizontalDirectionToNumber(direction), 0, this.lastIndex));
        };
        TuiPaginationComponent.prototype.focusActive = function () {
            var nativeFocusableElement = this.nativeFocusableElement;
            if (nativeFocusableElement) {
                nativeFocusableElement.focus();
            }
        };
        TuiPaginationComponent.prototype.updateIndex = function (index) {
            if (this.index === index) {
                return;
            }
            this.index = index;
            this.indexChange.emit(index);
        };
        return TuiPaginationComponent;
    }(i2.AbstractTuiInteractive));
    TuiPaginationComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPaginationComponent, deps: [{ token: i0.ElementRef }, { token: i1.TuiModeDirective, optional: true }, { token: tokens.TUI_PAGINATION_TEXTS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPaginationComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPaginationComponent, selector: "tui-pagination", inputs: { length: "length", size: "size", disabled: "disabled", activePadding: "activePadding", sidePadding: "sidePadding", content: "content", index: "index" }, outputs: { indexChange: "indexChange" }, providers: [i2.tuiAsFocusableItemAccessor(TuiPaginationComponent)], viewQueries: [{ propertyName: "els", predicate: ["element"], descendants: true, read: i2.TUI_FOCUSABLE_ITEM_ACCESSOR }], usesInheritance: true, ngImport: i0__namespace, template: "<div\n    class=\"t-content\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <ng-container *ngIf=\"sizeM; else smallButtons\">\n        <ng-container *ngIf=\"texts$ | async as texts\">\n            <button\n                tuiIconButton\n                type=\"button\"\n                tuiPreventDefault=\"mousedown\"\n                size=\"s\"\n                appearance=\"flat\"\n                icon=\"tuiIconChevronLeft\"\n                class=\"t-button\"\n                [title]=\"texts[0]\"\n                [disabled]=\"arrowIsDisabledLeft\"\n                [focusable]=\"false\"\n                (click)=\"onArrowClick('left')\"\n            ></button>\n            <ng-container *tuiRepeatTimes=\"let elementIndex of elementsLength\">\n                <ng-container *tuiLet=\"getItemIndexByElementIndex(elementIndex) as index\">\n                    <button\n                        *ngIf=\"index !== null; else dotsTemplate\"\n                        #element\n                        tuiButton\n                        type=\"button\"\n                        automation-id=\"tui-pagination__element\"\n                        shape=\"square\"\n                        size=\"s\"\n                        class=\"t-button\"\n                        [disabled]=\"disabled\"\n                        [focusable]=\"elementIsFocusable(index)\"\n                        [appearance]=\"getElementMode(index)\"\n                        (click)=\"onElementClick(index)\"\n                        (keydown.arrowLeft.prevent)=\"onElementKeyDownArrowLeft(element)\"\n                        (keydown.arrowRight.prevent)=\"onElementKeyDownArrowRight(element)\"\n                    >\n                        <ng-container *polymorpheusOutlet=\"content || index + 1 as text; context: {$implicit: index}\">\n                            {{ text }}\n                        </ng-container>\n                    </button>\n                </ng-container>\n            </ng-container>\n            <button\n                tuiIconButton\n                type=\"button\"\n                tuiPreventDefault=\"mousedown\"\n                size=\"s\"\n                appearance=\"flat\"\n                icon=\"tuiIconChevronRight\"\n                class=\"t-button\"\n                [title]=\"texts[1]\"\n                [disabled]=\"arrowIsDisabledRight\"\n                [focusable]=\"false\"\n                (click)=\"onArrowClick('right')\"\n            ></button>\n        </ng-container>\n    </ng-container>\n    <ng-template #smallButtons>\n        <button\n            *tuiRepeatTimes=\"let indexItem of length\"\n            #element\n            tuiButton\n            type=\"button\"\n            shape=\"square\"\n            class=\"t-button t-button_small\"\n            [class.t-button_active]=\"indexItem === index\"\n            [disabled]=\"disabled\"\n            [focusable]=\"elementIsFocusable(indexItem)\"\n            [appearance]=\"getSmallElementMode(indexItem, mode)\"\n            (click)=\"onElementClick(indexItem)\"\n            (keydown.arrowLeft.prevent)=\"onElementKeyDownArrowLeft(element)\"\n            (keydown.arrowRight.prevent)=\"onElementKeyDownArrowRight(element)\"\n        ></button>\n    </ng-template>\n    <ng-template #dotsTemplate>\n        <div\n            automation-id=\"tui-pagination__element\"\n            class=\"t-dots\"\n        ></div>\n    </ng-template>\n</div>\n", styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);display:block;text-align:center}.t-content{display:flex;justify-content:center}.t-button{margin:0 .125rem;flex-shrink:0}.t-button_active{background:currentColor}.t-button.t-button.t-button_small{width:.5rem;height:.5rem;margin:0}.t-button.t-button.t-button_small:not(:first-child){margin-left:.5rem}.t-dots{width:var(--tui-height-s);height:var(--tui-height-s);line-height:var(--tui-height-s);margin:0 .125rem;flex-shrink:0;color:var(--tui-text-03);text-align:center;cursor:default}.t-dots:before{content:\"\\2026\"}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2__namespace.TuiPreventDefaultDirective, selector: "[tuiPreventDefault]", inputs: ["tuiPreventDefault"] }, { type: i2__namespace.TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: ["tuiRepeatTimesOf"] }, { type: i2__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i3__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i2.tuiDefaultProp(nonNegativeInteger, 'Must be non-negative integer')
    ], TuiPaginationComponent.prototype, "length", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiPaginationComponent.prototype, "size", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiPaginationComponent.prototype, "disabled", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiPaginationComponent.prototype, "activePadding", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiPaginationComponent.prototype, "sidePadding", void 0);
    __decorate([
        i2.tuiDefaultProp(nonNegativeInteger, 'Must be non-negative integer')
    ], TuiPaginationComponent.prototype, "index", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPaginationComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-pagination',
                        templateUrl: './pagination.template.html',
                        styleUrls: ['./pagination.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [i2.tuiAsFocusableItemAccessor(TuiPaginationComponent)],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i1__namespace.TuiModeDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiModeDirective]
                        }] }, { type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PAGINATION_TEXTS]
                        }] }];
        }, propDecorators: { els: [{
                    type: i0.ViewChildren,
                    args: ['element', { read: i2.TUI_FOCUSABLE_ITEM_ACCESSOR }]
                }], length: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }], disabled: [{
                    type: i0.Input
                }], activePadding: [{
                    type: i0.Input
                }], sidePadding: [{
                    type: i0.Input
                }], content: [{
                    type: i0.Input
                }], index: [{
                    type: i0.Input
                }], indexChange: [{
                    type: i0.Output
                }] } });
    function nonNegativeInteger(length) {
        return Number.isInteger(length) && length >= 0;
    }

    var TuiPaginationModule = /** @class */ (function () {
        function TuiPaginationModule() {
        }
        return TuiPaginationModule;
    }());
    TuiPaginationModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPaginationModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPaginationModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPaginationModule, declarations: [TuiPaginationComponent], imports: [i3.CommonModule,
            i2.TuiRepeatTimesModule,
            i2.TuiLetModule,
            i4.PolymorpheusModule,
            i2.TuiPreventDefaultModule,
            i2.TuiActiveZoneModule,
            i1.TuiButtonModule], exports: [TuiPaginationComponent] });
    TuiPaginationModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPaginationModule, imports: [[
                i3.CommonModule,
                i2.TuiRepeatTimesModule,
                i2.TuiLetModule,
                i4.PolymorpheusModule,
                i2.TuiPreventDefaultModule,
                i2.TuiActiveZoneModule,
                i1.TuiButtonModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPaginationModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            i2.TuiRepeatTimesModule,
                            i2.TuiLetModule,
                            i4.PolymorpheusModule,
                            i2.TuiPreventDefaultModule,
                            i2.TuiActiveZoneModule,
                            i1.TuiButtonModule,
                        ],
                        declarations: [TuiPaginationComponent],
                        exports: [TuiPaginationComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPaginationComponent = TuiPaginationComponent;
    exports.TuiPaginationModule = TuiPaginationModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
