// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-range.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/components/input-number'), require('@taiga-ui/kit/components/range'), require('@taiga-ui/kit/constants'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-range', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/components/input-number', '@taiga-ui/kit/components/range', '@taiga-ui/kit/constants', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-range"] = {}), global.ng.core, global.ng.forms, global.i4, global.i3, global["taiga-ui"].kit.components["input-number"], global["taiga-ui"].kit.components.range, global["taiga-ui"].kit.constants, global.ng.common, global.i7));
})(this, (function (exports, i0, i5, i4, i3, i1, i2, constants, i6, i7) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiInputRangeComponent = /** @class */ (function (_super) {
        __extends(TuiInputRangeComponent, _super);
        function TuiInputRangeComponent(control, cdr, isMobile, el, controller) {
            var _this = _super.call(this, control, cdr) || this;
            _this.isMobile = isMobile;
            _this.el = el;
            _this.controller = controller;
            _this.inputNumberRefs = i4.EMPTY_QUERY;
            _this.rangeRef = null;
            _this.min = 0;
            _this.max = 100;
            _this.quantum = 1;
            _this.steps = 0;
            _this.segments = 1;
            _this.keySteps = null;
            _this.pluralize = null;
            _this.lastActiveSide = 'left';
            return _this;
        }
        Object.defineProperty(TuiInputRangeComponent.prototype, "leftFocusableElement", {
            get: function () {
                var _a;
                return ((_a = this.inputNumberRefs.first) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "rightFocusableElement", {
            get: function () {
                var _a;
                return ((_a = this.inputNumberRefs.last) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.disabled
                    ? null
                    : this.leftFocusableElement || this.rightFocusableElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "focused", {
            get: function () {
                return i4.tuiIsNativeFocusedIn(this.el.nativeElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "appearance", {
            get: function () {
                return this.controller.appearance;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "showLeftValueContent", {
            get: function () {
                var _a;
                return Boolean(this.leftValueContent &&
                    !i4.tuiIsNativeFocused(this.leftFocusableElement) &&
                    !(((_a = this.rangeRef) === null || _a === void 0 ? void 0 : _a.focused) && this.lastActiveSide === 'left'));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "showRightValueContent", {
            get: function () {
                var _a;
                return Boolean(this.rightValueContent &&
                    !i4.tuiIsNativeFocused(this.rightFocusableElement) &&
                    !(((_a = this.rangeRef) === null || _a === void 0 ? void 0 : _a.focused) && this.lastActiveSide === 'right'));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "precision", {
            get: function () {
                return i3.tuiGetFractionPartPadded(this.quantum).length;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "decimal", {
            get: function () {
                return this.precision ? 'not-zero' : 'never';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "computedSteps", {
            get: function () {
                return this.steps || (this.max - this.min) / this.quantum;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "step", {
            get: function () {
                return (this.max - this.min) / this.computedSteps;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputRangeComponent.prototype.computeKeySteps = function (keySteps, min, max) {
            return (keySteps || [
                [0, min],
                [100, max],
            ]);
        };
        TuiInputRangeComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
        };
        TuiInputRangeComponent.prototype.onTextInputFocused = function (focused, right) {
            if (focused) {
                return;
            }
            var _b = __read(this.inputNumberRefs, 2), leftTextInputRef = _b[0], rightTextInputRef = _b[1];
            var inputRef = right ? rightTextInputRef : leftTextInputRef;
            var valueIndex = right ? 1 : 0;
            if (!inputRef.nativeValue || inputRef.value !== this.value[valueIndex]) {
                this.updateTextInputValue(this.safeCurrentValue[valueIndex], right);
            }
        };
        TuiInputRangeComponent.prototype.changeByStep = function (event, _b) {
            var _c = __read(_b, 2), leftCoefficient = _c[0], rightCoefficient = _c[1];
            if (this.readOnly) {
                return;
            }
            event.preventDefault();
            var newValue = this.valueGuard([
                this.value[0] + leftCoefficient * this.step,
                this.value[1] + rightCoefficient * this.step,
            ]);
            var leftValueChanged = newValue[0] !== this.value[0];
            var rightValueChanged = newValue[1] !== this.value[1];
            if (leftValueChanged || rightValueChanged) {
                this.safelyUpdateValue(newValue);
                this.updateTextInputValue(newValue[rightValueChanged ? 1 : 0], rightValueChanged);
            }
        };
        TuiInputRangeComponent.prototype.onInputLeft = function (value) {
            this.safelyUpdateValue([value !== null && value !== void 0 ? value : this.safeCurrentValue[0], this.value[1]]);
        };
        TuiInputRangeComponent.prototype.onInputRight = function (value) {
            this.safelyUpdateValue([this.value[0], value !== null && value !== void 0 ? value : this.safeCurrentValue[1]]);
        };
        TuiInputRangeComponent.prototype.onRangeValue = function (value) {
            this.safelyUpdateValue(value);
            var rightValueChanged = this.lastActiveSide === 'right';
            this.updateTextInputValue(this.value[rightValueChanged ? 1 : 0], rightValueChanged);
        };
        TuiInputRangeComponent.prototype.focusToTextInput = function () {
            var element = this.lastActiveSide === 'left'
                ? this.leftFocusableElement
                : this.rightFocusableElement;
            if (!this.isMobile && element) {
                element.focus();
            }
        };
        TuiInputRangeComponent.prototype.onActiveThumbChange = function (activeThumb) {
            this.lastActiveSide = activeThumb;
        };
        TuiInputRangeComponent.prototype.getFallbackValue = function () {
            return [0, 0];
        };
        TuiInputRangeComponent.prototype.safelyUpdateValue = function (value) {
            this.value = this.valueGuard(value);
        };
        TuiInputRangeComponent.prototype.valueGuard = function (_b) {
            var _c = __read(_b, 2), leftValue = _c[0], rightValue = _c[1];
            var leftCalibratedValue = this.calibrate(leftValue);
            var rightCalibratedValue = this.calibrate(rightValue);
            return [
                Math.min(leftCalibratedValue, this.value[1]),
                Math.max(rightCalibratedValue, this.value[0]),
            ];
        };
        TuiInputRangeComponent.prototype.calibrate = function (value) {
            var roundedValue = i4.tuiRound(Math.round(value / this.quantum) * this.quantum, constants.TUI_FLOATING_PRECISION);
            return i4.tuiClamp(roundedValue, this.min, this.max);
        };
        TuiInputRangeComponent.prototype.updateTextInputValue = function (value, right) {
            var _b = __read(this.inputNumberRefs, 2), leftInputRef = _b[0], rightInputRef = _b[1];
            var textInputRef = right ? rightInputRef : leftInputRef;
            if (textInputRef) {
                textInputRef.nativeValue = textInputRef.getFormattedValue(value);
            }
        };
        return TuiInputRangeComponent;
    }(i4.AbstractTuiControl));
    TuiInputRangeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputRangeComponent, deps: [{ token: i5.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i4.TUI_IS_MOBILE }, { token: i0.ElementRef }, { token: i3.TUI_TEXTFIELD_WATCHED_CONTROLLER }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputRangeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputRangeComponent, selector: "tui-input-range", inputs: { min: "min", max: "max", quantum: "quantum", steps: "steps", segments: "segments", keySteps: "keySteps", leftValueContent: "leftValueContent", rightValueContent: "rightValueContent", pluralize: "pluralize" }, host: { properties: { "attr.data-size": "controller.size", "class._label-outside": "controller.labelOutside" } }, providers: [
            i4.tuiAsFocusableItemAccessor(TuiInputRangeComponent),
            i4.tuiAsControl(TuiInputRangeComponent),
            i3.TEXTFIELD_CONTROLLER_PROVIDER,
        ], viewQueries: [{ propertyName: "rangeRef", first: true, predicate: i2.TuiRangeComponent, descendants: true }, { propertyName: "inputNumberRefs", predicate: i1.TuiInputNumberComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<div\n    tuiWrapper\n    class=\"t-wrapper\"\n    [appearance]=\"appearance\"\n    [readOnly]=\"readOnly\"\n    [disabled]=\"disabled\"\n    [focus]=\"computedFocused\"\n    [hover]=\"pseudoHover\"\n    [invalid]=\"computedInvalid\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-input-number\n        tuiTextfieldAppearance=\"none\"\n        automation-id=\"tui-input-range__left-input\"\n        class=\"t-left t-text-input\"\n        [min]=\"min\"\n        [max]=\"value[1]\"\n        [precision]=\"precision\"\n        [decimal]=\"decimal\"\n        [postfix]=\"pluralize && !showLeftValueContent ? (value[0] | i18nPlural : pluralize) : ''\"\n        [disabled]=\"computedDisabled\"\n        [readOnly]=\"readOnly\"\n        [ngModel]=\"value[0]\"\n        (ngModelChange)=\"onInputLeft($event)\"\n        (focusedChange)=\"onTextInputFocused($event, false)\"\n        (keydown.arrowUp)=\"changeByStep($event, [1, 0])\"\n        (keydown.arrowDown)=\"changeByStep($event, [-1, 0])\"\n    >\n        <ng-content></ng-content>\n        <div\n            *ngIf=\"showLeftValueContent\"\n            ngProjectAs=\"tuiContent\"\n        >\n            <ng-container *polymorpheusOutlet=\"leftValueContent as text; context: {$implicit: value[0]}\">\n                {{ text }}\n            </ng-container>\n        </div>\n    </tui-input-number>\n\n    <div class=\"t-right\">\n        <tui-input-number\n            tuiTextfieldAppearance=\"none\"\n            automation-id=\"tui-input-range__right-input\"\n            class=\"t-text-input\"\n            [min]=\"value[0]\"\n            [max]=\"max\"\n            [precision]=\"precision\"\n            [decimal]=\"decimal\"\n            [disabled]=\"computedDisabled\"\n            [readOnly]=\"readOnly\"\n            [ngModel]=\"value[1]\"\n            (ngModelChange)=\"onInputRight($event)\"\n            (focusedChange)=\"onTextInputFocused($event, true)\"\n            (keydown.arrowUp)=\"changeByStep($event, [0, 1])\"\n            (keydown.arrowDown)=\"changeByStep($event, [0, -1])\"\n        >\n            <div\n                *ngIf=\"showRightValueContent\"\n                ngProjectAs=\"tuiContent\"\n            >\n                <ng-container *polymorpheusOutlet=\"rightValueContent as text; context: {$implicit: value[1]}\">\n                    {{ text }}\n                </ng-container>\n            </div>\n        </tui-input-number>\n\n        <!-- TODO replace by postfix of the right InputNumber (after fix https://github.com/Tinkoff/taiga-ui/issues/1193) -->\n        <span\n            *ngIf=\"!showRightValueContent && pluralize\"\n            automation-id=\"tui-input-range__pluralize-right\"\n            class=\"t-pluralize-right\"\n        >\n            &nbsp;{{ value[1] | i18nPlural : pluralize }}\n        </span>\n    </div>\n\n    <tui-range\n        class=\"t-range\"\n        [min]=\"min\"\n        [max]=\"max\"\n        [step]=\"step\"\n        [segments]=\"segments\"\n        [keySteps]=\"computeKeySteps(keySteps, min, max)\"\n        [focusable]=\"false\"\n        [disabled]=\"readOnly || disabled\"\n        [ngModel]=\"value\"\n        (ngModelChange)=\"onRangeValue($event)\"\n        (tuiPressedChange)=\"focusToTextInput()\"\n        (activeThumbChange)=\"onActiveThumbChange($event)\"\n    ></tui-range>\n</div>\n", styles: [".t-wrapper{position:relative}.t-left{width:100%}.t-right{position:absolute;right:0;top:0;display:flex;width:50%;height:100%;text-align:right}.t-right .t-text-input{flex:1}.t-text-input{text-align:inherit}.t-pluralize-right{display:flex;align-items:center;padding:1.125rem var(--tui-padding-m) 0 0;margin-left:-.75rem;font:var(--tui-font-text-s)}:host[data-size=l] .t-pluralize-right{font:var(--tui-font-text-m);padding-top:1.25rem;margin-left:-1rem;padding-right:var(--tui-padding-l)}:host._label-outside .t-pluralize-right{padding-top:0}:host._disabled .t-pluralize-right{color:var(--tui-text-03)}:host{display:block;border-radius:var(--tui-radius-m)}.t-range{position:absolute;top:100%;left:0;right:0;z-index:1;margin:-.125rem 0 0;background:transparent}\n"], components: [{ type: i1__namespace.TuiInputNumberComponent, selector: "tui-input-number", inputs: ["min", "max", "decimal", "precision", "prefix", "postfix"] }, { type: i2__namespace.TuiRangeComponent, selector: "tui-range", inputs: ["min", "max", "step", "size", "segments", "keySteps"] }], directives: [{ type: i3__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }, { type: i4__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i1__namespace.TuiInputNumberDirective, selector: "tui-input-number" }, { type: i3__namespace.TuiTextfieldAppearanceDirective, selector: "[tuiTextfieldAppearance]", inputs: ["tuiTextfieldAppearance"] }, { type: i5__namespace.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i5__namespace.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { type: i6__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i7__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i2__namespace.TuiRangeChangeDirective, selector: "tui-range", outputs: ["activeThumbChange"] }, { type: i4__namespace.TuiPressedDirective, selector: "[tuiPressedChange]", outputs: ["tuiPressedChange"] }], pipes: { "i18nPlural": i6__namespace.I18nPluralPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputRangeComponent.prototype, "min", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputRangeComponent.prototype, "max", void 0);
    __decorate([
        i4.tuiDefaultProp(function (q) { return q > 0; }, '[quantum] must be positive')
    ], TuiInputRangeComponent.prototype, "quantum", void 0);
    __decorate([
        i4.tuiDefaultProp(function (s) { return s >= 0 && Number.isInteger(s); }, '[steps] must be non-negative integer')
    ], TuiInputRangeComponent.prototype, "steps", void 0);
    __decorate([
        i4.tuiDefaultProp(function (s) { return s > 0 && Number.isInteger(s); }, '[segments] must be positive integer')
    ], TuiInputRangeComponent.prototype, "segments", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputRangeComponent.prototype, "keySteps", void 0);
    __decorate([
        i4.tuiDefaultProp()
    ], TuiInputRangeComponent.prototype, "pluralize", void 0);
    __decorate([
        i4.tuiPure
    ], TuiInputRangeComponent.prototype, "computeKeySteps", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputRangeComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-range',
                        templateUrl: './input-range.template.html',
                        styleUrls: ['./input-range.style.less'],
                        host: {
                            '[attr.data-size]': 'controller.size',
                            '[class._label-outside]': 'controller.labelOutside',
                        },
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i4.tuiAsFocusableItemAccessor(TuiInputRangeComponent),
                            i4.tuiAsControl(TuiInputRangeComponent),
                            i3.TEXTFIELD_CONTROLLER_PROVIDER,
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i5__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i5.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i4.TUI_IS_MOBILE]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i3__namespace.TuiTextfieldController, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_TEXTFIELD_WATCHED_CONTROLLER]
                        }] }];
        }, propDecorators: { inputNumberRefs: [{
                    type: i0.ViewChildren,
                    args: [i1.TuiInputNumberComponent]
                }], rangeRef: [{
                    type: i0.ViewChild,
                    args: [i2.TuiRangeComponent]
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], quantum: [{
                    type: i0.Input
                }], steps: [{
                    type: i0.Input
                }], segments: [{
                    type: i0.Input
                }], keySteps: [{
                    type: i0.Input
                }], leftValueContent: [{
                    type: i0.Input
                }], rightValueContent: [{
                    type: i0.Input
                }], pluralize: [{
                    type: i0.Input
                }], computeKeySteps: [] } });

    var TuiInputRangeModule = /** @class */ (function () {
        function TuiInputRangeModule() {
        }
        return TuiInputRangeModule;
    }());
    TuiInputRangeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputRangeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputRangeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputRangeModule, declarations: [TuiInputRangeComponent], imports: [i6.CommonModule,
            i5.FormsModule,
            i7.PolymorpheusModule,
            i4.TuiActiveZoneModule,
            i1.TuiInputNumberModule,
            i4.TuiPressedModule,
            i2.TuiRangeModule,
            i3.TuiWrapperModule,
            i3.TuiTextfieldControllerModule], exports: [TuiInputRangeComponent] });
    TuiInputRangeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputRangeModule, imports: [[
                i6.CommonModule,
                i5.FormsModule,
                i7.PolymorpheusModule,
                i4.TuiActiveZoneModule,
                i1.TuiInputNumberModule,
                i4.TuiPressedModule,
                i2.TuiRangeModule,
                i3.TuiWrapperModule,
                i3.TuiTextfieldControllerModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputRangeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i6.CommonModule,
                            i5.FormsModule,
                            i7.PolymorpheusModule,
                            i4.TuiActiveZoneModule,
                            i1.TuiInputNumberModule,
                            i4.TuiPressedModule,
                            i2.TuiRangeModule,
                            i3.TuiWrapperModule,
                            i3.TuiTextfieldControllerModule,
                        ],
                        declarations: [TuiInputRangeComponent],
                        exports: [TuiInputRangeComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiInputRangeComponent = TuiInputRangeComponent;
    exports.TuiInputRangeModule = TuiInputRangeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
