// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-date-time.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/constants'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/mask'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/kit/directives'), require('@tinkoff/ng-polymorpheus'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-date-time', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/constants', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/mask', 'rxjs', 'rxjs/operators', '@taiga-ui/kit/directives', '@tinkoff/ng-polymorpheus', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-date-time"] = {}), global.ng.core, global.ng.forms, global.i6, global.i1, global["taiga-ui"].kit.constants, global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils.mask, global.rxjs, global.rxjs.operators, global["taiga-ui"].kit.directives, global.i3, global.ng.common));
})(this, (function (exports, i0, i7, i6, i1, constants, tokens, mask, i8, operators, i2, i3, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i8__namespace = /*#__PURE__*/_interopNamespace(i8);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiNativeDateTimeDirective = /** @class */ (function () {
        function TuiNativeDateTimeDirective(host) {
            this.host = host;
        }
        Object.defineProperty(TuiNativeDateTimeDirective.prototype, "value", {
            get: function () {
                var _a, _b;
                if (!this.host.rawValue[0] || !this.host.rawValue[1]) {
                    return '';
                }
                return ((_a = this.host.rawValue[0]) === null || _a === void 0 ? void 0 : _a.toString('YMD', '-')) + "T" + ((_b = this.host.rawValue[1]) === null || _b === void 0 ? void 0 : _b.toString('HH:MM'));
            },
            enumerable: false,
            configurable: true
        });
        TuiNativeDateTimeDirective.prototype.onChange = function (value) {
            if (!value) {
                this.host.writeValue([null, null]);
                return;
            }
            var date = new Date(value);
            var day = i6.TuiDay.fromLocalNativeDate(date);
            var time = i6.TuiTime.fromLocalNativeDate(date);
            this.host.writeValue([day, time]);
        };
        return TuiNativeDateTimeDirective;
    }());
    TuiNativeDateTimeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeDateTimeDirective, deps: [{ token: i1.TUI_TEXTFIELD_HOST }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiNativeDateTimeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiNativeDateTimeDirective, selector: "input[tuiDateTime]", host: { attributes: { "type": "datetime-local" }, listeners: { "change.stop": "onChange($event.target.value)", "input.stop": "onChange($event.target.value)", "click.stop.silent": "0", "mousedown.stop.silent": "0" }, properties: { "tabIndex": "-1", "value": "value" } }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeDateTimeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'input[tuiDateTime]',
                        host: {
                            type: 'datetime-local',
                            '[tabIndex]': '-1',
                            '[value]': 'value',
                            '(change.stop)': 'onChange($event.target.value)',
                            '(input.stop)': 'onChange($event.target.value)',
                            '(click.stop.silent)': '0',
                            '(mousedown.stop.silent)': '0',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_HOST]
                        }] }];
        } });

    var TuiInputDateTimeComponent = /** @class */ (function (_super) {
        __extends(TuiInputDateTimeComponent, _super);
        function TuiInputDateTimeComponent(control, cdr, textfieldSize, dateFormat, dateSeparator, timeTexts$, dateTexts$, valueTransformer, options, isMobile, isIos) {
            var _this = _super.call(this, control, cdr, valueTransformer) || this;
            _this.textfieldSize = textfieldSize;
            _this.dateFormat = dateFormat;
            _this.dateSeparator = dateSeparator;
            _this.timeTexts$ = timeTexts$;
            _this.dateTexts$ = dateTexts$;
            _this.valueTransformer = valueTransformer;
            _this.options = options;
            _this.isMobile = isMobile;
            _this.isIos = isIos;
            _this.month = null;
            _this.min = _this.options.min;
            _this.max = _this.options.max;
            _this.disabledItemHandler = i6.ALWAYS_FALSE_HANDLER;
            _this.defaultActiveYearMonth = i6.TuiMonth.currentLocal();
            _this.timeMode = 'HH:MM';
            _this.open = false;
            _this.filler$ = i8.combineLatest([
                _this.dateTexts$.pipe(operators.map(function (dateTexts) { return i6.changeDateSeparator(dateTexts[_this.dateFormat], _this.dateSeparator); })),
                _this.timeTexts$.pipe(operators.map(function (texts) { return texts[_this.timeMode]; })),
            ]).pipe(operators.map(function (fillers) { return _this.getDateTimeString.apply(_this, __spreadArray([], __read(fillers))); }));
            return _this;
        }
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "fillerLength", {
            get: function () {
                return i6.DATE_FILLER_LENGTH + constants.DATE_TIME_SEPARATOR.length + this.timeMode.length;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "textMaskOptions", {
            get: function () {
                return this.calculateMask(this.value[0], this.calendarMinDay, this.calendarMaxDay, this.timeMode, this.dateFormat, this.dateSeparator);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "focused", {
            get: function () {
                return !!this.textfield && this.textfield.focused;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarIcon", {
            get: function () {
                return this.options.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "nativePicker", {
            get: function () {
                return !!this.options.nativePicker && this.isMobile;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "showNativePicker", {
            get: function () {
                return this.nativePicker && this.timeMode === 'HH:MM';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "computedValue", {
            get: function () {
                var _a = this, value = _a.value, nativeValue = _a.nativeValue, timeMode = _a.timeMode;
                var _b = __read(value, 2), date = _b[0], time = _b[1];
                var hasTimeInputChars = nativeValue.length > i6.DATE_FILLER_LENGTH;
                if (!date || (!time && hasTimeInputChars)) {
                    return nativeValue;
                }
                return this.getDateTimeString(date, time, timeMode);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarValue", {
            get: function () {
                return this.value[0];
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarMinDay", {
            get: function () {
                return Array.isArray(this.min) ? this.min[0] : this.min;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarMaxDay", {
            get: function () {
                return Array.isArray(this.max) ? this.max[0] : this.max;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "computedActiveYearMonth", {
            get: function () {
                return (this.month ||
                    this.value[0] ||
                    i6.tuiDateClamp(this.defaultActiveYearMonth, Array.isArray(this.min) ? this.min[0] : this.min, Array.isArray(this.max) ? this.max[0] : this.max));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.nativeFocusableElement) {
                    return;
                }
                this.nativeFocusableElement.value = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "size", {
            get: function () {
                return this.textfieldSize.size;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputDateTimeComponent.prototype.onClick = function () {
            this.open = !this.open;
        };
        TuiInputDateTimeComponent.prototype.onValueChange = function (value) {
            if (!value) {
                this.onOpenChange(true);
            }
            if (value.length < i6.DATE_FILLER_LENGTH) {
                this.value = [null, null];
                return;
            }
            var _a = __read(value.split(constants.DATE_TIME_SEPARATOR), 2), date = _a[0], time = _a[1];
            var parsedDate = i6.TuiDay.normalizeParse(date, this.dateFormat);
            var parsedTime = time && time.length === this.timeMode.length
                ? this.clampTime(i6.TuiTime.fromString(time), parsedDate)
                : null;
            this.open = false;
            this.value = [parsedDate, parsedTime];
        };
        TuiInputDateTimeComponent.prototype.onDayClick = function (day) {
            var modifiedTime = this.value[1] && this.clampTime(this.value[1], day);
            this.value = [day, modifiedTime];
            this.updateNativeValue(day);
            this.open = false;
        };
        TuiInputDateTimeComponent.prototype.onMonthChange = function (month) {
            this.month = month;
        };
        TuiInputDateTimeComponent.prototype.onOpenChange = function (open) {
            this.open = open;
        };
        TuiInputDateTimeComponent.prototype.onFocused = function (focused) {
            var _this = this;
            this.updateFocused(focused);
            if (focused ||
                this.value[0] === null ||
                this.value[1] !== null ||
                this.nativeValue.length <= this.fillerLength + constants.DATE_TIME_SEPARATOR.length ||
                this.timeMode === 'HH:MM') {
                return;
            }
            var _a = __read(this.nativeValue.split(constants.DATE_TIME_SEPARATOR), 2), time = _a[1];
            if (!time) {
                return;
            }
            var parsedTime = i6.TuiTime.fromString(time);
            this.value = [this.value[0], parsedTime];
            setTimeout(function () {
                if (_this.nativeValue.endsWith('.') || _this.nativeValue.endsWith(':')) {
                    _this.nativeValue = _this.nativeValue.slice(0, -1);
                }
            });
        };
        TuiInputDateTimeComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.open = false;
        };
        TuiInputDateTimeComponent.prototype.writeValue = function (value) {
            _super.prototype.writeValue.call(this, value);
            this.nativeValue =
                this.value && (this.value[0] || this.value[1]) ? this.computedValue : '';
        };
        TuiInputDateTimeComponent.prototype.getFallbackValue = function () {
            return [null, null];
        };
        TuiInputDateTimeComponent.prototype.valueIdenticalComparator = function (oldValue, newValue) {
            return (i6.tuiNullableSame(oldValue[0], newValue[0], function (a, b) { return a.daySame(b); }) &&
                i6.tuiNullableSame(oldValue[1], newValue[1], function (a, b) { return String(a) === String(b); }));
        };
        TuiInputDateTimeComponent.prototype.calculateMask = function (day, min, max, timeMode, dateFormat, dateSeparator) {
            return {
                mask: __spreadArray(__spreadArray(__spreadArray([], __read(mask.tuiCreateDateMask(dateFormat, dateSeparator))), [
                    ',',
                    ' '
                ]), __read(mask.tuiCreateTimeMask(timeMode))),
                pipe: mask.tuiCreateAutoCorrectedDateTimePipe({
                    value: day,
                    min: min,
                    max: max,
                    dateFormat: dateFormat,
                    dateSeparator: dateSeparator,
                    timeMode: timeMode,
                }),
                guide: false,
            };
        };
        TuiInputDateTimeComponent.prototype.getDateTimeString = function (date, time, timeMode) {
            if (timeMode === void 0) { timeMode = 'HH:MM'; }
            var dateString = date instanceof i6.TuiDay
                ? date.toString(this.dateFormat, this.dateSeparator)
                : date;
            var timeString = time instanceof i6.TuiTime ? time.toString(timeMode) : time || '';
            return "" + dateString + constants.DATE_TIME_SEPARATOR + timeString;
        };
        TuiInputDateTimeComponent.prototype.updateNativeValue = function (day) {
            var time = this.nativeValue.split(constants.DATE_TIME_SEPARATOR)[1] || '';
            this.nativeValue = this.getDateTimeString(day, time);
        };
        TuiInputDateTimeComponent.prototype.clampTime = function (time, day) {
            var ms = time.toAbsoluteMilliseconds();
            var min = Array.isArray(this.min) && day.daySame(this.calendarMinDay)
                ? this.min[1].toAbsoluteMilliseconds()
                : -Infinity;
            var max = Array.isArray(this.max) && day.daySame(this.calendarMaxDay)
                ? this.max[1].toAbsoluteMilliseconds()
                : Infinity;
            return i6.TuiTime.fromAbsoluteMilliseconds(i6.tuiClamp(ms, min, max));
        };
        return TuiInputDateTimeComponent;
    }(i6.AbstractTuiControl));
    TuiInputDateTimeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeComponent, deps: [{ token: i7.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i1.TUI_TEXTFIELD_SIZE }, { token: i6.TUI_DATE_FORMAT }, { token: i6.TUI_DATE_SEPARATOR }, { token: tokens.TUI_TIME_TEXTS }, { token: tokens.TUI_DATE_TEXTS }, { token: tokens.TUI_DATE_TIME_VALUE_TRANSFORMER, optional: true }, { token: tokens.TUI_INPUT_DATE_OPTIONS }, { token: i6.TUI_IS_MOBILE }, { token: i6.TUI_IS_IOS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputDateTimeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputDateTimeComponent, selector: "tui-input-date-time", inputs: { min: "min", max: "max", disabledItemHandler: "disabledItemHandler", defaultActiveYearMonth: "defaultActiveYearMonth", timeMode: "timeMode" }, host: { listeners: { "click": "onClick()" } }, providers: [
            i6.tuiAsFocusableItemAccessor(TuiInputDateTimeComponent),
            i6.tuiAsControl(TuiInputDateTimeComponent),
            tokens.tuiDateStreamWithTransformer(tokens.TUI_DATE_TIME_VALUE_TRANSFORMER),
        ], viewQueries: [{ propertyName: "textfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"interactive && !showNativePicker\"\n    [content]=\"dropdown\"\n    [open]=\"open && interactive\"\n    (openChange)=\"onOpenChange($event)\"\n>\n    <tui-primitive-textfield\n        automation-id=\"tui-input-date-time__textfield\"\n        tuiValueAccessor\n        class=\"t-textfield\"\n        [tuiTextfieldIcon]=\"calendarIcon && iconContent\"\n        [tuiTextfieldFiller]=\"(filler$ | async) || ''\"\n        [pseudoFocus]=\"pseudoFocus\"\n        [pseudoHover]=\"pseudoHover\"\n        [invalid]=\"computedInvalid\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [disabled]=\"computedDisabled\"\n        [textMask]=\"textMaskOptions\"\n        [value]=\"computedValue\"\n        (valueChange)=\"onValueChange($event)\"\n        (focusedChange)=\"onFocused($event)\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            select=\"input\"\n            ngProjectAs=\"input\"\n        ></ng-content>\n    </tui-primitive-textfield>\n\n    <ng-template #iconContent>\n        <tui-svg\n            *polymorpheusOutlet=\"calendarIcon as src; context: {$implicit: size}\"\n            tuiWrapper\n            appearance=\"icon\"\n            [class.t-icon]=\"!computedDisabled\"\n            [src]=\"src\"\n        ></tui-svg>\n        <input\n            *ngIf=\"showNativePicker\"\n            tuiDateTime\n            class=\"t-native-input\"\n        />\n    </ng-template>\n\n    <ng-template\n        #dropdown=\"polymorpheus\"\n        [polymorpheus]=\"type\"\n    >\n        <tui-calendar\n            tuiPreventDefault=\"mousedown\"\n            automation-id=\"tui-input-date-time__calendar\"\n            [min]=\"calendarMinDay\"\n            [max]=\"calendarMaxDay\"\n            [disabledItemHandler]=\"disabledItemHandler\"\n            [month]=\"computedActiveYearMonth\"\n            [value]=\"calendarValue\"\n            (dayClick)=\"onDayClick($event)\"\n            (monthChange)=\"onMonthChange($event)\"\n        ></tui-calendar>\n    </ng-template>\n</tui-hosted-dropdown>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}.t-icon{pointer-events:auto}.t-button{display:flex;height:2.75rem;justify-content:center;box-shadow:inset 0 1px var(--tui-base-03)}.t-native-input{position:absolute;top:0;left:0;width:100%;height:100%;opacity:0;pointer-events:auto;font-size:2rem}\n"], components: [{ type: i1__namespace.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i1__namespace.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }, { type: i1__namespace.TuiCalendarComponent, selector: "tui-calendar", inputs: ["month", "disabledItemHandler", "min", "max", "minViewedMonth", "maxViewedMonth", "hoveredItem", "showAdjacent", "markerHandler", "value"], outputs: ["dayClick", "monthChange", "hoveredItemChange"] }], directives: [{ type: i1__namespace.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i2__namespace.TuiValueAccessorDirective, selector: "[tuiValueAccessor]" }, { type: i1__namespace.TuiTextfieldIconDirective, selector: "[tuiTextfieldIcon]", inputs: ["tuiTextfieldIcon"] }, { type: i1__namespace.TuiTextfieldFillerDirective, selector: "[tuiTextfieldFiller]", inputs: ["tuiTextfieldFiller"] }, { type: i2__namespace.MaskedInputDirective, selector: "[textMask]", inputs: ["textMask"], exportAs: ["textMask"] }, { type: i3__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }, { type: i4__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: TuiNativeDateTimeDirective, selector: "input[tuiDateTime]" }, { type: i3__namespace.PolymorpheusTemplate, selector: "ng-template[polymorpheus]", inputs: ["polymorpheus"], exportAs: ["polymorpheus"] }, { type: i6__namespace.TuiPreventDefaultDirective, selector: "[tuiPreventDefault]", inputs: ["tuiPreventDefault"] }], pipes: { "async": i4__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i6.tuiDefaultProp()
    ], TuiInputDateTimeComponent.prototype, "min", void 0);
    __decorate([
        i6.tuiDefaultProp()
    ], TuiInputDateTimeComponent.prototype, "max", void 0);
    __decorate([
        i6.tuiDefaultProp()
    ], TuiInputDateTimeComponent.prototype, "disabledItemHandler", void 0);
    __decorate([
        i6.tuiDefaultProp()
    ], TuiInputDateTimeComponent.prototype, "defaultActiveYearMonth", void 0);
    __decorate([
        i6.tuiDefaultProp()
    ], TuiInputDateTimeComponent.prototype, "timeMode", void 0);
    __decorate([
        i6.tuiPure
    ], TuiInputDateTimeComponent.prototype, "calculateMask", null);
    __decorate([
        i6.tuiPure
    ], TuiInputDateTimeComponent.prototype, "getDateTimeString", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-date-time',
                        templateUrl: './input-date-time.template.html',
                        styleUrls: ['./input-date-time.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i6.tuiAsFocusableItemAccessor(TuiInputDateTimeComponent),
                            i6.tuiAsControl(TuiInputDateTimeComponent),
                            tokens.tuiDateStreamWithTransformer(tokens.TUI_DATE_TIME_VALUE_TRANSFORMER),
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i7__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i7.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i1__namespace.TuiTextfieldSizeDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_SIZE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_DATE_FORMAT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_DATE_SEPARATOR]
                        }] }, { type: i8__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_TIME_TEXTS]
                        }] }, { type: i8__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DATE_TEXTS]
                        }] }, { type: i6__namespace.AbstractTuiValueTransformer, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_DATE_TIME_VALUE_TRANSFORMER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_INPUT_DATE_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_IS_MOBILE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_IS_IOS]
                        }] }];
        }, propDecorators: { textfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], disabledItemHandler: [{
                    type: i0.Input
                }], defaultActiveYearMonth: [{
                    type: i0.Input
                }], timeMode: [{
                    type: i0.Input
                }], onClick: [{
                    type: i0.HostListener,
                    args: ['click']
                }], calculateMask: [], getDateTimeString: [] } });

    var TuiInputDateTimeDirective = /** @class */ (function (_super) {
        __extends(TuiInputDateTimeDirective, _super);
        function TuiInputDateTimeDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(TuiInputDateTimeDirective.prototype, "value", {
            get: function () {
                return this.host.computedValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeDirective.prototype, "rawValue", {
            get: function () {
                return this.host.value;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputDateTimeDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        TuiInputDateTimeDirective.prototype.writeValue = function (value) {
            this.host.writeValue(value);
        };
        TuiInputDateTimeDirective.prototype.process = function (input) {
            input.inputMode = 'numeric';
        };
        return TuiInputDateTimeDirective;
    }(i1.AbstractTuiTextfieldHost));
    TuiInputDateTimeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiInputDateTimeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputDateTimeDirective, selector: "tui-input-date-time", providers: [i1.tuiAsTextfieldHost(TuiInputDateTimeDirective)], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-input-date-time',
                        providers: [i1.tuiAsTextfieldHost(TuiInputDateTimeDirective)],
                    }]
            }] });

    var TuiInputDateTimeModule = /** @class */ (function () {
        function TuiInputDateTimeModule() {
        }
        return TuiInputDateTimeModule;
    }());
    TuiInputDateTimeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputDateTimeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeModule, declarations: [TuiInputDateTimeComponent,
            TuiInputDateTimeDirective,
            TuiNativeDateTimeDirective], imports: [i4.CommonModule,
            i2.TextMaskModule,
            i3.PolymorpheusModule,
            i1.TuiWrapperModule,
            i6.TuiPreventDefaultModule,
            i1.TuiCalendarModule,
            i1.TuiSvgModule,
            i1.TuiLinkModule,
            i1.TuiHostedDropdownModule,
            i1.TuiPrimitiveTextfieldModule,
            i2.TuiValueAccessorModule,
            i1.TuiTextfieldControllerModule], exports: [TuiInputDateTimeComponent,
            TuiInputDateTimeDirective,
            i1.TuiTextfieldComponent] });
    TuiInputDateTimeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeModule, imports: [[
                i4.CommonModule,
                i2.TextMaskModule,
                i3.PolymorpheusModule,
                i1.TuiWrapperModule,
                i6.TuiPreventDefaultModule,
                i1.TuiCalendarModule,
                i1.TuiSvgModule,
                i1.TuiLinkModule,
                i1.TuiHostedDropdownModule,
                i1.TuiPrimitiveTextfieldModule,
                i2.TuiValueAccessorModule,
                i1.TuiTextfieldControllerModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateTimeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i4.CommonModule,
                            i2.TextMaskModule,
                            i3.PolymorpheusModule,
                            i1.TuiWrapperModule,
                            i6.TuiPreventDefaultModule,
                            i1.TuiCalendarModule,
                            i1.TuiSvgModule,
                            i1.TuiLinkModule,
                            i1.TuiHostedDropdownModule,
                            i1.TuiPrimitiveTextfieldModule,
                            i2.TuiValueAccessorModule,
                            i1.TuiTextfieldControllerModule,
                        ],
                        declarations: [
                            TuiInputDateTimeComponent,
                            TuiInputDateTimeDirective,
                            TuiNativeDateTimeDirective,
                        ],
                        exports: [
                            TuiInputDateTimeComponent,
                            TuiInputDateTimeDirective,
                            i1.TuiTextfieldComponent,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiInputDateTimeComponent = TuiInputDateTimeComponent;
    exports.TuiInputDateTimeDirective = TuiInputDateTimeDirective;
    exports.TuiInputDateTimeModule = TuiInputDateTimeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
