// @taiga-ui/kit/bundles/taiga-ui-kit-directives-unfinished-validator.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/validators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/directives/unfinished-validator', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/kit/validators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.directives = global["taiga-ui"].kit.directives || {}, global["taiga-ui"].kit.directives["unfinished-validator"] = {}), global.ng.core, global.ng.forms, global.cdk, global["taiga-ui"].kit.validators));
})(this, (function (exports, i0, forms, cdk, validators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiUnfinishedValidatorDirective = /** @class */ (function () {
        function TuiUnfinishedValidatorDirective(injector, 
        // eslint-disable-next-line @angular-eslint/no-attribute-decorator
        message) {
            var _this = this;
            this.injector = injector;
            this.message = message;
            this.validate = validators.tuiCreateUnfinishedValidator(function () { return _this.injector.get(cdk.TUI_FOCUSABLE_ITEM_ACCESSOR); }, this.message || '');
        }
        return TuiUnfinishedValidatorDirective;
    }());
    TuiUnfinishedValidatorDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiUnfinishedValidatorDirective, deps: [{ token: i0.Injector }, { token: 'tuiUnfinishedValidator', attribute: true }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiUnfinishedValidatorDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiUnfinishedValidatorDirective, selector: "[tuiUnfinishedValidator]", providers: [
            {
                provide: forms.NG_VALIDATORS,
                useExisting: TuiUnfinishedValidatorDirective,
                multi: true,
            },
        ], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiUnfinishedValidatorDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiUnfinishedValidator]',
                        providers: [
                            {
                                provide: forms.NG_VALIDATORS,
                                useExisting: TuiUnfinishedValidatorDirective,
                                multi: true,
                            },
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.Injector, decorators: [{
                            type: i0.Inject,
                            args: [i0.Injector]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Attribute,
                            args: ['tuiUnfinishedValidator']
                        }] }];
        } });

    var TuiUnfinishedValidatorModule = /** @class */ (function () {
        function TuiUnfinishedValidatorModule() {
        }
        return TuiUnfinishedValidatorModule;
    }());
    TuiUnfinishedValidatorModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiUnfinishedValidatorModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiUnfinishedValidatorModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiUnfinishedValidatorModule, declarations: [TuiUnfinishedValidatorDirective], exports: [TuiUnfinishedValidatorDirective] });
    TuiUnfinishedValidatorModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiUnfinishedValidatorModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiUnfinishedValidatorModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiUnfinishedValidatorDirective],
                        exports: [TuiUnfinishedValidatorDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiUnfinishedValidatorDirective = TuiUnfinishedValidatorDirective;
    exports.TuiUnfinishedValidatorModule = TuiUnfinishedValidatorModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
