// @taiga-ui/kit/bundles/taiga-ui-kit-components-select.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core'), require('@taiga-ui/kit/abstract'), require('@taiga-ui/kit/components/data-list-wrapper'), require('@angular/common'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/components/arrow'), require('@taiga-ui/kit/components/select-option'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/tokens'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/select', ['exports', '@angular/core', '@taiga-ui/core', '@taiga-ui/kit/abstract', '@taiga-ui/kit/components/data-list-wrapper', '@angular/common', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/kit/components/arrow', '@taiga-ui/kit/components/select-option', '@taiga-ui/kit/providers', '@taiga-ui/kit/tokens', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.select = {}), global.ng.core, global.i1, global["taiga-ui"].kit.abstract, global["taiga-ui"].kit.components["data-list-wrapper"], global.ng.common, global.ng.forms, global.i2, global["taiga-ui"].kit.components.arrow, global["taiga-ui"].kit.components["select-option"], global["taiga-ui"].kit.providers, global["taiga-ui"].kit.tokens, global.i4));
})(this, (function (exports, i0, i1, abstract, i1$1, i3, i5, i2, arrow, selectOption, providers, tokens, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1$1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiNativeSelectComponent = /** @class */ (function (_super) {
        __extends(TuiNativeSelectComponent, _super);
        function TuiNativeSelectComponent() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.items = [];
            return _this;
        }
        return TuiNativeSelectComponent;
    }(abstract.AbstractTuiNativeSelect));
    TuiNativeSelectComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeSelectComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiNativeSelectComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiNativeSelectComponent, selector: "select[tuiSelect]:not([labels]):not([multiple])", inputs: { items: "items" }, host: { listeners: { "change": "host.onValueChange($event.target.value)" }, properties: { "attr.aria-invalid": "host.invalid", "disabled": "host.disabled || control.readOnly", "tabIndex": "host.focusable ? 0 : -1", "value": "host.value" } }, providers: [
            i1.tuiAsDataList(TuiNativeSelectComponent),
            {
                provide: i0.TemplateRef,
                deps: [TuiNativeSelectComponent],
                useFactory: function (_a) {
                    var datalist = _a.datalist;
                    return datalist;
                },
            },
            {
                provide: abstract.AbstractTuiNativeSelect,
                useExisting: TuiNativeSelectComponent,
            },
        ], usesInheritance: true, ngImport: i0__namespace, template: "<tui-data-list-wrapper\n    *tuiDataList\n    [items]=\"items\"\n    [disabledItemHandler]=\"disabledItemHandler || itemsHandlers.disabledItemHandler\"\n></tui-data-list-wrapper>\n<option\n    *ngFor=\"let option of items\"\n    [value]=\"option\"\n    [disabled]=\"disabledItemHandler ? disabledItemHandler(option) : itemsHandlers.disabledItemHandler(option)\"\n>\n    {{ option }}\n</option>\n", styles: [":host{position:absolute;top:0;left:0;width:100%;height:100%;opacity:0}\n"], components: [{ type: i1__namespace.TuiDataListWrapperComponent, selector: "tui-data-list-wrapper:not([labels])", inputs: ["items"] }], directives: [{ type: i1__namespace$1.TuiDataListDirective, selector: "ng-template[tuiDataList]" }, { type: i3__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeSelectComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'select[tuiSelect]:not([labels]):not([multiple])',
                        templateUrl: './native-select.template.html',
                        providers: [
                            i1.tuiAsDataList(TuiNativeSelectComponent),
                            {
                                provide: i0.TemplateRef,
                                deps: [TuiNativeSelectComponent],
                                useFactory: function (_a) {
                                    var datalist = _a.datalist;
                                    return datalist;
                                },
                            },
                            {
                                provide: abstract.AbstractTuiNativeSelect,
                                useExisting: TuiNativeSelectComponent,
                            },
                        ],
                        host: {
                            '[attr.aria-invalid]': 'host.invalid',
                            '[disabled]': 'host.disabled || control.readOnly',
                            '[tabIndex]': 'host.focusable ? 0 : -1',
                            '[value]': 'host.value',
                            '(change)': 'host.onValueChange($event.target.value)',
                        },
                        styleUrls: ['./native-select.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { items: [{
                    type: i0.Input
                }] } });

    var TuiNativeSelectGroupComponent = /** @class */ (function (_super) {
        __extends(TuiNativeSelectGroupComponent, _super);
        function TuiNativeSelectGroupComponent() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.items = [];
            _this.labels = [];
            return _this;
        }
        return TuiNativeSelectGroupComponent;
    }(abstract.AbstractTuiNativeSelect));
    TuiNativeSelectGroupComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeSelectGroupComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiNativeSelectGroupComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiNativeSelectGroupComponent, selector: "select[tuiSelect][labels]:not([multiple])", inputs: { items: "items", labels: "labels" }, host: { listeners: { "change": "host.onValueChange($event.target.value)" }, properties: { "attr.aria-invalid": "host.invalid", "disabled": "host.disabled || control.readOnly", "tabIndex": "host.focusable ? 0 : -1", "value": "host.value" } }, providers: [
            i1.tuiAsDataList(TuiNativeSelectGroupComponent),
            {
                provide: i0.TemplateRef,
                deps: [TuiNativeSelectGroupComponent],
                useFactory: function (_a) {
                    var datalist = _a.datalist;
                    return datalist;
                },
            },
            {
                provide: abstract.AbstractTuiNativeSelect,
                useExisting: TuiNativeSelectGroupComponent,
            },
        ], usesInheritance: true, ngImport: i0__namespace, template: "<tui-data-list-wrapper\n    *tuiDataList\n    [items]=\"items\"\n    [labels]=\"labels\"\n    [disabledItemHandler]=\"disabledItemHandler || itemsHandlers.disabledItemHandler\"\n></tui-data-list-wrapper>\n<optgroup\n    *ngFor=\"let group of items; let index = index\"\n    [label]=\"labels[index]\"\n>\n    <option\n        *ngFor=\"let option of group\"\n        [value]=\"option\"\n        [disabled]=\"disabledItemHandler ? disabledItemHandler(option) : itemsHandlers.disabledItemHandler(option)\"\n    >\n        {{ option }}\n    </option>\n</optgroup>\n", styles: [":host{position:absolute;top:0;left:0;width:100%;height:100%;opacity:0}\n"], components: [{ type: i1__namespace.TuiDataListGroupWrapperComponent, selector: "tui-data-list-wrapper[labels]", inputs: ["items", "labels"] }], directives: [{ type: i1__namespace$1.TuiDataListDirective, selector: "ng-template[tuiDataList]" }, { type: i3__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiNativeSelectGroupComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'select[tuiSelect][labels]:not([multiple])',
                        templateUrl: './native-select-group.template.html',
                        providers: [
                            i1.tuiAsDataList(TuiNativeSelectGroupComponent),
                            {
                                provide: i0.TemplateRef,
                                deps: [TuiNativeSelectGroupComponent],
                                useFactory: function (_a) {
                                    var datalist = _a.datalist;
                                    return datalist;
                                },
                            },
                            {
                                provide: abstract.AbstractTuiNativeSelect,
                                useExisting: TuiNativeSelectGroupComponent,
                            },
                        ],
                        host: {
                            '[attr.aria-invalid]': 'host.invalid',
                            '[disabled]': 'host.disabled || control.readOnly',
                            '[tabIndex]': 'host.focusable ? 0 : -1',
                            '[value]': 'host.value',
                            '(change)': 'host.onValueChange($event.target.value)',
                        },
                        styleUrls: ['./native-select.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { items: [{
                    type: i0.Input
                }], labels: [{
                    type: i0.Input
                }] } });

    var TUI_SELECT_DEFAULT_OPTIONS = {
        valueContent: "",
    };
    /**
     * Default parameters for Select component
     */
    var TUI_SELECT_OPTIONS = new i0.InjectionToken("[TUI_SELECT_OPTIONS]", {
        factory: function () { return TUI_SELECT_DEFAULT_OPTIONS; },
    });
    var tuiSelectOptionsProvider = function (options) { return ({
        provide: TUI_SELECT_OPTIONS,
        useValue: Object.assign(Object.assign({}, TUI_SELECT_DEFAULT_OPTIONS), options),
    }); };

    var TuiSelectComponent = /** @class */ (function (_super) {
        __extends(TuiSelectComponent, _super);
        function TuiSelectComponent(control, cdr, textfieldCleaner, arrowMode, itemsHandlers, options, isMobile) {
            var _this = _super.call(this, control, cdr) || this;
            _this.textfieldCleaner = textfieldCleaner;
            _this.arrowMode = arrowMode;
            _this.itemsHandlers = itemsHandlers;
            _this.options = options;
            _this.isMobile = isMobile;
            _this.stringify = _this.itemsHandlers.stringify;
            _this.identityMatcher = _this.itemsHandlers.identityMatcher;
            _this.valueContent = _this.options.valueContent;
            return _this;
        }
        Object.defineProperty(TuiSelectComponent.prototype, "arrow", {
            get: function () {
                return !this.interactive ? this.arrowMode.disabled : this.arrowMode.interactive;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSelectComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSelectComponent.prototype, "focused", {
            get: function () {
                return (i2.tuiIsNativeFocused(this.nativeFocusableElement) ||
                    (!!this.hostedDropdown && this.hostedDropdown.focused));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSelectComponent.prototype, "nativeDropdownMode", {
            get: function () {
                return !!this.nativeSelect && this.isMobile;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSelectComponent.prototype, "computedValue", {
            get: function () {
                return this.value === null ? '' : this.stringify(this.value) || ' ';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSelectComponent.prototype, "computedContent", {
            get: function () {
                return this.valueContent || this.computedValue;
            },
            enumerable: false,
            configurable: true
        });
        TuiSelectComponent.prototype.onValueChange = function (value) {
            if (!value) {
                this.value = null;
            }
            else {
                this.value = value || null;
            }
        };
        TuiSelectComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
        };
        TuiSelectComponent.prototype.onKeyDownDelete = function () {
            if (this.textfieldCleaner.cleaner) {
                this.value = null;
            }
        };
        TuiSelectComponent.prototype.handleOption = function (option) {
            var _a;
            this.focusInput();
            this.value = option;
            (_a = this.hostedDropdown) === null || _a === void 0 ? void 0 : _a.updateOpen(false);
        };
        TuiSelectComponent.prototype.focusInput = function (preventScroll) {
            if (preventScroll === void 0) { preventScroll = false; }
            if (this.nativeFocusableElement) {
                this.nativeFocusableElement.focus({ preventScroll: preventScroll });
            }
        };
        return TuiSelectComponent;
    }(i2.AbstractTuiNullableControl));
    TuiSelectComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectComponent, deps: [{ token: i5.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i1.TUI_TEXTFIELD_CLEANER }, { token: arrow.TUI_ARROW_MODE }, { token: tokens.TUI_ITEMS_HANDLERS }, { token: TUI_SELECT_OPTIONS }, { token: i2.TUI_IS_MOBILE }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSelectComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSelectComponent, selector: "tui-select", inputs: { stringify: "stringify", identityMatcher: "identityMatcher", valueContent: "valueContent" }, providers: [
            i2.tuiAsFocusableItemAccessor(TuiSelectComponent),
            i2.tuiAsControl(TuiSelectComponent),
            i1.tuiAsDataListHost(TuiSelectComponent),
            i1.tuiAsOptionContent(selectOption.TUI_SELECT_OPTION),
        ], queries: [{ propertyName: "nativeSelect", first: true, predicate: abstract.AbstractTuiNativeSelect, descendants: true, static: true }, { propertyName: "datalist", first: true, predicate: i1.TuiDataListDirective, descendants: true, read: i0.TemplateRef }], viewQueries: [{ propertyName: "textfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }, { propertyName: "hostedDropdown", first: true, predicate: i1.TuiHostedDropdownComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"interactive && !nativeDropdownMode\"\n    [content]=\"datalist || ''\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-primitive-textfield\n        automation-id=\"tui-select__textfield\"\n        class=\"t-textfield\"\n        [tuiTextfieldIcon]=\"arrow ? icon : ''\"\n        [editable]=\"false\"\n        [pseudoFocus]=\"computedFocused\"\n        [pseudoHover]=\"pseudoHover\"\n        [pseudoActive]=\"pseudoActive\"\n        [invalid]=\"computedInvalid\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [disabled]=\"computedDisabled\"\n        [focusable]=\"computedFocusable\"\n        [value]=\"computedValue\"\n        (valueChange)=\"onValueChange($any($event))\"\n        (keydown.delete.prevent)=\"onKeyDownDelete()\"\n        (keydown.backspace.prevent)=\"onKeyDownDelete()\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            select=\"input\"\n            ngProjectAs=\"input\"\n        ></ng-content>\n\n        <ng-template #select>\n            <ng-content\n                *ngIf=\"isMobile\"\n                select=\"select\"\n            ></ng-content>\n        </ng-template>\n\n        <ng-container\n            *ngTemplateOutlet=\"select\"\n            ngProjectAs=\"select\"\n        ></ng-container>\n\n        <div\n            *ngIf=\"value !== null\"\n            ngProjectAs=\"tuiContent\"\n            automation-id=\"tui-select__value\"\n            class=\"t-value\"\n        >\n            <div\n                *polymorpheusOutlet=\"computedContent as text; context: {$implicit: value, active: computedFocused}\"\n                class=\"t-primitive-value\"\n                [textContent]=\"text\"\n            ></div>\n        </div>\n    </tui-primitive-textfield>\n\n    <ng-template #icon>\n        <div\n            tuiWrapper\n            appearance=\"icon\"\n            class=\"t-icon\"\n        >\n            <ng-container *polymorpheusOutlet=\"arrow\"></ng-container>\n        </div>\n    </ng-template>\n</tui-hosted-dropdown>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}:host:not(._readonly) ::ng-deep input:not(:disabled){cursor:pointer}:host._readonly ::ng-deep input{cursor:default}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}.t-value{display:flex;width:100%;align-items:center}.t-primitive-value{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}\n"], components: [{ type: i1__namespace$1.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i1__namespace$1.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }], directives: [{ type: i2__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i1__namespace$1.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i1__namespace$1.TuiTextfieldIconDirective, selector: "[tuiTextfieldIcon]", inputs: ["tuiTextfieldIcon"] }, { type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3__namespace.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1__namespace$1.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }], viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i2.tuiDefaultProp()
    ], TuiSelectComponent.prototype, "stringify", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiSelectComponent.prototype, "identityMatcher", void 0);
    __decorate([
        i2.tuiDefaultProp()
    ], TuiSelectComponent.prototype, "valueContent", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-select',
                        templateUrl: './select.template.html',
                        styleUrls: ['./select.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i2.tuiAsFocusableItemAccessor(TuiSelectComponent),
                            i2.tuiAsControl(TuiSelectComponent),
                            i1.tuiAsDataListHost(TuiSelectComponent),
                            i1.tuiAsOptionContent(selectOption.TUI_SELECT_OPTION),
                        ],
                        viewProviders: [providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER],
                    }]
            }], ctorParameters: function () {
            return [{ type: i5__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i5.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i1__namespace$1.TuiTextfieldCleanerDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_CLEANER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [arrow.TUI_ARROW_MODE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ITEMS_HANDLERS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SELECT_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i2.TUI_IS_MOBILE]
                        }] }];
        }, propDecorators: { textfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], hostedDropdown: [{
                    type: i0.ViewChild,
                    args: [i1.TuiHostedDropdownComponent]
                }], nativeSelect: [{
                    type: i0.ContentChild,
                    args: [abstract.AbstractTuiNativeSelect, { static: true }]
                }], stringify: [{
                    type: i0.Input
                }], identityMatcher: [{
                    type: i0.Input
                }], valueContent: [{
                    type: i0.Input
                }], datalist: [{
                    type: i0.ContentChild,
                    args: [i1.TuiDataListDirective, { read: i0.TemplateRef }]
                }] } });

    var TuiSelectDirective = /** @class */ (function (_super) {
        __extends(TuiSelectDirective, _super);
        function TuiSelectDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(TuiSelectDirective.prototype, "readOnly", {
            get: function () {
                return true;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSelectDirective.prototype, "value", {
            get: function () {
                return this.host.computedValue;
            },
            enumerable: false,
            configurable: true
        });
        TuiSelectDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        return TuiSelectDirective;
    }(i1.AbstractTuiTextfieldHost));
    TuiSelectDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSelectDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSelectDirective, selector: "tui-select", providers: [i1.tuiAsTextfieldHost(TuiSelectDirective)], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-select',
                        providers: [i1.tuiAsTextfieldHost(TuiSelectDirective)],
                    }]
            }] });

    var TuiSelectModule = /** @class */ (function () {
        function TuiSelectModule() {
        }
        return TuiSelectModule;
    }());
    TuiSelectModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSelectModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectModule, declarations: [TuiSelectComponent,
            TuiSelectDirective,
            TuiNativeSelectComponent,
            TuiNativeSelectGroupComponent], imports: [i3.CommonModule,
            i4.PolymorpheusModule,
            i2.TuiActiveZoneModule,
            i1.TuiPrimitiveTextfieldModule,
            i1.TuiHostedDropdownModule,
            selectOption.TuiSelectOptionModule,
            arrow.TuiArrowModule,
            i1.TuiWrapperModule,
            i1.TuiTextfieldControllerModule,
            i1$1.TuiDataListWrapperModule,
            i1.TuiDataListModule], exports: [TuiSelectComponent,
            TuiSelectDirective,
            i1.TuiTextfieldComponent,
            TuiNativeSelectComponent,
            TuiNativeSelectGroupComponent] });
    TuiSelectModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectModule, imports: [[
                i3.CommonModule,
                i4.PolymorpheusModule,
                i2.TuiActiveZoneModule,
                i1.TuiPrimitiveTextfieldModule,
                i1.TuiHostedDropdownModule,
                selectOption.TuiSelectOptionModule,
                arrow.TuiArrowModule,
                i1.TuiWrapperModule,
                i1.TuiTextfieldControllerModule,
                i1$1.TuiDataListWrapperModule,
                i1.TuiDataListModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSelectModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            i4.PolymorpheusModule,
                            i2.TuiActiveZoneModule,
                            i1.TuiPrimitiveTextfieldModule,
                            i1.TuiHostedDropdownModule,
                            selectOption.TuiSelectOptionModule,
                            arrow.TuiArrowModule,
                            i1.TuiWrapperModule,
                            i1.TuiTextfieldControllerModule,
                            i1$1.TuiDataListWrapperModule,
                            i1.TuiDataListModule,
                        ],
                        declarations: [
                            TuiSelectComponent,
                            TuiSelectDirective,
                            TuiNativeSelectComponent,
                            TuiNativeSelectGroupComponent,
                        ],
                        exports: [
                            TuiSelectComponent,
                            TuiSelectDirective,
                            i1.TuiTextfieldComponent,
                            TuiNativeSelectComponent,
                            TuiNativeSelectGroupComponent,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_SELECT_DEFAULT_OPTIONS = TUI_SELECT_DEFAULT_OPTIONS;
    exports.TUI_SELECT_OPTIONS = TUI_SELECT_OPTIONS;
    exports.TuiNativeSelectComponent = TuiNativeSelectComponent;
    exports.TuiNativeSelectGroupComponent = TuiNativeSelectGroupComponent;
    exports.TuiSelectComponent = TuiSelectComponent;
    exports.TuiSelectDirective = TuiSelectDirective;
    exports.TuiSelectModule = TuiSelectModule;
    exports.tuiSelectOptionsProvider = tuiSelectOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
