// @taiga-ui/kit/bundles/taiga-ui-kit-components-range.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/components/slider'), require('@taiga-ui/kit/utils'), require('@angular/common'), require('@taiga-ui/kit/constants'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/range', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/kit/components/slider', '@taiga-ui/kit/utils', '@angular/common', '@taiga-ui/kit/constants', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.range = {}), global.ng.core, global.ng.forms, global.i3, global["taiga-ui"].kit.components.slider, global["taiga-ui"].kit.utils, global.ng.common, global["taiga-ui"].kit.constants, global.rxjs, global.rxjs.operators));
})(this, (function (exports, i0, i2, i3, i1, utils, common, constants, i2$1, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace$1 = /*#__PURE__*/_interopNamespace(i2$1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiRangeComponent = /** @class */ (function (_super) {
        __extends(TuiRangeComponent, _super);
        function TuiRangeComponent(control, cdr, el) {
            var _this = _super.call(this, control, cdr) || this;
            _this.el = el;
            _this.min = 0;
            _this.max = 100;
            _this.step = 1;
            _this.size = 'm';
            _this.segments = 1;
            _this.keySteps = null;
            _this.slidersRefs = i3.EMPTY_QUERY;
            _this.lastActiveThumb = 'right';
            return _this;
        }
        Object.defineProperty(TuiRangeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                var _a = __read(this.slidersRefs, 2), sliderLeftRef = _a[0], sliderRightRef = _a[1];
                if (this.computedDisabled ||
                    !this.focusable ||
                    !sliderLeftRef ||
                    !sliderRightRef) {
                    return null;
                }
                var isLeftThumbLocked = this.right === 100;
                return isLeftThumbLocked
                    ? sliderRightRef.nativeElement
                    : sliderLeftRef.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiRangeComponent.prototype, "focused", {
            get: function () {
                return i3.tuiIsNativeFocusedIn(this.el.nativeElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiRangeComponent.prototype, "fractionStep", {
            get: function () {
                return this.step / (this.max - this.min);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiRangeComponent.prototype, "computedKeySteps", {
            get: function () {
                return this.computePureKeySteps(this.keySteps, this.min, this.max);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiRangeComponent.prototype, "segmentWidthRatio", {
            get: function () {
                return 1 / this.segments;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiRangeComponent.prototype, "left", {
            get: function () {
                return this.getPercentageFromValue(this.value[0]);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiRangeComponent.prototype, "right", {
            get: function () {
                return 100 - this.getPercentageFromValue(this.value[1]);
            },
            enumerable: false,
            configurable: true
        });
        TuiRangeComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
        };
        TuiRangeComponent.prototype.changeByStep = function (coefficient, target) {
            var _a = __read(this.slidersRefs, 2), sliderLeftRef = _a[0], sliderRightRef = _a[1];
            var leftThumbElement = sliderLeftRef.nativeElement;
            var rightThumbElement = sliderRightRef.nativeElement;
            var isRightThumb = target === this.el.nativeElement
                ? this.lastActiveThumb === 'right'
                : target === rightThumbElement;
            var activeThumbElement = isRightThumb ? rightThumbElement : leftThumbElement;
            var previousValue = isRightThumb ? this.value[1] : this.value[0];
            /** @bad TODO think about a solution without twice conversion */
            var previousFraction = this.getPercentageFromValue(previousValue) / 100;
            var newFractionValue = previousFraction + coefficient * this.fractionStep;
            this.processValue(this.getValueFromFraction(newFractionValue), isRightThumb);
            if (activeThumbElement) {
                activeThumbElement.focus();
            }
        };
        TuiRangeComponent.prototype.processValue = function (value, right) {
            if (right) {
                this.updateEnd(value);
            }
            else {
                this.updateStart(value);
            }
            this.lastActiveThumb = right ? 'right' : 'left';
        };
        TuiRangeComponent.prototype.getValueFromFraction = function (fraction) {
            var guardedFraction = i3.tuiClamp(i3.tuiQuantize(fraction, this.fractionStep), 0, 1);
            return utils.tuiPercentageToKeyStepValue(guardedFraction * 100, this.computedKeySteps);
        };
        TuiRangeComponent.prototype.getPercentageFromValue = function (value) {
            return utils.tuiKeyStepValueToPercentage(value, this.computedKeySteps);
        };
        TuiRangeComponent.prototype.getFallbackValue = function () {
            return [0, 0];
        };
        TuiRangeComponent.prototype.computePureKeySteps = function (keySteps, min, max) {
            return (keySteps || [
                [0, min],
                [100, max],
            ]);
        };
        TuiRangeComponent.prototype.updateStart = function (value) {
            this.value = [Math.min(value, this.value[1]), this.value[1]];
        };
        TuiRangeComponent.prototype.updateEnd = function (value) {
            this.value = [this.value[0], Math.max(value, this.value[0])];
        };
        return TuiRangeComponent;
    }(i3.AbstractTuiControl));
    TuiRangeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeComponent, deps: [{ token: i2.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiRangeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRangeComponent, selector: "tui-range", inputs: { min: "min", max: "max", step: "step", size: "size", segments: "segments", keySteps: "keySteps" }, host: { listeners: { "focusin": "onFocused(true)", "focusout": "onFocused(false)", "keydown.arrowUp.prevent": "changeByStep(1,$event.target)", "keydown.arrowRight.prevent": "changeByStep(1,$event.target)", "keydown.arrowLeft.prevent": "changeByStep(-1,$event.target)", "keydown.arrowDown.prevent": "changeByStep(-1,$event.target)" }, properties: { "attr.tabindex": "-1", "attr.aria-disabled": "computedDisabled", "attr.data-size": "this.size", "style.--left.%": "this.left", "style.--right.%": "this.right" } }, viewQueries: [{ propertyName: "slidersRefs", predicate: i1.TuiSliderComponent, descendants: true, read: i0.ElementRef }], usesInheritance: true, ngImport: i0__namespace, template: "<div\n    class=\"t-track\"\n    [style.--segment-width.%]=\"segmentWidthRatio * 100\"\n    [style.--bg-size-ratio]=\"1 - segmentWidthRatio\"\n>\n    <input\n        tuiSlider\n        readonly\n        type=\"range\"\n        step=\"any\"\n        automation-id=\"tui-range__left\"\n        class=\"t-thumb\"\n        [min]=\"min\"\n        [max]=\"max\"\n        [ngModel]=\"value[0]\"\n        [size]=\"size\"\n        [keySteps]=\"computedKeySteps\"\n        [tuiFocusable]=\"focusable\"\n        [attr.disabled]=\"computedDisabled || null\"\n    />\n    <input\n        tuiSlider\n        readonly\n        type=\"range\"\n        step=\"any\"\n        automation-id=\"tui-range__right\"\n        class=\"t-thumb\"\n        [min]=\"min\"\n        [max]=\"max\"\n        [ngModel]=\"value[1]\"\n        [size]=\"size\"\n        [keySteps]=\"computedKeySteps\"\n        [tuiFocusable]=\"focusable\"\n        [attr.disabled]=\"computedDisabled || null\"\n    />\n</div>\n", styles: [":host{position:relative;display:block;height:.125rem;border-radius:var(--tui-radius-m);background:var(--tui-base-03);cursor:pointer;outline:none;margin:.4375rem 0;touch-action:pan-x}:host:active{cursor:ew-resize}:host:after{content:\"\";position:absolute;top:-.4375rem;bottom:-.4375rem;width:100%}:host._disabled{opacity:var(--tui-disabled-opacity);cursor:auto}:host[data-size=s] .t-track{position:relative;margin:0 .25rem;height:100%}:host[data-size=s] .t-track:before{content:\"\";position:absolute;top:0;left:var(--left);right:var(--right);height:100%;background:var(--tui-primary);margin:0 -.25rem}:host[data-size=s] .t-track:after{position:absolute;top:0;left:0;bottom:0;right:0;content:\"\";left:.125rem;right:.375rem;background-image:repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 calc(var(--segment-width) / var(--bg-size-ratio)));background-position-x:right;background-repeat:no-repeat;background-size:calc(100% * var(--bg-size-ratio))}:host[data-size=m] .t-track{position:relative;margin:0 .5rem;height:100%}:host[data-size=m] .t-track:before{content:\"\";position:absolute;top:0;left:var(--left);right:var(--right);height:100%;background:var(--tui-primary);margin:0 -.5rem}:host[data-size=m] .t-track:after{position:absolute;top:0;left:0;bottom:0;right:0;content:\"\";left:.375rem;right:.625rem;background-image:repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 calc(var(--segment-width) / var(--bg-size-ratio)));background-position-x:right;background-repeat:no-repeat;background-size:calc(100% * var(--bg-size-ratio))}.t-thumb{pointer-events:none;position:absolute;top:.0625rem;left:0;right:0;z-index:1;transform:translateY(-50%);margin:0}.t-thumb::-webkit-slider-thumb{pointer-events:all}.t-thumb::-moz-range-thumb{pointer-events:all}input[type=range].t-thumb::-webkit-slider-runnable-track{background:transparent}input[type=range].t-thumb::-moz-range-track{background:transparent}input[type=range].t-thumb::-moz-range-progress{background:transparent}input[type=range].t-thumb::-ms-track{background:transparent}input[type=range].t-thumb::-ms-fill-lower{background:transparent}.t-thumb:last-of-type::-webkit-slider-thumb{transform:translate(50%)}.t-thumb:first-of-type::-webkit-slider-thumb{transform:translate(-50%)}.t-thumb:last-of-type::-moz-range-thumb{transform:translate(50%)}.t-thumb:first-of-type::-moz-range-thumb{transform:translate(-50%)}:host._disabled .t-thumb{opacity:1}\n"], components: [{ type: i1__namespace.TuiSliderComponent, selector: "input[type=range][tuiSlider]", inputs: ["size", "segments"] }], directives: [{ type: i1__namespace.TuiSliderReadonlyDirective, selector: "input[tuiSlider][readonly]", inputs: ["readonly"] }, { type: i1__namespace.TuiSliderKeyStepsDirective, selector: "input[tuiSlider][keySteps]", inputs: ["keySteps"] }, { type: i2__namespace.RangeValueAccessor, selector: "input[type=range][formControlName],input[type=range][formControl],input[type=range][ngModel]" }, { type: i2__namespace.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { type: i2__namespace.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i2__namespace.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { type: i3__namespace.TuiFocusableDirective, selector: "[tuiFocusable]", inputs: ["tuiFocusable"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i3.tuiDefaultProp()
    ], TuiRangeComponent.prototype, "min", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiRangeComponent.prototype, "max", void 0);
    __decorate([
        i3.tuiDefaultProp(function (s) { return s > 0; }, 'Step must be a non-negative number')
    ], TuiRangeComponent.prototype, "step", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiRangeComponent.prototype, "size", void 0);
    __decorate([
        i3.tuiDefaultProp(function (s) { return s > 0 && Number.isInteger(s); }, 'Segments must be positive integer')
    ], TuiRangeComponent.prototype, "segments", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiRangeComponent.prototype, "keySteps", void 0);
    __decorate([
        i3.tuiPure
    ], TuiRangeComponent.prototype, "computePureKeySteps", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-range',
                        templateUrl: './range.template.html',
                        styleUrls: ['./range.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        host: {
                            '[attr.tabindex]': '-1',
                            '[attr.aria-disabled]': 'computedDisabled',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i2__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i2.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], step: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], segments: [{
                    type: i0.Input
                }], keySteps: [{
                    type: i0.Input
                }], slidersRefs: [{
                    type: i0.ViewChildren,
                    args: [i1.TuiSliderComponent, { read: i0.ElementRef }]
                }], left: [{
                    type: i0.HostBinding,
                    args: ['style.--left.%']
                }], right: [{
                    type: i0.HostBinding,
                    args: ['style.--right.%']
                }], onFocused: [{
                    type: i0.HostListener,
                    args: ['focusin', ['true']]
                }, {
                    type: i0.HostListener,
                    args: ['focusout', ['false']]
                }], changeByStep: [{
                    type: i0.HostListener,
                    args: ['keydown.arrowUp.prevent', ['1', '$event.target']]
                }, {
                    type: i0.HostListener,
                    args: ['keydown.arrowRight.prevent', ['1', '$event.target']]
                }, {
                    type: i0.HostListener,
                    args: ['keydown.arrowLeft.prevent', ['-1', '$event.target']]
                }, {
                    type: i0.HostListener,
                    args: ['keydown.arrowDown.prevent', ['-1', '$event.target']]
                }], computePureKeySteps: [] } });

    var TuiRangeChangeDirective = /** @class */ (function () {
        function TuiRangeChangeDirective(doc, el, range, destroy$) {
            var _this = this;
            this.doc = doc;
            this.el = el;
            this.range = range;
            /**
             * TODO replace with pointer events (when all supported browsers can handle them).
             * Don't forget to use setPointerCapture instead of listening all doc events
             */
            this.pointerDown$ = i2$1.merge(i3.tuiTypedFromEvent(this.el.nativeElement, 'touchstart', {
                passive: true,
            }).pipe(operators.filter(function (_a) {
                var touches = _a.touches;
                return touches.length === 1;
            }), operators.map(function (_a) {
                var touches = _a.touches;
                return touches[0];
            })), i3.tuiTypedFromEvent(this.el.nativeElement, 'mousedown', { passive: true }));
            this.pointerMove$ = i2$1.merge(i3.tuiTypedFromEvent(this.doc, 'touchmove').pipe(operators.filter(function (_a) {
                var touches = _a.touches;
                return touches.length === 1;
            }), operators.map(function (_a) {
                var touches = _a.touches;
                return touches[0];
            })), i3.tuiTypedFromEvent(this.doc, 'mousemove'));
            this.pointerUp$ = i2$1.merge(i3.tuiTypedFromEvent(this.doc, 'touchend', { passive: true }), i3.tuiTypedFromEvent(this.doc, 'mouseup', { passive: true }));
            this.activeThumbChange = new i0.EventEmitter();
            var activeThumb;
            this.pointerDown$
                .pipe(operators.tap(function (_a) {
                var clientX = _a.clientX, target = _a.target;
                activeThumb = _this.detectActiveThumb(clientX, target);
                _this.activeThumbChange.emit(activeThumb);
                if (_this.range.focusable) {
                    el.nativeElement.focus();
                }
            }), operators.switchMap(function (event) { return _this.pointerMove$.pipe(operators.startWith(event)); }), operators.map(function (_a) {
                var clientX = _a.clientX;
                return _this.getFractionFromEvents(clientX);
            }), operators.takeUntil(this.pointerUp$), operators.repeat(), operators.takeUntil(destroy$))
                .subscribe(function (fraction) {
                var value = _this.range.getValueFromFraction(fraction);
                _this.range.processValue(value, activeThumb === 'right');
            });
        }
        TuiRangeChangeDirective.prototype.getFractionFromEvents = function (clickClientX) {
            var hostRect = this.el.nativeElement.getBoundingClientRect();
            var value = clickClientX - hostRect.left;
            var total = hostRect.width;
            return i3.tuiClamp(i3.tuiRound(value / total, constants.TUI_FLOATING_PRECISION), 0, 1);
        };
        TuiRangeChangeDirective.prototype.detectActiveThumb = function (clientX, target) {
            var _a = __read(this.range.slidersRefs, 2), leftSliderRef = _a[0], rightSliderRef = _a[1];
            switch (target) {
                case leftSliderRef.nativeElement:
                    return 'left';
                case rightSliderRef.nativeElement:
                    return 'right';
                default:
                    return this.findNearestActiveThumb(clientX);
            }
        };
        TuiRangeChangeDirective.prototype.findNearestActiveThumb = function (clientX) {
            var fraction = this.getFractionFromEvents(clientX);
            var deltaLeft = fraction * 100 - this.range.left;
            var deltaRight = fraction * 100 - 100 + this.range.right;
            return Math.abs(deltaLeft) > Math.abs(deltaRight) ||
                deltaRight > 0 ||
                (this.range.left === 0 && this.range.right === 100)
                ? 'right'
                : 'left';
        };
        return TuiRangeChangeDirective;
    }());
    TuiRangeChangeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeChangeDirective, deps: [{ token: common.DOCUMENT }, { token: i0.ElementRef }, { token: TuiRangeComponent }, { token: i3.TuiDestroyService, self: true }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiRangeChangeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiRangeChangeDirective, selector: "tui-range", outputs: { activeThumbChange: "activeThumbChange" }, providers: [i3.TuiDestroyService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeChangeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-range',
                        providers: [i3.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: Document, decorators: [{
                            type: i0.Inject,
                            args: [common.DOCUMENT]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: TuiRangeComponent, decorators: [{
                            type: i0.Inject,
                            args: [TuiRangeComponent]
                        }] }, { type: i2__namespace$1.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDestroyService]
                        }] }];
        }, propDecorators: { activeThumbChange: [{
                    type: i0.Output
                }] } });

    var TuiRangeModule = /** @class */ (function () {
        function TuiRangeModule() {
        }
        return TuiRangeModule;
    }());
    TuiRangeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiRangeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeModule, declarations: [TuiRangeComponent, TuiRangeChangeDirective], imports: [common.CommonModule, i3.TuiFocusableModule, i1.TuiSliderModule, i2.FormsModule], exports: [TuiRangeComponent, TuiRangeChangeDirective] });
    TuiRangeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeModule, imports: [[common.CommonModule, i3.TuiFocusableModule, i1.TuiSliderModule, i2.FormsModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRangeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [common.CommonModule, i3.TuiFocusableModule, i1.TuiSliderModule, i2.FormsModule],
                        declarations: [TuiRangeComponent, TuiRangeChangeDirective],
                        exports: [TuiRangeComponent, TuiRangeChangeDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiRangeChangeDirective = TuiRangeChangeDirective;
    exports.TuiRangeComponent = TuiRangeComponent;
    exports.TuiRangeModule = TuiRangeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
