// @taiga-ui/kit/bundles/taiga-ui-kit-pipes-to-year.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/pipes/to-year', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.pipes = global["taiga-ui"].kit.pipes || {}, global["taiga-ui"].kit.pipes["to-year"] = {}), global.ng.core, global.cdk));
})(this, (function (exports, i0, cdk) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiToYearPipe = /** @class */ (function () {
        function TuiToYearPipe() {
        }
        TuiToYearPipe.prototype.transform = function (value) {
            return new cdk.TuiYear(value);
        };
        return TuiToYearPipe;
    }());
    TuiToYearPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiToYearPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipe, name: "tuiToYear" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: "tuiToYear",
                    }]
            }] });

    var TuiToYearPipeModule = /** @class */ (function () {
        function TuiToYearPipeModule() {
        }
        return TuiToYearPipeModule;
    }());
    TuiToYearPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiToYearPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipeModule, declarations: [TuiToYearPipe], exports: [TuiToYearPipe] });
    TuiToYearPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToYearPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiToYearPipe],
                        exports: [TuiToYearPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiToYearPipe = TuiToYearPipe;
    exports.TuiToYearPipeModule = TuiToYearPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
