// @taiga-ui/kit/bundles/taiga-ui-kit-pipes-stringify.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/pipes/stringify', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.pipes = global["taiga-ui"].kit.pipes || {}, global["taiga-ui"].kit.pipes.stringify = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiStringifyPipe = /** @class */ (function () {
        function TuiStringifyPipe() {
        }
        TuiStringifyPipe.prototype.transform = function (key) {
            return function (value) { var _a; return String((_a = value[key]) !== null && _a !== void 0 ? _a : ""); };
        };
        return TuiStringifyPipe;
    }());
    TuiStringifyPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiStringifyPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipe, name: "tuiStringify" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: "tuiStringify",
                    }]
            }] });

    var TuiStringifyPipeModule = /** @class */ (function () {
        function TuiStringifyPipeModule() {
        }
        return TuiStringifyPipeModule;
    }());
    TuiStringifyPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiStringifyPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipeModule, declarations: [TuiStringifyPipe], exports: [TuiStringifyPipe] });
    TuiStringifyPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStringifyPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiStringifyPipe],
                        exports: [TuiStringifyPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiStringifyPipe = TuiStringifyPipe;
    exports.TuiStringifyPipeModule = TuiStringifyPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
