// @taiga-ui/kit/bundles/taiga-ui-kit-components-files.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/tokens'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/kit/utils'), require('@taiga-ui/core'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/files', ['exports', '@angular/core', '@angular/platform-browser', '@taiga-ui/cdk', '@taiga-ui/kit/tokens', 'rxjs', 'rxjs/operators', '@taiga-ui/kit/utils', '@taiga-ui/core', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.files = {}), global.ng.core, global.ng.platformBrowser, global.i3, global["taiga-ui"].kit.tokens, global.rxjs, global.rxjs.operators, global["taiga-ui"].kit.utils, global.i1, global.ng.common, global.i4));
})(this, (function (exports, i0, i5, i3, tokens, i3$1, operators, utils, i1, i2, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i3__namespace$1 = /*#__PURE__*/_interopNamespace(i3$1);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_FILE_DEFAULT_OPTIONS = {
        formatSize: utils.tuiFormatSize,
    };
    /**
     * Default parameters for file component
     */
    var TUI_FILE_OPTIONS = new i0.InjectionToken("[TUI_FILE_OPTIONS]", {
        factory: function () { return TUI_FILE_DEFAULT_OPTIONS; },
    });

    var TuiFileComponent = /** @class */ (function () {
        function TuiFileComponent(sanitizer, fileTexts$, options, units$) {
            this.sanitizer = sanitizer;
            this.fileTexts$ = fileTexts$;
            this.options = options;
            this.units$ = units$;
            this.file = { name: '' };
            this.state = 'normal';
            this.size = 'm';
            this.showDelete = true;
            this.showSize = true;
            this.removed = new i0.EventEmitter();
            this.focused = false;
        }
        Object.defineProperty(TuiFileComponent.prototype, "preview", {
            get: function () {
                return this.isBig ? this.createPreview(this.file, this.sanitizer) : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "isBig", {
            get: function () {
                return this.size === 'l';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "isLoading", {
            get: function () {
                return this.state === 'loading';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "isError", {
            get: function () {
                return this.state === 'error';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "isDeleted", {
            get: function () {
                return this.state === 'deleted';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "allowDelete", {
            get: function () {
                return this.showDelete && i3.tuiIsObserved(this.removed);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "icon", {
            get: function () {
                if (this.state === 'normal' && this.isBig) {
                    return 'tuiIconFileLarge';
                }
                switch (this.state) {
                    case 'deleted':
                        return 'tuiIconTrashLarge';
                    case 'error':
                        return 'tuiIconAlertCircleLarge';
                    default:
                        return 'tuiIconCheckCircleLarge';
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "src", {
            get: function () {
                return this.file.src || '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "name", {
            get: function () {
                return this.getName(this.file);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "type", {
            get: function () {
                return this.getType(this.file);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "content$", {
            get: function () {
                return this.calculateContent$(this.state, this.file, this.fileTexts$);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiFileComponent.prototype, "fileSize$", {
            get: function () {
                return this.calculateFileSize$(this.file, this.units$);
            },
            enumerable: false,
            configurable: true
        });
        TuiFileComponent.prototype.onRemoveClick = function () {
            this.removed.emit();
        };
        TuiFileComponent.prototype.onFocusVisible = function (focusVisible) {
            this.focused = focusVisible;
        };
        TuiFileComponent.prototype.calculateContent$ = function (state, file, fileTexts$) {
            return state === 'error' && !file.content
                ? fileTexts$.pipe(operators.map(function (texts) { return texts.loadingError; }))
                : i3$1.of(this.file.content || '');
        };
        TuiFileComponent.prototype.calculateFileSize$ = function (file, units$) {
            var _this = this;
            return units$.pipe(operators.map(function (units) { return _this.options.formatSize(units, file.size); }));
        };
        TuiFileComponent.prototype.createPreview = function (file, sanitizer) {
            if (file.src) {
                return file.src;
            }
            // TODO: iframe warning
            if (file instanceof File && file.type && file.type.startsWith('image/')) {
                return sanitizer.bypassSecurityTrustUrl(URL.createObjectURL(file));
            }
            return '';
        };
        TuiFileComponent.prototype.getName = function (file) {
            return file.name.split('.').slice(0, -1).join('.');
        };
        TuiFileComponent.prototype.getType = function (file) {
            return "." + file.name.split('.').pop() || '';
        };
        return TuiFileComponent;
    }());
    TuiFileComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFileComponent, deps: [{ token: i5.DomSanitizer }, { token: tokens.TUI_FILE_TEXTS }, { token: TUI_FILE_OPTIONS }, { token: tokens.TUI_DIGITAL_INFORMATION_UNITS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiFileComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiFileComponent, selector: "tui-file", inputs: { file: "file", state: "state", size: "size", showDelete: "showDelete", showSize: "showSize", leftContent: "leftContent" }, outputs: { removed: "removed" }, host: { properties: { "class._focused": "this.focused", "class._link": "this.src" } }, ngImport: i0__namespace, template: "<ng-container\n    *ngIf=\"!src; else withLink\"\n    [ngTemplateOutlet]=\"fileInfoTemplate\"\n></ng-container>\n\n<ng-template #withLink>\n    <a\n        rel=\"noreferrer noopener\"\n        target=\"_blank\"\n        class=\"t-link\"\n        [href]=\"src\"\n        (tuiFocusVisibleChange)=\"onFocusVisible($event)\"\n    >\n        <ng-container [ngTemplateOutlet]=\"fileInfoTemplate\"></ng-container>\n    </a>\n</ng-template>\n\n<ng-template #fileInfoTemplate>\n    <div\n        class=\"t-preview\"\n        [class.t-preview_big]=\"isBig\"\n    >\n        <ng-container *polymorpheusOutlet=\"leftContent || defaultLeftContent as text\">\n            {{ text }}\n        </ng-container>\n    </div>\n    <div class=\"t-wrapper\">\n        <div class=\"t-text\">\n            <div\n                automation-id=\"tui-file__name\"\n                class=\"t-name\"\n            >\n                {{ name }}\n            </div>\n            <div\n                automation-id=\"tui-file__type\"\n                class=\"t-type\"\n            >\n                {{ type }}\n            </div>\n            <div\n                *ngIf=\"showSize && (fileSize$ | async) as fileSize\"\n                automation-id=\"tui-file__size\"\n                class=\"t-size\"\n            >\n                {{ fileSize }}\n            </div>\n        </div>\n        <div\n            *ngIf=\"content$ | async as content\"\n            automation-id=\"tui-file__content\"\n            class=\"t-content\"\n        >\n            <ng-container *polymorpheusOutlet=\"content as text\">\n                {{ text }}\n            </ng-container>\n        </div>\n        <ng-content></ng-content>\n    </div>\n    <ng-container *ngIf=\"allowDelete\">\n        <button\n            *ngIf=\"fileTexts$ | async as texts\"\n            automation-id=\"tui-file__remove\"\n            tuiIconButton\n            type=\"button\"\n            size=\"xs\"\n            icon=\"tuiIconCloseLarge\"\n            appearance=\"icon\"\n            tuiPreventDefault=\"mousedown\"\n            class=\"t-remove\"\n            [title]=\"texts.remove\"\n            (click.prevent)=\"onRemoveClick()\"\n        ></button>\n    </ng-container>\n</ng-template>\n\n<ng-template #defaultLeftContent>\n    <img\n        *ngIf=\"preview; else loader\"\n        automation-id=\"tui-file__preview\"\n        alt=\"file preview\"\n        class=\"t-image\"\n        [src]=\"preview\"\n    />\n    <ng-template #loader>\n        <tui-loader\n            *ngIf=\"isLoading; else svg\"\n            automation-id=\"tui-file__loader\"\n            class=\"t-loader\"\n            [inheritColor]=\"isBig\"\n        ></tui-loader>\n    </ng-template>\n    <ng-template #svg>\n        <tui-svg\n            automation-id=\"tui-file__icon\"\n            class=\"t-icon\"\n            [class.t-icon_blank]=\"isBig || isDeleted\"\n            [class.t-icon_error]=\"isError\"\n            [src]=\"icon\"\n        ></tui-svg>\n    </ng-template>\n</ng-template>\n", styles: [":host{position:relative;display:flex;align-items:center;font:var(--tui-font-text-m);background:var(--tui-base-01);border:1px solid var(--tui-base-03);border-radius:var(--tui-radius-m)}:host[data-mode=onDark]{color:var(--tui-text-01-night)}:host:not(._link){padding:.5625rem 2.1875rem .5625rem .5625rem}:host:hover .t-remove{opacity:1}:host._focused{border-color:var(--tui-focus);box-shadow:0 0 0 1px inset var(--tui-focus)}.t-link{display:flex;flex:1;padding:.5625rem 2.1875rem .5625rem .5625rem;text-decoration:none;outline:none;cursor:pointer;color:var(--tui-text-01);max-width:calc(100% - 2.75rem)}.t-link:hover{background-color:var(--tui-base-02)}.t-preview{position:relative;display:flex;align-items:center;justify-content:center;flex-shrink:0;width:1.5rem;height:1.5rem;margin-right:.75rem;border-radius:var(--tui-radius-m);overflow:hidden;color:var(--tui-base-06)}.t-preview_big{width:4rem;height:4rem;margin-right:1rem}.t-preview_big:before{position:absolute;top:0;left:0;width:100%;height:100%;content:\"\";background:var(--tui-base-08);opacity:.08}.t-image{max-width:100%;max-height:100%}.t-loader{position:absolute;top:0;left:0;width:100%;height:100%}.t-icon{position:absolute;top:0;left:0;width:100%;height:100%;color:var(--tui-success-fill)}.t-icon_blank{color:var(--tui-base-06)}.t-icon_error{color:var(--tui-error-fill)}.t-remove{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:absolute;top:.625rem;right:.625rem}.t-remove:focus{opacity:1}.t-remove:focus-visible{box-shadow:inset 0 0 0 2px var(--tui-focus)}@media (hover: hover) and (pointer: fine){.t-remove{opacity:0}}.t-wrapper{display:flex;flex-direction:column;justify-content:center;overflow:hidden}.t-text{display:flex}.t-size{flex-shrink:0;opacity:var(--tui-disabled-opacity);margin-left:.5rem}.t-type{flex-shrink:0}.t-name{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.t-content{font:var(--tui-font-text-s);color:var(--tui-error-fill)}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }, { type: i1__namespace.TuiLoaderComponent, selector: "tui-loader", inputs: ["size", "inheritColor", "overlay", "textContent", "showLoader"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2__namespace.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }, { type: i3__namespace.TuiFocusVisibleDirective, selector: "[tuiFocusVisibleChange]", outputs: ["tuiFocusVisibleChange"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i3__namespace.TuiPreventDefaultDirective, selector: "[tuiPreventDefault]", inputs: ["tuiPreventDefault"] }], pipes: { "async": i2__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i3.tuiDefaultProp()
    ], TuiFileComponent.prototype, "file", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiFileComponent.prototype, "state", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiFileComponent.prototype, "size", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiFileComponent.prototype, "showDelete", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiFileComponent.prototype, "showSize", void 0);
    __decorate([
        i3.tuiPure
    ], TuiFileComponent.prototype, "calculateContent$", null);
    __decorate([
        i3.tuiPure
    ], TuiFileComponent.prototype, "calculateFileSize$", null);
    __decorate([
        i3.tuiPure
    ], TuiFileComponent.prototype, "createPreview", null);
    __decorate([
        i3.tuiPure
    ], TuiFileComponent.prototype, "getName", null);
    __decorate([
        i3.tuiPure
    ], TuiFileComponent.prototype, "getType", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFileComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-file',
                        templateUrl: './file.template.html',
                        styleUrls: ['./file.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i5__namespace.DomSanitizer, decorators: [{
                            type: i0.Inject,
                            args: [i5.DomSanitizer]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_FILE_TEXTS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_FILE_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DIGITAL_INFORMATION_UNITS]
                        }] }];
        }, propDecorators: { file: [{
                    type: i0.Input
                }], state: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }], showDelete: [{
                    type: i0.Input
                }], showSize: [{
                    type: i0.Input
                }], leftContent: [{
                    type: i0.Input
                }], removed: [{
                    type: i0.Output
                }], focused: [{
                    type: i0.HostBinding,
                    args: ['class._focused']
                }], src: [{
                    type: i0.HostBinding,
                    args: ['class._link']
                }], calculateContent$: [], calculateFileSize$: [], createPreview: [], getName: [], getType: [] } });

    var TuiFilesComponent = /** @class */ (function () {
        function TuiFilesComponent(hideText$, showAllText$) {
            this.hideText$ = hideText$;
            this.showAllText$ = showAllText$;
            this.items = i3.EMPTY_QUERY;
            this.max = 0;
            this.hidden = true;
        }
        Object.defineProperty(TuiFilesComponent.prototype, "hasExtraItems", {
            get: function () {
                return !!this.max && this.items.length > this.max;
            },
            enumerable: false,
            configurable: true
        });
        TuiFilesComponent.prototype.toggle = function () {
            this.hidden = !this.hidden;
        };
        return TuiFilesComponent;
    }());
    TuiFilesComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFilesComponent, deps: [{ token: tokens.TUI_HIDE_TEXT }, { token: tokens.TUI_SHOW_ALL_TEXT }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiFilesComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiFilesComponent, selector: "tui-files", inputs: { max: "max" }, queries: [{ propertyName: "items", predicate: i3.TuiItemDirective, read: i0.TemplateRef }], ngImport: i0__namespace, template: "<ng-container *ngIf=\"items?.changes | async\"></ng-container>\n\n<section\n    tuiGroup\n    orientation=\"vertical\"\n    class=\"t-files\"\n    [collapsed]=\"true\"\n>\n    <ng-content></ng-content>\n    <ng-container *ngFor=\"let item of items; let index = index\">\n        <ng-container\n            *ngIf=\"!max || index < max\"\n            [ngTemplateOutlet]=\"item\"\n        ></ng-container>\n    </ng-container>\n    <tui-expand\n        *ngIf=\"hasExtraItems\"\n        [expanded]=\"!hidden\"\n    >\n        <div class=\"t-extra-items\">\n            <ng-container *ngFor=\"let item of items; let index = index\">\n                <ng-container\n                    *ngIf=\"max && index >= max\"\n                    [ngTemplateOutlet]=\"item\"\n                ></ng-container>\n            </ng-container>\n        </div>\n    </tui-expand>\n    <button\n        *ngIf=\"hasExtraItems\"\n        tuiButton\n        appearance=\"outline\"\n        size=\"m\"\n        type=\"button\"\n        class=\"t-button\"\n        [class.t-button_collapsed]=\"hidden\"\n        (click)=\"toggle()\"\n    >\n        {{ (hidden ? showAllText$ : hideText$) | async }}\n    </button>\n</section>\n", styles: ["tui-files{display:block}tui-files .t-files{position:relative;display:block;width:100%;height:100%;border-radius:var(--tui-radius-m);overflow:hidden}tui-files .t-button{z-index:3;width:100%;background:var(--tui-base-01)}tui-files .t-button_collapsed{box-shadow:0 1.125rem 1.875rem #0000007a;margin-top:-1.5rem}tui-files .t-extra-items tui-file{border-radius:0;margin-top:0;margin-bottom:-.0625rem}\n"], components: [{ type: i1__namespace.TuiExpandComponent, selector: "tui-expand", inputs: ["async", "expanded"] }, { type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1__namespace.TuiGroupDirective, selector: "[tuiGroup]:not(ng-container)", inputs: ["orientation", "adaptive", "collapsed", "rounded", "size"] }, { type: i2__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2__namespace.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], pipes: { "async": i2__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
    __decorate([
        i3.tuiDefaultProp()
    ], TuiFilesComponent.prototype, "max", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFilesComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-files',
                        templateUrl: './files.component.html',
                        styleUrls: ['./files.component.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        encapsulation: i0.ViewEncapsulation.None,
                    }]
            }], ctorParameters: function () {
            return [{ type: i3__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_HIDE_TEXT]
                        }] }, { type: i3__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_SHOW_ALL_TEXT]
                        }] }];
        }, propDecorators: { items: [{
                    type: i0.ContentChildren,
                    args: [i3.TuiItemDirective, { read: i0.TemplateRef }]
                }], max: [{
                    type: i0.Input
                }] } });

    var TuiFilesModule = /** @class */ (function () {
        function TuiFilesModule() {
        }
        return TuiFilesModule;
    }());
    TuiFilesModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFilesModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFilesModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFilesModule, declarations: [TuiFilesComponent, TuiFileComponent], imports: [i2.CommonModule,
            i1.TuiGroupModule,
            i1.TuiLoaderModule,
            i4.PolymorpheusModule,
            i1.TuiSvgModule,
            i3.TuiFocusVisibleModule,
            i1.TuiButtonModule,
            i3.TuiItemModule,
            i1.TuiExpandModule,
            i3.TuiPreventDefaultModule], exports: [TuiFilesComponent, TuiFileComponent, i3.TuiItemDirective] });
    TuiFilesModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFilesModule, imports: [[
                i2.CommonModule,
                i1.TuiGroupModule,
                i1.TuiLoaderModule,
                i4.PolymorpheusModule,
                i1.TuiSvgModule,
                i3.TuiFocusVisibleModule,
                i1.TuiButtonModule,
                i3.TuiItemModule,
                i1.TuiExpandModule,
                i3.TuiPreventDefaultModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFilesModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFilesComponent, TuiFileComponent],
                        imports: [
                            i2.CommonModule,
                            i1.TuiGroupModule,
                            i1.TuiLoaderModule,
                            i4.PolymorpheusModule,
                            i1.TuiSvgModule,
                            i3.TuiFocusVisibleModule,
                            i1.TuiButtonModule,
                            i3.TuiItemModule,
                            i1.TuiExpandModule,
                            i3.TuiPreventDefaultModule,
                        ],
                        exports: [TuiFilesComponent, TuiFileComponent, i3.TuiItemDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_FILE_DEFAULT_OPTIONS = TUI_FILE_DEFAULT_OPTIONS;
    exports.TUI_FILE_OPTIONS = TUI_FILE_OPTIONS;
    exports.TuiFileComponent = TuiFileComponent;
    exports.TuiFilesComponent = TuiFilesComponent;
    exports.TuiFilesModule = TuiFilesModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
