// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-date-range.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/constants'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/mask'), require('@tinkoff/ng-polymorpheus'), require('rxjs/operators'), require('@taiga-ui/kit/components/calendar-range'), require('@taiga-ui/kit/directives'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-date-range', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/constants', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/mask', '@tinkoff/ng-polymorpheus', 'rxjs/operators', '@taiga-ui/kit/components/calendar-range', '@taiga-ui/kit/directives', '@angular/common', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-date-range"] = {}), global.ng.core, global.ng.forms, global.i3, global.i1, global["taiga-ui"].kit.constants, global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils.mask, global.i5, global.rxjs.operators, global["taiga-ui"].kit.components["calendar-range"], global["taiga-ui"].kit.directives, global.ng.common, global.rxjs));
})(this, (function (exports, i0, i7, i3, i1, constants, tokens, mask, i5, operators, i2, i4, i6, i8) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i8__namespace = /*#__PURE__*/_interopNamespace(i8);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiInputDateRangeComponent = /** @class */ (function (_super) {
        __extends(TuiInputDateRangeComponent, _super);
        function TuiInputDateRangeComponent(control, cdr, injector, isMobile, dialogs, mobileCalendar, textfieldSize, dateFormat, dateSeparator, dateTexts$, valueTransformer, options) {
            var _this = _super.call(this, control, cdr, valueTransformer) || this;
            _this.injector = injector;
            _this.isMobile = isMobile;
            _this.dialogs = dialogs;
            _this.mobileCalendar = mobileCalendar;
            _this.textfieldSize = textfieldSize;
            _this.dateFormat = dateFormat;
            _this.dateSeparator = dateSeparator;
            _this.dateTexts$ = dateTexts$;
            _this.valueTransformer = valueTransformer;
            _this.options = options;
            _this.textMaskOptions = {
                mask: mask.tuiCreateDateRangeMask(_this.dateFormat, _this.dateSeparator),
                pipe: mask.tuiCreateAutoCorrectedDateRangePipe(_this),
                guide: false,
            };
            _this.disabledItemHandler = i3.ALWAYS_FALSE_HANDLER;
            _this.markerHandler = i1.TUI_DEFAULT_MARKER_HANDLER;
            _this.defaultViewedMonth = i3.TuiMonth.currentLocal();
            _this.items = [];
            _this.min = _this.options.min;
            _this.max = _this.options.max;
            _this.minLength = null;
            _this.maxLength = null;
            _this.open = false;
            _this.maxLengthMapper = constants.MAX_DAY_RANGE_LENGTH_MAPPER;
            _this.dateFiller$ = _this.dateTexts$.pipe(operators.map(function (dateTexts) { return i3.changeDateSeparator(dateTexts[_this.dateFormat], _this.dateSeparator); }));
            return _this;
        }
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "focused", {
            get: function () {
                return !!this.textfield && this.textfield.focused;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "computedMobile", {
            get: function () {
                return this.isMobile && !!this.mobileCalendar;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "calendarIcon", {
            get: function () {
                return this.options.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "canOpen", {
            get: function () {
                return this.interactive && !this.computedMobile;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "computedExampleText", {
            get: function () {
                var _a, _b;
                return this.items.length
                    ? ((_b = (_a = this.textfield) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) === null || _b === void 0 ? void 0 : _b.placeholder) || ''
                    : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "computedMask", {
            get: function () {
                return this.activePeriod ? constants.EMPTY_MASK : this.textMaskOptions;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "activePeriod", {
            get: function () {
                var _this = this;
                return (this.items.find(function (item) { return i3.tuiNullableSame(_this.value, item.range, function (a, b) { return a.from.daySame(b.from.dayLimit(_this.min, _this.max)) &&
                    a.to.daySame(b.to.dayLimit(_this.min, _this.max)); }); }) || null);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "computedValue", {
            get: function () {
                var _c = this, value = _c.value, nativeValue = _c.nativeValue, activePeriod = _c.activePeriod;
                if (activePeriod) {
                    return String(activePeriod);
                }
                return value
                    ? value.getFormattedDayRange(this.dateFormat, this.dateSeparator)
                    : nativeValue;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "innerPseudoFocused", {
            get: function () {
                if (this.pseudoFocus === false) {
                    return false;
                }
                if (this.open || this.computedFocused) {
                    return true;
                }
                return null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.nativeFocusableElement) {
                    return;
                }
                this.nativeFocusableElement.value = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "size", {
            get: function () {
                return this.textfieldSize.size;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputDateRangeComponent.prototype.onClick = function () {
            if (!this.isMobile) {
                this.toggle();
            }
        };
        TuiInputDateRangeComponent.prototype.getComputedRangeFiller = function (dateFiller) {
            return this.activePeriod ? '' : this.getDateRangeFiller(dateFiller);
        };
        TuiInputDateRangeComponent.prototype.onIconClick = function () {
            var _this = this;
            if (!this.computedMobile || !this.mobileCalendar) {
                return;
            }
            this.dialogs
                .open(new i5.PolymorpheusComponent(this.mobileCalendar, this.injector), {
                size: 'fullscreen',
                closeable: false,
                data: {
                    single: false,
                    min: this.maxLengthMapper(this.min, this.value, this.maxLength, true),
                    max: this.maxLengthMapper(this.max, this.value, this.maxLength, false),
                    disabledItemHandler: this.disabledItemHandler,
                },
            })
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (value) {
                _this.value = value;
            });
        };
        TuiInputDateRangeComponent.prototype.onOpenChange = function (open) {
            this.open = open;
        };
        TuiInputDateRangeComponent.prototype.onValueChange = function (value) {
            if (this.control) {
                this.control.updateValueAndValidity({ emitEvent: false });
            }
            if (!value) {
                this.onOpenChange(true);
            }
            if (value.length !== i3.DATE_RANGE_FILLER_LENGTH) {
                this.value = null;
                return;
            }
            var parsedValue = i3.TuiDayRange.normalizeParse(value, this.dateFormat);
            this.value =
                !this.minLength && !this.maxLength
                    ? parsedValue
                    : this.clampValue(parsedValue);
        };
        TuiInputDateRangeComponent.prototype.onRangeChange = function (range) {
            this.toggle();
            this.focusInput();
            if (!range) {
                this.nativeValue = '';
            }
            this.value = range;
        };
        TuiInputDateRangeComponent.prototype.onItemSelect = function (item) {
            this.toggle();
            this.focusInput();
            if (typeof item !== 'string') {
                this.value = item.range.dayLimit(this.min, this.max);
                return;
            }
            if (this.activePeriod === null) {
                return;
            }
            this.value = null;
            this.nativeValue = '';
        };
        TuiInputDateRangeComponent.prototype.onActiveZone = function (focused) {
            this.updateFocused(focused);
            if (!focused &&
                !this.itemSelected &&
                (this.nativeValue.length === i3.DATE_FILLER_LENGTH ||
                    this.nativeValue.length ===
                        i3.DATE_FILLER_LENGTH + i3.RANGE_SEPARATOR_CHAR.length)) {
                this.value = i3.TuiDayRange.normalizeParse(this.nativeValue, this.dateFormat);
            }
        };
        TuiInputDateRangeComponent.prototype.writeValue = function (value) {
            _super.prototype.writeValue.call(this, value);
            this.nativeValue = value ? this.computedValue : '';
        };
        TuiInputDateRangeComponent.prototype.valueIdenticalComparator = function (oldValue, newValue) {
            return i3.tuiNullableSame(oldValue, newValue, function (a, b) { return a.daySame(b); });
        };
        Object.defineProperty(TuiInputDateRangeComponent.prototype, "itemSelected", {
            get: function () {
                var _this = this;
                return this.items.findIndex(function (item) { return String(item) === _this.nativeValue; }) !== -1;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputDateRangeComponent.prototype.toggle = function () {
            this.open = !this.open;
        };
        TuiInputDateRangeComponent.prototype.focusInput = function (preventScroll) {
            if (preventScroll === void 0) { preventScroll = false; }
            if (this.nativeFocusableElement) {
                this.nativeFocusableElement.focus({ preventScroll: preventScroll });
            }
        };
        TuiInputDateRangeComponent.prototype.clampValue = function (value) {
            var clampedBottom = this.minLength && value.from.append(this.minLength).dayAfter(value.to)
                ? new i3.TuiDayRange(value.from, value.from.append(this.minLength).append({ day: -1 }))
                : value;
            var availableMax = this.maxLength
                ? clampedBottom.from.append(this.maxLength).append({ day: -1 })
                : this.max;
            return clampedBottom.to.dayAfter(availableMax)
                ? new i3.TuiDayRange(clampedBottom.from, availableMax)
                : clampedBottom;
        };
        TuiInputDateRangeComponent.prototype.getDateRangeFiller = function (dateFiller) {
            return "" + dateFiller + i3.RANGE_SEPARATOR_CHAR + dateFiller;
        };
        return TuiInputDateRangeComponent;
    }(i3.AbstractTuiNullableControl));
    TuiInputDateRangeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeComponent, deps: [{ token: i7.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i0.Injector }, { token: i3.TUI_IS_MOBILE }, { token: i1.TuiDialogService }, { token: tokens.TUI_MOBILE_CALENDAR, optional: true }, { token: i1.TUI_TEXTFIELD_SIZE }, { token: i3.TUI_DATE_FORMAT }, { token: i3.TUI_DATE_SEPARATOR }, { token: tokens.TUI_DATE_TEXTS }, { token: tokens.TUI_DATE_RANGE_VALUE_TRANSFORMER, optional: true }, { token: tokens.TUI_INPUT_DATE_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputDateRangeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputDateRangeComponent, selector: "tui-input-date-range", inputs: { disabledItemHandler: "disabledItemHandler", markerHandler: "markerHandler", defaultViewedMonth: "defaultViewedMonth", items: "items", min: "min", max: "max", minLength: "minLength", maxLength: "maxLength" }, host: { listeners: { "click": "onClick()" } }, providers: [
            i3.tuiAsFocusableItemAccessor(TuiInputDateRangeComponent),
            i3.tuiAsControl(TuiInputDateRangeComponent),
            tokens.tuiDateStreamWithTransformer(tokens.TUI_DATE_RANGE_VALUE_TRANSFORMER),
        ], viewQueries: [{ propertyName: "textfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"canOpen\"\n    [content]=\"dropdown\"\n    [open]=\"open && canOpen\"\n    (openChange)=\"onOpenChange($event)\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-primitive-textfield\n        *tuiLet=\"dateFiller$ | async as dateFiller\"\n        automation-id=\"tui-input-date-range__textfield\"\n        tuiValueAccessor\n        class=\"t-textfield\"\n        [tuiTextfieldIcon]=\"calendarIcon && iconContent\"\n        [tuiTextfieldFiller]=\"getComputedRangeFiller(dateFiller || '')\"\n        [pseudoFocus]=\"innerPseudoFocused\"\n        [pseudoHover]=\"pseudoHover\"\n        [pseudoActive]=\"pseudoActive\"\n        [invalid]=\"computedInvalid\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [disabled]=\"computedDisabled\"\n        [textMask]=\"computedMask\"\n        [value]=\"computedValue\"\n        (valueChange)=\"onValueChange($event)\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            select=\"input\"\n            ngProjectAs=\"input\"\n        ></ng-content>\n    </tui-primitive-textfield>\n\n    <ng-template #iconContent>\n        <tui-svg\n            *polymorpheusOutlet=\"calendarIcon as src; context: {$implicit: size}\"\n            tuiWrapper\n            appearance=\"icon\"\n            automation-id=\"tui-input-date-range__icon\"\n            [class.t-icon]=\"!computedDisabled\"\n            [src]=\"src\"\n            (click)=\"onIconClick()\"\n        ></tui-svg>\n    </ng-template>\n\n    <ng-template #dropdown>\n        <tui-calendar-range\n            [defaultViewedMonth]=\"defaultViewedMonth\"\n            [disabledItemHandler]=\"disabledItemHandler\"\n            [items]=\"items\"\n            [min]=\"min\"\n            [max]=\"max\"\n            [markerHandler]=\"markerHandler\"\n            [minLength]=\"minLength\"\n            [maxLength]=\"maxLength\"\n            [value]=\"value\"\n            (valueChange)=\"onRangeChange($event)\"\n        ></tui-calendar-range>\n    </ng-template>\n</tui-hosted-dropdown>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}:host._disabled{pointer-events:none}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}.t-icon{pointer-events:auto}\n"], components: [{ type: i1__namespace.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i1__namespace.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }, { type: i2__namespace.TuiCalendarRangeComponent, selector: "tui-calendar-range", inputs: ["defaultViewedMonth", "disabledItemHandler", "markerHandler", "items", "min", "max", "minLength", "maxLength", "value"], outputs: ["valueChange"] }], directives: [{ type: i3__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i3__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i1__namespace.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i4__namespace.TuiValueAccessorDirective, selector: "[tuiValueAccessor]" }, { type: i1__namespace.TuiTextfieldIconDirective, selector: "[tuiTextfieldIcon]", inputs: ["tuiTextfieldIcon"] }, { type: i1__namespace.TuiTextfieldFillerDirective, selector: "[tuiTextfieldFiller]", inputs: ["tuiTextfieldFiller"] }, { type: i4__namespace.MaskedInputDirective, selector: "[textMask]", inputs: ["textMask"], exportAs: ["textMask"] }, { type: i5__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }], pipes: { "async": i6__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "disabledItemHandler", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "markerHandler", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "defaultViewedMonth", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "items", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "min", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "max", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "minLength", void 0);
    __decorate([
        i3.tuiDefaultProp()
    ], TuiInputDateRangeComponent.prototype, "maxLength", void 0);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-date-range',
                        templateUrl: './input-date-range.template.html',
                        styleUrls: ['./input-date-range.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i3.tuiAsFocusableItemAccessor(TuiInputDateRangeComponent),
                            i3.tuiAsControl(TuiInputDateRangeComponent),
                            tokens.tuiDateStreamWithTransformer(tokens.TUI_DATE_RANGE_VALUE_TRANSFORMER),
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i7__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i7.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i0__namespace.Injector, decorators: [{
                            type: i0.Inject,
                            args: [i0.Injector]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_IS_MOBILE]
                        }] }, { type: i1__namespace.TuiDialogService, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiDialogService]
                        }] }, { type: i0__namespace.Type, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_MOBILE_CALENDAR]
                        }] }, { type: i1__namespace.TuiTextfieldSizeDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_SIZE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_DATE_FORMAT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_DATE_SEPARATOR]
                        }] }, { type: i8__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DATE_TEXTS]
                        }] }, { type: i3__namespace.AbstractTuiValueTransformer, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_DATE_RANGE_VALUE_TRANSFORMER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_INPUT_DATE_OPTIONS]
                        }] }];
        }, propDecorators: { textfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], disabledItemHandler: [{
                    type: i0.Input
                }], markerHandler: [{
                    type: i0.Input
                }], defaultViewedMonth: [{
                    type: i0.Input
                }], items: [{
                    type: i0.Input
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], minLength: [{
                    type: i0.Input
                }], maxLength: [{
                    type: i0.Input
                }], onClick: [{
                    type: i0.HostListener,
                    args: ['click']
                }] } });

    var TuiInputDateRangeDirective = /** @class */ (function (_super) {
        __extends(TuiInputDateRangeDirective, _super);
        function TuiInputDateRangeDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(TuiInputDateRangeDirective.prototype, "value", {
            get: function () {
                return this.host.computedValue;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputDateRangeDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        TuiInputDateRangeDirective.prototype.process = function (input) {
            input.inputMode = 'numeric';
        };
        TuiInputDateRangeDirective.prototype.ngDoCheck = function () {
            if (this.host.nativeFocusableElement) {
                this.host.nativeFocusableElement.placeholder = this.host.computedExampleText;
            }
        };
        return TuiInputDateRangeDirective;
    }(i1.AbstractTuiTextfieldHost));
    TuiInputDateRangeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiInputDateRangeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputDateRangeDirective, selector: "tui-input-date-range", providers: [i1.tuiAsTextfieldHost(TuiInputDateRangeDirective)], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-input-date-range',
                        providers: [i1.tuiAsTextfieldHost(TuiInputDateRangeDirective)],
                    }]
            }] });

    var TuiInputDateRangeModule = /** @class */ (function () {
        function TuiInputDateRangeModule() {
        }
        return TuiInputDateRangeModule;
    }());
    TuiInputDateRangeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputDateRangeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeModule, declarations: [TuiInputDateRangeComponent, TuiInputDateRangeDirective], imports: [i6.CommonModule,
            i4.TextMaskModule,
            i3.TuiActiveZoneModule,
            i3.TuiLetModule,
            i5.PolymorpheusModule,
            i1.TuiWrapperModule,
            i1.TuiPrimitiveTextfieldModule,
            i1.TuiTextfieldControllerModule,
            i1.TuiHostedDropdownModule,
            i1.TuiSvgModule,
            i2.TuiCalendarRangeModule,
            i4.TuiValueAccessorModule], exports: [TuiInputDateRangeComponent,
            TuiInputDateRangeDirective,
            i1.TuiTextfieldComponent] });
    TuiInputDateRangeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeModule, imports: [[
                i6.CommonModule,
                i4.TextMaskModule,
                i3.TuiActiveZoneModule,
                i3.TuiLetModule,
                i5.PolymorpheusModule,
                i1.TuiWrapperModule,
                i1.TuiPrimitiveTextfieldModule,
                i1.TuiTextfieldControllerModule,
                i1.TuiHostedDropdownModule,
                i1.TuiSvgModule,
                i2.TuiCalendarRangeModule,
                i4.TuiValueAccessorModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateRangeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i6.CommonModule,
                            i4.TextMaskModule,
                            i3.TuiActiveZoneModule,
                            i3.TuiLetModule,
                            i5.PolymorpheusModule,
                            i1.TuiWrapperModule,
                            i1.TuiPrimitiveTextfieldModule,
                            i1.TuiTextfieldControllerModule,
                            i1.TuiHostedDropdownModule,
                            i1.TuiSvgModule,
                            i2.TuiCalendarRangeModule,
                            i4.TuiValueAccessorModule,
                        ],
                        declarations: [TuiInputDateRangeComponent, TuiInputDateRangeDirective],
                        exports: [
                            TuiInputDateRangeComponent,
                            TuiInputDateRangeDirective,
                            i1.TuiTextfieldComponent,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiInputDateRangeComponent = TuiInputDateRangeComponent;
    exports.TuiInputDateRangeDirective = TuiInputDateRangeDirective;
    exports.TuiInputDateRangeModule = TuiInputDateRangeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
