// @taiga-ui/core/esm2015/components/group/group.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiGroupDirective } from './group.directive';
import { TuiGroupStylesComponent } from './group-styles.component';
import * as i0 from "@angular/core";
export class TuiGroupModule {
}
TuiGroupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiGroupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiGroupModule, declarations: [TuiGroupDirective, TuiGroupStylesComponent], imports: [CommonModule], exports: [TuiGroupDirective, TuiGroupStylesComponent] });
TuiGroupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiGroupModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiGroupDirective, TuiGroupStylesComponent],
                    exports: [TuiGroupDirective, TuiGroupStylesComponent],
                }]
        }] });
