// @taiga-ui/core/esm2015/components/calendar/calendar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiPrimitiveCalendarModule } from '@taiga-ui/core/components/primitive-calendar';
import { TuiPrimitiveYearPickerModule } from '@taiga-ui/core/components/primitive-year-picker';
import { TuiScrollbarModule } from '@taiga-ui/core/components/scrollbar';
import { TuiPrimitiveYearMonthPaginationModule } from '@taiga-ui/core/internal/primitive-year-month-pagination';
import { TuiCalendarComponent } from './calendar.component';
import * as i0 from "@angular/core";
export class TuiCalendarModule {
}
TuiCalendarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCalendarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarModule, declarations: [TuiCalendarComponent], imports: [CommonModule,
        TuiPrimitiveYearMonthPaginationModule,
        TuiPrimitiveCalendarModule,
        TuiPrimitiveYearPickerModule,
        TuiScrollbarModule,
        TuiMapperPipeModule], exports: [TuiCalendarComponent] });
TuiCalendarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarModule, imports: [[
            CommonModule,
            TuiPrimitiveYearMonthPaginationModule,
            TuiPrimitiveCalendarModule,
            TuiPrimitiveYearPickerModule,
            TuiScrollbarModule,
            TuiMapperPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiPrimitiveYearMonthPaginationModule,
                        TuiPrimitiveCalendarModule,
                        TuiPrimitiveYearPickerModule,
                        TuiScrollbarModule,
                        TuiMapperPipeModule,
                    ],
                    declarations: [TuiCalendarComponent],
                    exports: [TuiCalendarComponent],
                }]
        }] });
