// @taiga-ui/core/esm2015/components/hints-host/hints-host.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiHintsHostComponent } from './hints-host.component';
import * as i0 from "@angular/core";
export class TuiHintsHostModule {
}
TuiHintsHostModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintsHostModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiHintsHostModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintsHostModule, declarations: [TuiHintsHostComponent], imports: [CommonModule, PolymorpheusModule, TuiActiveZoneModule], exports: [TuiHintsHostComponent] });
TuiHintsHostModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintsHostModule, imports: [[CommonModule, PolymorpheusModule, TuiActiveZoneModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintsHostModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiActiveZoneModule],
                    declarations: [TuiHintsHostComponent],
                    exports: [TuiHintsHostComponent],
                }]
        }] });
