// @taiga-ui/core/esm2015/directives/hint/hint-describe.directive.js
import { __decorate } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, Inject, Input, NgZone } from '@angular/core';
import { tuiIsNativeFocused, tuiPure, tuiTypedFromEvent, tuiZoneOptimized, } from '@taiga-ui/cdk';
import { tuiAsDriver, TuiDriver } from '@taiga-ui/core/abstract';
import { merge, of, timer } from 'rxjs';
import { debounce, distinctUntilChanged, map, skip, startWith, switchMap, } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiHintDescribeDirective extends TuiDriver {
    constructor(ngZone, doc, el) {
        super(subscriber => this.stream$.subscribe(subscriber));
        this.ngZone = ngZone;
        this.doc = doc;
        this.el = el;
        this.stream$ = tuiTypedFromEvent(this.doc, 'keydown', {
            capture: true,
        }).pipe(switchMap(() => this.focused
            ? of(false)
            : merge(tuiTypedFromEvent(this.doc, 'keyup'), tuiTypedFromEvent(this.element, 'blur')).pipe(map(() => this.focused))), debounce(visible => (visible ? timer(1000) : of(null))), startWith(false), distinctUntilChanged(), skip(1), tuiZoneOptimized(this.ngZone));
        this.tuiHintDescribe = '';
        this.type = 'hint';
    }
    get focused() {
        return tuiIsNativeFocused(this.element);
    }
    get element() {
        return this.doc.getElementById(this.tuiHintDescribe) || this.el.nativeElement;
    }
}
TuiHintDescribeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintDescribeDirective, deps: [{ token: NgZone }, { token: DOCUMENT }, { token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiHintDescribeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintDescribeDirective, selector: "[tuiHintDescribe]", inputs: { tuiHintDescribe: "tuiHintDescribe" }, providers: [tuiAsDriver(TuiHintDescribeDirective)], usesInheritance: true, ngImport: i0 });
__decorate([
    tuiPure
], TuiHintDescribeDirective.prototype, "element", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintDescribeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHintDescribe]',
                    providers: [tuiAsDriver(TuiHintDescribeDirective)],
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { tuiHintDescribe: [{
                type: Input
            }], element: [] } });
