// @taiga-ui/core/esm2015/directives/hint/hint-host.directive.js
import { Directive, Input } from '@angular/core';
import { EMPTY_CLIENT_RECT } from '@taiga-ui/cdk';
import { tuiAsRectAccessor, TuiRectAccessor } from '@taiga-ui/core/abstract';
import * as i0 from "@angular/core";
export class TuiHintHostDirective extends TuiRectAccessor {
    constructor() {
        super(...arguments);
        this.type = 'hint';
    }
    getClientRect() {
        var _a;
        return ((_a = this.tuiHintHost) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect()) || EMPTY_CLIENT_RECT;
    }
}
TuiHintHostDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintHostDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiHintHostDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintHostDirective, selector: "[tuiHint][tuiHintHost]", inputs: { tuiHintHost: "tuiHintHost" }, providers: [tuiAsRectAccessor(TuiHintHostDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintHostDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHint][tuiHintHost]',
                    providers: [tuiAsRectAccessor(TuiHintHostDirective)],
                }]
        }], propDecorators: { tuiHintHost: [{
                type: Input
            }] } });
