// @taiga-ui/core/esm2015/directives/hint/hint-options.directive.js
import { __decorate } from "tslib";
import { Directive, forwardRef, Inject, InjectionToken, Input, Optional, SkipSelf, } from '@angular/core';
import { AbstractTuiController, tuiDefaultProp } from '@taiga-ui/cdk';
import { tuiOverrideOptions } from '@taiga-ui/core/utils';
import * as i0 from "@angular/core";
/** Default values for hint options */
export const TUI_HINT_DEFAULT_OPTIONS = {
    direction: 'bottom-left',
    showDelay: 500,
    hideDelay: 200,
    appearance: '',
    icon: 'tuiIconHelpCircle',
};
/**
 * Default parameters for hint directive
 */
export const TUI_HINT_OPTIONS = new InjectionToken('[TUI_HINT_OPTIONS]', {
    factory: () => TUI_HINT_DEFAULT_OPTIONS,
});
export const tuiHintOptionsProvider = (override) => ({
    provide: TUI_HINT_OPTIONS,
    deps: [
        [new Optional(), TuiHintOptionsDirective],
        [new SkipSelf(), TUI_HINT_OPTIONS],
    ],
    useFactory: tuiOverrideOptions(override),
});
export class TuiHintOptionsDirective extends AbstractTuiController {
    constructor(options) {
        super();
        this.options = options;
        this.direction = this.options.direction;
        this.appearance = this.options.appearance;
        this.showDelay = this.options.showDelay;
        this.hideDelay = this.options.hideDelay;
        this.icon = this.options.icon;
    }
}
TuiHintOptionsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintOptionsDirective, deps: [{ token: TUI_HINT_OPTIONS, skipSelf: true }], target: i0.ɵɵFactoryTarget.Directive });
TuiHintOptionsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintOptionsDirective, selector: "[tuiHintContent]", inputs: { content: ["tuiHintContent", "content"], direction: ["tuiHintDirection", "direction"], appearance: ["tuiHintAppearance", "appearance"], showDelay: ["tuiHintShowDelay", "showDelay"], hideDelay: ["tuiHintHideDelay", "hideDelay"] }, providers: [
        {
            provide: TUI_HINT_OPTIONS,
            useExisting: forwardRef(() => TuiHintOptionsDirective),
        },
    ], usesInheritance: true, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiHintOptionsDirective.prototype, "direction", void 0);
__decorate([
    tuiDefaultProp()
], TuiHintOptionsDirective.prototype, "appearance", void 0);
__decorate([
    tuiDefaultProp()
], TuiHintOptionsDirective.prototype, "showDelay", void 0);
__decorate([
    tuiDefaultProp()
], TuiHintOptionsDirective.prototype, "hideDelay", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintOptionsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHintContent]',
                    providers: [
                        {
                            provide: TUI_HINT_OPTIONS,
                            useExisting: forwardRef(() => TuiHintOptionsDirective),
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: SkipSelf
                }, {
                    type: Inject,
                    args: [TUI_HINT_OPTIONS]
                }] }]; }, propDecorators: { content: [{
                type: Input,
                args: ['tuiHintContent']
            }], direction: [{
                type: Input,
                args: ['tuiHintDirection']
            }], appearance: [{
                type: Input,
                args: ['tuiHintAppearance']
            }], showDelay: [{
                type: Input,
                args: ['tuiHintShowDelay']
            }], hideDelay: [{
                type: Input,
                args: ['tuiHintHideDelay']
            }] } });
