// @taiga-ui/core/esm2015/directives/hint/hint-manual.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { tuiAsDriver, TuiDriver } from '@taiga-ui/core/abstract';
import { BehaviorSubject } from 'rxjs';
import { TuiHintHoverDirective } from './hint-hover.directive';
import * as i0 from "@angular/core";
import * as i1 from "./hint-hover.directive";
export class TuiHintManualDirective extends TuiDriver {
    constructor(hover) {
        super(subscriber => this.stream$.subscribe(subscriber));
        this.stream$ = new BehaviorSubject(false);
        this.tuiHintManual = false;
        this.type = 'hint';
        hover.enabled = false;
    }
    ngOnChanges() {
        this.stream$.next(this.tuiHintManual);
    }
}
TuiHintManualDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintManualDirective, deps: [{ token: TuiHintHoverDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiHintManualDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintManualDirective, selector: "[tuiHint][tuiHintManual]", inputs: { tuiHintManual: "tuiHintManual" }, providers: [tuiAsDriver(TuiHintManualDirective)], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintManualDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHint][tuiHintManual]',
                    providers: [tuiAsDriver(TuiHintManualDirective)],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiHintHoverDirective, decorators: [{
                    type: Inject,
                    args: [TuiHintHoverDirective]
                }] }]; }, propDecorators: { tuiHintManual: [{
                type: Input
            }] } });
