// @taiga-ui/core/esm2015/directives/hint/hint-pointer.directive.js
import { Directive, HostListener } from '@angular/core';
import { EMPTY_CLIENT_RECT, tuiPointToClientRect } from '@taiga-ui/cdk';
import { tuiAsDriver, tuiAsRectAccessor } from '@taiga-ui/core/abstract';
import { TuiHintHoverDirective } from './hint-hover.directive';
import * as i0 from "@angular/core";
export class TuiHintPointerDirective extends TuiHintHoverDirective {
    constructor() {
        super(...arguments);
        this.currentRect = EMPTY_CLIENT_RECT;
    }
    onMove({ clientX, clientY }) {
        this.currentRect = tuiPointToClientRect(clientX, clientY);
    }
    getClientRect() {
        return this.currentRect;
    }
}
TuiHintPointerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintPointerDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiHintPointerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintPointerDirective, selector: "[tuiHint][tuiHintPointer]", host: { listeners: { "mousemove.silent": "onMove($event)" } }, providers: [
        tuiAsRectAccessor(TuiHintPointerDirective),
        tuiAsDriver(TuiHintPointerDirective),
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintPointerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHint][tuiHintPointer]',
                    providers: [
                        tuiAsRectAccessor(TuiHintPointerDirective),
                        tuiAsDriver(TuiHintPointerDirective),
                    ],
                }]
        }], propDecorators: { onMove: [{
                type: HostListener,
                args: ['mousemove.silent', ['$event']]
            }] } });
