// @taiga-ui/core/esm2015/components/error/error.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiErrorComponent } from './error.component';
import * as i0 from "@angular/core";
export class TuiErrorModule {
}
TuiErrorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiErrorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiErrorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiErrorModule, declarations: [TuiErrorComponent], imports: [CommonModule, PolymorpheusModule], exports: [TuiErrorComponent] });
TuiErrorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiErrorModule, imports: [[CommonModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiErrorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule],
                    declarations: [TuiErrorComponent],
                    exports: [TuiErrorComponent],
                }]
        }] });
