// @taiga-ui/core/esm2015/directives/dropdown/dropdown-host.directive.js
import { Directive, Input } from '@angular/core';
import { EMPTY_CLIENT_RECT } from '@taiga-ui/cdk';
import { tuiAsRectAccessor, TuiRectAccessor } from '@taiga-ui/core/abstract';
import * as i0 from "@angular/core";
export class TuiDropdownHostDirective extends TuiRectAccessor {
    constructor() {
        super(...arguments);
        this.type = 'dropdown';
    }
    getClientRect() {
        var _a;
        return ((_a = this.tuiDropdownHost) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect()) || EMPTY_CLIENT_RECT;
    }
}
TuiDropdownHostDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownHostDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiDropdownHostDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownHostDirective, selector: "[tuiDropdown][tuiDropdownHost]", inputs: { tuiDropdownHost: "tuiDropdownHost" }, providers: [tuiAsRectAccessor(TuiDropdownHostDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownHostDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiDropdown][tuiDropdownHost]',
                    providers: [tuiAsRectAccessor(TuiDropdownHostDirective)],
                }]
        }], propDecorators: { tuiDropdownHost: [{
                type: Input
            }] } });
