// @taiga-ui/core/esm2015/directives/dropdown/dropdown.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Inject, INJECTOR, Input, } from '@angular/core';
import { TuiDropdownPortalService, tuiPure, } from '@taiga-ui/cdk';
import { tuiAsRectAccessor, tuiAsVehicle, } from '@taiga-ui/core/abstract';
import { tuiCheckFixedPosition } from '@taiga-ui/core/utils';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
// TODO: find the best way for prevent cycle
// eslint-disable-next-line import/no-cycle
import { TUI_DROPDOWN_COMPONENT } from './dropdown.providers';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiDropdownDirective {
    constructor(el, hapica, injector, dropdownService) {
        this.el = el;
        this.hapica = hapica;
        this.injector = injector;
        this.dropdownService = dropdownService;
        this.dropdownBoxRef = null;
        this.type = 'dropdown';
        this.component = new PolymorpheusComponent(this.hapica, this.injector);
    }
    get position() {
        return tuiCheckFixedPosition(this.el.nativeElement) ? 'fixed' : 'absolute';
    }
    ngAfterViewChecked() {
        var _a, _b;
        (_a = this.dropdownBoxRef) === null || _a === void 0 ? void 0 : _a.changeDetectorRef.detectChanges();
        (_b = this.dropdownBoxRef) === null || _b === void 0 ? void 0 : _b.changeDetectorRef.markForCheck();
    }
    ngOnChanges() {
        if (!this.content) {
            this.toggle(false);
        }
    }
    ngOnDestroy() {
        this.toggle(false);
    }
    getClientRect() {
        return this.el.nativeElement.getBoundingClientRect();
    }
    toggle(show) {
        if (show && this.content && !this.dropdownBoxRef) {
            this.dropdownBoxRef = this.dropdownService.add(this.component);
        }
        else if (!show && this.dropdownBoxRef) {
            this.dropdownService.remove(this.dropdownBoxRef);
            this.dropdownBoxRef = null;
        }
    }
}
TuiDropdownDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownDirective, deps: [{ token: ElementRef }, { token: TUI_DROPDOWN_COMPONENT }, { token: INJECTOR }, { token: TuiDropdownPortalService }], target: i0.ɵɵFactoryTarget.Directive });
TuiDropdownDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownDirective, selector: "[tuiDropdown]:not(ng-container)", inputs: { content: ["tuiDropdown", "content"] }, providers: [
        tuiAsRectAccessor(TuiDropdownDirective),
        tuiAsVehicle(TuiDropdownDirective),
    ], exportAs: ["tuiDropdown"], usesOnChanges: true, ngImport: i0 });
__decorate([
    tuiPure
], TuiDropdownDirective.prototype, "position", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiDropdown]:not(ng-container)',
                    exportAs: 'tuiDropdown',
                    providers: [
                        tuiAsRectAccessor(TuiDropdownDirective),
                        tuiAsVehicle(TuiDropdownDirective),
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i0.Type, decorators: [{
                    type: Inject,
                    args: [TUI_DROPDOWN_COMPONENT]
                }] }, { type: i0.Injector, decorators: [{
                    type: Inject,
                    args: [INJECTOR]
                }] }, { type: i1.TuiDropdownPortalService, decorators: [{
                    type: Inject,
                    args: [TuiDropdownPortalService]
                }] }]; }, propDecorators: { content: [{
                type: Input,
                args: ['tuiDropdown']
            }], position: [] } });
