// @taiga-ui/core/esm2015/directives/dropdown/dropdown-manual.directive.js
import { Directive, Input } from '@angular/core';
import { tuiAsDriver, TuiDriver } from '@taiga-ui/core/abstract';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class TuiDropdownManualDirective extends TuiDriver {
    constructor() {
        super(subscriber => this.stream$.subscribe(subscriber));
        this.stream$ = new BehaviorSubject(false);
        this.tuiDropdownManual = false;
        this.type = 'dropdown';
    }
    ngOnChanges() {
        this.stream$.next(this.tuiDropdownManual);
    }
}
TuiDropdownManualDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownManualDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiDropdownManualDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownManualDirective, selector: "[tuiDropdown][tuiDropdownManual]", inputs: { tuiDropdownManual: "tuiDropdownManual" }, providers: [tuiAsDriver(TuiDropdownManualDirective)], usesInheritance: true, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownManualDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiDropdown][tuiDropdownManual]',
                    providers: [tuiAsDriver(TuiDropdownManualDirective)],
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { tuiDropdownManual: [{
                type: Input
            }] } });
