// @taiga-ui/cdk/esm2015/directives/focusable/focusable.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Abstraction over `tabindex`
 */
export class TuiFocusableDirective {
    constructor() {
        /**
         * Element can be focused with keyboard
         */
        this.focusable = true;
    }
    get tabIndex() {
        return this.focusable ? 0 : -1;
    }
}
TuiFocusableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFocusableDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiFocusableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiFocusableDirective, selector: "[tuiFocusable]", inputs: { focusable: ["tuiFocusable", "focusable"] }, host: { properties: { "tabIndex": "this.tabIndex" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFocusableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiFocusable]',
                }]
        }], propDecorators: { focusable: [{
                type: Input,
                args: ['tuiFocusable']
            }], tabIndex: [{
                type: HostBinding,
                args: ['tabIndex']
            }] } });
