// @taiga-ui/kit/esm2015/components/input-files/input-files.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiDroppableModule, TuiFocusableModule, TuiFocusedModule, TuiFocusVisibleModule, TuiHoveredModule, TuiItemDirective, TuiLetModule, TuiPressedModule, } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiGroupModule, TuiLinkModule, TuiLoaderModule, TuiSvgModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiFileComponent, TuiFilesComponent, TuiFilesModule, } from '@taiga-ui/kit/components/files';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputFilesComponent } from './input-files.component';
import { TuiInputFilesDirective } from './input-files.directive';
import { TuiMaxSizeRejectionErrorPipe } from './max-size-rejection-error.pipe';
import * as i0 from "@angular/core";
export class TuiInputFilesModule {
}
TuiInputFilesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputFilesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputFilesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputFilesModule, declarations: [TuiInputFilesComponent,
        TuiInputFilesDirective,
        TuiMaxSizeRejectionErrorPipe], imports: [CommonModule,
        PolymorpheusModule,
        TuiLetModule,
        TuiFocusedModule,
        TuiFocusVisibleModule,
        TuiPressedModule,
        TuiHoveredModule,
        TuiFocusableModule,
        TuiDroppableModule,
        TuiWrapperModule,
        TuiSvgModule,
        TuiLinkModule,
        TuiLoaderModule,
        TuiButtonModule,
        TuiGroupModule,
        TuiFilesModule], exports: [TuiInputFilesComponent,
        TuiInputFilesDirective,
        TuiFilesComponent,
        TuiFileComponent,
        TuiItemDirective] });
TuiInputFilesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputFilesModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiLetModule,
            TuiFocusedModule,
            TuiFocusVisibleModule,
            TuiPressedModule,
            TuiHoveredModule,
            TuiFocusableModule,
            TuiDroppableModule,
            TuiWrapperModule,
            TuiSvgModule,
            TuiLinkModule,
            TuiLoaderModule,
            TuiButtonModule,
            TuiGroupModule,
            TuiFilesModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputFilesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiLetModule,
                        TuiFocusedModule,
                        TuiFocusVisibleModule,
                        TuiPressedModule,
                        TuiHoveredModule,
                        TuiFocusableModule,
                        TuiDroppableModule,
                        TuiWrapperModule,
                        TuiSvgModule,
                        TuiLinkModule,
                        TuiLoaderModule,
                        TuiButtonModule,
                        TuiGroupModule,
                        TuiFilesModule,
                    ],
                    declarations: [
                        TuiInputFilesComponent,
                        TuiInputFilesDirective,
                        TuiMaxSizeRejectionErrorPipe,
                    ],
                    exports: [
                        TuiInputFilesComponent,
                        TuiInputFilesDirective,
                        TuiFilesComponent,
                        TuiFileComponent,
                        TuiItemDirective,
                    ],
                }]
        }] });
