// @taiga-ui/kit/esm2015/components/input-files/input-files.directive.js
import { Directive, ElementRef, HostBinding, Inject } from '@angular/core';
import { TuiIdService } from '@taiga-ui/cdk';
// eslint-disable-next-line import/no-cycle
import { TuiInputFilesComponent } from './input-files.component';
import { TUI_INPUT_FILES_OPTIONS } from './input-files.options';
import * as i0 from "@angular/core";
import * as i1 from "./input-files.component";
import * as i2 from "@taiga-ui/cdk";
export class TuiInputFilesDirective {
    constructor(host, el, idService, options) {
        this.host = host;
        this.el = el;
        this.idService = idService;
        this.options = options;
    }
    get tabIndex() {
        return this.host.focusable ? 0 : -1;
    }
    get id() {
        return this.el.nativeElement.id || this.idService.generate();
    }
    get accept() {
        var _a;
        return (_a = this.el.nativeElement.accept) !== null && _a !== void 0 ? _a : this.options.accepts;
    }
    get multiple() {
        var _a;
        return (_a = this.el.nativeElement.multiple) !== null && _a !== void 0 ? _a : this.options.multiple;
    }
    get capture() {
        var _a;
        return (_a = this.el.nativeElement.getAttribute('capture')) !== null && _a !== void 0 ? _a : this.options.capture;
    }
    get input() {
        return this.el.nativeElement;
    }
}
TuiInputFilesDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputFilesDirective, deps: [{ token: TuiInputFilesComponent }, { token: ElementRef }, { token: TuiIdService }, { token: TUI_INPUT_FILES_OPTIONS }], target: i0.ɵɵFactoryTarget.Directive });
TuiInputFilesDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputFilesDirective, selector: "input[tuiInputFiles]", host: { attributes: { "type": "file" }, properties: { "tabIndex": "this.tabIndex", "id": "this.id", "accept": "this.accept", "multiple": "this.multiple", "attr.capture": "this.capture" }, classAttribute: "t-native" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputFilesDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[tuiInputFiles]',
                    host: {
                        type: 'file',
                        class: 't-native',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiInputFilesComponent, decorators: [{
                    type: Inject,
                    args: [TuiInputFilesComponent]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i2.TuiIdService, decorators: [{
                    type: Inject,
                    args: [TuiIdService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_INPUT_FILES_OPTIONS]
                }] }]; }, propDecorators: { tabIndex: [{
                type: HostBinding,
                args: ['tabIndex']
            }], id: [{
                type: HostBinding,
                args: ['id']
            }], accept: [{
                type: HostBinding,
                args: ['accept']
            }], multiple: [{
                type: HostBinding,
                args: ['multiple']
            }], capture: [{
                type: HostBinding,
                args: ['attr.capture']
            }] } });
