// @taiga-ui/cdk/esm2015/directives/autofilled/autofilled.directive.js
import { Directive, EventEmitter, HostBinding, HostListener, Inject, Output, } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk/services';
import { TuiAutofilledStyleComponent } from './autofilled-style.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/services";
export class TuiAutofilledDirective {
    constructor(directiveStyles) {
        this.autofilled = false;
        this.tuiAutofilledChange = new EventEmitter();
        directiveStyles.addComponent(TuiAutofilledStyleComponent);
    }
    transitionStartHandler({ propertyName, target }) {
        var _a;
        const matchedAutofill = propertyName.includes('box-shadow') && ((_a = target) === null || _a === void 0 ? void 0 : _a.matches('input'));
        if (matchedAutofill) {
            this.autofilled = !this.autofilled;
            this.tuiAutofilledChange.emit(this.autofilled);
        }
    }
}
TuiAutofilledDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutofilledDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiAutofilledDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAutofilledDirective, selector: "[tuiAutofilledChange]", outputs: { tuiAutofilledChange: "tuiAutofilledChange" }, host: { listeners: { "transitionstart": "transitionStartHandler($event)" }, properties: { "class._autofilled": "this.autofilled" }, classAttribute: "tui-autofill" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutofilledDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiAutofilledChange]',
                    host: { class: 'tui-autofill' },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { autofilled: [{
                type: HostBinding,
                args: ['class._autofilled']
            }], tuiAutofilledChange: [{
                type: Output
            }], transitionStartHandler: [{
                type: HostListener,
                args: ['transitionstart', ['$event']]
            }] } });
