// @taiga-ui/kit/esm2015/components/line-clamp/line-clamp.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLetModule, TuiResizeModule } from '@taiga-ui/cdk';
import { TuiHintModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiLineClampComponent } from './line-clamp.component';
import { TuiLineClampBoxComponent } from './line-clamp-box.component';
import { TuiLineClampPositionDirective } from './line-clamp-position.directive';
import * as i0 from "@angular/core";
export class TuiLineClampModule {
}
TuiLineClampModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLineClampModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampModule, declarations: [TuiLineClampComponent,
        TuiLineClampBoxComponent,
        TuiLineClampPositionDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiHintModule,
        TuiResizeModule,
        TuiLetModule], exports: [TuiLineClampComponent] });
TuiLineClampModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiHintModule,
            TuiResizeModule,
            TuiLetModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiHintModule,
                        TuiResizeModule,
                        TuiLetModule,
                    ],
                    declarations: [
                        TuiLineClampComponent,
                        TuiLineClampBoxComponent,
                        TuiLineClampPositionDirective,
                    ],
                    exports: [TuiLineClampComponent],
                }]
        }] });
