// @taiga-ui/kit/esm2015/components/line-clamp/line-clamp-box.component.js
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import { TuiDestroyService, TuiHoveredService } from '@taiga-ui/cdk';
import { tuiFadeIn, TuiHintComponent, TuiHintDirective, tuiPositionAccessorFor, TuiPositionService, tuiRectAccessorFor, } from '@taiga-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@tinkoff/ng-polymorpheus";
export class TuiLineClampBoxComponent extends TuiHintComponent {
    get width() {
        return this.accessor.getClientRect().width;
    }
}
TuiLineClampBoxComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampBoxComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TuiLineClampBoxComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLineClampBoxComponent, selector: "tui-line-clamp-box", host: { properties: { "style.minWidth.px": "this.width" } }, providers: [
        TuiDestroyService,
        TuiPositionService,
        TuiHoveredService,
        tuiPositionAccessorFor('hint'),
        tuiRectAccessorFor('hint', TuiHintDirective),
    ], usesInheritance: true, ngImport: i0, template: `
        <ng-container *polymorpheusOutlet="content as text">{{ text }}</ng-container>
    `, isInline: true, styles: [":host{box-shadow:0 1.5rem 1rem #00000008,0 .75rem .75rem #0000000a,0 .25rem .375rem #0000000d;position:absolute;width:-webkit-min-content;width:min-content;padding:.75rem 1rem;margin-left:calc(-1px - 1rem);margin-top:calc(-1px - .75rem);border-radius:var(--tui-radius-l);box-sizing:content-box;border:1px solid var(--tui-base-03);background:var(--tui-base-01);color:var(--tui-text-01);word-wrap:break-word}\n"], directives: [{ type: i1.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [tuiFadeIn], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampBoxComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-line-clamp-box',
                    template: `
        <ng-container *polymorpheusOutlet="content as text">{{ text }}</ng-container>
    `,
                    styleUrls: ['./line-clamp-box.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        TuiDestroyService,
                        TuiPositionService,
                        TuiHoveredService,
                        tuiPositionAccessorFor('hint'),
                        tuiRectAccessorFor('hint', TuiHintDirective),
                    ],
                    animations: [tuiFadeIn],
                }]
        }], propDecorators: { width: [{
                type: HostBinding,
                args: ['style.minWidth.px']
            }] } });
