// @taiga-ui/kit/esm2015/components/line-clamp/line-clamp-position.directive.js
import { Directive, Inject } from '@angular/core';
import { tuiAsPositionAccessor, TuiHintDirective, TuiPositionAccessor, } from '@taiga-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
export class TuiLineClampPositionDirective extends TuiPositionAccessor {
    constructor(accessor) {
        super();
        this.accessor = accessor;
        this.type = 'hint';
    }
    getPosition() {
        const { top, left } = this.accessor.getClientRect();
        return [top, left];
    }
}
TuiLineClampPositionDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampPositionDirective, deps: [{ token: TuiHintDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiLineClampPositionDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiLineClampPositionDirective, selector: "[tuiHintCustomPosition]", providers: [tuiAsPositionAccessor(TuiLineClampPositionDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineClampPositionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHintCustomPosition]',
                    providers: [tuiAsPositionAccessor(TuiLineClampPositionDirective)],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiRectAccessor, decorators: [{
                    type: Inject,
                    args: [TuiHintDirective]
                }] }]; } });
