// @taiga-ui/kit/esm2015/components/progress/progress.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiProgressBarComponent } from './progress-bar/progress-bar.component';
import { TuiProgressColorSegmentsDirective } from './progress-bar/progress-color-segments.directive';
import { TuiProgressCircleComponent } from './progress-circle/progress-circle.component';
import { TuiProgressLabelComponent } from './progress-label/progress-label.component';
import { TuiProgressSegmentedComponent } from './progress-segmented/progress-segmented.component';
import * as i0 from "@angular/core";
export class TuiProgressModule {
}
TuiProgressModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiProgressModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressModule, declarations: [TuiProgressBarComponent,
        TuiProgressColorSegmentsDirective,
        TuiProgressCircleComponent,
        TuiProgressLabelComponent,
        TuiProgressSegmentedComponent], imports: [CommonModule, TuiRepeatTimesModule], exports: [TuiProgressBarComponent,
        TuiProgressCircleComponent,
        TuiProgressColorSegmentsDirective,
        TuiProgressLabelComponent,
        TuiProgressSegmentedComponent] });
TuiProgressModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressModule, imports: [[CommonModule, TuiRepeatTimesModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiRepeatTimesModule],
                    declarations: [
                        TuiProgressBarComponent,
                        TuiProgressColorSegmentsDirective,
                        TuiProgressCircleComponent,
                        TuiProgressLabelComponent,
                        TuiProgressSegmentedComponent,
                    ],
                    exports: [
                        TuiProgressBarComponent,
                        TuiProgressCircleComponent,
                        TuiProgressColorSegmentsDirective,
                        TuiProgressLabelComponent,
                        TuiProgressSegmentedComponent,
                    ],
                }]
        }] });
