// @taiga-ui/core/esm2015/components/dialog/dialog-close.service.js
import { DOCUMENT } from '@angular/common';
import { ElementRef, Inject, Injectable } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { tuiContainsOrAfter, tuiGetActualTarget, tuiIsCurrentTarget, tuiIsElement, tuiTypedFromEvent, } from '@taiga-ui/cdk';
import { tuiGetViewportWidth } from '@taiga-ui/core/utils';
import { merge, Observable } from 'rxjs';
import { filter, map, switchMap, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
const SCROLLBAR_PLACEHOLDER = 17;
export class TuiDialogCloseService extends Observable {
    constructor(win, doc, el) {
        super(subscriber => merge(this.click$, this.esc$, this.mousedown$).subscribe(subscriber));
        this.win = win;
        this.doc = doc;
        this.el = el;
        this.click$ = tuiTypedFromEvent(this.element, `click`).pipe(filter(tuiIsCurrentTarget));
        this.esc$ = tuiTypedFromEvent(this.doc, `keydown`).pipe(filter(event => {
            const key = event.key;
            const target = tuiGetActualTarget(event);
            return (key === `Escape` &&
                tuiIsElement(target) &&
                (this.element.contains(target) ||
                    !tuiContainsOrAfter(this.element, target)));
        }));
        this.mousedown$ = tuiTypedFromEvent(this.doc, `mousedown`).pipe(filter(event => {
            const target = tuiGetActualTarget(event);
            const clientX = event.clientX;
            return (tuiIsElement(target) &&
                tuiGetViewportWidth(this.win) - clientX > SCROLLBAR_PLACEHOLDER &&
                !tuiContainsOrAfter(this.element, target));
        }), switchMap(() => tuiTypedFromEvent(this.doc, `mouseup`).pipe(take(1), map(tuiGetActualTarget), filter(target => tuiIsElement(target) && !tuiContainsOrAfter(this.element, target)))));
    }
    get element() {
        return this.el.nativeElement;
    }
}
TuiDialogCloseService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogCloseService, deps: [{ token: WINDOW }, { token: DOCUMENT }, { token: ElementRef }], target: i0.ɵɵFactoryTarget.Injectable });
TuiDialogCloseService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogCloseService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogCloseService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; } });
