// @taiga-ui/core/esm2015/components/dialog/dialog.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { tuiAsDialog, TuiPreventDefaultModule } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core/components/button';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDialogComponent } from './dialog.component';
import { TuiDialogDirective } from './dialog.directive';
import { TuiDialogService } from './dialog.service';
import * as i0 from "@angular/core";
export class TuiDialogModule {
}
TuiDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogModule, declarations: [TuiDialogComponent, TuiDialogDirective], imports: [PolymorpheusModule, TuiButtonModule, CommonModule, TuiPreventDefaultModule], exports: [TuiDialogComponent, TuiDialogDirective] });
TuiDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogModule, providers: [tuiAsDialog(TuiDialogService)], imports: [[PolymorpheusModule, TuiButtonModule, CommonModule, TuiPreventDefaultModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PolymorpheusModule, TuiButtonModule, CommonModule, TuiPreventDefaultModule],
                    declarations: [TuiDialogComponent, TuiDialogDirective],
                    exports: [TuiDialogComponent, TuiDialogDirective],
                    providers: [tuiAsDialog(TuiDialogService)],
                }]
        }] });
