// @taiga-ui/kit/esm2015/components/multi-select-option/multi-select-option.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiPrimitiveCheckboxModule } from '@taiga-ui/core';
import { TuiMultiSelectOptionComponent } from './multi-select-option.component';
import * as i0 from "@angular/core";
export class TuiMultiSelectOptionModule {
}
TuiMultiSelectOptionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMultiSelectOptionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMultiSelectOptionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMultiSelectOptionModule, declarations: [TuiMultiSelectOptionComponent], imports: [CommonModule, TuiPrimitiveCheckboxModule], exports: [TuiMultiSelectOptionComponent] });
TuiMultiSelectOptionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMultiSelectOptionModule, imports: [[CommonModule, TuiPrimitiveCheckboxModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMultiSelectOptionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiPrimitiveCheckboxModule],
                    declarations: [TuiMultiSelectOptionComponent],
                    exports: [TuiMultiSelectOptionComponent],
                }]
        }] });
