// @taiga-ui/kit/esm2015/components/elastic-container/elastic-container.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import { shouldCall } from '@tinkoff/ng-event-plugins';
import * as i0 from "@angular/core";
import * as i1 from "./elastic-container.directive";
export class TuiElasticContainerComponent {
    constructor() {
        this.height = NaN;
        this.transitions = 0;
    }
    onAnimation(_name, count) {
        this.transitions += count;
    }
}
TuiElasticContainerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElasticContainerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiElasticContainerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiElasticContainerComponent, selector: "tui-elastic-container", host: { properties: { "style.height.px": "this.height", "class._transitioning": "this.transitions" } }, ngImport: i0, template: "<div\n    class=\"t-wrapper\"\n    (transitionstart.silent)=\"onAnimation($any($event).propertyName, 1)\"\n    (transitionend.silent)=\"onAnimation($any($event).propertyName, -1)\"\n    (transitioncancel.silent)=\"onAnimation($any($event).propertyName, -1)\"\n    (tuiElasticContainer)=\"height = $event\"\n>\n    <ng-content></ng-content>\n</div>\n", styles: [":host{transition-property:height;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;display:block;overflow:hidden}:host._transitioning{height:auto!important}.t-wrapper{padding-top:1px;margin-top:-1px}\n"], directives: [{ type: i1.TuiElasticContainerDirective, selector: "[tuiElasticContainer]", outputs: ["tuiElasticContainer"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
__decorate([
    shouldCall(name => name === 'height')
], TuiElasticContainerComponent.prototype, "onAnimation", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElasticContainerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-elastic-container',
                    templateUrl: './elastic-container.component.html',
                    styleUrls: ['./elastic-container.component.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { height: [{
                type: HostBinding,
                args: ['style.height.px']
            }], transitions: [{
                type: HostBinding,
                args: ['class._transitioning']
            }], onAnimation: [] } });
