// @taiga-ui/kit/esm2015/components/carousel/carousel.directive.js
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { PAGE_VISIBILITY } from '@ng-web-apis/common';
import { ALWAYS_FALSE_HANDLER, ALWAYS_TRUE_HANDLER, tuiIfMap, tuiTypedFromEvent, } from '@taiga-ui/cdk';
import { BehaviorSubject, combineLatest, interval, merge, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiCarouselDirective extends Observable {
    constructor(el, visible$) {
        super(subscriber => this.output$.subscribe(subscriber));
        this.el = el;
        this.visible$ = visible$;
        this.duration$ = new BehaviorSubject(0);
        this.running$ = merge(tuiTypedFromEvent(this.el.nativeElement, 'mouseenter').pipe(map(ALWAYS_FALSE_HANDLER)), tuiTypedFromEvent(this.el.nativeElement, 'touchstart').pipe(map(ALWAYS_FALSE_HANDLER)), tuiTypedFromEvent(this.el.nativeElement, 'touchend').pipe(map(ALWAYS_TRUE_HANDLER)), tuiTypedFromEvent(this.el.nativeElement, 'mouseleave').pipe(map(ALWAYS_TRUE_HANDLER)), this.visible$);
        this.output$ = combineLatest([this.duration$, this.running$]).pipe(tuiIfMap(([duration]) => interval(duration), values => values.every(Boolean)));
    }
    set duration(duration) {
        this.duration$.next(duration);
    }
    set index(_) {
        this.duration$.next(this.duration$.value);
    }
}
TuiCarouselDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselDirective, deps: [{ token: ElementRef }, { token: PAGE_VISIBILITY }], target: i0.ɵɵFactoryTarget.Directive });
TuiCarouselDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCarouselDirective, selector: "tui-carousel", inputs: { duration: "duration", index: "index" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-carousel',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [PAGE_VISIBILITY]
                }] }]; }, propDecorators: { duration: [{
                type: Input
            }], index: [{
                type: Input
            }] } });
