// @taiga-ui/kit/esm2015/components/carousel/carousel-scroll.directive.js
import { Directive, ElementRef, Inject, Output } from '@angular/core';
import { tuiTypedFromEvent } from '@taiga-ui/cdk';
import { filter, map, tap, throttleTime } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiCarouselScrollDirective {
    constructor(el) {
        this.el = el;
        this.tuiCarouselScroll = tuiTypedFromEvent(this.el.nativeElement, 'wheel').pipe(filter(({ deltaX }) => Math.abs(deltaX) > 20), throttleTime(500), map(({ deltaX }) => Math.sign(deltaX)), tap(() => {
            // So we always have space to scroll and overflow-behavior saves us from back nav
            this.el.nativeElement.scrollLeft = 10;
        }));
    }
}
TuiCarouselScrollDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselScrollDirective, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiCarouselScrollDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCarouselScrollDirective, selector: "[tuiCarouselScroll]", outputs: { tuiCarouselScroll: "tuiCarouselScroll" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselScrollDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCarouselScroll]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { tuiCarouselScroll: [{
                type: Output
            }] } });
