// @taiga-ui/kit/esm2015/components/carousel/carousel.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { IntersectionObserverModule } from '@ng-web-apis/intersection-observer';
import { TuiItemDirective, TuiItemModule, TuiPanModule, TuiSwipeModule, } from '@taiga-ui/cdk';
import { TuiCarouselComponent } from './carousel.component';
import { TuiCarouselDirective } from './carousel.directive';
import { TuiCarouselAutoscrollDirective } from './carousel-autoscroll.directive';
import { TuiCarouselButtonsDirective } from './carousel-buttons.directive';
import { TuiCarouselScrollDirective } from './carousel-scroll.directive';
import * as i0 from "@angular/core";
export class TuiCarouselModule {
}
TuiCarouselModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCarouselModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselModule, declarations: [TuiCarouselComponent,
        TuiCarouselDirective,
        TuiCarouselButtonsDirective,
        TuiCarouselAutoscrollDirective,
        TuiCarouselScrollDirective], imports: [CommonModule,
        IntersectionObserverModule,
        TuiPanModule,
        TuiSwipeModule,
        TuiItemModule], exports: [TuiCarouselComponent,
        TuiCarouselDirective,
        TuiCarouselButtonsDirective,
        TuiItemDirective] });
TuiCarouselModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselModule, imports: [[
            CommonModule,
            IntersectionObserverModule,
            TuiPanModule,
            TuiSwipeModule,
            TuiItemModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        IntersectionObserverModule,
                        TuiPanModule,
                        TuiSwipeModule,
                        TuiItemModule,
                    ],
                    declarations: [
                        TuiCarouselComponent,
                        TuiCarouselDirective,
                        TuiCarouselButtonsDirective,
                        TuiCarouselAutoscrollDirective,
                        TuiCarouselScrollDirective,
                    ],
                    exports: [
                        TuiCarouselComponent,
                        TuiCarouselDirective,
                        TuiCarouselButtonsDirective,
                        TuiItemDirective,
                    ],
                }]
        }] });
