// @taiga-ui/kit/esm2015/components/carousel/carousel-autoscroll.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiCarouselDirective } from './carousel.directive';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiCarouselAutoscrollDirective {
    constructor(tuiCarouselAutoscroll) {
        this.tuiCarouselAutoscroll = tuiCarouselAutoscroll;
    }
}
TuiCarouselAutoscrollDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselAutoscrollDirective, deps: [{ token: TuiCarouselDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiCarouselAutoscrollDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCarouselAutoscrollDirective, selector: "[tuiCarouselAutoscroll]", outputs: { tuiCarouselAutoscroll: "tuiCarouselAutoscroll" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselAutoscrollDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCarouselAutoscroll]',
                    outputs: ['tuiCarouselAutoscroll'],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiCarouselDirective]
                }] }]; } });
