// @taiga-ui/core/esm2015/components/alert/alert.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { tuiAsAlerts } from '@taiga-ui/cdk';
import { TuiNotificationModule } from '@taiga-ui/core/components/notification';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiAlertComponent } from './alert.component';
import { TuiAlertService } from './alert.service';
import * as i0 from "@angular/core";
export class TuiAlertModule {
}
TuiAlertModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAlertModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertModule, declarations: [TuiAlertComponent], imports: [CommonModule, PolymorpheusModule, TuiNotificationModule], exports: [TuiAlertComponent] });
TuiAlertModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertModule, providers: [tuiAsAlerts(TuiAlertService)], imports: [[CommonModule, PolymorpheusModule, TuiNotificationModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiNotificationModule],
                    declarations: [TuiAlertComponent],
                    exports: [TuiAlertComponent],
                    providers: [tuiAsAlerts(TuiAlertService)],
                }]
        }] });
