// @taiga-ui/cdk/esm2015/directives/control/control.directive.js
import { Directive, Inject } from '@angular/core';
import { NgControl } from '@angular/forms';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
export class TuiControlDirective {
    constructor(ngControl) {
        this.ngControl = ngControl;
    }
    get control() {
        return this.ngControl.control;
    }
}
TuiControlDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlDirective, deps: [{ token: NgControl }], target: i0.ɵɵFactoryTarget.Directive });
TuiControlDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiControlDirective, selector: "[tuiControl]", exportAs: ["ngControl"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiControl]',
                    exportAs: 'ngControl',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgControl, decorators: [{
                    type: Inject,
                    args: [NgControl]
                }] }]; } });
