// @taiga-ui/kit/esm2015/components/items-with-more/items-with-more.directive.js
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { AbstractTuiController, tuiDefaultProp } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class TuiItemsWithMoreDirective extends AbstractTuiController {
    constructor() {
        super(...arguments);
        this.itemsLimit = Infinity;
        this.required = -1;
    }
}
TuiItemsWithMoreDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemsWithMoreDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiItemsWithMoreDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiItemsWithMoreDirective, selector: "tui-items-with-more", inputs: { itemsLimit: "itemsLimit", required: "required" }, usesInheritance: true, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiItemsWithMoreDirective.prototype, "itemsLimit", void 0);
__decorate([
    tuiDefaultProp()
], TuiItemsWithMoreDirective.prototype, "required", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemsWithMoreDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-items-with-more',
                }]
        }], propDecorators: { itemsLimit: [{
                type: Input
            }], required: [{
                type: Input
            }] } });
