// @taiga-ui/kit/esm2015/components/items-with-more/items-with-more.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiItemDirective, TuiItemModule, TuiLetModule } from '@taiga-ui/cdk';
import { TuiProjectClassModule } from '@taiga-ui/kit/directives';
import { TuiItemsWithMoreComponent } from './items-with-more.component';
import { TuiItemsWithMoreDirective } from './items-with-more.directive';
import { TuiMoreDirective } from './more.directive';
import * as i0 from "@angular/core";
export class TuiItemsWithMoreModule {
}
TuiItemsWithMoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemsWithMoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiItemsWithMoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemsWithMoreModule, declarations: [TuiItemsWithMoreComponent,
        TuiItemsWithMoreDirective,
        TuiMoreDirective], imports: [CommonModule, TuiItemModule, TuiLetModule, TuiProjectClassModule], exports: [TuiItemsWithMoreComponent,
        TuiItemsWithMoreDirective,
        TuiMoreDirective,
        TuiItemDirective] });
TuiItemsWithMoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemsWithMoreModule, imports: [[CommonModule, TuiItemModule, TuiLetModule, TuiProjectClassModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemsWithMoreModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiItemModule, TuiLetModule, TuiProjectClassModule],
                    declarations: [
                        TuiItemsWithMoreComponent,
                        TuiItemsWithMoreDirective,
                        TuiMoreDirective,
                    ],
                    exports: [
                        TuiItemsWithMoreComponent,
                        TuiItemsWithMoreDirective,
                        TuiMoreDirective,
                        TuiItemDirective,
                    ],
                }]
        }] });
