// @taiga-ui/kit/esm2015/components/data-list-wrapper/data-list-wrapper.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiElementModule } from '@taiga-ui/cdk';
import { TuiDataListDirective, TuiDataListModule, TuiLoaderModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDataListGroupWrapperComponent } from './data-list-group-wrapper.component';
import { TuiDataListWrapperComponent } from './data-list-wrapper.component';
import * as i0 from "@angular/core";
export class TuiDataListWrapperModule {
}
TuiDataListWrapperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListWrapperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDataListWrapperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListWrapperModule, declarations: [TuiDataListWrapperComponent, TuiDataListGroupWrapperComponent], imports: [CommonModule,
        PolymorpheusModule,
        TuiElementModule,
        TuiDataListModule,
        TuiLoaderModule], exports: [TuiDataListWrapperComponent,
        TuiDataListGroupWrapperComponent,
        TuiDataListDirective] });
TuiDataListWrapperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListWrapperModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiElementModule,
            TuiDataListModule,
            TuiLoaderModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDataListWrapperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiElementModule,
                        TuiDataListModule,
                        TuiLoaderModule,
                    ],
                    declarations: [TuiDataListWrapperComponent, TuiDataListGroupWrapperComponent],
                    exports: [
                        TuiDataListWrapperComponent,
                        TuiDataListGroupWrapperComponent,
                        TuiDataListDirective,
                    ],
                }]
        }] });
