// @taiga-ui/cdk/esm2015/directives/pan/pan.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiPanService } from '@taiga-ui/cdk/services';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiPanDirective {
    constructor(tuiPan) {
        this.tuiPan = tuiPan;
    }
}
TuiPanDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanDirective, deps: [{ token: TuiPanService }], target: i0.ɵɵFactoryTarget.Directive });
TuiPanDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPanDirective, selector: "[tuiPan]", outputs: { tuiPan: "tuiPan" }, providers: [TuiPanService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPan]',
                    outputs: ['tuiPan'],
                    providers: [TuiPanService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiPanService]
                }] }]; } });
