// @taiga-ui/kit/esm2015/components/input-inline/input-inline.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiFocusableModule, TuiFocusedModule } from '@taiga-ui/cdk';
import { TuiMaskAccessorModule } from '@taiga-ui/core';
import { TuiInputInlineComponent } from './input-inline.component';
import * as i0 from "@angular/core";
export class TuiInputInlineModule {
}
TuiInputInlineModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputInlineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputInlineModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputInlineModule, declarations: [TuiInputInlineComponent], imports: [CommonModule,
        FormsModule,
        TuiFocusableModule,
        TuiFocusedModule,
        TuiMaskAccessorModule], exports: [TuiInputInlineComponent] });
TuiInputInlineModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputInlineModule, imports: [[
            CommonModule,
            FormsModule,
            TuiFocusableModule,
            TuiFocusedModule,
            TuiMaskAccessorModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputInlineModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiFocusableModule,
                        TuiFocusedModule,
                        TuiMaskAccessorModule,
                    ],
                    declarations: [TuiInputInlineComponent],
                    exports: [TuiInputInlineComponent],
                }]
        }] });
