// @taiga-ui/core/esm2015/components/label/label.module.js
import { NgModule } from '@angular/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiLabelComponent } from './label.component';
import * as i0 from "@angular/core";
export class TuiLabelModule {
}
TuiLabelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLabelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLabelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLabelModule, declarations: [TuiLabelComponent], imports: [PolymorpheusModule], exports: [TuiLabelComponent] });
TuiLabelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLabelModule, imports: [[PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLabelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PolymorpheusModule],
                    declarations: [TuiLabelComponent],
                    exports: [TuiLabelComponent],
                }]
        }] });
