// @taiga-ui/cdk/esm2015/directives/focus-visible/focus-visible.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiDestroyService, TuiFocusVisibleService } from '@taiga-ui/cdk/services';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
/**
 * Directive to imitate :focus-visible
 * (https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible)
 * in browsers that do not support it
 */
export class TuiFocusVisibleDirective {
    constructor(tuiFocusVisibleChange) {
        this.tuiFocusVisibleChange = tuiFocusVisibleChange;
    }
}
TuiFocusVisibleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFocusVisibleDirective, deps: [{ token: TuiFocusVisibleService }], target: i0.ɵɵFactoryTarget.Directive });
TuiFocusVisibleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiFocusVisibleDirective, selector: "[tuiFocusVisibleChange]", outputs: { tuiFocusVisibleChange: "tuiFocusVisibleChange" }, providers: [TuiDestroyService, TuiFocusVisibleService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFocusVisibleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiFocusVisibleChange]',
                    outputs: ['tuiFocusVisibleChange'],
                    providers: [TuiDestroyService, TuiFocusVisibleService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiFocusVisibleService]
                }] }]; } });
