// @taiga-ui/kit/esm2015/components/input-tag/input-tag.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiActiveZoneModule, TuiFocusableModule, TuiHoveredModule, TuiScrollService, } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiScrollbarModule, TuiSvgModule, TuiTooltipModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiTagModule } from '@taiga-ui/kit/components/tag';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputTagComponent } from './input-tag.component';
import * as i0 from "@angular/core";
export class TuiInputTagModule {
}
TuiInputTagModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTagModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputTagModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTagModule, declarations: [TuiInputTagComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiFocusableModule,
        TuiHoveredModule,
        TuiActiveZoneModule,
        TuiSvgModule,
        TuiScrollbarModule,
        TuiTooltipModule,
        TuiHostedDropdownModule,
        TuiTagModule,
        TuiWrapperModule], exports: [TuiInputTagComponent] });
TuiInputTagModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTagModule, providers: [TuiScrollService], imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiFocusableModule,
            TuiHoveredModule,
            TuiActiveZoneModule,
            TuiSvgModule,
            TuiScrollbarModule,
            TuiTooltipModule,
            TuiHostedDropdownModule,
            TuiTagModule,
            TuiWrapperModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTagModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiFocusableModule,
                        TuiHoveredModule,
                        TuiActiveZoneModule,
                        TuiSvgModule,
                        TuiScrollbarModule,
                        TuiTooltipModule,
                        TuiHostedDropdownModule,
                        TuiTagModule,
                        TuiWrapperModule,
                    ],
                    providers: [TuiScrollService],
                    declarations: [TuiInputTagComponent],
                    exports: [TuiInputTagComponent],
                }]
        }] });
