// @taiga-ui/kit/esm2015/components/island/island.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiIslandComponent } from './island.component';
import * as i0 from "@angular/core";
export class TuiIslandModule {
}
TuiIslandModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIslandModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiIslandModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIslandModule, declarations: [TuiIslandComponent], imports: [CommonModule], exports: [TuiIslandComponent] });
TuiIslandModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIslandModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIslandModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiIslandComponent],
                    exports: [TuiIslandComponent],
                }]
        }] });
