// @taiga-ui/kit/esm2015/components/input-phone-international/input-phone-international.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiActiveZoneModule, TuiLetModule, TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiDataListModule, TuiFlagPipeModule, TuiGroupModule, TuiHintModule, TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiArrowModule } from '@taiga-ui/kit/components/arrow';
import { TuiInputPhoneModule } from '@taiga-ui/kit/components/input-phone';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputPhoneInternationalComponent } from './input-phone-international.component';
import * as i0 from "@angular/core";
export class TuiInputPhoneInternationalModule {
}
TuiInputPhoneInternationalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneInternationalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputPhoneInternationalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneInternationalModule, declarations: [TuiInputPhoneInternationalComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiSvgModule,
        TuiInputPhoneModule,
        TuiGroupModule,
        TuiPrimitiveTextfieldModule,
        TuiTextfieldControllerModule,
        TuiHintModule,
        TuiHostedDropdownModule,
        TuiDataListModule,
        TuiArrowModule,
        TuiWrapperModule,
        TuiLetModule,
        TuiMapperPipeModule,
        TuiActiveZoneModule,
        TuiFlagPipeModule], exports: [TuiInputPhoneInternationalComponent] });
TuiInputPhoneInternationalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneInternationalModule, imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiSvgModule,
            TuiInputPhoneModule,
            TuiGroupModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiHintModule,
            TuiHostedDropdownModule,
            TuiDataListModule,
            TuiArrowModule,
            TuiWrapperModule,
            TuiLetModule,
            TuiMapperPipeModule,
            TuiActiveZoneModule,
            TuiFlagPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneInternationalModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiSvgModule,
                        TuiInputPhoneModule,
                        TuiGroupModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiHintModule,
                        TuiHostedDropdownModule,
                        TuiDataListModule,
                        TuiArrowModule,
                        TuiWrapperModule,
                        TuiLetModule,
                        TuiMapperPipeModule,
                        TuiActiveZoneModule,
                        TuiFlagPipeModule,
                    ],
                    declarations: [TuiInputPhoneInternationalComponent],
                    exports: [TuiInputPhoneInternationalComponent],
                }]
        }] });
