// @taiga-ui/kit/fesm2015/taiga-ui-kit-components-routable-dialog.js
import { __rest } from 'tslib';
import * as i0 from '@angular/core';
import { Injector, Component, ChangeDetectionStrategy, Inject, Self, NgModule } from '@angular/core';
import * as i1 from '@angular/router';
import { ActivatedRoute, Router } from '@angular/router';
import * as i3 from '@taiga-ui/cdk';
import { TuiDestroyService } from '@taiga-ui/cdk';
import * as i2 from '@taiga-ui/core';
import { TuiDialogService } from '@taiga-ui/core';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { takeUntil } from 'rxjs/operators';

class TuiRoutableDialogComponent {
    constructor(route, router, dialogs, injector, destroy$) {
        this.route = route;
        this.router = router;
        dialogs
            .open(new PolymorpheusComponent(this.route.snapshot.data['dialog'], injector), this.route.snapshot.data['dialogOptions'])
            .pipe(takeUntil(destroy$))
            .subscribe({
            complete: () => this.navigateToParent(),
        });
    }
    navigateToParent() {
        const isLazy = this.route.snapshot.data['isLazy'];
        const backUrl = isLazy
            ? this.getLazyLoadedBackUrl()
            : this.route.snapshot.data['backUrl'];
        void this.router.navigate([backUrl], {
            relativeTo: this.route,
        });
    }
    getLazyLoadedBackUrl() {
        var _a;
        const urlSegments = ((_a = this.route.parent) === null || _a === void 0 ? void 0 : _a.snapshot.url) || [];
        return urlSegments.map(() => '..').join('/');
    }
}
TuiRoutableDialogComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogComponent, deps: [{ token: ActivatedRoute }, { token: Router }, { token: TuiDialogService }, { token: Injector }, { token: TuiDestroyService, self: true }], target: i0.ɵɵFactoryTarget.Component });
TuiRoutableDialogComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRoutableDialogComponent, selector: "tui-routable-dialog", providers: [TuiDestroyService], ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-routable-dialog',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i1.ActivatedRoute, decorators: [{
                    type: Inject,
                    args: [ActivatedRoute]
                }] }, { type: i1.Router, decorators: [{
                    type: Inject,
                    args: [Router]
                }] }, { type: i2.TuiDialogService, decorators: [{
                    type: Inject,
                    args: [TuiDialogService]
                }] }, { type: i0.Injector, decorators: [{
                    type: Inject,
                    args: [Injector]
                }] }, { type: i3.TuiDestroyService, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }]; } });

function tuiGenerateDialogableRoute(component, _a = {}) {
    var { path = `` } = _a, dialogOptions = __rest(_a, ["path"]);
    return {
        path,
        component: TuiRoutableDialogComponent,
        data: {
            dialog: component,
            backUrl: path
                .split(`/`)
                .map(() => `..`)
                .join(`/`),
            isLazy: path === ``,
            dialogOptions,
        },
    };
}

class TuiRoutableDialogModule {
}
TuiRoutableDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRoutableDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogModule, declarations: [TuiRoutableDialogComponent], exports: [TuiRoutableDialogComponent] });
TuiRoutableDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiRoutableDialogComponent],
                    exports: [TuiRoutableDialogComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiRoutableDialogComponent, TuiRoutableDialogModule, tuiGenerateDialogableRoute };
