// @taiga-ui/kit/fesm2015/taiga-ui-kit-directives-value-accessor.js
import * as i0 from '@angular/core';
import { Directive, NgModule } from '@angular/core';
import { TUI_VALUE_ACCESSOR_PROVIDER } from '@taiga-ui/kit/providers';

class TuiValueAccessorDirective {
}
TuiValueAccessorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValueAccessorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiValueAccessorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiValueAccessorDirective, selector: "[tuiValueAccessor]", providers: [TUI_VALUE_ACCESSOR_PROVIDER], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValueAccessorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiValueAccessor]',
                    providers: [TUI_VALUE_ACCESSOR_PROVIDER],
                }]
        }] });

class TuiValueAccessorModule {
}
TuiValueAccessorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValueAccessorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiValueAccessorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValueAccessorModule, declarations: [TuiValueAccessorDirective], exports: [TuiValueAccessorDirective] });
TuiValueAccessorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValueAccessorModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValueAccessorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiValueAccessorDirective],
                    exports: [TuiValueAccessorDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiValueAccessorDirective, TuiValueAccessorModule };
