// @taiga-ui/kit/fesm2015/taiga-ui-kit-pipes-to-year.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';
import { TuiYear } from '@taiga-ui/cdk';

class TuiToYearPipe {
    transform(value) {
        return new TuiYear(value);
    }
}
TuiToYearPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiToYearPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipe, name: "tuiToYear" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: `tuiToYear`,
                }]
        }] });

class TuiToYearPipeModule {
}
TuiToYearPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiToYearPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipeModule, declarations: [TuiToYearPipe], exports: [TuiToYearPipe] });
TuiToYearPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiToYearPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiToYearPipe],
                    exports: [TuiToYearPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiToYearPipe, TuiToYearPipeModule };
