// @taiga-ui/kit/fesm2015/taiga-ui-kit-pipes-stringify.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';

class TuiStringifyPipe {
    transform(key) {
        return value => { var _a; return String((_a = value[key]) !== null && _a !== void 0 ? _a : ``); };
    }
}
TuiStringifyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiStringifyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipe, name: "tuiStringify" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: `tuiStringify`,
                }]
        }] });

class TuiStringifyPipeModule {
}
TuiStringifyPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiStringifyPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipeModule, declarations: [TuiStringifyPipe], exports: [TuiStringifyPipe] });
TuiStringifyPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStringifyPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiStringifyPipe],
                    exports: [TuiStringifyPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiStringifyPipe, TuiStringifyPipeModule };
