// @taiga-ui/kit/fesm2015/taiga-ui-kit-components-breadcrumbs.js
import { __decorate } from 'tslib';
import * as i0 from '@angular/core';
import { forwardRef, TemplateRef, Component, ChangeDetectionStrategy, Input, HostBinding, ContentChildren, NgModule } from '@angular/core';
import { EMPTY_QUERY, TuiItemDirective, tuiDefaultProp, TuiItemModule } from '@taiga-ui/cdk';
import * as i1 from '@taiga-ui/core';
import { TuiModeDirective, TuiSvgModule } from '@taiga-ui/core';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';

class TuiBreadcrumbsComponent extends TuiModeDirective {
    constructor() {
        super(...arguments);
        this.size = 'm';
        this.items = EMPTY_QUERY;
        this.mode = 'onLight';
    }
}
TuiBreadcrumbsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TuiBreadcrumbsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBreadcrumbsComponent, selector: "tui-breadcrumbs", inputs: { size: "size" }, host: { properties: { "attr.data-size": "this.size" } }, providers: [
        {
            provide: TuiModeDirective,
            useExisting: forwardRef(() => TuiBreadcrumbsComponent),
        },
    ], queries: [{ propertyName: "items", predicate: TuiItemDirective, read: TemplateRef }], usesInheritance: true, ngImport: i0, template: "<ng-container *ngIf=\"items.changes | async\"></ng-container>\n<ng-container *ngFor=\"let item of items; let last = last\">\n    <ng-container [ngTemplateOutlet]=\"item\"></ng-container>\n    <tui-svg\n        *ngIf=\"!last\"\n        src=\"tuiIconChevronRight\"\n        class=\"t-icon\"\n    ></tui-svg>\n</ng-container>\n", styles: [":host{display:flex;align-items:center;height:2.5rem;font:var(--tui-font-text-s)}:host[data-size=l]{font:var(--tui-font-text-m)}.t-icon{margin:0 .25rem;color:var(--tui-text-03)}\n"], components: [{ type: i1.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], pipes: { "async": i2.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
__decorate([
    tuiDefaultProp()
], TuiBreadcrumbsComponent.prototype, "size", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-breadcrumbs',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    templateUrl: './breadcrumbs.template.html',
                    styleUrls: ['./breadcrumbs.style.less'],
                    providers: [
                        {
                            provide: TuiModeDirective,
                            useExisting: forwardRef(() => TuiBreadcrumbsComponent),
                        },
                    ],
                }]
        }], propDecorators: { size: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-size']
            }], items: [{
                type: ContentChildren,
                args: [TuiItemDirective, { read: TemplateRef }]
            }] } });

class TuiBreadcrumbsModule {
}
TuiBreadcrumbsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBreadcrumbsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, declarations: [TuiBreadcrumbsComponent], imports: [CommonModule, TuiSvgModule, TuiItemModule], exports: [TuiBreadcrumbsComponent, TuiItemDirective] });
TuiBreadcrumbsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, imports: [[CommonModule, TuiSvgModule, TuiItemModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiItemModule],
                    declarations: [TuiBreadcrumbsComponent],
                    exports: [TuiBreadcrumbsComponent, TuiItemDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBreadcrumbsComponent, TuiBreadcrumbsModule };
