// @taiga-ui/kit/fesm2015/taiga-ui-kit-directives-project-class.js
import { __decorate } from 'tslib';
import * as i0 from '@angular/core';
import { ElementRef, Directive, Inject, Input, NgModule } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';

/**
 * A directive for projecting classes from nested children to host
 */
class TuiProjectClassDirective {
    constructor(el) {
        this.el = el;
        this.classNames = [];
    }
    ngAfterViewChecked() {
        this.classNames.forEach(className => {
            const hostElement = this.el.nativeElement;
            hostElement.classList.toggle(className, !!hostElement.querySelector(`.${className}`));
        });
    }
}
TuiProjectClassDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassDirective, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiProjectClassDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiProjectClassDirective, selector: "[tuiProjectClass]", inputs: { classNames: ["tuiProjectClass", "classNames"] }, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiProjectClassDirective.prototype, "classNames", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiProjectClass]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { classNames: [{
                type: Input,
                args: ['tuiProjectClass']
            }] } });

class TuiProjectClassModule {
}
TuiProjectClassModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiProjectClassModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassModule, declarations: [TuiProjectClassDirective], exports: [TuiProjectClassDirective] });
TuiProjectClassModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiProjectClassDirective],
                    exports: [TuiProjectClassDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiProjectClassDirective, TuiProjectClassModule };
