// @taiga-ui/kit/esm2015/components/checkbox/checkbox.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiCheckedModule, TuiFocusableModule, TuiFocusedModule, TuiFocusVisibleModule, } from '@taiga-ui/cdk';
import { TuiPrimitiveCheckboxModule } from '@taiga-ui/core';
import { TuiCheckboxComponent } from './checkbox.component';
import * as i0 from "@angular/core";
export class TuiCheckboxModule {
}
TuiCheckboxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCheckboxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxModule, declarations: [TuiCheckboxComponent], imports: [CommonModule,
        TuiCheckedModule,
        TuiFocusableModule,
        TuiFocusedModule,
        TuiFocusVisibleModule,
        TuiPrimitiveCheckboxModule], exports: [TuiCheckboxComponent] });
TuiCheckboxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxModule, imports: [[
            CommonModule,
            TuiCheckedModule,
            TuiFocusableModule,
            TuiFocusedModule,
            TuiFocusVisibleModule,
            TuiPrimitiveCheckboxModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiCheckedModule,
                        TuiFocusableModule,
                        TuiFocusedModule,
                        TuiFocusVisibleModule,
                        TuiPrimitiveCheckboxModule,
                    ],
                    declarations: [TuiCheckboxComponent],
                    exports: [TuiCheckboxComponent],
                }]
        }] });
