// @taiga-ui/kit/esm2015/components/combo-box/combo-box-strict.directive.js
import { __decorate } from "tslib";
import { Directive, HostListener, Inject, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiComboBoxComponent } from './combo-box.component';
import * as i0 from "@angular/core";
import * as i1 from "./combo-box.component";
export class TuiComboBoxStrictDirective {
    constructor(comboBox) {
        this.comboBox = comboBox;
        this.strict = true;
    }
    onInput() {
        if (!this.strict && this.comboBox.search) {
            this.comboBox.value = this.comboBox.search;
        }
    }
}
TuiComboBoxStrictDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiComboBoxStrictDirective, deps: [{ token: TuiComboBoxComponent }], target: i0.ɵɵFactoryTarget.Directive });
TuiComboBoxStrictDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiComboBoxStrictDirective, selector: "tui-combo-box[strict]", inputs: { strict: "strict" }, host: { listeners: { "input": "onInput()" } }, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiComboBoxStrictDirective.prototype, "strict", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiComboBoxStrictDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-combo-box[strict]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiComboBoxComponent, decorators: [{
                    type: Inject,
                    args: [TuiComboBoxComponent]
                }] }]; }, propDecorators: { strict: [{
                type: Input
            }], onInput: [{
                type: HostListener,
                args: ['input']
            }] } });
