// @taiga-ui/kit/esm2015/components/combo-box/combo-box.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiTextfieldComponent, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiArrowModule } from '@taiga-ui/kit/components/arrow';
import { TuiSelectOptionModule } from '@taiga-ui/kit/components/select-option';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiComboBoxComponent } from './combo-box.component';
import { TuiComboBoxDirective } from './combo-box.directive';
import { TuiComboBoxStrictDirective } from './combo-box-strict.directive';
import * as i0 from "@angular/core";
export class TuiComboBoxModule {
}
TuiComboBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiComboBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiComboBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiComboBoxModule, declarations: [TuiComboBoxComponent,
        TuiComboBoxStrictDirective,
        TuiComboBoxDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiActiveZoneModule,
        TuiPrimitiveTextfieldModule,
        TuiHostedDropdownModule,
        TuiSelectOptionModule,
        TuiArrowModule,
        TuiWrapperModule,
        TuiTextfieldControllerModule], exports: [TuiComboBoxComponent,
        TuiComboBoxStrictDirective,
        TuiComboBoxDirective,
        TuiTextfieldComponent] });
TuiComboBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiComboBoxModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiActiveZoneModule,
            TuiPrimitiveTextfieldModule,
            TuiHostedDropdownModule,
            TuiSelectOptionModule,
            TuiArrowModule,
            TuiWrapperModule,
            TuiTextfieldControllerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiComboBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiActiveZoneModule,
                        TuiPrimitiveTextfieldModule,
                        TuiHostedDropdownModule,
                        TuiSelectOptionModule,
                        TuiArrowModule,
                        TuiWrapperModule,
                        TuiTextfieldControllerModule,
                    ],
                    declarations: [
                        TuiComboBoxComponent,
                        TuiComboBoxStrictDirective,
                        TuiComboBoxDirective,
                    ],
                    exports: [
                        TuiComboBoxComponent,
                        TuiComboBoxStrictDirective,
                        TuiComboBoxDirective,
                        TuiTextfieldComponent,
                    ],
                }]
        }] });
