// @taiga-ui/core/esm2015/components/scroll-controls/scrollbar-wrapper.directive.js
import { Directive, ElementRef } from '@angular/core';
import { TUI_ELEMENT_REF } from '@taiga-ui/core/tokens';
import * as i0 from "@angular/core";
export class TuiScrollbarWrapperDirective {
}
TuiScrollbarWrapperDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollbarWrapperDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiScrollbarWrapperDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiScrollbarWrapperDirective, selector: "[tuiScrollbarWrapper]", providers: [
        {
            provide: TUI_ELEMENT_REF,
            useExisting: ElementRef,
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollbarWrapperDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiScrollbarWrapper]',
                    providers: [
                        {
                            provide: TUI_ELEMENT_REF,
                            useExisting: ElementRef,
                        },
                    ],
                }]
        }] });
