// @taiga-ui/core/esm2015/components/hosted-dropdown/accessor-proxy.directive.js
import { Directive, Optional, SkipSelf } from '@angular/core';
import { TuiPositionAccessor } from '@taiga-ui/core/abstract';
import * as i0 from "@angular/core";
export class TuiAccessorProxyDirective extends TuiPositionAccessor {
    constructor() {
        super(...arguments);
        this.type = 'dummy';
    }
    getPosition() {
        return [0, 0];
    }
}
TuiAccessorProxyDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAccessorProxyDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiAccessorProxyDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAccessorProxyDirective, selector: "[tuiAccessorProxy]", providers: [
        {
            provide: TuiPositionAccessor,
            multi: true,
            deps: [
                [new Optional(), new SkipSelf(), TuiAccessorProxyDirective],
                TuiAccessorProxyDirective,
            ],
            useFactory: (accessor, fallback) => accessor || fallback,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAccessorProxyDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiAccessorProxy]',
                    providers: [
                        {
                            provide: TuiPositionAccessor,
                            multi: true,
                            deps: [
                                [new Optional(), new SkipSelf(), TuiAccessorProxyDirective],
                                TuiAccessorProxyDirective,
                            ],
                            useFactory: (accessor, fallback) => accessor || fallback,
                        },
                    ],
                }]
        }] });
