// @taiga-ui/core/esm2015/components/hosted-dropdown/hosted-dropdown.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule, TuiLetModule, TuiObscuredModule } from '@taiga-ui/cdk';
import { TuiDropdownModule, TuiDropdownOptionsDirective, } from '@taiga-ui/core/directives/dropdown';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiAccessorProxyDirective } from './accessor-proxy.directive';
import { TuiHostedDropdownComponent } from './hosted-dropdown.component';
import { TuiHostedDropdownConnectorDirective } from './hosted-dropdown-connector.directive';
import * as i0 from "@angular/core";
export class TuiHostedDropdownModule {
}
TuiHostedDropdownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHostedDropdownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiHostedDropdownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHostedDropdownModule, declarations: [TuiAccessorProxyDirective,
        TuiHostedDropdownComponent,
        TuiHostedDropdownConnectorDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiLetModule,
        TuiObscuredModule,
        TuiActiveZoneModule,
        TuiDropdownModule], exports: [TuiHostedDropdownComponent,
        TuiHostedDropdownConnectorDirective,
        TuiDropdownOptionsDirective] });
TuiHostedDropdownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHostedDropdownModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiLetModule,
            TuiObscuredModule,
            TuiActiveZoneModule,
            TuiDropdownModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHostedDropdownModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiLetModule,
                        TuiObscuredModule,
                        TuiActiveZoneModule,
                        TuiDropdownModule,
                    ],
                    declarations: [
                        TuiAccessorProxyDirective,
                        TuiHostedDropdownComponent,
                        TuiHostedDropdownConnectorDirective,
                    ],
                    exports: [
                        TuiHostedDropdownComponent,
                        TuiHostedDropdownConnectorDirective,
                        TuiDropdownOptionsDirective,
                    ],
                }]
        }] });
