// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-pipes-replace.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';

class TuiReplacePipe {
    transform(value, search, replaceValue) {
        var _a;
        return ((_a = value === null || value === void 0 ? void 0 : value.replace(search, 
        // TS bug: https://github.com/microsoft/TypeScript/issues/22378
        replaceValue)) !== null && _a !== void 0 ? _a : ``);
    }
}
TuiReplacePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiReplacePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipe, name: "tuiReplace" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipe, decorators: [{
            type: Pipe,
            args: [{ name: `tuiReplace` }]
        }] });

class TuiReplacePipeModule {
}
TuiReplacePipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiReplacePipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipeModule, declarations: [TuiReplacePipe], exports: [TuiReplacePipe] });
TuiReplacePipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReplacePipeModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [TuiReplacePipe],
                    declarations: [TuiReplacePipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiReplacePipe, TuiReplacePipeModule };
