// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-pipes-mapper.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';

class TuiMapperPipe {
    /**
     * Maps object to an arbitrary result through a mapper function
     *
     * @param value an item to transform
     * @param mapper a mapping function
     * @param args arbitrary number of additional arguments
     */
    transform(value, mapper, ...args) {
        return mapper(value, ...args);
    }
}
TuiMapperPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiMapperPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipe, name: "tuiMapper" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipe, decorators: [{
            type: Pipe,
            args: [{ name: `tuiMapper` }]
        }] });

class TuiMapperPipeModule {
}
TuiMapperPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMapperPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipeModule, declarations: [TuiMapperPipe], exports: [TuiMapperPipe] });
TuiMapperPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMapperPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [TuiMapperPipe],
                    declarations: [TuiMapperPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMapperPipe, TuiMapperPipeModule };
