// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-portal.js
import * as i0 from '@angular/core';
import { TemplateRef, Directive, Inject, Input, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk/components/dropdown-host';
import { TuiDropdownPortalService } from '@taiga-ui/cdk/components/dropdown-host';

class TuiPortalDirective {
    constructor(templateRef, portalService) {
        this.templateRef = templateRef;
        this.portalService = portalService;
    }
    set tuiPortal(show) {
        var _a;
        (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
        if (show) {
            this.viewRef = this.portalService.addTemplate(this.templateRef);
        }
    }
    ngOnDestroy() {
        var _a;
        (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
    }
}
TuiPortalDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPortalDirective, deps: [{ token: TemplateRef }, { token: TuiDropdownPortalService }], target: i0.ɵɵFactoryTarget.Directive });
TuiPortalDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPortalDirective, selector: "[tuiPortal]", inputs: { tuiPortal: "tuiPortal" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPortalDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPortal]',
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }, { type: i1.TuiDropdownPortalService, decorators: [{
                    type: Inject,
                    args: [TuiDropdownPortalService]
                }] }]; }, propDecorators: { tuiPortal: [{
                type: Input
            }] } });

class TuiPortalModule {
}
TuiPortalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPortalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPortalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPortalModule, declarations: [TuiPortalDirective], exports: [TuiPortalDirective] });
TuiPortalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPortalModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPortalModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiPortalDirective],
                    exports: [TuiPortalDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPortalDirective, TuiPortalModule };
