// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-control.js
import * as i0 from '@angular/core';
import { Directive, Inject, NgModule } from '@angular/core';
import * as i1 from '@angular/forms';
import { NgControl } from '@angular/forms';

class TuiControlDirective {
    constructor(ngControl) {
        this.ngControl = ngControl;
    }
    get control() {
        return this.ngControl.control;
    }
}
TuiControlDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlDirective, deps: [{ token: NgControl }], target: i0.ɵɵFactoryTarget.Directive });
TuiControlDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiControlDirective, selector: "[tuiControl]", exportAs: ["ngControl"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiControl]',
                    exportAs: 'ngControl',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgControl, decorators: [{
                    type: Inject,
                    args: [NgControl]
                }] }]; } });

class TuiControlModule {
}
TuiControlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiControlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlModule, declarations: [TuiControlDirective], exports: [TuiControlDirective] });
TuiControlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiControlModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiControlDirective],
                    exports: [TuiControlDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiControlDirective, TuiControlModule };
