// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-item.js
import * as i0 from '@angular/core';
import { Directive, NgModule } from '@angular/core';

/**
 * Blank directive for queries via `@ContentChildren` / `@ViewChildren` / `querySelector`
 */
class TuiItemDirective {
}
TuiItemDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiItemDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiItemDirective, selector: "[tuiItem]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiItem]',
                }]
        }] });

class TuiItemModule {
}
TuiItemModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiItemModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemModule, declarations: [TuiItemDirective], exports: [TuiItemDirective] });
TuiItemModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiItemModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiItemDirective],
                    exports: [TuiItemDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiItemDirective, TuiItemModule };
