// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-element.js
import * as i0 from '@angular/core';
import { ElementRef, Directive, Inject, NgModule } from '@angular/core';

class TuiElementDirective {
    constructor({ nativeElement }) {
        /**
         * @note:
         * Typically, when your constructor is invoked with new,
         * an object is created, its constructor is assigned to
         * the invoked constructor and the object is then assigned
         * to this before executing any operations specified
         * in your constructor method.
         *
         * ERROR TypeError: Class constructor ElementRef cannot be invoked without 'new'
         * https://github.com/Tinkoff/taiga-ui/issues/3072
         *
         * This way we can instantiate object creation
         * without additional prototype chain for possible fix bug.
         */
        return new ElementRef(nativeElement);
    }
}
TuiElementDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElementDirective, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiElementDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiElementDirective, selector: "[tuiElement]", exportAs: ["elementRef"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElementDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiElement]',
                    exportAs: 'elementRef',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; } });

class TuiElementModule {
}
TuiElementModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElementModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiElementModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElementModule, declarations: [TuiElementDirective], exports: [TuiElementDirective] });
TuiElementModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElementModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElementModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiElementDirective],
                    exports: [TuiElementDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiElementDirective, TuiElementModule };
