// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-pipes-keys.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';

class TuiKeysPipe {
    transform(object) {
        return Object.keys(object);
    }
}
TuiKeysPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiKeysPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipe, name: "tuiKeys" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipe, decorators: [{
            type: Pipe,
            args: [{ name: `tuiKeys` }]
        }] });

class TuiKeysPipeModule {
}
TuiKeysPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiKeysPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipeModule, declarations: [TuiKeysPipe], exports: [TuiKeysPipe] });
TuiKeysPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeysPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [TuiKeysPipe],
                    declarations: [TuiKeysPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiKeysPipe, TuiKeysPipeModule };
