// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-validator.js
import { __decorate } from 'tslib';
import * as i0 from '@angular/core';
import { forwardRef, Directive, Input, NgModule } from '@angular/core';
import { Validators, NG_VALIDATORS } from '@angular/forms';
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';
import { tuiDefaultProp } from '@taiga-ui/cdk/decorators';

class TuiValidatorDirective {
    constructor() {
        this.onChange = EMPTY_FUNCTION;
        this.tuiValidator = Validators.nullValidator;
    }
    validate(control) {
        return this.tuiValidator(control);
    }
    registerOnValidatorChange(onChange) {
        this.onChange = onChange;
    }
    ngOnChanges() {
        this.onChange();
    }
    ngOnDestroy() {
        this.tuiValidator = Validators.nullValidator;
        this.onChange();
    }
}
TuiValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValidatorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiValidatorDirective, selector: "[tuiValidator]", inputs: { tuiValidator: "tuiValidator" }, providers: [
        {
            provide: NG_VALIDATORS,
            useExisting: forwardRef(() => TuiValidatorDirective),
            multi: true,
        },
    ], usesOnChanges: true, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiValidatorDirective.prototype, "tuiValidator", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiValidator]',
                    providers: [
                        {
                            provide: NG_VALIDATORS,
                            useExisting: forwardRef(() => TuiValidatorDirective),
                            multi: true,
                        },
                    ],
                }]
        }], propDecorators: { tuiValidator: [{
                type: Input
            }] } });

class TuiValidatorModule {
}
TuiValidatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValidatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiValidatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValidatorModule, declarations: [TuiValidatorDirective], exports: [TuiValidatorDirective] });
TuiValidatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValidatorModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiValidatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiValidatorDirective],
                    exports: [TuiValidatorDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiValidatorDirective, TuiValidatorModule };
