// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-high-dpi.js
import * as i0 from '@angular/core';
import { ViewContainerRef, TemplateRef, Directive, Inject, NgModule } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';

/**
 * Only adds current content if user has High DPI display
 */
class TuiHighDpiDirective {
    constructor({ devicePixelRatio }, viewContainer, templateRef) {
        if (devicePixelRatio > 1) {
            viewContainer.createEmbeddedView(templateRef);
        }
    }
}
TuiHighDpiDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHighDpiDirective, deps: [{ token: WINDOW }, { token: ViewContainerRef }, { token: TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiHighDpiDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHighDpiDirective, selector: "[tuiHighDpi]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHighDpiDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHighDpi]',
                }]
        }], ctorParameters: function () { return [{ type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }, { type: i0.ViewContainerRef, decorators: [{
                    type: Inject,
                    args: [ViewContainerRef]
                }] }, { type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }]; } });

class TuiHighDpiModule {
}
TuiHighDpiModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHighDpiModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiHighDpiModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHighDpiModule, declarations: [TuiHighDpiDirective], exports: [TuiHighDpiDirective] });
TuiHighDpiModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHighDpiModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHighDpiModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiHighDpiDirective],
                    exports: [TuiHighDpiDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiHighDpiDirective, TuiHighDpiModule };
