// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-pan.js
import * as i0 from '@angular/core';
import { Directive, Inject, NgModule } from '@angular/core';
import { TuiPanService } from '@taiga-ui/cdk/services';
import * as i1 from 'rxjs';

class TuiPanDirective {
    constructor(tuiPan) {
        this.tuiPan = tuiPan;
    }
}
TuiPanDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanDirective, deps: [{ token: TuiPanService }], target: i0.ɵɵFactoryTarget.Directive });
TuiPanDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPanDirective, selector: "[tuiPan]", outputs: { tuiPan: "tuiPan" }, providers: [TuiPanService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPan]',
                    outputs: ['tuiPan'],
                    providers: [TuiPanService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiPanService]
                }] }]; } });

/**
 * @experimental
 */
class TuiPanModule {
}
TuiPanModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPanModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanModule, declarations: [TuiPanDirective], exports: [TuiPanDirective] });
TuiPanModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPanModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiPanDirective],
                    exports: [TuiPanDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPanDirective, TuiPanModule };
