// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-pipes-is-present.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';
import { tuiIsPresent } from '@taiga-ui/cdk/utils';

class TuiIsPresentPipe {
    transform(value) {
        return tuiIsPresent(value);
    }
}
TuiIsPresentPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiIsPresentPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipe, name: "tuiIsPresent" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipe, decorators: [{
            type: Pipe,
            args: [{ name: `tuiIsPresent` }]
        }] });

class TuiIsPresentPipeModule {
}
TuiIsPresentPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiIsPresentPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipeModule, declarations: [TuiIsPresentPipe], exports: [TuiIsPresentPipe] });
TuiIsPresentPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsPresentPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiIsPresentPipe],
                    exports: [TuiIsPresentPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiIsPresentPipe, TuiIsPresentPipeModule };
