// @taiga-ui/cdk/esm2015/directives/zoom/zoom.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiZoomService } from '@taiga-ui/cdk/services';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiZoomDirective {
    constructor(tuiZoom) {
        this.tuiZoom = tuiZoom;
    }
}
TuiZoomDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiZoomDirective, deps: [{ token: TuiZoomService }], target: i0.ɵɵFactoryTarget.Directive });
TuiZoomDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiZoomDirective, selector: "[tuiZoom]", outputs: { tuiZoom: "tuiZoom" }, host: { properties: { "style.touch-action": "\"none\"" } }, providers: [TuiZoomService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiZoomDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiZoom]',
                    outputs: ['tuiZoom'],
                    providers: [TuiZoomService],
                    host: {
                        '[style.touch-action]': '"none"',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiZoomService]
                }] }]; } });
