// @taiga-ui/kit/esm2015/components/pdf-viewer/pdf-viewer.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { tuiAsDialog } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPdfViewerComponent } from './pdf-viewer.component';
import { TuiPdfViewerDirective } from './pdf-viewer.directive';
import { TuiPdfViewerService } from './pdf-viewer.service';
import * as i0 from "@angular/core";
export class TuiPdfViewerModule {
}
TuiPdfViewerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPdfViewerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPdfViewerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPdfViewerModule, declarations: [TuiPdfViewerComponent, TuiPdfViewerDirective], imports: [CommonModule, TuiButtonModule, PolymorpheusModule], exports: [TuiPdfViewerComponent, TuiPdfViewerDirective] });
TuiPdfViewerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPdfViewerModule, providers: [tuiAsDialog(TuiPdfViewerService)], imports: [[CommonModule, TuiButtonModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPdfViewerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiButtonModule, PolymorpheusModule],
                    declarations: [TuiPdfViewerComponent, TuiPdfViewerDirective],
                    exports: [TuiPdfViewerComponent, TuiPdfViewerDirective],
                    providers: [tuiAsDialog(TuiPdfViewerService)],
                }]
        }] });
