// @taiga-ui/kit/esm2015/components/pdf-viewer/pdf-viewer.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiDialogDirective, AbstractTuiDialogService } from '@taiga-ui/cdk';
import { TuiPdfViewerService } from './pdf-viewer.service';
import * as i0 from "@angular/core";
export class TuiPdfViewerDirective extends AbstractTuiDialogDirective {
}
TuiPdfViewerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPdfViewerDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiPdfViewerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPdfViewerDirective, selector: "ng-template[tuiPdfViewer]", inputs: { options: ["tuiPdfViewerOptions", "options"], open: ["tuiPdfViewer", "open"] }, outputs: { openChange: "tuiPdfViewerChange" }, providers: [
        {
            provide: AbstractTuiDialogService,
            useExisting: TuiPdfViewerService,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPdfViewerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiPdfViewer]',
                    providers: [
                        {
                            provide: AbstractTuiDialogService,
                            useExisting: TuiPdfViewerService,
                        },
                    ],
                    inputs: ['options: tuiPdfViewerOptions', 'open: tuiPdfViewer'],
                    outputs: ['openChange: tuiPdfViewerChange'],
                }]
        }] });
