// @taiga-ui/kit/esm2015/components/breadcrumbs/breadcrumbs.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiItemDirective, TuiItemModule } from '@taiga-ui/cdk';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiBreadcrumbsComponent } from './breadcrumbs.component';
import * as i0 from "@angular/core";
export class TuiBreadcrumbsModule {
}
TuiBreadcrumbsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBreadcrumbsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, declarations: [TuiBreadcrumbsComponent], imports: [CommonModule, TuiSvgModule, TuiItemModule], exports: [TuiBreadcrumbsComponent, TuiItemDirective] });
TuiBreadcrumbsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, imports: [[CommonModule, TuiSvgModule, TuiItemModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreadcrumbsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiItemModule],
                    declarations: [TuiBreadcrumbsComponent],
                    exports: [TuiBreadcrumbsComponent, TuiItemDirective],
                }]
        }] });
