// @taiga-ui/cdk/esm2015/directives/resize/resize.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiDestroyService, TuiResizeService } from '@taiga-ui/cdk/services';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiResizeDirective {
    constructor(tuiResize) {
        this.tuiResize = tuiResize;
    }
}
TuiResizeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResizeDirective, deps: [{ token: TuiResizeService }], target: i0.ɵɵFactoryTarget.Directive });
TuiResizeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiResizeDirective, selector: "[tuiResize]", outputs: { tuiResize: "tuiResize" }, providers: [TuiDestroyService, TuiResizeService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResizeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiResize]',
                    outputs: ['tuiResize'],
                    providers: [TuiDestroyService, TuiResizeService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }]; } });
