// @taiga-ui/kit/esm2015/components/arrow/arrow-options.js
import { InjectionToken } from '@angular/core';
// TODO: find the best way for prevent cycle
// eslint-disable-next-line import/no-cycle
import { TUI_ARROW } from './arrow.component';
/** Default values for arrow options */
export const TUI_ARROW_DEFAULT_OPTIONS = {
    iconSmall: `tuiIconChevronDown`,
    iconLarge: `tuiIconChevronDownLarge`,
};
/**
 * Default parameters for arrow component
 */
export const TUI_ARROW_OPTIONS = new InjectionToken(`[TUI_ARROW_OPTIONS]`, {
    factory: () => TUI_ARROW_DEFAULT_OPTIONS,
});
export const tuiArrowOptionsProvider = (options) => ({
    provide: TUI_ARROW_OPTIONS,
    useValue: Object.assign(Object.assign({}, TUI_ARROW_DEFAULT_OPTIONS), options),
});
/**
 * Type of icon in dropdowns for interactive or disable mode
 */
export const TUI_ARROW_MODE = new InjectionToken(`[TUI_ARROW_MODE]`, {
    factory: () => ({
        interactive: TUI_ARROW,
        disabled: TUI_ARROW,
    }),
});
export const tuiArrowModeProvider = (options) => ({
    provide: TUI_ARROW_MODE,
    useValue: Object.assign({ interactive: TUI_ARROW, disabled: TUI_ARROW }, options),
});
