// @taiga-ui/kit/esm2015/components/arrow/arrow.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiArrowComponent } from './arrow.component';
import * as i0 from "@angular/core";
export class TuiArrowModule {
}
TuiArrowModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArrowModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiArrowModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArrowModule, declarations: [TuiArrowComponent], imports: [CommonModule, TuiSvgModule, PolymorpheusModule], exports: [TuiArrowComponent] });
TuiArrowModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArrowModule, imports: [[CommonModule, TuiSvgModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArrowModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, PolymorpheusModule],
                    declarations: [TuiArrowComponent],
                    exports: [TuiArrowComponent],
                }]
        }] });
