// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-appearance.directive.js
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { AbstractTuiController } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
// TODO: rename to TUI_TEXTFIELD_APPEARANCE in v4
export const TUI_TEXTFIELD_APPEARANCE_DIRECTIVE = new InjectionToken('[TUI_TEXTFIELD_APPEARANCE_DIRECTIVE]', {
    factory: () => new TuiTextfieldAppearanceDirective(),
});
export class TuiTextfieldAppearanceDirective extends AbstractTuiController {
    constructor() {
        super(...arguments);
        // it's an empty string by default for backward compatibility
        // (see comment https://github.com/Tinkoff/taiga-ui/pull/3007#issuecomment-1315179508)
        this.appearance = '';
    }
}
TuiTextfieldAppearanceDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextfieldAppearanceDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiTextfieldAppearanceDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTextfieldAppearanceDirective, selector: "[tuiTextfieldAppearance]", inputs: { appearance: ["tuiTextfieldAppearance", "appearance"] }, providers: [
        {
            provide: TUI_TEXTFIELD_APPEARANCE_DIRECTIVE,
            useExisting: forwardRef(() => TuiTextfieldAppearanceDirective),
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextfieldAppearanceDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTextfieldAppearance]',
                    providers: [
                        {
                            provide: TUI_TEXTFIELD_APPEARANCE_DIRECTIVE,
                            useExisting: forwardRef(() => TuiTextfieldAppearanceDirective),
                        },
                    ],
                }]
        }], propDecorators: { appearance: [{
                type: Input,
                args: ['tuiTextfieldAppearance']
            }] } });
