// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-postfix.directive.js
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { AbstractTuiController } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export const TUI_TEXTFIELD_POSTFIX = new InjectionToken('[TUI_TEXTFIELD_POSTFIX]', {
    factory: () => new TuiTextfieldPostfixDirective(),
});
export class TuiTextfieldPostfixDirective extends AbstractTuiController {
    constructor() {
        super(...arguments);
        this.postfix = '';
    }
}
TuiTextfieldPostfixDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextfieldPostfixDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiTextfieldPostfixDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTextfieldPostfixDirective, selector: "[tuiTextfieldPostfix]", inputs: { postfix: ["tuiTextfieldPostfix", "postfix"] }, providers: [
        {
            provide: TUI_TEXTFIELD_POSTFIX,
            useExisting: forwardRef(() => TuiTextfieldPostfixDirective),
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextfieldPostfixDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTextfieldPostfix]',
                    providers: [
                        {
                            provide: TUI_TEXTFIELD_POSTFIX,
                            useExisting: forwardRef(() => TuiTextfieldPostfixDirective),
                        },
                    ],
                }]
        }], propDecorators: { postfix: [{
                type: Input,
                args: ['tuiTextfieldPostfix']
            }] } });
