// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-custom-content.directive.js
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { AbstractTuiController } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export const TUI_TEXTFIELD_CUSTOM_CONTENT = new InjectionToken('[TUI_TEXTFIELD_CUSTOM_CONTENT]', {
    factory: () => new TuiTextfieldCustomContentDirective(),
});
export class TuiTextfieldCustomContentDirective extends AbstractTuiController {
}
TuiTextfieldCustomContentDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextfieldCustomContentDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiTextfieldCustomContentDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTextfieldCustomContentDirective, selector: "[tuiTextfieldCustomContent]", inputs: { customContent: ["tuiTextfieldCustomContent", "customContent"] }, providers: [
        {
            provide: TUI_TEXTFIELD_CUSTOM_CONTENT,
            useExisting: forwardRef(() => TuiTextfieldCustomContentDirective),
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextfieldCustomContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTextfieldCustomContent]',
                    providers: [
                        {
                            provide: TUI_TEXTFIELD_CUSTOM_CONTENT,
                            useExisting: forwardRef(() => TuiTextfieldCustomContentDirective),
                        },
                    ],
                }]
        }], propDecorators: { customContent: [{
                type: Input,
                args: ['tuiTextfieldCustomContent']
            }] } });
