// @taiga-ui/kit/esm2015/directives/project-class/project-class.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
/**
 * A directive for projecting classes from nested children to host
 */
export class TuiProjectClassDirective {
    constructor(el) {
        this.el = el;
        this.classNames = [];
    }
    ngAfterViewChecked() {
        this.classNames.forEach(className => {
            const hostElement = this.el.nativeElement;
            hostElement.classList.toggle(className, !!hostElement.querySelector(`.${className}`));
        });
    }
}
TuiProjectClassDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassDirective, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiProjectClassDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiProjectClassDirective, selector: "[tuiProjectClass]", inputs: { classNames: ["tuiProjectClass", "classNames"] }, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiProjectClassDirective.prototype, "classNames", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProjectClassDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiProjectClass]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { classNames: [{
                type: Input,
                args: ['tuiProjectClass']
            }] } });
