// @taiga-ui/kit/esm2015/components/range/range.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiFocusableModule } from '@taiga-ui/cdk';
import { TuiSliderModule } from '@taiga-ui/kit/components/slider';
import { TuiRangeComponent } from './range.component';
import { TuiRangeChangeDirective } from './range-change.directive';
import * as i0 from "@angular/core";
export class TuiRangeModule {
}
TuiRangeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRangeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRangeModule, declarations: [TuiRangeComponent, TuiRangeChangeDirective], imports: [CommonModule, TuiFocusableModule, TuiSliderModule, FormsModule], exports: [TuiRangeComponent, TuiRangeChangeDirective] });
TuiRangeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRangeModule, imports: [[CommonModule, TuiFocusableModule, TuiSliderModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiFocusableModule, TuiSliderModule, FormsModule],
                    declarations: [TuiRangeComponent, TuiRangeChangeDirective],
                    exports: [TuiRangeComponent, TuiRangeChangeDirective],
                }]
        }] });
