// @taiga-ui/core/esm2015/components/primitive-spin-button/primitive-spin-button.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiPreventDefaultModule } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core/components/button';
import { TuiPrimitiveSpinButtonComponent } from './primitive-spin-button.component';
import * as i0 from "@angular/core";
export class TuiPrimitiveSpinButtonModule {
}
TuiPrimitiveSpinButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveSpinButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveSpinButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveSpinButtonModule, declarations: [TuiPrimitiveSpinButtonComponent], imports: [CommonModule, TuiPreventDefaultModule, TuiButtonModule], exports: [TuiPrimitiveSpinButtonComponent] });
TuiPrimitiveSpinButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveSpinButtonModule, imports: [[CommonModule, TuiPreventDefaultModule, TuiButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveSpinButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiPreventDefaultModule, TuiButtonModule],
                    declarations: [TuiPrimitiveSpinButtonComponent],
                    exports: [TuiPrimitiveSpinButtonComponent],
                }]
        }] });
