// @taiga-ui/kit/esm2015/components/input-copy/input-copy.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiAlertModule, TuiHintModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldComponent, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputCopyComponent } from './input-copy.component';
import { TuiInputCopyDirective } from './input-copy.directive';
import * as i0 from "@angular/core";
export class TuiInputCopyModule {
}
TuiInputCopyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCopyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputCopyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCopyModule, declarations: [TuiInputCopyComponent, TuiInputCopyDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiWrapperModule,
        TuiSvgModule,
        TuiHintModule,
        TuiPrimitiveTextfieldModule,
        TuiAlertModule,
        TuiTextfieldControllerModule], exports: [TuiInputCopyComponent, TuiInputCopyDirective, TuiTextfieldComponent] });
TuiInputCopyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCopyModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiWrapperModule,
            TuiSvgModule,
            TuiHintModule,
            TuiPrimitiveTextfieldModule,
            TuiAlertModule,
            TuiTextfieldControllerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCopyModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiWrapperModule,
                        TuiSvgModule,
                        TuiHintModule,
                        TuiPrimitiveTextfieldModule,
                        TuiAlertModule,
                        TuiTextfieldControllerModule,
                    ],
                    declarations: [TuiInputCopyComponent, TuiInputCopyDirective],
                    exports: [TuiInputCopyComponent, TuiInputCopyDirective, TuiTextfieldComponent],
                }]
        }] });
