// @taiga-ui/core/esm2015/pipes/format-date/format-date.pipe.js
import { Inject, Pipe } from '@angular/core';
import { TuiFormatDateService } from '@taiga-ui/core/services';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/services";
export class TuiFormatDatePipe {
    constructor(service) {
        this.service = service;
    }
    transform(timestampOrDate) {
        return this.service.format(timestampOrDate.valueOf());
    }
}
TuiFormatDatePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatDatePipe, deps: [{ token: TuiFormatDateService }], target: i0.ɵɵFactoryTarget.Pipe });
TuiFormatDatePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatDatePipe, name: "tuiFormatDate" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFormatDatePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: `tuiFormatDate`,
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiFormatDateService, decorators: [{
                    type: Inject,
                    args: [TuiFormatDateService]
                }] }]; } });
