// @taiga-ui/cdk/esm2015/components/alert-host/alert-host.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiMapperPipeModule } from '@taiga-ui/cdk/pipes';
import { TuiAlertHostComponent } from './alert-host.component';
import * as i0 from "@angular/core";
export class TuiAlertHostModule {
}
TuiAlertHostModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertHostModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAlertHostModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertHostModule, declarations: [TuiAlertHostComponent], imports: [CommonModule, TuiMapperPipeModule], exports: [TuiAlertHostComponent] });
TuiAlertHostModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertHostModule, imports: [[CommonModule, TuiMapperPipeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertHostModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiMapperPipeModule],
                    declarations: [TuiAlertHostComponent],
                    exports: [TuiAlertHostComponent],
                }]
        }] });
