// @taiga-ui/kit/esm2015/components/accordion/accordion.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusableModule, TuiFocusedModule, TuiFocusVisibleModule } from '@taiga-ui/cdk';
import { TuiExpandModule, TuiGroupModule, TuiSvgModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiAccordionComponent } from './accordion.component';
import { TuiAccordionItemComponent } from './accordion-item/accordion-item.component';
import { TuiAccordionItemContentDirective } from './accordion-item/accordion-item-content.directive';
import { TuiAccordionItemEagerContentDirective } from './accordion-item/accordion-item-eager-content.directive';
import * as i0 from "@angular/core";
export class TuiAccordionModule {
}
TuiAccordionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAccordionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAccordionModule, declarations: [TuiAccordionComponent,
        TuiAccordionItemComponent,
        TuiAccordionItemContentDirective,
        TuiAccordionItemEagerContentDirective], imports: [CommonModule,
        TuiFocusedModule,
        TuiFocusVisibleModule,
        TuiFocusableModule,
        PolymorpheusModule,
        TuiSvgModule,
        TuiGroupModule,
        TuiExpandModule], exports: [TuiAccordionComponent,
        TuiAccordionItemComponent,
        TuiAccordionItemContentDirective,
        TuiAccordionItemEagerContentDirective] });
TuiAccordionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAccordionModule, imports: [[
            CommonModule,
            TuiFocusedModule,
            TuiFocusVisibleModule,
            TuiFocusableModule,
            PolymorpheusModule,
            TuiSvgModule,
            TuiGroupModule,
            TuiExpandModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiFocusedModule,
                        TuiFocusVisibleModule,
                        TuiFocusableModule,
                        PolymorpheusModule,
                        TuiSvgModule,
                        TuiGroupModule,
                        TuiExpandModule,
                    ],
                    declarations: [
                        TuiAccordionComponent,
                        TuiAccordionItemComponent,
                        TuiAccordionItemContentDirective,
                        TuiAccordionItemEagerContentDirective,
                    ],
                    exports: [
                        TuiAccordionComponent,
                        TuiAccordionItemComponent,
                        TuiAccordionItemContentDirective,
                        TuiAccordionItemEagerContentDirective,
                    ],
                }]
        }] });
