// @taiga-ui/kit/esm2015/components/tabs/tabs.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusableModule, TuiItemDirective, TuiItemModule } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiSvgModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTabComponent } from './tab/tab.component';
import { TuiTabsDirective } from './tabs.directive';
import { TuiTabsComponent } from './tabs/tabs.component';
import { TuiTabsVerticalComponent } from './tabs-vertical/tabs-vertical.component';
import { TuiTabsWithMoreComponent } from './tabs-with-more/tabs-with-more.component';
import { TuiUnderlineComponent } from './underline/underline.component';
import * as i0 from "@angular/core";
export class TuiTabsModule {
}
TuiTabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabsModule, declarations: [TuiTabsWithMoreComponent,
        TuiTabsComponent,
        TuiTabsDirective,
        TuiTabsVerticalComponent,
        TuiTabComponent,
        TuiUnderlineComponent], imports: [CommonModule,
        PolymorpheusModule,
        TuiHostedDropdownModule,
        TuiSvgModule,
        TuiFocusableModule,
        TuiItemModule], exports: [TuiTabsWithMoreComponent,
        TuiTabsComponent,
        TuiTabsDirective,
        TuiTabsVerticalComponent,
        TuiTabComponent,
        TuiItemDirective] });
TuiTabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabsModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiHostedDropdownModule,
            TuiSvgModule,
            TuiFocusableModule,
            TuiItemModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiHostedDropdownModule,
                        TuiSvgModule,
                        TuiFocusableModule,
                        TuiItemModule,
                    ],
                    declarations: [
                        TuiTabsWithMoreComponent,
                        TuiTabsComponent,
                        TuiTabsDirective,
                        TuiTabsVerticalComponent,
                        TuiTabComponent,
                        TuiUnderlineComponent,
                    ],
                    exports: [
                        TuiTabsWithMoreComponent,
                        TuiTabsComponent,
                        TuiTabsDirective,
                        TuiTabsVerticalComponent,
                        TuiTabComponent,
                        TuiItemDirective,
                    ],
                }]
        }] });
