// @taiga-ui/kit/esm2015/components/checkbox-block/checkbox-block.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiHoveredModule, TuiPressedModule } from '@taiga-ui/cdk';
import { TuiWrapperModule } from '@taiga-ui/core';
import { TuiCheckboxModule } from '@taiga-ui/kit/components/checkbox';
import { TuiCheckboxBlockComponent } from './checkbox-block.component';
import * as i0 from "@angular/core";
export class TuiCheckboxBlockModule {
}
TuiCheckboxBlockModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxBlockModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCheckboxBlockModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxBlockModule, declarations: [TuiCheckboxBlockComponent], imports: [FormsModule,
        TuiHoveredModule,
        TuiPressedModule,
        TuiCheckboxModule,
        TuiWrapperModule], exports: [TuiCheckboxBlockComponent] });
TuiCheckboxBlockModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxBlockModule, imports: [[
            FormsModule,
            TuiHoveredModule,
            TuiPressedModule,
            TuiCheckboxModule,
            TuiWrapperModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCheckboxBlockModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        FormsModule,
                        TuiHoveredModule,
                        TuiPressedModule,
                        TuiCheckboxModule,
                        TuiWrapperModule,
                    ],
                    declarations: [TuiCheckboxBlockComponent],
                    exports: [TuiCheckboxBlockComponent],
                }]
        }] });
