// @taiga-ui/cdk/esm2015/observables/drag-and-drop-from.js
import { TuiOwnerDocumentException } from '@taiga-ui/cdk/exceptions';
import { tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { concat, merge } from 'rxjs';
import { endWith, map, repeat, take, takeWhile } from 'rxjs/operators';
import { tuiMouseDragFinishFrom } from './mouse-drag-finish-from';
import { tuiTypedFromEvent } from './typed-from-event';
export var TuiDragStage;
(function (TuiDragStage) {
    TuiDragStage[TuiDragStage["Start"] = 0] = "Start";
    TuiDragStage[TuiDragStage["Continues"] = 1] = "Continues";
    TuiDragStage[TuiDragStage["End"] = 2] = "End";
})(TuiDragStage || (TuiDragStage = {}));
export class TuiDragState {
    constructor(stage, event) {
        this.stage = stage;
        this.event = event;
    }
}
export function tuiDragAndDropFrom(element) {
    const { ownerDocument } = element;
    if (!ownerDocument) {
        throw new TuiOwnerDocumentException();
    }
    return concat(tuiTypedFromEvent(element, `mousedown`).pipe(take(1), map(event => new TuiDragState(0 /* Start */, event))), merge(tuiTypedFromEvent(ownerDocument, `mousemove`).pipe(map(event => new TuiDragState(1 /* Continues */, event))), tuiMouseDragFinishFrom(ownerDocument).pipe(take(1), map(event => new TuiDragState(2 /* End */, event)), endWith(null))).pipe(takeWhile(tuiIsPresent))).pipe(repeat());
}
