// @taiga-ui/kit/esm2015/components/text-area/text-area.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiFocusableModule } from '@taiga-ui/cdk';
import { TuiPrimitiveTextfieldModule, TuiScrollbarModule, TuiSvgModule, TuiTextfieldComponent, TuiTooltipModule, TuiWrapperModule, } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTextAreaComponent } from './text-area.component';
import { TuiTextAreaDirective } from './text-area.directive';
import * as i0 from "@angular/core";
export class TuiTextAreaModule {
}
TuiTextAreaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextAreaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTextAreaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextAreaModule, declarations: [TuiTextAreaComponent, TuiTextAreaDirective], imports: [CommonModule,
        FormsModule,
        TuiFocusableModule,
        TuiScrollbarModule,
        TuiTooltipModule,
        TuiWrapperModule,
        TuiSvgModule,
        TuiPrimitiveTextfieldModule,
        PolymorpheusModule], exports: [TuiTextAreaComponent, TuiTextAreaDirective, TuiTextfieldComponent] });
TuiTextAreaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextAreaModule, imports: [[
            CommonModule,
            FormsModule,
            TuiFocusableModule,
            TuiScrollbarModule,
            TuiTooltipModule,
            TuiWrapperModule,
            TuiSvgModule,
            TuiPrimitiveTextfieldModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextAreaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiFocusableModule,
                        TuiScrollbarModule,
                        TuiTooltipModule,
                        TuiWrapperModule,
                        TuiSvgModule,
                        TuiPrimitiveTextfieldModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiTextAreaComponent, TuiTextAreaDirective],
                    exports: [TuiTextAreaComponent, TuiTextAreaDirective, TuiTextfieldComponent],
                }]
        }] });
