// @taiga-ui/kit/esm2015/components/tabs/tab/tab.providers.js
import { ElementRef, InjectionToken, Optional } from '@angular/core';
import { RouterLinkActive } from '@angular/router';
import { MutationObserverService } from '@ng-web-apis/mutation-observer';
import { TuiDestroyService, TuiFocusVisibleService, tuiTypedFromEvent, } from '@taiga-ui/cdk';
import { MODE_PROVIDER, TuiRouterLinkActiveService } from '@taiga-ui/core';
import { EMPTY, identity, merge } from 'rxjs';
import { filter, map } from 'rxjs/operators';
/**
 * Stream of tab activation events
 */
export const TUI_TAB_EVENT = new InjectionToken(`[TUI_TAB_EVENT]`);
export const TUI_TAB_ACTIVATE = `tui-tab-activate`;
export const TUI_TAB_PROVIDERS = [
    TuiDestroyService,
    TuiFocusVisibleService,
    TuiRouterLinkActiveService,
    {
        provide: TUI_TAB_EVENT,
        deps: [
            ElementRef,
            TuiRouterLinkActiveService,
            [new Optional(), MutationObserverService],
            [new Optional(), RouterLinkActive],
        ],
        useFactory: ({ nativeElement }, routerLinkActiveService, mutationObserverService, routerLinkActive) => {
            const mutationObserver = routerLinkActive && mutationObserverService
                ? mutationObserverService.pipe(filter(() => routerLinkActive.isActive))
                : EMPTY;
            return merge(mutationObserver, routerLinkActiveService.pipe(filter(identity)), nativeElement.matches(`button`)
                ? tuiTypedFromEvent(nativeElement, `click`)
                : EMPTY).pipe(map(() => nativeElement.dispatchEvent(new CustomEvent(TUI_TAB_ACTIVATE, { bubbles: true }))));
        },
    },
    MODE_PROVIDER,
];
