// @taiga-ui/core/esm2015/components/root/root.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiAlertHostModule, TuiDialogHostModule, TuiDragModule, TuiDropdownHostModule, } from '@taiga-ui/cdk';
import { TuiHintsHostModule } from '@taiga-ui/core/components/hints-host';
import { TuiScrollControlsModule } from '@taiga-ui/core/components/scroll-controls';
import { TuiSvgDefsHostModule } from '@taiga-ui/core/internal/svg-defs-host';
import { EventPluginsModule } from '@tinkoff/ng-event-plugins';
import { TuiRootComponent } from './root.component';
import * as i0 from "@angular/core";
export class TuiRootModule {
}
TuiRootModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRootModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRootModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRootModule, declarations: [TuiRootComponent], imports: [CommonModule,
        EventPluginsModule,
        TuiDragModule,
        TuiDropdownHostModule,
        TuiSvgDefsHostModule,
        TuiHintsHostModule,
        TuiDialogHostModule,
        TuiAlertHostModule,
        TuiScrollControlsModule], exports: [TuiRootComponent] });
TuiRootModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRootModule, imports: [[
            CommonModule,
            EventPluginsModule,
            TuiDragModule,
            TuiDropdownHostModule,
            TuiSvgDefsHostModule,
            TuiHintsHostModule,
            TuiDialogHostModule,
            TuiAlertHostModule,
            TuiScrollControlsModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRootModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        EventPluginsModule,
                        TuiDragModule,
                        TuiDropdownHostModule,
                        TuiSvgDefsHostModule,
                        TuiHintsHostModule,
                        TuiDialogHostModule,
                        TuiAlertHostModule,
                        TuiScrollControlsModule,
                    ],
                    declarations: [TuiRootComponent],
                    exports: [TuiRootComponent],
                }]
        }] });
