// @taiga-ui/cdk/esm2015/directives/swipe/swipe.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiSwipeService } from '@taiga-ui/cdk/services';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiSwipeDirective {
    constructor(tuiSwipe) {
        this.tuiSwipe = tuiSwipe;
    }
}
TuiSwipeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeDirective, deps: [{ token: TuiSwipeService }], target: i0.ɵɵFactoryTarget.Directive });
TuiSwipeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSwipeDirective, selector: "[tuiSwipe]", outputs: { tuiSwipe: "tuiSwipe" }, providers: [TuiSwipeService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSwipe]',
                    outputs: ['tuiSwipe'],
                    providers: [TuiSwipeService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiSwipeService]
                }] }]; } });
