// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-item.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/item', ['exports', '@angular/core'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.item = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

        /**
         * Blank directive for queries via `@ContentChildren` / `@ViewChildren` / `querySelector`
         */
        var TuiItemDirective = /** @class */ (function () {
            function TuiItemDirective() {
            }
            return TuiItemDirective;
        }());
        TuiItemDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiItemDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiItemDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiItemDirective, selector: "[tuiItem]", ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiItemDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiItem]',
                        }]
                }] });

        var TuiItemModule = /** @class */ (function () {
            function TuiItemModule() {
            }
            return TuiItemModule;
        }());
        TuiItemModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiItemModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiItemModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiItemModule, declarations: [TuiItemDirective], exports: [TuiItemDirective] });
        TuiItemModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiItemModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiItemModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiItemDirective],
                            exports: [TuiItemDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiItemDirective = TuiItemDirective;
        exports.TuiItemModule = TuiItemModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
