// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-focusable.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/focusable', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.focusable = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    /**
     * Abstraction over `tabindex`
     */
    var TuiFocusableDirective = /** @class */ (function () {
        function TuiFocusableDirective() {
            /**
             * Element can be focused with keyboard
             */
            this.focusable = true;
        }
        Object.defineProperty(TuiFocusableDirective.prototype, "tabIndex", {
            get: function () {
                return this.focusable ? 0 : -1;
            },
            enumerable: false,
            configurable: true
        });
        return TuiFocusableDirective;
    }());
    TuiFocusableDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusableDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiFocusableDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiFocusableDirective, selector: "[tuiFocusable]", inputs: { focusable: ["tuiFocusable", "focusable"] }, host: { properties: { "tabIndex": "this.tabIndex" } }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusableDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiFocusable]',
                    }]
            }], propDecorators: { focusable: [{
                    type: i0.Input,
                    args: ['tuiFocusable']
                }], tabIndex: [{
                    type: i0.HostBinding,
                    args: ['tabIndex']
                }] } });

    var TuiFocusableModule = /** @class */ (function () {
        function TuiFocusableModule() {
        }
        return TuiFocusableModule;
    }());
    TuiFocusableModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusableModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFocusableModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusableModule, declarations: [TuiFocusableDirective], exports: [TuiFocusableDirective] });
    TuiFocusableModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusableModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusableModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFocusableDirective],
                        exports: [TuiFocusableDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiFocusableDirective = TuiFocusableDirective;
    exports.TuiFocusableModule = TuiFocusableModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
