// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-control.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/control', ['exports', '@angular/core', '@angular/forms'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.control = {}), global.ng.core, global.ng.forms));
})(this, (function (exports, i0, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiControlDirective = /** @class */ (function () {
        function TuiControlDirective(ngControl) {
            this.ngControl = ngControl;
        }
        Object.defineProperty(TuiControlDirective.prototype, "control", {
            get: function () {
                return this.ngControl.control;
            },
            enumerable: false,
            configurable: true
        });
        return TuiControlDirective;
    }());
    TuiControlDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiControlDirective, deps: [{ token: i1.NgControl }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiControlDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiControlDirective, selector: "[tuiControl]", exportAs: ["ngControl"], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiControlDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiControl]',
                        exportAs: 'ngControl',
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.NgControl, decorators: [{
                            type: i0.Inject,
                            args: [i1.NgControl]
                        }] }];
        } });

    var TuiControlModule = /** @class */ (function () {
        function TuiControlModule() {
        }
        return TuiControlModule;
    }());
    TuiControlModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiControlModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiControlModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiControlModule, declarations: [TuiControlDirective], exports: [TuiControlDirective] });
    TuiControlModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiControlModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiControlModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiControlDirective],
                        exports: [TuiControlDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiControlDirective = TuiControlDirective;
    exports.TuiControlModule = TuiControlModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
