// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-swipe.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/services'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/swipe', ['exports', '@angular/core', '@taiga-ui/cdk/services', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.swipe = {}), global.ng.core, global["taiga-ui"].cdk.services, global.rxjs));
})(this, (function (exports, i0, services, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiSwipeDirective = /** @class */ (function () {
        function TuiSwipeDirective(tuiSwipe) {
            this.tuiSwipe = tuiSwipe;
        }
        return TuiSwipeDirective;
    }());
    TuiSwipeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSwipeDirective, deps: [{ token: services.TuiSwipeService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSwipeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSwipeDirective, selector: "[tuiSwipe]", outputs: { tuiSwipe: "tuiSwipe" }, providers: [services.TuiSwipeService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSwipeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiSwipe]',
                        outputs: ['tuiSwipe'],
                        providers: [services.TuiSwipeService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [services.TuiSwipeService]
                        }] }];
        } });

    /**
     * @experimental
     */
    var TuiSwipeModule = /** @class */ (function () {
        function TuiSwipeModule() {
        }
        return TuiSwipeModule;
    }());
    TuiSwipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSwipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSwipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSwipeModule, declarations: [TuiSwipeDirective], exports: [TuiSwipeDirective] });
    TuiSwipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSwipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSwipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiSwipeDirective],
                        exports: [TuiSwipeDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiSwipeDirective = TuiSwipeDirective;
    exports.TuiSwipeModule = TuiSwipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
