// @taiga-ui/cdk/bundles/taiga-ui-cdk-pipes-is-present.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/pipes/is-present', ['exports', '@angular/core', '@taiga-ui/cdk/utils'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.pipes = global["taiga-ui"].cdk.pipes || {}, global["taiga-ui"].cdk.pipes["is-present"] = {}), global.ng.core, global["taiga-ui"].cdk.utils));
})(this, (function (exports, i0, utils) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiIsPresentPipe = /** @class */ (function () {
        function TuiIsPresentPipe() {
        }
        TuiIsPresentPipe.prototype.transform = function (value) {
            return utils.tuiIsPresent(value);
        };
        return TuiIsPresentPipe;
    }());
    TuiIsPresentPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiIsPresentPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipe, name: "tuiIsPresent" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: "tuiIsPresent" }]
            }] });

    var TuiIsPresentPipeModule = /** @class */ (function () {
        function TuiIsPresentPipeModule() {
        }
        return TuiIsPresentPipeModule;
    }());
    TuiIsPresentPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiIsPresentPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipeModule, declarations: [TuiIsPresentPipe], exports: [TuiIsPresentPipe] });
    TuiIsPresentPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsPresentPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiIsPresentPipe],
                        exports: [TuiIsPresentPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiIsPresentPipe = TuiIsPresentPipe;
    exports.TuiIsPresentPipeModule = TuiIsPresentPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
