// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-pan.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/services'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/pan', ['exports', '@angular/core', '@taiga-ui/cdk/services', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.pan = {}), global.ng.core, global["taiga-ui"].cdk.services, global.rxjs));
})(this, (function (exports, i0, services, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiPanDirective = /** @class */ (function () {
        function TuiPanDirective(tuiPan) {
            this.tuiPan = tuiPan;
        }
        return TuiPanDirective;
    }());
    TuiPanDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPanDirective, deps: [{ token: services.TuiPanService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPanDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPanDirective, selector: "[tuiPan]", outputs: { tuiPan: "tuiPan" }, providers: [services.TuiPanService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPanDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPan]',
                        outputs: ['tuiPan'],
                        providers: [services.TuiPanService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [services.TuiPanService]
                        }] }];
        } });

    /**
     * @experimental
     */
    var TuiPanModule = /** @class */ (function () {
        function TuiPanModule() {
        }
        return TuiPanModule;
    }());
    TuiPanModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPanModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPanModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPanModule, declarations: [TuiPanDirective], exports: [TuiPanDirective] });
    TuiPanModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPanModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPanModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiPanDirective],
                        exports: [TuiPanDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPanDirective = TuiPanDirective;
    exports.TuiPanModule = TuiPanModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
