// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-resize.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/services'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/resize', ['exports', '@angular/core', '@taiga-ui/cdk/services', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.resize = {}), global.ng.core, global["taiga-ui"].cdk.services, global.rxjs));
})(this, (function (exports, i0, services, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiResizeDirective = /** @class */ (function () {
        function TuiResizeDirective(tuiResize) {
            this.tuiResize = tuiResize;
        }
        return TuiResizeDirective;
    }());
    TuiResizeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiResizeDirective, deps: [{ token: services.TuiResizeService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiResizeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiResizeDirective, selector: "[tuiResize]", outputs: { tuiResize: "tuiResize" }, providers: [services.TuiDestroyService, services.TuiResizeService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiResizeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiResize]',
                        outputs: ['tuiResize'],
                        providers: [services.TuiDestroyService, services.TuiResizeService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [services.TuiResizeService]
                        }] }];
        } });

    /**
     * @experimental
     */
    var TuiResizeModule = /** @class */ (function () {
        function TuiResizeModule() {
        }
        return TuiResizeModule;
    }());
    TuiResizeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiResizeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiResizeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiResizeModule, declarations: [TuiResizeDirective], exports: [TuiResizeDirective] });
    TuiResizeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiResizeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiResizeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiResizeDirective],
                        exports: [TuiResizeDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiResizeDirective = TuiResizeDirective;
    exports.TuiResizeModule = TuiResizeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
