// @taiga-ui/cdk/bundles/taiga-ui-cdk-tokens.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk/observables'), require('@taiga-ui/cdk/utils'), require('rxjs'), require('rxjs/operators'), require('@angular/animations/browser'), require('@taiga-ui/cdk/constants')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/tokens', ['exports', '@angular/common', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk/observables', '@taiga-ui/cdk/utils', 'rxjs', 'rxjs/operators', '@angular/animations/browser', '@taiga-ui/cdk/constants'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.tokens = {}), global.ng.common, global.ng.core, global.common, global["taiga-ui"].cdk.observables, global["taiga-ui"].cdk.utils, global.rxjs, global.rxjs.operators, global.ng.animations.browser, global["taiga-ui"].cdk.constants));
})(this, (function (exports, common$1, core, common, observables, utils, rxjs, operators, browser, constants) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    /**
     * Element currently being removed by AnimationEngine
     */
    var TUI_REMOVED_ELEMENT = new core.InjectionToken("[TUI_REMOVED_ELEMENT]", {
        factory: function () {
            var stub = { onRemovalComplete: function () { } };
            var element$ = new rxjs.BehaviorSubject(null);
            var engine = core.inject(browser["ɵAnimationEngine"], core.InjectFlags.Optional) || stub;
            var _a = engine.onRemovalComplete, onRemovalComplete = _a === void 0 ? stub.onRemovalComplete : _a;
            engine.onRemovalComplete = function (element, context) {
                element$.next(element);
                onRemovalComplete(element, context);
            };
            return element$.pipe(operators.switchMap(function (element) { return rxjs.timer(0).pipe(operators.map(function () { return null; }), operators.startWith(element)); }), operators.share());
        },
    });

    /**
     * Active element on the document for ActiveZone
     */
    var TUI_ACTIVE_ELEMENT = new core.InjectionToken("[TUI_ACTIVE_ELEMENT]", {
        factory: function () {
            var removedElement$ = core.inject(TUI_REMOVED_ELEMENT);
            var win = core.inject(common.WINDOW);
            var doc = core.inject(common$1.DOCUMENT);
            var focusout$ = observables.tuiTypedFromEvent(win, "focusout");
            var focusin$ = observables.tuiTypedFromEvent(win, "focusin");
            var blur$ = observables.tuiTypedFromEvent(win, "blur");
            var mousedown$ = observables.tuiTypedFromEvent(win, "mousedown");
            var mouseup$ = observables.tuiTypedFromEvent(win, "mouseup");
            return rxjs.merge(focusout$.pipe(
            // eslint-disable-next-line rxjs/no-unsafe-takeuntil
            operators.takeUntil(mousedown$), 
            /**
             * TODO: replace to
             * repeat({delay: () => mouseup$})
             * in RxJS 7
             */
            // eslint-disable-next-line rxjs/no-ignored-notifier
            operators.repeatWhen(function () { return mouseup$; }), operators.withLatestFrom(removedElement$), operators.filter(function (_a) {
                var _b = __read(_a, 2), event = _b[0], removedElement = _b[1];
                return isValidFocusout(utils.tuiGetActualTarget(event), removedElement);
            }), operators.map(function (_a) {
                var _b = __read(_a, 1), relatedTarget = _b[0].relatedTarget;
                return relatedTarget;
            })), blur$.pipe(operators.map(function () { return doc.activeElement; }), operators.filter(function (element) { return !!(element === null || element === void 0 ? void 0 : element.matches("iframe")); })), focusin$.pipe(operators.switchMap(function (event) {
                var target = utils.tuiGetActualTarget(event);
                var root = utils.tuiGetDocumentOrShadowRoot(target);
                return root === doc
                    ? rxjs.of(target)
                    : shadowRootActiveElement(root).pipe(operators.startWith(target));
            })), mousedown$.pipe(operators.switchMap(function (event) {
                var actualTargetInCurrentTime = utils.tuiGetActualTarget(event);
                return !doc.activeElement || doc.activeElement === doc.body
                    ? rxjs.of(actualTargetInCurrentTime)
                    : focusout$.pipe(operators.take(1), operators.map(
                    /**
                     * Do not use `map(() => tuiGetActualTarget(event))`
                     * because we have different result in runtime
                     */
                    function () { return actualTargetInCurrentTime; }), operators.takeUntil(rxjs.timer(0)));
            }))).pipe(operators.distinctUntilChanged(), operators.share());
        },
    });
    // Checks if focusout event should be considered leaving active zone
    function isValidFocusout(target, removedElement) {
        if (removedElement === void 0) { removedElement = null; }
        return (
        // Not due to switching tabs/going to DevTools
        utils.tuiGetDocumentOrShadowRoot(target).activeElement !== target &&
            // Not due to button/input becoming disabled
            !target.disabled &&
            // Not due to element being removed from DOM
            !(removedElement === null || removedElement === void 0 ? void 0 : removedElement.contains(target)));
    }
    function shadowRootActiveElement(root) {
        return rxjs.merge(observables.tuiTypedFromEvent(root, "focusin").pipe(operators.map(function (_a) {
            var target = _a.target;
            return target;
        })), observables.tuiTypedFromEvent(root, "focusout").pipe(operators.filter(function (_a) {
            var target = _a.target, relatedTarget = _a.relatedTarget;
            return !!relatedTarget && isValidFocusout(target);
        }), operators.map(function (_a) {
            var relatedTarget = _a.relatedTarget;
            return relatedTarget;
        })));
    }

    /**
     * A stream of alerts
     */
    var TUI_ALERTS = new core.InjectionToken("[TUI_ALERTS]", {
        factory: function () { return []; },
    });
    function tuiAsAlerts(useExisting) {
        return {
            provide: TUI_ALERTS,
            multi: true,
            useExisting: useExisting,
        };
    }

    /**
     * A Renderer2 for global services
     * https://github.com/angular/angular/issues/17824#issuecomment-353239017
     * @deprecated unnecessary
     * TODO: remove in 4.0
     */
    var TUI_DEFAULT_RENDERER = new core.InjectionToken("[TUI_DEFAULT_RENDERER]", {
        // @ts-ignore
        factory: function () { return core.inject(core.RendererFactory2).createRenderer(null, null); },
    });

    /**
     * A stream of dialogs
     */
    var TUI_DIALOGS = new core.InjectionToken("[TUI_DIALOGS]", {
        factory: function () { return []; },
    });
    function tuiAsDialog(useExisting) {
        return {
            provide: TUI_DIALOGS,
            multi: true,
            useExisting: useExisting,
        };
    }

    var TUI_ENSURE_BASE_HREF = new core.InjectionToken("[TUI_ENSURE_BASE_HREF]", {
        factory: function () {
            var link = core.inject(common$1.DOCUMENT).createElement("a");
            link.href = "";
            return link.pathname;
        },
    });

    /**
     * A component that can be focused
     */
    var TUI_FOCUSABLE_ITEM_ACCESSOR = new core.InjectionToken("[TUI_FOCUSABLE_ITEM_ACCESSOR]");
    function tuiAsFocusableItemAccessor(useExisting) {
        return {
            provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
            useExisting: useExisting,
        };
    }

    /**
     * iOS browser detection
     */
    var TUI_IS_IOS = new core.InjectionToken("[TUI_IS_IOS]", {
        factory: function () { return utils.tuiIsIos(core.inject(common.NAVIGATOR)); },
    });

    // https://stackoverflow.com/a/11381730/2706426 http://detectmobilebrowsers.com/
    var firstRegex = /(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/;
    var secondRegex = /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw-(n|u)|c55\/|capi|ccwa|cdm-|cell|chtm|cldc|cmd-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc-s|devi|dica|dmob|do(c|p)o|ds(12|-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(-|_)|g1 u|g560|gene|gf-5|g-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd-(m|p|t)|hei-|hi(pt|ta)|hp( i|ip)|hs-c|ht(c(-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i-(20|go|ma)|i230|iac( |-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|-[a-w])|libw|lynx|m1-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|-([1-8]|c))|phil|pire|pl(ay|uc)|pn-2|po(ck|rt|se)|prox|psio|pt-g|qa-a|qc(07|12|21|32|60|-[2-7]|i-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h-|oo|p-)|sdk\/|se(c(-|0|1)|47|mc|nd|ri)|sgh-|shar|sie(-|m)|sk-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h-|v-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl-|tdg-|tel(i|m)|tim-|t-mo|to(pl|sh)|ts(70|m-|m3|m5)|tx-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas-|your|zeto|zte-/;
    /**
     * Mobile browser detection
     */
    var TUI_IS_MOBILE = new core.InjectionToken("[TUI_IS_MOBILE]", {
        factory: function () { return firstRegex.test(core.inject(common.USER_AGENT).toLowerCase()) ||
            secondRegex.test(core.inject(common.USER_AGENT).slice(0, 4).toLowerCase()); },
    });

    /**
     * Mobile browser that is not iOS (technically includes Windows Phone, Blackberry etc.)
     */
    var TUI_IS_ANDROID = new core.InjectionToken("[TUI_IS_ANDROID]", {
        factory: function () { return core.inject(TUI_IS_MOBILE) && !core.inject(TUI_IS_IOS); },
    });

    /**
     * Apple(safari/webkit) detection
     */
    var TUI_IS_APPLE = new core.InjectionToken("[TUI_IS_APPLE]", {
        factory: function () { return utils.tuiIsApple(core.inject(common.NAVIGATOR)); },
    });

    /**
     * Chromium browser engine detection
     */
    var TUI_IS_CHROMIUM = new core.InjectionToken("[TUI_IS_CHROMIUM]", {
        factory: function () { return !!core.inject(common.WINDOW).chrome; },
    });

    /**
     * Detect if app is running under Cypress
     * {@link https://docs.cypress.io/faq/questions/using-cypress-faq#Is-there-any-way-to-detect-if-my-app-is-running-under-Cypress Cypress docs}
     */
    var TUI_IS_CYPRESS = new core.InjectionToken("[TUI_IS_CYPRESS]", {
        factory: function () { return !!core.inject(common.WINDOW).Cypress; },
    });

    /**
     * Firefox browser engine detection
     */
    var TUI_IS_FIREFOX = new core.InjectionToken("[TUI_IS_FIREFOX]", {
        factory: function () {
            var _a;
            return typeof ((_a = core.inject(common.WINDOW)) === null || _a === void 0 ? void 0 : _a.mozCancelFullScreen) !== "undefined";
        },
    });

    var TUI_IS_STACKBLITZ = new core.InjectionToken("[TUI_IS_STACKBLITZ]", {
        factory: function () { return core.inject(common.WINDOW).location.host.endsWith("stackblitz.io"); },
    });

    /**
     * Webkit browser engine detection
     */
    var TUI_IS_WEBKIT = new core.InjectionToken("[TUI_IS_WEBKIT]", {
        factory: function () {
            var _a;
            return !!((_a = core.inject(common.WINDOW)) === null || _a === void 0 ? void 0 : _a.webkitConvertPointFromNodeToPage);
        },
    });

    /**
     * SSR safe default empty Range
     */
    var TUI_RANGE = new core.InjectionToken("[TUI_RANGE]", {
        factory: function () { return common$1.isPlatformBrowser(core.inject(core.PLATFORM_ID)) ? new Range() : {}; },
    });

    /**
     * Swipe options
     */
    var TUI_SWIPE_OPTIONS = new core.InjectionToken("[TUI_SWIPE_OPTIONS]", {
        factory: function () { return ({ timeout: 500, threshold: 30 }); },
    });

    /**
     * @description:
     * The isTrusted read-only property of the Event interface is a boolean value that is true
     * when the event was generated by a user action, and false when the event was created or
     * modified by a script or dispatched via EventTarget.dispatchEvent().
     */
    var TUI_TAKE_ONLY_TRUSTED_EVENTS = new core.InjectionToken("[TUI_TAKE_ONLY_TRUSTED_EVENTS]", {
        factory: constants.ALWAYS_TRUE_HANDLER,
    });

    var TUI_TOUCH_SUPPORTED = new core.InjectionToken("[TUI_TOUCH_SUPPORTED]", {
        factory: function () { return core.inject(common.WINDOW).matchMedia("(any-pointer: coarse)").matches; },
    });

    /**
     * Window height accounting for disappearing address bar
     */
    var TUI_WINDOW_HEIGHT = new core.InjectionToken("[TUI_WINDOW_HEIGHT]", {
        factory: function () {
            var win = core.inject(common.WINDOW);
            return observables.tuiTypedFromEvent(win, "resize").pipe(operators.startWith(null), operators.map(function () { return win.innerHeight; }), operators.shareReplay({ bufferSize: 1, refCount: true }));
        },
    });

    /**
     * Zoom options
     */
    var TUI_ZOOM_OPTIONS = new core.InjectionToken("[TUI_ZOOM_OPTIONS]", {
        factory: function () { return ({
            wheelSensitivity: 0.01,
        }); },
    });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_ACTIVE_ELEMENT = TUI_ACTIVE_ELEMENT;
    exports.TUI_ALERTS = TUI_ALERTS;
    exports.TUI_DEFAULT_RENDERER = TUI_DEFAULT_RENDERER;
    exports.TUI_DIALOGS = TUI_DIALOGS;
    exports.TUI_ENSURE_BASE_HREF = TUI_ENSURE_BASE_HREF;
    exports.TUI_FOCUSABLE_ITEM_ACCESSOR = TUI_FOCUSABLE_ITEM_ACCESSOR;
    exports.TUI_IS_ANDROID = TUI_IS_ANDROID;
    exports.TUI_IS_APPLE = TUI_IS_APPLE;
    exports.TUI_IS_CHROMIUM = TUI_IS_CHROMIUM;
    exports.TUI_IS_CYPRESS = TUI_IS_CYPRESS;
    exports.TUI_IS_FIREFOX = TUI_IS_FIREFOX;
    exports.TUI_IS_IOS = TUI_IS_IOS;
    exports.TUI_IS_MOBILE = TUI_IS_MOBILE;
    exports.TUI_IS_STACKBLITZ = TUI_IS_STACKBLITZ;
    exports.TUI_IS_WEBKIT = TUI_IS_WEBKIT;
    exports.TUI_RANGE = TUI_RANGE;
    exports.TUI_REMOVED_ELEMENT = TUI_REMOVED_ELEMENT;
    exports.TUI_SWIPE_OPTIONS = TUI_SWIPE_OPTIONS;
    exports.TUI_TAKE_ONLY_TRUSTED_EVENTS = TUI_TAKE_ONLY_TRUSTED_EVENTS;
    exports.TUI_TOUCH_SUPPORTED = TUI_TOUCH_SUPPORTED;
    exports.TUI_WINDOW_HEIGHT = TUI_WINDOW_HEIGHT;
    exports.TUI_ZOOM_OPTIONS = TUI_ZOOM_OPTIONS;
    exports.tuiAsAlerts = tuiAsAlerts;
    exports.tuiAsDialog = tuiAsDialog;
    exports.tuiAsFocusableItemAccessor = tuiAsFocusableItemAccessor;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
