// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-zoom.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/services'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/zoom', ['exports', '@angular/core', '@taiga-ui/cdk/services', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.zoom = {}), global.ng.core, global["taiga-ui"].cdk.services, global.rxjs));
})(this, (function (exports, i0, services, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiZoomDirective = /** @class */ (function () {
        function TuiZoomDirective(tuiZoom) {
            this.tuiZoom = tuiZoom;
        }
        return TuiZoomDirective;
    }());
    TuiZoomDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiZoomDirective, deps: [{ token: services.TuiZoomService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiZoomDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiZoomDirective, selector: "[tuiZoom]", outputs: { tuiZoom: "tuiZoom" }, host: { properties: { "style.touch-action": "\"none\"" } }, providers: [services.TuiZoomService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiZoomDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiZoom]',
                        outputs: ['tuiZoom'],
                        providers: [services.TuiZoomService],
                        host: {
                            '[style.touch-action]': '"none"',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [services.TuiZoomService]
                        }] }];
        } });

    /**
     * @experimental
     */
    var TuiZoomModule = /** @class */ (function () {
        function TuiZoomModule() {
        }
        return TuiZoomModule;
    }());
    TuiZoomModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiZoomModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiZoomModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiZoomModule, declarations: [TuiZoomDirective], exports: [TuiZoomDirective] });
    TuiZoomModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiZoomModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiZoomModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiZoomDirective],
                        exports: [TuiZoomDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiZoomDirective = TuiZoomDirective;
    exports.TuiZoomModule = TuiZoomModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
