// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-focus-visible.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/services'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/focus-visible', ['exports', '@angular/core', '@taiga-ui/cdk/services', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives["focus-visible"] = {}), global.ng.core, global["taiga-ui"].cdk.services, global.rxjs));
})(this, (function (exports, i0, services, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    /**
     * Directive to imitate :focus-visible
     * (https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible)
     * in browsers that do not support it
     */
    var TuiFocusVisibleDirective = /** @class */ (function () {
        function TuiFocusVisibleDirective(tuiFocusVisibleChange) {
            this.tuiFocusVisibleChange = tuiFocusVisibleChange;
        }
        return TuiFocusVisibleDirective;
    }());
    TuiFocusVisibleDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusVisibleDirective, deps: [{ token: services.TuiFocusVisibleService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiFocusVisibleDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiFocusVisibleDirective, selector: "[tuiFocusVisibleChange]", outputs: { tuiFocusVisibleChange: "tuiFocusVisibleChange" }, providers: [services.TuiDestroyService, services.TuiFocusVisibleService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusVisibleDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiFocusVisibleChange]',
                        outputs: ['tuiFocusVisibleChange'],
                        providers: [services.TuiDestroyService, services.TuiFocusVisibleService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [services.TuiFocusVisibleService]
                        }] }];
        } });

    var TuiFocusVisibleModule = /** @class */ (function () {
        function TuiFocusVisibleModule() {
        }
        return TuiFocusVisibleModule;
    }());
    TuiFocusVisibleModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusVisibleModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFocusVisibleModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusVisibleModule, declarations: [TuiFocusVisibleDirective], exports: [TuiFocusVisibleDirective] });
    TuiFocusVisibleModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusVisibleModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFocusVisibleModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFocusVisibleDirective],
                        exports: [TuiFocusVisibleDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiFocusVisibleDirective = TuiFocusVisibleDirective;
    exports.TuiFocusVisibleModule = TuiFocusVisibleModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
