// @taiga-ui/cdk/bundles/taiga-ui-cdk-pipes-keys.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/pipes/keys', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.pipes = global["taiga-ui"].cdk.pipes || {}, global["taiga-ui"].cdk.pipes.keys = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiKeysPipe = /** @class */ (function () {
        function TuiKeysPipe() {
        }
        TuiKeysPipe.prototype.transform = function (object) {
            return Object.keys(object);
        };
        return TuiKeysPipe;
    }());
    TuiKeysPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiKeysPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipe, name: "tuiKeys" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: "tuiKeys" }]
            }] });

    var TuiKeysPipeModule = /** @class */ (function () {
        function TuiKeysPipeModule() {
        }
        return TuiKeysPipeModule;
    }());
    TuiKeysPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiKeysPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipeModule, declarations: [TuiKeysPipe], exports: [TuiKeysPipe] });
    TuiKeysPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeysPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        exports: [TuiKeysPipe],
                        declarations: [TuiKeysPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiKeysPipe = TuiKeysPipe;
    exports.TuiKeysPipeModule = TuiKeysPipeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
