// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-high-dpi.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ng-web-apis/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/high-dpi', ['exports', '@angular/core', '@ng-web-apis/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives["high-dpi"] = {}), global.ng.core, global.common));
})(this, (function (exports, i0, common) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    /**
     * Only adds current content if user has High DPI display
     */
    var TuiHighDpiDirective = /** @class */ (function () {
        function TuiHighDpiDirective(_a, viewContainer, templateRef) {
            var devicePixelRatio = _a.devicePixelRatio;
            if (devicePixelRatio > 1) {
                viewContainer.createEmbeddedView(templateRef);
            }
        }
        return TuiHighDpiDirective;
    }());
    TuiHighDpiDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHighDpiDirective, deps: [{ token: common.WINDOW }, { token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiHighDpiDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHighDpiDirective, selector: "[tuiHighDpi]", ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHighDpiDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiHighDpi]',
                    }]
            }], ctorParameters: function () {
            return [{ type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }, { type: i0__namespace.ViewContainerRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ViewContainerRef]
                        }] }, { type: i0__namespace.TemplateRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.TemplateRef]
                        }] }];
        } });

    var TuiHighDpiModule = /** @class */ (function () {
        function TuiHighDpiModule() {
        }
        return TuiHighDpiModule;
    }());
    TuiHighDpiModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHighDpiModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiHighDpiModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHighDpiModule, declarations: [TuiHighDpiDirective], exports: [TuiHighDpiDirective] });
    TuiHighDpiModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHighDpiModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHighDpiModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiHighDpiDirective],
                        exports: [TuiHighDpiDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiHighDpiDirective = TuiHighDpiDirective;
    exports.TuiHighDpiModule = TuiHighDpiModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
