// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-element.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/element', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.element = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiElementDirective = /** @class */ (function () {
        function TuiElementDirective(_a) {
            var nativeElement = _a.nativeElement;
            /**
             * @note:
             * Typically, when your constructor is invoked with new,
             * an object is created, its constructor is assigned to
             * the invoked constructor and the object is then assigned
             * to this before executing any operations specified
             * in your constructor method.
             *
             * ERROR TypeError: Class constructor ElementRef cannot be invoked without 'new'
             * https://github.com/Tinkoff/taiga-ui/issues/3072
             *
             * This way we can instantiate object creation
             * without additional prototype chain for possible fix bug.
             */
            return new i0.ElementRef(nativeElement);
        }
        return TuiElementDirective;
    }());
    TuiElementDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiElementDirective, deps: [{ token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiElementDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiElementDirective, selector: "[tuiElement]", exportAs: ["elementRef"], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiElementDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiElement]',
                        exportAs: 'elementRef',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        } });

    var TuiElementModule = /** @class */ (function () {
        function TuiElementModule() {
        }
        return TuiElementModule;
    }());
    TuiElementModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiElementModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiElementModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiElementModule, declarations: [TuiElementDirective], exports: [TuiElementDirective] });
    TuiElementModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiElementModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiElementModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiElementDirective],
                        exports: [TuiElementDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiElementDirective = TuiElementDirective;
    exports.TuiElementModule = TuiElementModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
