// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-portal.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/components/dropdown-host')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/portal', ['exports', '@angular/core', '@taiga-ui/cdk/components/dropdown-host'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.portal = {}), global.ng.core, global["taiga-ui"].cdk.components["dropdown-host"]));
})(this, (function (exports, i0, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiPortalDirective = /** @class */ (function () {
        function TuiPortalDirective(templateRef, portalService) {
            this.templateRef = templateRef;
            this.portalService = portalService;
        }
        Object.defineProperty(TuiPortalDirective.prototype, "tuiPortal", {
            set: function (show) {
                var _a;
                (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
                if (show) {
                    this.viewRef = this.portalService.addTemplate(this.templateRef);
                }
            },
            enumerable: false,
            configurable: true
        });
        TuiPortalDirective.prototype.ngOnDestroy = function () {
            var _a;
            (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
        };
        return TuiPortalDirective;
    }());
    TuiPortalDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPortalDirective, deps: [{ token: i0.TemplateRef }, { token: i1.TuiDropdownPortalService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPortalDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPortalDirective, selector: "[tuiPortal]", inputs: { tuiPortal: "tuiPortal" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPortalDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPortal]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.TemplateRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.TemplateRef]
                        }] }, { type: i1__namespace.TuiDropdownPortalService, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiDropdownPortalService]
                        }] }];
        }, propDecorators: { tuiPortal: [{
                    type: i0.Input
                }] } });

    var TuiPortalModule = /** @class */ (function () {
        function TuiPortalModule() {
        }
        return TuiPortalModule;
    }());
    TuiPortalModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPortalModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPortalModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPortalModule, declarations: [TuiPortalDirective], exports: [TuiPortalDirective] });
    TuiPortalModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPortalModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPortalModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiPortalDirective],
                        exports: [TuiPortalDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPortalDirective = TuiPortalDirective;
    exports.TuiPortalModule = TuiPortalModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
