// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-pressed.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/observables'), require('@taiga-ui/cdk/tokens')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/pressed', ['exports', '@angular/core', '@taiga-ui/cdk/observables', '@taiga-ui/cdk/tokens'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.pressed = {}), global.ng.core, global["taiga-ui"].cdk.observables, global["taiga-ui"].cdk.tokens));
})(this, (function (exports, i0, observables, tokens) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiPressedDirective = /** @class */ (function () {
        function TuiPressedDirective(el, takeOnlyTrustedEvents) {
            this.el = el;
            this.takeOnlyTrustedEvents = takeOnlyTrustedEvents;
            this.tuiPressedChange = observables.tuiPressedObservable(this.el.nativeElement, {
                onlyTrusted: this.takeOnlyTrustedEvents,
            });
        }
        return TuiPressedDirective;
    }());
    TuiPressedDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPressedDirective, deps: [{ token: i0.ElementRef }, { token: tokens.TUI_TAKE_ONLY_TRUSTED_EVENTS }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPressedDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPressedDirective, selector: "[tuiPressedChange]", outputs: { tuiPressedChange: "tuiPressedChange" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPressedDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPressedChange]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_TAKE_ONLY_TRUSTED_EVENTS]
                        }] }];
        }, propDecorators: { tuiPressedChange: [{
                    type: i0.Output
                }] } });

    var TuiPressedModule = /** @class */ (function () {
        function TuiPressedModule() {
        }
        return TuiPressedModule;
    }());
    TuiPressedModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPressedModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPressedModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPressedModule, declarations: [TuiPressedDirective], exports: [TuiPressedDirective] });
    TuiPressedModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPressedModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPressedModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiPressedDirective],
                        exports: [TuiPressedDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPressedDirective = TuiPressedDirective;
    exports.TuiPressedModule = TuiPressedModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
