// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-autofilled.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/services')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/autofilled', ['exports', '@angular/core', '@taiga-ui/cdk/services'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.autofilled = {}), global.ng.core, global["taiga-ui"].cdk.services));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiAutofilledStyleComponent = /** @class */ (function () {
            function TuiAutofilledStyleComponent() {
            }
            return TuiAutofilledStyleComponent;
        }());
        TuiAutofilledStyleComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledStyleComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiAutofilledStyleComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAutofilledStyleComponent, selector: "tui-autofilled-style", ngImport: i0__namespace, template: '', isInline: true, styles: [".tui-autofill input,input.tui-autofill{transition:box-shadow .01s}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledStyleComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            selector: 'tui-autofilled-style',
                            template: '',
                            styleUrls: ['./autofilled.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        }]
                }] });

        var TuiAutofilledDirective = /** @class */ (function () {
            function TuiAutofilledDirective(directiveStyles) {
                this.autofilled = false;
                this.tuiAutofilledChange = new i0.EventEmitter();
                directiveStyles.addComponent(TuiAutofilledStyleComponent);
            }
            TuiAutofilledDirective.prototype.transitionStartHandler = function (_b) {
                var propertyName = _b.propertyName, target = _b.target;
                var _a;
                var matchedAutofill = propertyName.includes('box-shadow') && ((_a = target) === null || _a === void 0 ? void 0 : _a.matches('input'));
                if (matchedAutofill) {
                    this.autofilled = !this.autofilled;
                    this.tuiAutofilledChange.emit(this.autofilled);
                }
            };
            return TuiAutofilledDirective;
        }());
        TuiAutofilledDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiAutofilledDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAutofilledDirective, selector: "[tuiAutofilledChange]", outputs: { tuiAutofilledChange: "tuiAutofilledChange" }, host: { listeners: { "transitionstart": "transitionStartHandler($event)" }, properties: { "class._autofilled": "this.autofilled" }, classAttribute: "tui-autofill" }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiAutofilledChange]',
                            host: { class: 'tui-autofill' },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { autofilled: [{
                        type: i0.HostBinding,
                        args: ['class._autofilled']
                    }], tuiAutofilledChange: [{
                        type: i0.Output
                    }], transitionStartHandler: [{
                        type: i0.HostListener,
                        args: ['transitionstart', ['$event']]
                    }] } });

        var TuiAutofilledModule = /** @class */ (function () {
            function TuiAutofilledModule() {
            }
            return TuiAutofilledModule;
        }());
        TuiAutofilledModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiAutofilledModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledModule, declarations: [TuiAutofilledDirective, TuiAutofilledStyleComponent], exports: [TuiAutofilledDirective] });
        TuiAutofilledModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutofilledModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            exports: [TuiAutofilledDirective],
                            declarations: [TuiAutofilledDirective, TuiAutofilledStyleComponent],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiAutofilledDirective = TuiAutofilledDirective;
        exports.TuiAutofilledModule = TuiAutofilledModule;
        exports.TuiAutofilledStyleComponent = TuiAutofilledStyleComponent;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
