// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-for.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/for', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives["for"] = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiForDirective = /** @class */ (function () {
        function TuiForDirective(vcr) {
            this.vcr = vcr;
            this.ngForOf = [];
        }
        TuiForDirective.prototype.ngOnChanges = function () {
            var _a, _b;
            (_a = this.ref) === null || _a === void 0 ? void 0 : _a.destroy();
            if (((_b = this.ngForOf) === null || _b === void 0 ? void 0 : _b.length) === 0 && this.ngForEmpty) {
                this.ref = this.vcr.createEmbeddedView(this.ngForEmpty);
            }
            else if (!this.ngForOf && this.ngForElse) {
                this.ref = this.vcr.createEmbeddedView(this.ngForElse);
            }
        };
        return TuiForDirective;
    }());
    TuiForDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiForDirective, deps: [{ token: i0.ViewContainerRef }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiForDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiForDirective, selector: "[ngFor][ngForOf][ngForElse],[ngFor][ngForOf][ngForEmpty]", inputs: { ngForOf: "ngForOf", ngForElse: "ngForElse", ngForEmpty: "ngForEmpty" }, usesOnChanges: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiForDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[ngFor][ngForOf][ngForElse],[ngFor][ngForOf][ngForEmpty]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ViewContainerRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ViewContainerRef]
                        }] }];
        }, propDecorators: { ngForOf: [{
                    type: i0.Input
                }], ngForElse: [{
                    type: i0.Input
                }], ngForEmpty: [{
                    type: i0.Input
                }] } });

    var TuiForModule = /** @class */ (function () {
        function TuiForModule() {
        }
        return TuiForModule;
    }());
    TuiForModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiForModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiForModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiForModule, declarations: [TuiForDirective], exports: [TuiForDirective] });
    TuiForModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiForModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiForModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiForDirective],
                        exports: [TuiForDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiForDirective = TuiForDirective;
    exports.TuiForModule = TuiForModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
