// @taiga-ui/kit/esm2015/components/pagination/pagination.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule, TuiLetModule, TuiPreventDefaultModule, TuiRepeatTimesModule, } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPaginationComponent } from './pagination.component';
import * as i0 from "@angular/core";
export class TuiPaginationModule {
}
TuiPaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPaginationModule, declarations: [TuiPaginationComponent], imports: [CommonModule,
        TuiRepeatTimesModule,
        TuiLetModule,
        PolymorpheusModule,
        TuiPreventDefaultModule,
        TuiActiveZoneModule,
        TuiButtonModule], exports: [TuiPaginationComponent] });
TuiPaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPaginationModule, imports: [[
            CommonModule,
            TuiRepeatTimesModule,
            TuiLetModule,
            PolymorpheusModule,
            TuiPreventDefaultModule,
            TuiActiveZoneModule,
            TuiButtonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiRepeatTimesModule,
                        TuiLetModule,
                        PolymorpheusModule,
                        TuiPreventDefaultModule,
                        TuiActiveZoneModule,
                        TuiButtonModule,
                    ],
                    declarations: [TuiPaginationComponent],
                    exports: [TuiPaginationComponent],
                }]
        }] });
