// @taiga-ui/core/esm2015/components/loader/loader.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiLoaderComponent } from './loader.component';
import * as i0 from "@angular/core";
export class TuiLoaderModule {
}
TuiLoaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLoaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLoaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLoaderModule, declarations: [TuiLoaderComponent], imports: [CommonModule, PolymorpheusModule], exports: [TuiLoaderComponent] });
TuiLoaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLoaderModule, imports: [[CommonModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLoaderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule],
                    declarations: [TuiLoaderComponent],
                    exports: [TuiLoaderComponent],
                }]
        }] });
