// @taiga-ui/kit/esm2015/components/tree/directives/tree-node.directive.js
import { Directive, Inject, Input, Optional } from '@angular/core';
import { TuiTreeItemComponent } from '../components/tree-item/tree-item.component';
import { TUI_TREE_ACCESSOR } from '../misc/tree.tokens';
import * as i0 from "@angular/core";
import * as i1 from "../components/tree-item/tree-item.component";
export class TuiTreeNodeDirective {
    constructor(directive, component) {
        this.directive = directive;
        this.component = component;
    }
    set value(value) {
        var _a;
        (_a = this.directive) === null || _a === void 0 ? void 0 : _a.register(this.component, value);
    }
    ngOnDestroy() {
        var _a;
        (_a = this.directive) === null || _a === void 0 ? void 0 : _a.unregister(this.component);
    }
}
TuiTreeNodeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeNodeDirective, deps: [{ token: TUI_TREE_ACCESSOR, optional: true }, { token: TuiTreeItemComponent }], target: i0.ɵɵFactoryTarget.Directive });
TuiTreeNodeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTreeNodeDirective, selector: "tui-tree-item[tuiTreeNode]", inputs: { value: ["tuiTreeNode", "value"] }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeNodeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-tree-item[tuiTreeNode]',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TUI_TREE_ACCESSOR]
                }] }, { type: i1.TuiTreeItemComponent, decorators: [{
                    type: Inject,
                    args: [TuiTreeItemComponent]
                }] }]; }, propDecorators: { value: [{
                type: Input,
                args: ['tuiTreeNode']
            }] } });
