// @taiga-ui/kit/esm2015/components/tree/directives/tree-children.directive.js
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { EMPTY_ARRAY, tuiDefaultProp } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class TuiTreeChildrenDirective {
    constructor() {
        this.childrenHandler = TuiTreeChildrenDirective.defaultHandler;
    }
    static defaultHandler(item) {
        return Array.isArray(item) ? item : EMPTY_ARRAY;
    }
}
TuiTreeChildrenDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeChildrenDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiTreeChildrenDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTreeChildrenDirective, selector: "tui-tree[childrenHandler]", inputs: { childrenHandler: "childrenHandler" }, ngImport: i0 });
__decorate([
    tuiDefaultProp()
], TuiTreeChildrenDirective.prototype, "childrenHandler", void 0);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTreeChildrenDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-tree[childrenHandler]',
                }]
        }], propDecorators: { childrenHandler: [{
                type: Input
            }] } });
