// @taiga-ui/kit/esm2015/components/filter/filter.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiModeModule } from '@taiga-ui/core';
import { TuiBadgeModule } from '@taiga-ui/kit/components/badge';
import { TuiCheckboxBlockModule } from '@taiga-ui/kit/components/checkbox-block';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiFilterComponent } from './filter.component';
import * as i0 from "@angular/core";
export class TuiFilterModule {
}
TuiFilterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiFilterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterModule, declarations: [TuiFilterComponent], imports: [CommonModule,
        FormsModule,
        PolymorpheusModule,
        TuiCheckboxBlockModule,
        TuiBadgeModule,
        TuiModeModule], exports: [TuiFilterComponent] });
TuiFilterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterModule, imports: [[
            CommonModule,
            FormsModule,
            PolymorpheusModule,
            TuiCheckboxBlockModule,
            TuiBadgeModule,
            TuiModeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusModule,
                        TuiCheckboxBlockModule,
                        TuiBadgeModule,
                        TuiModeModule,
                    ],
                    declarations: [TuiFilterComponent],
                    exports: [TuiFilterComponent],
                }]
        }] });
