// @taiga-ui/kit/esm2015/components/input-month-range/input-month-range.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule, TuiMapperPipeModule, TuiPreventDefaultModule, } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldComponent, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TuiCalendarMonthModule } from '@taiga-ui/kit/components/calendar-month';
import { TuiInputMonthRangeComponent } from './input-month-range.component';
import { TuiInputMonthRangeDirective } from './input-month-range.directive';
import * as i0 from "@angular/core";
export class TuiInputMonthRangeModule {
}
TuiInputMonthRangeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputMonthRangeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthRangeModule, declarations: [TuiInputMonthRangeComponent, TuiInputMonthRangeDirective], imports: [CommonModule,
        TuiCalendarMonthModule,
        TuiHostedDropdownModule,
        TuiPrimitiveTextfieldModule,
        TuiSvgModule,
        TuiPreventDefaultModule,
        TuiActiveZoneModule,
        TuiMapperPipeModule,
        TuiTextfieldControllerModule], exports: [TuiInputMonthRangeComponent,
        TuiInputMonthRangeDirective,
        TuiTextfieldComponent] });
TuiInputMonthRangeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthRangeModule, imports: [[
            CommonModule,
            TuiCalendarMonthModule,
            TuiHostedDropdownModule,
            TuiPrimitiveTextfieldModule,
            TuiSvgModule,
            TuiPreventDefaultModule,
            TuiActiveZoneModule,
            TuiMapperPipeModule,
            TuiTextfieldControllerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiCalendarMonthModule,
                        TuiHostedDropdownModule,
                        TuiPrimitiveTextfieldModule,
                        TuiSvgModule,
                        TuiPreventDefaultModule,
                        TuiActiveZoneModule,
                        TuiMapperPipeModule,
                        TuiTextfieldControllerModule,
                    ],
                    declarations: [TuiInputMonthRangeComponent, TuiInputMonthRangeDirective],
                    exports: [
                        TuiInputMonthRangeComponent,
                        TuiInputMonthRangeDirective,
                        TuiTextfieldComponent,
                    ],
                }]
        }] });
