// @taiga-ui/core/esm2015/components/scrollbar/scrollbar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiScrollControlsModule } from '@taiga-ui/core/components/scroll-controls';
import { TuiScrollRefDirective } from './scroll-ref.directive';
import { TuiScrollableDirective } from './scrollable.directive';
import { TuiScrollbarComponent } from './scrollbar.component';
import * as i0 from "@angular/core";
export class TuiScrollbarModule {
}
TuiScrollbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiScrollbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollbarModule, declarations: [TuiScrollbarComponent, TuiScrollRefDirective, TuiScrollableDirective], imports: [CommonModule, TuiScrollControlsModule], exports: [TuiScrollbarComponent, TuiScrollRefDirective, TuiScrollableDirective] });
TuiScrollbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollbarModule, imports: [[CommonModule, TuiScrollControlsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollbarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiScrollControlsModule],
                    declarations: [TuiScrollbarComponent, TuiScrollRefDirective, TuiScrollableDirective],
                    exports: [TuiScrollbarComponent, TuiScrollRefDirective, TuiScrollableDirective],
                }]
        }] });
