// @taiga-ui/core/esm2015/components/scrollbar/scrollable.directive.js
import { Directive, ElementRef, Inject } from '@angular/core';
import { TUI_SCROLLABLE } from '@taiga-ui/core/constants';
import * as i0 from "@angular/core";
export class TuiScrollableDirective {
    constructor(el) {
        this.el = el;
    }
    ngOnInit() {
        this.el.nativeElement.dispatchEvent(new CustomEvent(TUI_SCROLLABLE, {
            bubbles: true,
            detail: this.el.nativeElement,
        }));
    }
}
TuiScrollableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollableDirective, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiScrollableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiScrollableDirective, selector: "[tuiScrollable]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiScrollable]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; } });
